---
title: "DownloadProcessed"
output: html_document
date: "2023-07-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

`scfetch` provides functions for users to download processed single-cell RNA-seq data, such as **count matrix**, **cell type annotation and composition** and **RDS/RData/h5ad objects**. For count matrix, `scfetch` supports loading to `Seurat` after downloading.

Until now, the public resources supported and the returned results:

<table>
<thead>
  <tr>
    <th>Type</th>
    <th>Resources</th>
    <th>URL</th>
    <th>Download Type</th>
    <th>Returned results</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td rowspan="3">Count matrices</td>
    <td>GEO</td>
    <td>https://www.ncbi.nlm.nih.gov/geo/</td>
    <td>count matrix</td>
    <td>SeuratObject</td>
  </tr>
  <tr>
    <td>PanglaoDB</td>
    <td>https://panglaodb.se/index.html</td>
    <td>count matrix</td>
    <td>SeuratObject</td>
  </tr>
  <tr>
    <td>UCSC Cell Browser</td>
    <td>https://cells.ucsc.edu/</td>
    <td>count matrix</td>
    <td>SeuratObject</td>
  </tr>
  <tr>
    <td rowspan="2">Objects</td>
    <td>Zenodo</td>
    <td>https://zenodo.org/</td>
    <td>count matrix, rds, rdata, h5ad, et al.</td>
    <td>NULL</td>
  </tr>
  <tr>
    <td>CELLxGENE</td>
    <td>https://cellxgene.cziscience.com/</td>
    <td>rds, h5ad</td>
    <td>NULL</td>
  </tr>
</tbody>
</table>

<hr />

## Count matrices

### GEO

[GEO is an international public repository that archives and freely distributes microarray, next-generation sequencing, and other forms of high-throughput functional genomics data submitted by the research community.](https://www.ncbi.nlm.nih.gov/geo/info/overview.html) It provides a very convenient way for users to explore and select interested scRNA-seq datasets. 

#### Extract metadata

`scfetch` provides `ExtractGEOMeta` to extract sample metadata, including sample title, source name/tissue, description, cell type, treatment, paper title, paper abstract, organism, protocol, data processing methods, et al.

```{r geo_meta}
# extract metadata of specified platform
GSE200257.meta <- ExtractGEOMeta(acce = "GSE200257", platform = "GPL24676")
# extract metadata of all platforms
GSE94820.meta <- ExtractGEOMeta(acce = "GSE94820", platform = NULL)
head(GSE94820.meta)
```

<hr />

#### Download matrix and load to Seurat

After manually check the extracted metadata, users can **download count matrix** and **load the count matrix** to Seurat with `ParseGEO`. 

For count matrix, `ParseGEO` supports downloading the matrix from **supplementary files** and extracting from `ExpressionSet`, users can control the source by specifying `down.supp` or detecting automatically (`ParseGEO` will extract the count matrix from `ExpressionSet` first, if the count matrix is NULL or contains non-integer values, `ParseGEO` will download supplementary files). While the supplementary files have two main types: single count matrix file containing all cells and CellRanger-style outputs (barcode, matrix, feature/gene), users are required to choose the type of supplementary files with `supp.type`.

With the count matrix, `ParseGEO` will load the matrix to Seurat automatically. If multiple samples available, users can choose to merge the SeuratObject with `merge`.

```{r geo_parse, eval=FALSE}
# for cellranger output
GSE200257.seu <- ParseGEO(
  acce = "GSE200257", platform = NULL, supp.idx = 1, down.supp = TRUE, supp.type = "10x",
  out.folder = "/Users/soyabean/Desktop/tmp/scdown/dwonload_geo"
)
# for count matrix, no need to specify out.folder, download count matrix to tmp folder
GSE94820.seu <- ParseGEO(acce = "GSE94820", platform = NULL, supp.idx = 1, down.supp = TRUE, supp.type = "count")
```

The structure of downloaded matrix for 10x:
```{bash geo_parse_structure}
tree /Users/soyabean/Desktop/tmp/scdown/dwonload_geo
```

**For bulk RNA-seq**, set `data.type = "bulk"` in `ParseGEO`, this will return count matrix.

<hr />

### PanglaoDB

[PanglaoDB](https://panglaodb.se/index.html) is a database which contains scRNA-seq datasets from mouse and human. Up to now, it contains **5,586,348 cells** from **1368 datasets (1063 from Mus musculus and 305 from	Homo sapiens)**. It has well organized metadata for every dataset, including tissue, protocol, species, number of cells and cell type annotation (computationally identified). Daniel Osorio has developed [rPanglaoDB](https://github.com/dosorio/rPanglaoDB/) to access [PanglaoDB](https://panglaodb.se/index.html) data, the functions of `scfetch` here are based on [rPanglaoDB](https://github.com/dosorio/rPanglaoDB/).

#### Extract metadata

`scfetch` provides `ExtractPanglaoDBMeta` to show all available datasets or select interested datasets with specified species, protocol, tissue and cell number. User can also choose to whether to add cell type annotation to every dataset with `show.cell.type`.

Since [PanglaoDB](https://panglaodb.se/about.html) is no longer maintained, `scfetch` has cached all metadata and sample composition and use these cached data by default to accelerate, users can change this by setting `local.data = FALSE`.

```{r panglaodb_meta}
hsa.meta <- ExtractPanglaoDBMeta(species = "Homo sapiens", protocol = c("Smart-seq2", "10x chromium"), show.cell.type = TRUE, cell.num = c(1000, 2000))
head(hsa.meta)
```

<hr />

#### Extract cell type composition

`scfetch` provides `ExtractPanglaoDBComposition` to extract cell type annotation and composition (use cached data by default to accelerate, users can change this by setting `local.data = FALSE`).

```{r panglaodb_celltype}
hsa.composition <- ExtractPanglaoDBComposition(species = "Homo sapiens", protocol = c("Smart-seq2", "10x chromium"))
head(hsa.composition)
```

<hr />

#### Download matrix and load to Seurat

After manually check the extracted metadata, `scfetch` provides `ParsePanglaoDB` to **download count matrix** and **load the count matrix** to Seurat. With available cell type annotation, uses can filter samples without specified cell type with `cell.type`. Users can also include/exclude cells expressing specified genes with `include.gene`/`exclude.gene`. 

With the count matrix, `ParsePanglaoDB` will load the matrix to Seurat automatically. If multiple samples available, users can choose to merge the SeuratObject with `merge`.

```{r panglaodb_parse, eval=FALSE}
hsa.seu <- ParsePanglaoDB(hsa.meta, merge = TRUE)
```

<hr />

### UCSC Cell Browser

The [UCSC Cell Browser](https://cells.ucsc.edu/?#) is a web-based tool that allows scientists to interactively visualize scRNA-seq datasets. It contains **1040 single cell datasets** from **17 different species**. And, it is **organized with the hierarchical structure**, which can help users quickly locate the samples they are interested in.

#### Show available datasets

`scfetch` provides `ShowCBDatasets` to show all available datasets. Due to the large number of datasets, `ShowCBDatasets` enables users to perform *lazy load* of dataset json files instead of downloading the json files online (time-consuming!!!). This *lazy load* requires users to provide `json.folder` to save json files and set `lazy = TRUE` (for the first time of run, `ShowCBDatasets` will download current json files to `json.folder`, for next time of run, with `lazy = TRUE`, `ShowCBDatasets` will load the downloaded json files from `json.folder`.). And, `ShowCBDatasets` supports updating the local datasets with `update = TRUE`.

```{r cb_show}
# first time run, the json files are stored under json.folder
# ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "/Users/soyabean/Desktop/tmp/scdown/cell_browser/json", update = TRUE)

# second time run, load the downloaded json files
ucsc.cb.samples <- ShowCBDatasets(lazy = TRUE, json.folder = "/Users/soyabean/Desktop/tmp/scdown/cell_browser/json", update = FALSE)
head(ucsc.cb.samples)

# always read online
# ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)
```

The number of datasets and all available species:

```{r cb_show_detail}
# the number of datasets
nrow(ucsc.cb.samples)

# available species
unique(unlist(sapply(unique(gsub(pattern = "\\|parent", replacement = "", x = ucsc.cb.samples$organisms)), function(x) {
  unlist(strsplit(x = x, split = ", "))
})))
```

<hr />

#### Extract metadata

`scfetch` provides `ExtractCBDatasets` to filter metadata with collection, sub-collection, organ, disease status, organism, project and cell number. All attributes except cell number support fuzzy match with `fuzzy.match`, this is useful when selecting datasets.

```{r cb_extract}
hbb.sample.df <- ExtractCBDatasets(all.samples.df = ucsc.cb.samples, organ = c("brain", "blood"), organism = "Human (H. sapiens)", cell.num = c(1000, 2000))
head(hbb.sample.df)
```

<hr />

#### Extract cell type composition

`scfetch` provides `ExtractCBComposition` to extract cell type annotation and composition.

```{r cb_celltype}
hbb.sample.ct <- ExtractCBComposition(json.folder = "/Users/soyabean/Desktop/tmp/scdown/cell_browser/json", sample.df = hbb.sample.df)
head(hbb.sample.ct)
```

<hr />

#### Load the online datasets to Seurat

After manually check the extracted metadata, `scfetch` provides `ParseCBDatasets` to **load the online count matrix** to Seurat. All the attributes available in `ExtractCBDatasets` are also same here. Please note that the loading process provided by `ParseCBDatasets` will load the online count matrix instead of downloading it to local. If multiple samples available, users can choose to merge the SeuratObject with `merge`.

```{r cb_parse, eval=FALSE}
hbb.sample.seu <- ParseCBDatasets(sample.df = hbb.sample.df)
```

<hr />

## Objects

### Zenodo

Different from above resources which mainly contains count matrix, [Zenodo](https://zenodo.org/) mainly contains processed objects, such as `SeuratObject` which has been clustered and annotated, `AnnData` which contains processed results generated by `scanpy`.

#### Extract metadata

`scfetch` provides `ExtractZenodoMeta` to extract sample metadata, including sample title, description, available files and corresponding md5. Please note that when the dataset is restricted access, the returned dataframe will be empty.

```{r zenodo_meta}
# single doi
zebrafish.df <- ExtractZenodoMeta(doi = "10.5281/zenodo.7243603")
head(zebrafish.df)

# vector dois
multi.dois <- ExtractZenodoMeta(doi = c("1111", "10.5281/zenodo.7243603", "10.5281/zenodo.7244441"))
head(multi.dois)
```

<hr />

#### Download object

After manually check the extracted metadata, users can **download the specified objects** with `ParseZenodo`. The downloaded objects are controlled by `file.ext` and **the provided object formats should be in lower case (e.g. rds/rdata/h5ad).**

The returned result is a dataframe containing failed objects. If not `NULL`, users can re-run `ParseZenodo` by setting `doi` to the returned result.

```{r zenodo_parse, eval=FALSE}
multi.dois.parse <- ParseZenodo(
  doi = c("1111", "10.5281/zenodo.7243603", "10.5281/zenodo.7244441"),
  file.ext = c("rdata", "rds"), out.folder = "/Users/soyabean/Desktop/tmp/scdown/download_zenodo"
)
```

The structure of downloaded objects:
```{bash zenodo_parse_structure}
tree /Users/soyabean/Desktop/tmp/scdown/download_zenodo
```

<hr />

### CELLxGENE

The [CELLxGENE](https://cellxgene.cziscience.com/) is a web server contains **910** single-cell datasets, users can explore, download and upload own datasets. The downloaded datasets provided by [CELLxGENE](https://cellxgene.cziscience.com/) have two formats: `h5ad (AnnData v0.8)` and `rds (Seurat v4)`.

#### Show available datasets

`scfetch` provides `ShowCELLxGENEDatasets` to extract sample metadata, including sample title, description, contact, organism, ethnicity, sex, tissue, disease, assay, suspension type, cell type, et al.

```{r cellxgene_all}
# all available datasets
all.cellxgene.datasets <- ShowCELLxGENEDatasets()
```

<hr />

#### Extract metadata

`scfetch` provides `ExtractCELLxGENEMeta` to filter sample metadata:

```{r cellxgene_meta}
# human 10x v2 and v3 datasets
human.10x.cellxgene.meta <- ExtractCELLxGENEMeta(
  all.samples.df = all.cellxgene.datasets,
  assay = c("10x 3' v2", "10x 3' v3"), organism = "Homo sapiens"
)
head(human.10x.cellxgene.meta)
```

<hr />

#### Download object

After manually check the extracted metadata, users can **download the specified objects** with `ParseCELLxGENE`. The downloaded objects are controlled by `file.ext` (choose from `"rds"` and `"h5ad"`).

```{r cellxgene_parse, eval=FALSE}
ParseCELLxGENE(
  meta = human.10x.cellxgene.meta[1:5, ], file.ext = "rds",
  out.folder = "/Users/soyabean/Desktop/tmp/scdown/download_cellxgene"
)
```

The structure of downloaded objects:
```{bash cellxgene_parse_structure}
tree /Users/soyabean/Desktop/tmp/scdown/download_cellxgene
```

<hr />


