---
title: "DownloadMatrices"
output: html_document
date: "2023-07-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

`GEfetch2R` provides functions for users to download **count matrices** and **annotations** (e.g. cell type annotation and composition) from GEO and some single-cell databases (e.g. [PanglaoDB](https://panglaodb.se/index.html) and [UCSC Cell Browser](https://cells.ucsc.edu/?#)). `GEfetch2R` also supports loading the downloaded data to `Seurat`.

Until now, the public resources supported and the returned values:

<table>
<thead>
  <tr>
    <th>Resources</th>
    <th>URL</th>
    <th>Download Type</th>
    <th>Returned values</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>GEO</td>
    <td>https://www.ncbi.nlm.nih.gov/geo/</td>
    <td>count matrix</td>
    <td>SeuratObject or count matrix for bulk RNA-seq</td>
  </tr>
  <tr>
    <td>PanglaoDB</td>
    <td>https://panglaodb.se/index.html</td>
    <td>count matrix and annotation</td>
    <td>SeuratObject</td>
  </tr>
  <tr>
    <td>UCSC Cell Browser</td>
    <td>https://cells.ucsc.edu/</td>
    <td>count matrix and annotation</td>
    <td>SeuratObject</td>
  </tr>
</tbody>
</table>

<hr />

## GEO

[GEO is an international public repository that archives and freely distributes microarray, next-generation sequencing, and other forms of high-throughput functional genomics data submitted by the research community.](https://www.ncbi.nlm.nih.gov/geo/info/overview.html) It provides a very convenient way for users to explore and select interested scRNA-seq datasets. 

```{r echo=FALSE, results='hide',message=FALSE}
library(GEfetch2R)
load("/Volumes/soyabean/GEfetch2R/GEfetch2R_scrna-seq_download.RData")
```

### Extract metadata

`GEfetch2R` provides `ExtractGEOMeta` to extract sample metadata, including sample title, source name/tissue, description, cell type, treatment, paper title, paper abstract, organism, protocol, data processing methods, et al.

```{r geo_meta, eval=FALSE}
# library
library(GEfetch2R)

# extract metadata of specified platform
GSE200257.meta <- ExtractGEOMeta(acce = "GSE200257", platform = "GPL24676")
# set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough
Sys.setenv("VROOM_CONNECTION_SIZE" = 131072 * 60)
# extract metadata of all platforms
GSE94820.meta <- ExtractGEOMeta(acce = "GSE94820", platform = NULL)
```

Show the metadata:
```{r geo_meta_show, eval=FALSE}
head(GSE94820.meta)
```

<hr />

### Download matrix and load to Seurat

After manually check the extracted metadata, users can **download count matrix** and **load the count matrix** to Seurat with `ParseGEO`. 

For count matrix, `ParseGEO` supports downloading the matrix from **supplementary files** and extracting from `ExpressionSet`, users can control the source by specifying `down.supp` or detecting automatically (`ParseGEO` will extract the count matrix from `ExpressionSet` first, if the count matrix is NULL or contains non-integer values, `ParseGEO` will download supplementary files). While the supplementary files have two main types: single count matrix file containing all cells and CellRanger-style outputs (barcode, matrix, feature/gene), users are required to choose the type of supplementary files with `supp.type`.

With the count matrix, `ParseGEO` will load the matrix to Seurat automatically. If multiple samples available, users can choose to merge the SeuratObject with `merge`.

```{r geo_parse, eval=FALSE}
# cellranger out
GSE200257.seu <- ParseGEO(
  acce = "GSE200257", platform = NULL, supp.idx = 1,
  down.supp = TRUE, supp.type = "10x",
  out.folder = "/Volumes/soyabean/GEfetch2R/dwonload_geo"
)
# count matrix
GSE94820.seu <- ParseGEO(
  acce = "GSE94820", platform = NULL,
  supp.idx = 1, down.supp = TRUE,
  supp.type = "count"
)
```

Show the `GSE200257.seu`:
```{r geo_parse_show1, eval=FALSE}
# for cellranger output
GSE200257.seu
```

Show the `GSE94820.seu`:
```{r geo_parse_show2, eval=FALSE}
# for count matrix
GSE94820.seu
```

The structure of downloaded count matrix for 10x:
```{bash geo_parse_structure}
tree /Volumes/soyabean/GEfetch2R/dwonload_geo
```

**For bulk RNA-seq**, set `data.type = "bulk"` in `ParseGEO`, this will return count matrix.

<hr />

## PanglaoDB

[PanglaoDB](https://panglaodb.se/index.html) is a database which contains scRNA-seq datasets from mouse and human. Up to now, it contains **5,586,348 cells** from **1368 datasets (1063 from Mus musculus and 305 from	Homo sapiens)**. It has well organized metadata for every dataset, including tissue, protocol, species, number of cells and cell type annotation (computationally identified). Daniel Osorio has developed [rPanglaoDB](https://github.com/dosorio/rPanglaoDB/) to access [PanglaoDB](https://panglaodb.se/index.html) data, the functions of `GEfetch2R` here are based on [rPanglaoDB](https://github.com/dosorio/rPanglaoDB/).

Since [PanglaoDB](https://panglaodb.se/about.html) is no longer maintained, `GEfetch2R` has cached all metadata and cell type composition and use these cached data by default to accelerate, users can access the cached data with `PanglaoDBMeta` (all metadata) and `PanglaoDBComposition` (all cell type composition).

### Summary attributes

`GEfetch2R` provides `StatDBAttribute` to summary attributes of [PanglaoDB](https://panglaodb.se/index.html):

```{r panglaodb_summary}
# use cached metadata
StatDBAttribute(df = PanglaoDBMeta, filter = c("species", "protocol"), database = "PanglaoDB")
```

<hr />

### Extract metadata

`GEfetch2R` provides `ExtractPanglaoDBMeta` to select interested datasets with specified **species**, **protocol**, **tissue** and **cell number** (The available values of these attributes can be obtained with `StatDBAttribute`). User can also choose to whether to add cell type annotation to every dataset with `show.cell.type`.

`GEfetch2R` uses cached metadata and cell type composition by default, users can change this by setting `local.data = FALSE`.

```{r panglaodb_meta, eval=FALSE}
hsa.meta <- ExtractPanglaoDBMeta(
  species = "Homo sapiens", protocol = c("Smart-seq2", "10x chromium"),
  show.cell.type = TRUE, cell.num = c(1000, 2000)
)
```

Show the metadata:
```{r panglaodb_meta_show}
head(hsa.meta)
```

<hr />

### Extract cell type composition

`GEfetch2R` provides `ExtractPanglaoDBComposition` to extract cell type annotation and composition (use cached data by default to accelerate, users can change this by setting `local.data = FALSE`).

```{r panglaodb_celltype, eval=FALSE}
hsa.composition <- ExtractPanglaoDBComposition(
  species = "Homo sapiens",
  protocol = c("Smart-seq2", "10x chromium")
)
```

Show the extracted cell type annotation and composition:
```{r panglaodb_celltype_show}
head(hsa.composition)
```

<hr />

### Download matrix and load to Seurat

After manually check the extracted metadata, `GEfetch2R` provides `ParsePanglaoDB` to **download count matrix** and **load the count matrix** to Seurat. With available cell type annotation, uses can filter datasets without specified cell type with `cell.type`. Users can also include/exclude cells expressing specified genes with `include.gene`/`exclude.gene`. 

With the count matrix, `ParsePanglaoDB` will load the matrix to Seurat automatically. If multiple datasets available, users can choose to merge the SeuratObject with `merge`.

```{r panglaodb_parse, eval=FALSE}
hsa.seu <- ParsePanglaoDB(hsa.meta[1:3, ], merge = TRUE)
```

Show the returned `SeuratObject`:
```{r panglaodb_parse_show}
hsa.seu
```

<hr />

## UCSC Cell Browser

The [UCSC Cell Browser](https://cells.ucsc.edu/?#) is a web-based tool that allows scientists to interactively visualize scRNA-seq datasets. It contains **1267 single cell datasets** from **25 different species**. And, it is **organized with the hierarchical structure**, which can help users quickly locate the datasets they are interested in.

### Show available datasets

`GEfetch2R` provides `ShowCBDatasets` to show all available datasets. Due to the large number of datasets, `ShowCBDatasets` enables users to perform *lazy load* of dataset json files instead of downloading the json files online (time-consuming!!!). This *lazy load* requires users to provide `json.folder` to save json files and set `lazy = TRUE` (for the first time of run, `ShowCBDatasets` will download current json files to `json.folder`, for next time of run, with `lazy = TRUE`, `ShowCBDatasets` will load the downloaded json files from `json.folder`.). And, `ShowCBDatasets` supports updating the local datasets with `update = TRUE`.

```{r cb_show, eval=FALSE}
# first time run, the json files are stored under json.folder
# ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "/Volumes/soyabean/GEfetch2R/cell_browser/json", update = TRUE)

# second time run, load the downloaded json files
ucsc.cb.samples <- ShowCBDatasets(lazy = TRUE, json.folder = "/Volumes/soyabean/GEfetch2R/cell_browser/json", update = FALSE)

# always read online
# ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)
```

Show the metadata:
```{r cb_show_show}
head(ucsc.cb.samples)
```

The number of datasets and all available species:

```{r cb_show_detail}
# the number of datasets
nrow(ucsc.cb.samples)

# available species
unique(unlist(sapply(unique(gsub(pattern = "\\|parent", replacement = "", x = ucsc.cb.samples$organisms)), function(x) {
  unlist(strsplit(x = x, split = ", "))
})))
```

<hr />

### Summary attributes

`GEfetch2R` provides `StatDBAttribute` to summary attributes of [UCSC Cell Browser](https://cells.ucsc.edu/?#):

```{r cb_summary}
StatDBAttribute(
  df = ucsc.cb.samples, filter = c("organism", "organ"),
  database = "UCSC", combine = TRUE
)
```

<hr />

### Extract metadata

`GEfetch2R` provides `ExtractCBDatasets` to filter metadata with **collection**, **sub-collection**, **organ**, **disease status**, **organism**, **project** and **cell number** (The available values of these attributes can be obtained with `StatDBAttribute` except **cell number**). All attributes except cell number support fuzzy match with `fuzzy.match`, this is useful when selecting datasets.

```{r cb_extract, eval=FALSE}
hbb.sample.df <- ExtractCBDatasets(
  all.samples.df = ucsc.cb.samples, organ = c("skeletal muscle"),
  organism = "Human (H. sapiens)", cell.num = c(1000, 2000)
)
```

Show the metadata:
```{r cb_extract_show}
head(hbb.sample.df)
```

<hr />

### Extract cell type composition

`GEfetch2R` provides `ExtractCBComposition` to extract cell type annotation and composition.

```{r cb_celltype, eval=FALSE}
hbb.sample.ct <- ExtractCBComposition(
  json.folder = "/Volumes/soyabean/GEfetch2R/cell_browser/json",
  sample.df = hbb.sample.df
)
```

Show the extracted cell type annotation and composition:
```{r cb_celltype_show}
head(hbb.sample.ct)
```

<hr />

### Load the online datasets to Seurat

After manually check the extracted metadata, `GEfetch2R` provides `ParseCBDatasets` to **load the online count matrix** to Seurat. All the attributes available in `ExtractCBDatasets` are also same here. Please note that the loading process provided by `ParseCBDatasets` will load the online count matrix instead of downloading it to local. If multiple datasets available, users can choose to merge the SeuratObject with `merge`.

`ParseCBDatasets` supports **extracting subset with metadata and gene**:

```{r cb_parse, eval=FALSE}
# parse the whole datasets
hbb.sample.seu <- ParseCBDatasets(sample.df = hbb.sample.df)
# subset metadata and gene
hbb.sample.seu <- ParseCBDatasets(
  sample.df = hbb.sample.df, obs.value.filter = "Cell.Type == 'MP' & Phase == 'G2M'",
  include.genes = c(
    "PAX7", "MYF5", "C1QTNF3", "MYOD1", "MYOG", "RASSF4", "MYH3", "MYL4",
    "TNNT3", "PDGFRA", "OGN", "COL3A1"
  )
)
```

Show the returned `SeuratObject`:
```{r cb_parse_show}
hbb.sample.seu
```

<hr />
