---
title: >
  GEfetch2R - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools
output: 
  html_document:
    toc_depth: 4
    toc_float: true
fig_caption: TRUE
vignette: >
  %\VignetteIndexEntry{GEfetch2R}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE, warning=FALSE}
library(knitr)
htmltools::tagList(rmarkdown::html_dependency_font_awesome())

# set dpi
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  dpi = 60
)
```

# [<mark>Quick start</mark>](https://showteeth.github.io/GEfetch2R/articles/QuickStart.html)

# Introduction

`GEfetch2R` is designed to accelerate users' downloading and preparation of single-cell/bulk RNA-seq datasets from public resources. It can be used to:

For **bulk RNA-seq**:

* Download **raw data (`sra/fastq/bam`)** from `SRA/ENA` with **GEO accession**:
  * **Parallel** download `sra/fastq/bam` files from `SRA/ENA`
  * **Support Aspera** (`ENA`)
  * **Parallel** split `sra` to `fastq` files
  * **Convert `bam` to `fastq` files**
  * **Mapping with `STAR` and load the output to `DESeq2`**
* Download **count matrix** from `GEO` with **GEO accession**:
  * **Generate count matrix from supplementary files**
  * **Load the count matrix to `DESeq2`**

For **scRNA-seq**:

* Download **raw data (`sra/fastq/bam`)** from `SRA/ENA` with **GEO accession**:
  * **Parallel** download `sra/fastq/bam` files from `SRA/ENA`
  * **Foramt downloaded `fastq` files** to standard style that can be identified by 10x softwares (e.g. `CellRanger`).
  * **Support Aspera** (`ENA`)
  * **Parallel** split `sra` to `fastq` files
  * **Download original 10x generated `bam` files (with custom tags)**
  * **Convert `bam` to `fastq` files (`samtools/bamtofastq_linux`)**
  * **Mapping with `STAR/CellRanger` and load the output to `Seurat`**
* Download **count matrix**:
  * Download **count matrix** from `GEO` with **GEO accession**
  * Download **count matrix** and **annotation (e.g. cell type)** information from `PanglaoDB` and `UCSC Cell Browser` with **key words (filter criteria)**
  * **Extract subset with annotation and gene (`PanglaoDB` and `UCSC Cell Browser`)**  
  * **Load the count matrix and annotation to `Seurat`**
* Download **processed object**:
  * Download **processed object** from `GEO` with **GEO accession**
  * Download **processed object** from `Zenodo` with **DOI**
  * Download **processed object** from `CELLxGENE` and `Human Cell Atlas` with **key words (filter criteria)**
  * **Parallel** download
  * **Extract subset with annotation and gene (`CELLxGENE`)**
  * **Load the processed object (`rds`/`rdata`) to `Seurat`**
  * **Extract count matrix and metadata from processed object (`rds`/`rdata`)**, including `Seurat`, `seurat`, `SingleCellExperiment`, `cell_data_set`, `CellDataSet`, `DESeqDataSet`, `DGEList`.
* **Format conversion and benchmark related tools**:
  * **Format conversion between widely used single cell objects** (`SeuratObject`, `AnnData`, `SingleCellExperiment`, `CellDataSet/cell_data_set` and `loom`)
  * **Benchmark the tools (more than two tools available for the conversion)**:
    * `SeuratObject` to `AnnData`: `SeuratDisk`, `sceasy`, `scDIOR`
    * `SingleCellExperiment` to `AnnData`: `sceasy`, `scDIOR`, `zellkonverter`
    * `AnnData` to `SeuratObject`: `SeuratDisk`, `sceasy`, `scDIOR`, `schard`
    * `AnnData` to `SingleCellExperiment`: `schard`, `scDIOR`, `zellkonverter`

<hr />

# Framework

<div align="center">
<img src="man/figures/GEfetch2R_overview.png"  title="GEfetch2R_framework"  alt="GEfetch2R_framework" />
</div>

<hr />

# Vignette
Detailed usage is available in [website](https://showteeth.github.io/GEfetch2R/):

* [Download raw data (SRA/ENA)](https://showteeth.github.io/scfetch/articles/DownloadRaw.html)
* [Download count matrix and annotaion (GEO, PanglaoDB, UCSC Cell Browser)](https://showteeth.github.io/scfetch/articles/DownloadMatrices.html)
* [Download processed data (GEO, Zenodo, CELLxGENE, Human Cell Atlas)](https://showteeth.github.io/scfetch/articles/DownloadObjects.html)
* [Format conversion and benchmark (AnnData, SeuratObject, SingleCellExperiment, CellDataSet/cell_data_set, loom)](https://showteeth.github.io/scfetch/articles/ObjectConversion.html)

<hr />

# Comparison with similar tools

<div align="center">
<img src="man/figures/GEfetch2R_compare.png"  title="GEfetch2R_comparison"  alt="GEfetch2R_comparison" />
</div>

<hr />
