<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog â€¢ GEfetch2R</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.3" id="gefetch2r-073">GEfetch2R 0.7.3<a class="anchor" aria-label="anchor" href="#gefetch2r-073"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-7-3">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-7-3"></a></h3>
<ul><li>Added <code>LoadRData</code> to load objects, count matrix, and metadata from RData.</li>
<li>Added <code>CheckAPI</code> to check the availability of APIs.</li>
<li>Supported loading and downloading a given dataset/collection/project from <code>CELLxGENE</code>, <code>Human Cell Atlas</code>, <code>UCSC Cell Browser</code>, <code>PanglaoDB</code>.</li>
<li>Supported multiple runs in <code>RunCellRanger</code>, <code>RunSTAR</code>, <code>Fastq2R</code> .</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-7-3">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-7-3"></a></h3>
<ul><li>Fix bug in <code>ImportSeurat</code>.</li>
<li>Fix bug in <code>LoadRData</code>.</li>
<li>Fix API changes in <code>CELLxGENE</code> and <code>Human Cell Atlas</code>.</li>
<li>Fix bugs in <code>ParseCELLxGENE</code>.</li>
<li>Fix bugs in <code>ExtractRun</code> when a sample has multiple runs.</li>
<li>Fix bugs in <code>Fastq2R</code> when a sample has multiple runs.</li>
<li>Add <code>wget</code> in <code>DownloadBam</code>, <code>DownloadFastq</code> and <code>DownloadSRA</code>.</li>
<li>Fix bug in <code>ParseGEO</code> when 10X data contains more than one type.</li>
<li>Fix bug in scientific notation (<code>36000000</code> -&gt; <code>3.6e+07</code>).</li>
<li>Add <code>skip.gsm</code> in <code>DownloadFastq2R</code> to control used <code>GSM</code>.</li>
<li>Fix bug in <code>Bam2Fastq</code>.</li>
<li>Fix bug in <code>ImportSeurat</code>.</li>
<li>Fix bug in <code>Fastq2R</code> when loading to <code>Seurat</code>.</li>
<li>Change <code>PanglanDB</code> to <code>PanglaoDB</code> (spelling mistake).</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.2" id="gefetch2r-072">GEfetch2R 0.7.2<a class="anchor" aria-label="anchor" href="#gefetch2r-072"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-7-2">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-7-2"></a></h3>
<ul><li>Large-scale evaluation:
<ul><li>Load count matrix generated by <code>featureCounts</code> from GEO (smart-seq2)</li>
<li>Load count matrix with extra columns from GEO (smart-seq2)</li>
<li>Load count matrix generated by <code>STAR</code> from GEO (smart-seq2)</li>
<li>Load count matrix in mixed file formats from GEO (smart-seq2)</li>
<li>Load count matrix in h5 format from GEO (10x)</li>
<li>Compressed files (zip, tar.gz, tar) in downloaded supplemental files (GEO, 10x)</li>
<li>Sample name is after the fixed name (GEO, 10x)</li>
</ul></li>
<li>Change the way of downloading supplemental files with index from GEO.</li>
<li>Extract metadata from supplemental files (GEO).</li>
<li>Added <code>ParseGEOProcessed</code> to download processed objects from GEO.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-7-2">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-7-2"></a></h3>
<ul><li>Removed <code>cellxgene.census</code> dependency to avoid possible installation issue on Windows.</li>
<li>Fix bug in <code>MoveFileRecursively</code>.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.1" id="gefetch2r-071">GEfetch2R 0.7.1<a class="anchor" aria-label="anchor" href="#gefetch2r-071"></a></h2>
<div class="section level3">
<h3 id="new-features-0-7-1">New features<a class="anchor" aria-label="anchor" href="#new-features-0-7-1"></a></h3>
<ul><li>Added <code>Seu2AD</code>/<code>SCE2AD</code>/<code>AD2Seu</code>/<code>AD2SCE</code> for benchmark (data IO).</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-7-1">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-7-1"></a></h3>
<ul><li>Added <code>timeout</code> in <code>ShowCBDatasets</code>.</li>
<li>Added <code>Gunzip</code> to deal with the removal of <code>gunzip</code> in package <code>GEOquery</code>.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.0" id="gefetch2r-070">GEfetch2R 0.7.0<a class="anchor" aria-label="anchor" href="#gefetch2r-070"></a></h2>
<div class="section level3">
<h3 id="new-features-0-7-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-7-0"></a></h3>
<ul><li>Added <code>DownloadFastq</code> to download FASTQ files directly from ENA.</li>
<li>Updated <code>DownloadSRA</code> to download sra files directly from ENA.</li>
<li>Updated <code>DownloadBam</code> to download bam files directly from ENA.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-7-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-7-0"></a></h3>
<ul><li>Updated <code>DownloadSRA</code>, <code>DownloadFastq</code>, <code>DownloadBam</code> to keep the same file structure.</li>
<li>Updated <code>DownloadFastq</code> to format the fastqs to 10x format.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.1" id="gefetch2r-061">GEfetch2R 0.6.1<a class="anchor" aria-label="anchor" href="#gefetch2r-061"></a></h2>
<ul><li>Fix bugs in <code>LoadRDS2Seurat</code>.</li>
<li>Give a message when there is no file with <code>file.ext</code> in <code>ParseHCA</code>, <code>ParseHCA</code>, <code>ExtractZenodoMeta</code>, <code>ParseZenodo</code>.</li>
<li>Return SeuratObject in <code>ParseHCA</code>.</li>
<li>Fix bug in <code>Load2Seurat</code>.</li>
<li>Support <code>pigz</code> in <code>Bam2Fastq</code>.</li>
<li>Return full dataframe in <code>ExtractRun</code>.</li>
<li>Fix bug in <code>ParseCELLxGENE</code>, <code>ParseHCA</code>.</li>
<li>Add <code>use.cores</code> in <code>ParseCELLxGENE</code>, <code>ParseHCA</code>, <code>ParseZenodo</code>.</li>
<li>Fix bug in <code>ExtractGEOMeta</code>.</li>
</ul><hr></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="gefetch2r-060">GEfetch2R 0.6.0<a class="anchor" aria-label="anchor" href="#gefetch2r-060"></a></h2>
<div class="section level3">
<h3 id="new-features-0-6-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-6-0"></a></h3>
<ul><li>Added <code>RunCellRanger</code> to run CellRanger on downloaded FASTQ files.</li>
<li>Added <code>RunSTAR</code> to run STAR on downloaded FASTQ files.</li>
<li>Added <code>Fastq2R</code> to pipe FASTQ files to <code>SeuratObject</code> and <code>DESeqDataSet</code>.</li>
<li>Updated <code>StatDBAttribute</code> to support <code>cellxgene.census</code>.</li>
<li>Updated <code>ParseCELLxGENE</code> to support <code>cellxgene.census</code> and return SeuratObject.</li>
<li>Updated <code>ParseCBDatasets</code> to support subset.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-6-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-6-0"></a></h3>
<ul><li>Return <code>DESeqDataSet</code> when <code>data.type</code> is bulk in <code>ParseGEO</code>.</li>
<li>Support summarising multiple attributes in <code>StatDBAttribute</code>.</li>
<li>Fix bugs in <code>mergeExperiments</code>.</li>
<li>Fix bugs in <code>ParseHCA</code> (no file with extension specified by <code>file.ext</code>).</li>
<li>Optimize <code>ParseHCA</code> to return metadata of downloaded files.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.1" id="gefetch2r-051">GEfetch2R 0.5.1<a class="anchor" aria-label="anchor" href="#gefetch2r-051"></a></h2>
<ul><li>Create output directory automatically.</li>
<li>CELLxGENE new <a href="https://api.cellxgene.cziscience.com/curation/ui/" class="external-link">API</a>.</li>
</ul><hr></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.0" id="gefetch2r-050">GEfetch2R 0.5.0<a class="anchor" aria-label="anchor" href="#gefetch2r-050"></a></h2>
<div class="section level3">
<h3 id="new-features-0-5-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-5-0"></a></h3>
<ul><li>Added <code>ExtractCELLxGENEMeta</code> to extract metadata of CELLxGENE datasets.</li>
<li>Added <code>ParseCELLxGENE</code> to download objects from CELLxGENE.</li>
<li>Added <code>ExtractHCAMeta</code> to extract metadata of Human Cell Atlas projects.</li>
<li>Added <code>ShowHCAProjects</code> to show all available Human Cell Atlas projects.</li>
<li>Added <code>ShowCELLxGENEDatasets</code> to show all available CELLxGENE datasets.</li>
<li>Added <code>StatDBAttribute</code>to stat database attributes.</li>
<li>Added <code>ParseHCA</code> to download objects from Human Cell Atlas.</li>
<li>docker image added.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-5-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-5-0"></a></h3>
<ul><li>Added filters to <code>ExtractCELLxGENEMeta</code>.</li>
<li>Added <code>data.type</code> in <code>ParseGEO</code> to return count matrix for bulk RNA-seq.</li>
<li>Added filters to <code>ExtractHCAMeta</code>.</li>
<li>Added <code>catalog</code> in <code>ShowHCAProjects</code>.</li>
<li>Fixed bug in <code>ParseCELLxGENE</code> and returned failed dataframe when downloading error.</li>
<li>Fixed bug in <code>ImportSeurat</code>.</li>
<li>Fixed bug in <code>ExportSeurat</code>.</li>
<li>Resolved installation.</li>
<li>Fixed bug in <code>ExtractZenodoMeta</code> (API changed).</li>
<li>Fixed bug in <code>ShowCBDatasets</code> (added <code>--no-check-certificate</code> when downloading json files).</li>
<li>Fixed bug in <code>ShowCELLxGENEDatasets</code>.</li>
<li>Supported GEO of 10x (separate files).</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.0" id="gefetch2r-040">GEfetch2R 0.4.0<a class="anchor" aria-label="anchor" href="#gefetch2r-040"></a></h2>
<div class="section level3">
<h3 id="new-features-0-4-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-4-0"></a></h3>
<ul><li>Added <code>ExportSeurat</code> to convert SeuratObject to other scRNA-seq formats.</li>
<li>Added <code>ImportSeurat</code> to convert other scRNA-seq formats to SeuratObject.</li>
<li>Added <code>SCEaAnnData</code> to perform data format conversion between SingleCellExperiment and AnnData.</li>
<li>Added <code>SCELoom</code> to perform data format conversion between SingleCellExperiment and loom.</li>
<li>Added <code>ExtractPanglaoDBComposition</code> to extract cell type composition of PanglaoDB datasets.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-4-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-4-0"></a></h3>
<ul><li>Added <code>conda.path</code> in <code>ExportSeurat</code> to specify conda enviroment used.</li>
<li>Fixed bug in <code>SCELoom</code> to keep gene names and cell ID.</li>
<li>
<code>Bam2Fastq</code> supported normal bam files (non-10x bam files).</li>
<li>Changed function names: <code>ShowPanglaoDBMeta</code> to <code>ExtractPanglaoDBMeta</code>, <code>PrepareZenodo</code> to <code>ExtractZenodoMeta</code>.</li>
<li>Supported downloadding normal (non-10x) bam file in <code>DownloadBam</code>.</li>
<li>
<code>ExtractZenodoMeta</code> supported a vector of Zenodo dois.</li>
<li>Added <code>fastq.type</code> in <code>SplitSRA</code> to deal with fastq files from 10x, other scRNA-seq protocols and bulk RNA-seq.</li>
<li>Added check for <code>split.cmd.paras</code> in <code>SplitSRA</code>.</li>
<li>Added <code>local.data</code> in <code>ExtractPanglaoDBMeta</code> and <code>ExtractPanglaoDBComposition</code> to use cached sample metadata and composition.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.0" id="gefetch2r-030">GEfetch2R 0.3.0<a class="anchor" aria-label="anchor" href="#gefetch2r-030"></a></h2>
<div class="section level3">
<h3 id="new-features-0-3-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-3-0"></a></h3>
<ul><li>Added <code>ExtractRun</code> to extract run number from GEO.</li>
<li>Added <code>DownloadSRA</code> to download SRA according to run number.</li>
<li>Added <code>SplitSRA</code> to split SRA to fastqs and format the fastqs to 10x format.</li>
<li>Added <code>DownloadBam</code> to download 10x bam files from GEO.</li>
<li>Added <code>Bam2Fastq</code> to convert bam to 10x formatted fastqs.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-3-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-3-0"></a></h3>
<ul><li>Fixed bug in <code>DownloadSRA</code> when downloading bam files.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="gefetch2r-020">GEfetch2R 0.2.0<a class="anchor" aria-label="anchor" href="#gefetch2r-020"></a></h2>
<div class="section level3">
<h3 id="new-features-0-2-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-2-0"></a></h3>
<ul><li>Provided APIs for GEO.</li>
<li>Added <code>Read10XOnline</code> to load cellranger output.</li>
<li>Added <code>ExtractCBComposition</code> to extract cell type composition of UCSC Cell Browser datasets.</li>
<li>
<code>ParseGEO</code> support creating Seurat object.</li>
<li>Added <code>ExtractGEOMeta</code> to support extract metadata from GEO.</li>
</ul><hr></div>
<div class="section level3">
<h3 id="minor-changes-0-2-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-0"></a></h3>
<ul><li>Supported lazy mode in <code>ShowCBDatasets</code> to save time for multi-runs.</li>
<li>Optimized the paras for <code>ExtractCBDatasets</code> and <code>ParseCBDatasets</code>.</li>
<li>Added <code>cell.num</code> filter to <code>ParseCBDatasets</code>, <code>ExtractCBDatasets</code> and <code>ShowPanglaoDBMeta</code>.</li>
<li>Extracted matrix, barcode and feature info from dataset.json (comprehensive) instead of desc.json.</li>
<li>Fixed bugs in <code>ParseCBDatasets</code> when dealing with cellranger output.</li>
<li>Added <code>timeout</code> in <code>ParseCBDatasets</code> to avoid possible timeout error.</li>
<li>Changed <code>time.out</code> in GEO related functions to <code>timeout</code>.</li>
<li>Simplified the output of <code>ParseGEO</code>.</li>
<li>Fixed bug in <code>ExtractGEOMeta</code>.</li>
</ul><hr></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="gefetch2r-010">GEfetch2R 0.1.0<a class="anchor" aria-label="anchor" href="#gefetch2r-010"></a></h2>
<ul><li>Provided APIs for Zenodo, PanglaoDB and UCSC Cell Browser.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

