<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="The goal of GEfetch2R is to access and format single-cell RNA-seq datasets. It can be used to download single-cell RNA-seq
    datasets from widely used public resources, including GEO &lt;https://www.ncbi.nlm.nih.gov/geo/&gt;, Zenodo &lt;https://zenodo.org/&gt;, 
    CELLxGENE &lt;https://cellxgene.cziscience.com/&gt;, Human Cell Atlas &lt;https://www.humancellatlas.org/&gt;, PanglaoDB &lt;https://panglaodb.se/index.html&gt; 
    and UCSC Cell Browser &lt;https://cells.ucsc.edu/&gt;. And, it can also be used to perform object conversion between SeuratObject &lt;https://satijalab.org/seurat/&gt;, 
    loom &lt;http://loompy.org/&gt;, h5ad &lt;https://scanpy.readthedocs.io/en/stable/&gt;, SingleCellExperiment &lt;https://bioconductor.org/packages/release/bioc/html/scran.html&gt;, 
    CellDataSet &lt;http://cole-trapnell-lab.github.io/monocle-release/&gt; and cell_data_set &lt;https://cole-trapnell-lab.github.io/monocle3/&gt;.">
<title>Access and Format Single-Cell RNA-Seq Datasets from Public Resources  • GEfetch2R</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Access and Format Single-Cell RNA-Seq Datasets from Public Resources ">
<meta property="og:description" content="The goal of GEfetch2R is to access and format single-cell RNA-seq datasets. It can be used to download single-cell RNA-seq
    datasets from widely used public resources, including GEO &lt;https://www.ncbi.nlm.nih.gov/geo/&gt;, Zenodo &lt;https://zenodo.org/&gt;, 
    CELLxGENE &lt;https://cellxgene.cziscience.com/&gt;, Human Cell Atlas &lt;https://www.humancellatlas.org/&gt;, PanglaoDB &lt;https://panglaodb.se/index.html&gt; 
    and UCSC Cell Browser &lt;https://cells.ucsc.edu/&gt;. And, it can also be used to perform object conversion between SeuratObject &lt;https://satijalab.org/seurat/&gt;, 
    loom &lt;http://loompy.org/&gt;, h5ad &lt;https://scanpy.readthedocs.io/en/stable/&gt;, SingleCellExperiment &lt;https://bioconductor.org/packages/release/bioc/html/scran.html&gt;, 
    CellDataSet &lt;http://cole-trapnell-lab.github.io/monocle-release/&gt; and cell_data_set &lt;https://cole-trapnell-lab.github.io/monocle3/&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a>
    <a class="dropdown-item" href="articles/DownloadMatrices.html">Download count matrix and annotation</a>
    <a class="dropdown-item" href="articles/DownloadObjects.html">Download processed object</a>
    <a class="dropdown-item" href="articles/ObjectConversion.html">Format conversion and benchmark</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="gefetch2r---access-single-cellbulk-rna-seq-data-from-public-repositories-and-benchmark-the-subsequent-format-conversion-tools">GEfetch2R - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools<a class="anchor" aria-label="anchor" href="#gefetch2r---access-single-cellbulk-rna-seq-data-from-public-repositories-and-benchmark-the-subsequent-format-conversion-tools"></a>
</h1></div>
<p><img src="reference/figures/GEfetch2R.png" align="right" width="200"></p>
<p><a href="https://cran.r-project.org/web/licenses/GPLv3" class="external-link"><img src="https://img.shields.io/badge/license-GPLv3-blue.svg" alt="License: GPLv3"></a> <a href="https://github.com/showteeth/GEfetch2R" class="external-link"><img src="https://img.shields.io/github/languages/code-size/showteeth/GEfetch2R.svg" alt="CODE_SIZE"></a></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> is designed to accelerate users’ downloading and preparation of single-cell/bulk RNA-seq datasets from public resources. It can be used to:</p>
<p>For <strong>bulk RNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from <code>SRA/ENA</code> with <strong>GEO accessions</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files from <code>ENA</code>
</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to <code>fastq</code> files</li>
<li><strong>Convert <code>bam</code> to <code>fastq</code> files</strong></li>
<li><strong>Mapping with <code>STAR</code> and load the output to <code>DESeq2</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with <strong>GEO accessions</strong>:
<ul>
<li><strong>Generate count matrix from supplementary files</strong></li>
<li><strong>Load the count matrix to <code>DESeq2</code></strong></li>
</ul>
</li>
</ul>
<p>For <strong>scRNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from <code>SRA/ENA</code> with <strong>GEO accessions</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files from <code>ENA</code>
</li>
<li>
<strong>Foramt downloaded <code>fastq</code> files</strong> to standard style that can be identified by 10x softwares (e.g. <code>CellRanger</code>).</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to <code>fastq</code> files</li>
<li><strong>Download original 10x generated <code>bam</code> files (with custom tags)</strong></li>
<li><strong>Convert <code>bam</code> to <code>fastq</code> files (<code>samtools/bamtofastq_linux</code>)</strong></li>
<li><strong>Mapping with <code>STAR/CellRanger</code> and load the output to <code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong>:
<ul>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with <strong>GEO accessions</strong>
</li>
<li>Download <strong>count matrix</strong> and <strong>annotation (e.g. cell type)</strong> information from <code>PanglanDB</code> and <code>UCSC Cell Browser</code> with <strong>key words (filter criteria)</strong>
</li>
<li>
<strong>Extract subset with annotation and gene (<code>PanglanDB</code> and <code>UCSC Cell Browser</code>)</strong><br>
</li>
<li><strong>Load the count matrix and annotation to <code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>processed object</strong>:
<ul>
<li>Download <strong>processed object</strong> from <code>Zenodo</code> with <strong>DOI</strong>
</li>
<li>Download <strong>processed object</strong> from <code>CELLxGENE</code> and <code>Human Cell Atlas</code> with <strong>key words (filter criteria)</strong>
</li>
<li>
<strong>Parallel</strong> download</li>
<li><strong>Extract subset with annotation and gene (<code>CELLxGENE</code>)</strong></li>
<li><strong>Load the processed object (<code>rds</code>) to <code>Seurat</code></strong></li>
</ul>
</li>
<li>
<strong>Format conversion and benchmark related tools</strong>:
<ul>
<li>
<strong>Format conversion between widely used single cell objects</strong> (<code>SeuratObject</code>, <code>AnnData</code>, <code>SingleCellExperiment</code>, <code>CellDataSet/cell_data_set</code> and <code>loom</code>)</li>
<li>
<strong>Benchmark the tools (more than two tools available for the conversion)</strong>:
<ul>
<li>
<code>SeuratObject</code> to <code>AnnData</code>: <code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>
</li>
<li>
<code>SingleCellExperiment</code> to <code>AnnData</code>: <code>sceasy</code>, <code>scDIOR</code>, <code>zellkonverter</code>
</li>
<li>
<code>AnnData</code> to <code>SeuratObject</code>: <code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>, <code>schard</code>
</li>
<li>
<code>AnnData</code> to <code>SingleCellExperiment</code>: <code>schard</code>, <code>scDIOR</code>, <code>zellkonverter</code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="framework">Framework<a class="anchor" aria-label="anchor" href="#framework"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_overview.png" title="GEfetch2R_framework" alt="GEfetch2R_framework"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install <code>GEfetch2R</code>, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install via CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"GEfetch2R"</span><span class="op">)</span> <span class="co"># coming soon</span></span>
<span></span>
<span><span class="co"># you can also install the development version from GitHub</span></span>
<span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"showteeth/GEfetch2R"</span><span class="op">)</span></span></code></pre></div>
<p>There are some conditionally used packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools") #In case you have not installed it.</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"alexvpickering/GEOfastq"</span><span class="op">)</span> <span class="co"># download fastq</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'tiledbsoma'</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'https://tiledb-inc.r-universe.dev'</span>, <span class="st">'https://cloud.r-project.org'</span><span class="op">)</span><span class="op">)</span> <span class="co"># download from CELLxGENE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'cellxgene.census'</span>, repos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'https://chanzuckerberg.r-universe.dev'</span>, <span class="st">'https://cloud.r-project.org'</span><span class="op">)</span><span class="op">)</span> <span class="co"># download from CELLxGENE</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cellgeni/sceasy"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mojaveazure/seurat-disk"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/seurat-wrappers"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'theislab/zellkonverter@7b118653a471330b3734dcfee60c3537352ecb8d'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'cellgeni/schard'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'JiekaiLab/dior'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span>  <span class="co"># format conversion</span></span></code></pre></div>
<p><strong>For possible issues about installation, please refer <code>INSTALL.md</code>.</strong></p>
<p>For data structures conversion and downloading fastq/bam files, <code>GEfetch2R</code> requires additional tools, you can install with:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install additional packages for format conversion</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install diopy</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda loompy anndata </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install anndata loompy</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># install additional packages for downloading fastq/bam files</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda <span class="st">'parallel-fastq-dump'</span> <span class="st">'sra-tools&gt;=3.0.0'</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># install bamtofastq, the following installs linux version</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--quiet</span> https://github.com/10XGenomics/bamtofastq/releases/download/v1.4.1/bamtofastq_linux <span class="kw">&amp;&amp;</span> <span class="fu">chmod</span> +x bamtofastq_linux</span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a>
</h2>
<p>We also provide a <a href="https://hub.docker.com/repository/docker/soyabean/GEfetch2R" class="external-link">docker image</a> to use:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pull the image</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull soyabean/GEfetch2R:1.6</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># run the image</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-p</span> 8888:8787 <span class="at">-e</span> PASSWORD=passwd <span class="at">-e</span> ROOT=TRUE <span class="at">-it</span> soyabean/GEfetch2R:1.6</span></code></pre></div>
<p><strong>Notes</strong>:</p>
<ul>
<li>After running the above codes, open browser and enter <code>http://localhost:8888/</code>, the user name is <code>rstudio</code>, the password is <code>passwd</code> (set by <code>-e PASSWORD=passwd</code>)</li>
<li>If port <code>8888</code> is in use, change <code>-p 8888:8787</code>
</li>
<li>The <code>conda.path</code> in <code>ExportSeurat</code> and <code>ImportSeurat</code> can be set <code>/opt/conda</code>.</li>
<li>The <strong>sra-tools</strong> can be found in <code>/opt/sratoolkit.3.0.6-ubuntu64/bin</code>.</li>
<li>The <code>parallel-fastq-dump</code> path: <code>/opt/conda/bin/parallel-fastq-dump</code>.</li>
<li>The <code>bamtofastq_linux</code> path: <code>/opt/bamtofastq_linux</code>.</li>
<li>The <code>samtools</code> path: <code>/opt/conda/bin/samtools</code>.</li>
<li>The <code>STAR</code> and <code>Cell Ranger</code> is not available in the image because customized reference genome is required.</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="vignette">Vignette<a class="anchor" aria-label="anchor" href="#vignette"></a>
</h2>
<p>Detailed usage is available in <a href="https://showteeth.github.io/GEfetch2R/">website</a>:</p>
<ul>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadRaw.html" class="external-link">Download raw data</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadMatrices.html" class="external-link">Download count matrix and annotaion</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadObjects.html" class="external-link">Download processed data</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/ObjectConversion.html" class="external-link">Format conversion and benchmark</a></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="comparison-with-similar-tools">Comparison with similar tools<a class="anchor" aria-label="anchor" href="#comparison-with-similar-tools"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_compare.png" title="GEfetch2R_comparison" alt="GEfetch2R_comparison"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="download-raw-data-srafastqbam">Download raw data (sra/fastq/bam)<a class="anchor" aria-label="anchor" href="#download-raw-data-srafastqbam"></a>
</h3>
<blockquote>
<p><code>GEfetch2R</code> supports downloading raw data (<code>sra/fastq/bam</code>) from <code>SRA</code> and <code>ENA</code> with <strong>GEO accessions</strong>. In general, downloading raw data from ENA is much faster than SRA, because of <code>ascp</code> and <strong>parallel</strong> support.</p>
</blockquote>
<p>Since the downloading process is time-consuming, we provide the commands used to illustrate the usage.</p>
<div class="section level4">
<h4 id="extract-all-samples-runs">Extract all samples (runs)<a class="anchor" aria-label="anchor" href="#extract-all-samples-runs"></a>
</h4>
<p>For <code>fastq</code> files stored in <code>SRA/ENA</code>, <code>GEfetch2R</code> can extract sample information and run number with <strong>GEO accessions</strong> or users can also <strong>provide a dataframe contains the run number of interested samples</strong>.</p>
<p>Extract all samples under <code>GSE130636</code> and the platform is <code>GPL20301</code> (use <code>platform = NULL</code> for all platforms):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE130636"</span>, platform <span class="op">=</span> <span class="st">"GPL20301"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-sra">Download sra<a class="anchor" aria-label="anchor" href="#download-sra"></a>
</h4>
<p>With the dataframe contains gsm and run number, <code>GEfetch2R</code> will use <code>prefetch</code> to download <code>sra</code> files from <code>SRA</code> or using <code>ascp/download.file</code> to download <code>sra</code> files from <code>ENA</code>. The returned value is a dataframe contains failed runs. If not <code>NULL</code>, users can re-run <code>DownloadSRA</code> by setting <code>gsm.df</code> to the returned value.</p>
<p>Download from <code>SRA</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="va">GSE130636.runs</span><span class="op">[</span><span class="va">GSE130636.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR9004346"</span>, <span class="st">"SRR9004351"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download, you may need to set prefetch.path</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>Download from <code>ENA</code> (<strong>parallel</strong>):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># download from ENA using download.file</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, download.method <span class="op">=</span> <span class="st">"download.file"</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">3600</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download from ENA using ascp</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, download.method <span class="op">=</span> <span class="st">"ascp"</span>,</span>
<span>  ascp.path <span class="op">=</span> <span class="st">"/path/to/ascp"</span>, max.rate <span class="op">=</span> <span class="st">"300m"</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be: <code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level4">
<h4 id="download-fastq">Download fastq<a class="anchor" aria-label="anchor" href="#download-fastq"></a>
</h4>
<div class="section level5">
<h5 id="split-sra-to-generate-fastq">Split sra to generate fastq<a class="anchor" aria-label="anchor" href="#split-sra-to-generate-fastq"></a>
</h5>
<p>After obtaining the <code>sra</code> files, <code>GEfetch2R</code> provides function <code>SplitSRA</code> to split <code>sra</code> files to fastq files using <code>parallel-fastq-dump</code> (<strong>parallel, fastest and gzip output</strong>), <code>fasterq-dump</code> (<strong>parallel, fast but unzipped output</strong>) and <code>fastq-dump</code> (<strong>slowest and gzip output</strong>).</p>
<p>For fastqs generated with 10x Genomics, <code>SplitSRA</code> can identify read1, read2 and index files and format the read1 and read2 to 10x required format (<code>sample1_S1_L001_R1_001.fastq.gz</code> and <code>sample1_S1_L001_R2_001.fastq.gz</code>). In detail, the file with read length 26 or 28 is considered as read1, the files with read length 8 or 10 are considered as index files and the remain file is considered as read2. The read length rules is from <a href="https://www.10xgenomics.com/cn/support/single-cell-gene-expression/documentation/steps/sequencing/sequencing-requirements-for-single-cell-3" class="external-link">Sequencing Requirements for Single Cell 3’</a> and <a href="https://www.10xgenomics.com/cn/support/single-cell-immune-profiling/documentation/steps/sequencing/sequencing-requirements-for-single-cell-v-d-j" class="external-link">Sequencing Requirements for Single Cell V(D)J</a>.</p>
<p>The returned value is a vector of failed <code>sra</code> files. If not <code>NULL</code>, users can re-run <code>SplitSRA</code> by setting <code>sra.path</code> to the returned value.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parallel-fastq-dump requires sratools.path</span></span>
<span><span class="co"># you may need to set split.cmd.path and sratools.path</span></span>
<span><span class="va">sra.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE130636.split</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SplitSRA.html">SplitSRA</a></span><span class="op">(</span></span>
<span>  sra.folder <span class="op">=</span> <span class="va">sra.folder</span>,</span>
<span>  fastq.type <span class="op">=</span> <span class="st">"10x"</span>, split.cmd.threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-fastq-directly-from-ena">Download fastq directly from ENA<a class="anchor" aria-label="anchor" href="#download-fastq-directly-from-ena"></a>
</h5>
<p>Alternatively, <code>GEfetch2R</code> provides function <code>DownloadFastq</code> to download <code>fastq</code> files directly from <code>ENA</code> (<strong>parallel, faster than the above method</strong>). The returned value is a dataframe contains failed runs. If not <code>NULL</code>, users can re-run <code>DownloadFastq</code> by setting <code>gsm.df</code> to the returned value.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use download.file</span></span>
<span><span class="va">download.file.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">gsm.df</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"download.file"</span>, </span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># use ascp</span></span>
<span><span class="va">ascp.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">gsm.df</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"ascp"</span>, ascp.path <span class="op">=</span> <span class="st">"/path/to/ascp"</span>, </span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="download-bam">Download bam<a class="anchor" aria-label="anchor" href="#download-bam"></a>
</h4>
<div class="section level5">
<h5 id="extract-all-samples-runs-1">Extract all samples (runs)<a class="anchor" aria-label="anchor" href="#extract-all-samples-runs-1"></a>
</h5>
<p><code>GEfetch2R</code> can extract sample information and run number with <strong>GEO accessions</strong> or users can also <strong>provide a dataframe contains the run number of interested samples</strong>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE138266"</span>, platform <span class="op">=</span> <span class="st">"GPL18573"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-bam-from-sra">Download bam from SRA<a class="anchor" aria-label="anchor" href="#download-bam-from-sra"></a>
</h5>
<p>With the dataframe contains gsm and run number, <code>GEfetch2R</code> provides <code>DownloadBam</code> to download <code>bam</code> files using <code>prefetch</code>. It supports 10x generated <code>bam</code> files and normal <code>bam</code> files.</p>
<ul>
<li>
<strong>10x generated bam</strong>: While <code>bam</code> files generated from 10x softwares (e.g. CellRanger) contain custom tags which are not kept when using default parameters of <code>prefetch</code>, <code>GEfetch2R</code> adds <code>--type TenX</code> to make sure the downloaded <code>bam</code> files contain these tags.</li>
<li>
<strong>normal bam</strong>: For normal bam files, <code>DownloadBam</code> will download <code>sra</code> files first and then convert <code>sra</code> files to <code>bam</code> files with <code>sam-dump</code>. After testing the efficiency of <code>prefetch</code> + <code>sam-dump</code> and <code>sam-dump</code>, the former is much faster than the latter (52G <code>sra</code> and 72G <code>bam</code> files):</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # use prefetch to download sra file</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># prefetch -X 60G SRR1976036</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    117m26.334s</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    16m42.062s</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 3m28.295s</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # use sam-dump to convert sra to bam</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036.sra | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    536m2.721s</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    749m41.421s</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 20m49.069s</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="co"># use sam-dump to download bam directly</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036 | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co"># # more than 36hrs only get ~3G bam files, too slow</span></span></code></pre></div>
<p>The returned value is a dataframe containing failed runs (either failed to download <code>sra</code> files or failed to convert to <code>bam</code> files for normal <code>bam</code>; failed to download <code>bam</code> files for 10x generated <code>bam</code>). If not <code>NULL</code>, users can re-run <code>DownloadBam</code> by setting <code>gsm.df</code> to the returned value. The following is an example to download 10x generated <code>bam</code> file:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="va">GSE138266.runs</span><span class="op">[</span><span class="va">GSE138266.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR10211566"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download, you may need to set prefetch.path</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE138266.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be: <code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level5">
<h5 id="download-bam-from-ena">Download bam from ENA<a class="anchor" aria-label="anchor" href="#download-bam-from-ena"></a>
</h5>
<p>The returned value is a dataframe containing failed runs. If not <code>NULL</code>, users can re-run <code>DownloadBam</code> by setting <code>gsm.df</code> to the returned value. The following is an example to download 10x generated <code>bam</code> file from <code>ENA</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download.file</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>, download.method <span class="op">=</span> <span class="st">"download.file"</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">3600</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># ascp</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>, download.method <span class="op">=</span> <span class="st">"ascp"</span>,</span>
<span>  ascp.path <span class="op">=</span> <span class="st">"/path/to/ascp"</span>, max.rate <span class="op">=</span> <span class="st">"300m"</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="convert-bam-to-fastq">Convert bam to fastq<a class="anchor" aria-label="anchor" href="#convert-bam-to-fastq"></a>
</h5>
<p>With downloaded <code>bam</code> files, <code>GEfetch2R</code> provides function <code>Bam2Fastq</code> to convert <code>bam</code> files to <code>fastq</code> files. For <code>bam</code> files generated from 10x softwares, <code>Bam2Fastq</code> utilizes <code>bamtofastq</code> tool developed by 10x Genomics, otherwise, <code>samtools</code> is utilized.</p>
<p>The returned value is a vector of <code>bam</code> files failed to convert to <code>fastq</code> files. If not <code>NULL</code>, users can re-run <code>Bam2Fastq</code> by setting <code>bam.path</code> to the returned value.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># you may need to set bamtofastq.path and bamtofastq.paras</span></span>
<span><span class="va">GSE138266.convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Bam2Fastq.html">Bam2Fastq</a></span><span class="op">(</span></span>
<span>  bam.folder <span class="op">=</span> <span class="va">bam.folder</span>,</span>
<span>  bamtofastq.path <span class="op">=</span> <span class="st">"/path/to/bamtofastq"</span>, </span>
<span>  bamtofastq.paras <span class="op">=</span> <span class="st">"--nthreads 4"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level3">
<h3 id="download-count-matrix">Download count matrix<a class="anchor" aria-label="anchor" href="#download-count-matrix"></a>
</h3>
<p><code>GEfetch2R</code> provides functions for users to download <strong>count matrices</strong> and <strong>annotations</strong> (e.g. cell type annotation and composition) from <code>GEO</code> and some single-cell databases (e.g. <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> and <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>). <code>GEfetch2R</code> also supports loading the downloaded data to <code>Seurat</code>.</p>
<p>Until now, the public resources supported and the returned values:</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="26%">
<col width="23%">
<col width="35%">
</colgroup>
<thead><tr class="header">
<th>Resources</th>
<th>URL</th>
<th>Download Type</th>
<th>returned values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>GEO</td>
<td><a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a></td>
<td>count matrix</td>
<td>SeuratObject or count matrix for bulk RNA-seq</td>
</tr>
<tr class="even">
<td>PanglaoDB</td>
<td><a href="https://panglaodb.se/index.html" class="external-link uri">https://panglaodb.se/index.html</a></td>
<td>count matrix and annotations</td>
<td>SeuratObject</td>
</tr>
<tr class="odd">
<td>UCSC Cell Browser</td>
<td><a href="https://cells.ucsc.edu/" class="external-link uri">https://cells.ucsc.edu/</a></td>
<td>count matrix and annotations</td>
<td>SeuratObject</td>
</tr>
</tbody>
</table>
<hr>
<div class="section level4">
<h4 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h4>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is an international public repository that archives and freely distributes microarray, next-generation sequencing, and other forms of high-throughput functional genomics data submitted by the research community.</a> It provides a very convenient way for users to explore and select interested scRNA-seq datasets.</p>
<div class="section level5">
<h5 id="extract-metadata">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractGEOMeta</code> to extract sample metadata, including sample title, source name/tissue, description, cell type, treatment, paper title, paper abstract, organism, protocol, data processing methods, et al.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract metadata of specified platform</span></span>
<span><span class="va">GSE200257.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="st">"GPL24676"</span><span class="op">)</span></span>
<span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata of all platforms</span></span>
<span><span class="va">GSE94820.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-matrix-and-load-to-seurat">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat"></a>
</h5>
<p>After manually check the extracted metadata, users can <strong>download count matrix</strong> and <strong>load the count matrix</strong> to Seurat with <code>ParseGEO</code>.</p>
<p>For count matrix, <code>ParseGEO</code> supports downloading the matrix from <strong>supplementary files</strong> and extracting from <code>ExpressionSet</code>, users can control the source by specifying <code>down.supp</code> or detecting automatically (<code>ParseGEO</code> will extract the count matrix from <code>ExpressionSet</code> first, if the count matrix is NULL or contains non-integer values, <code>ParseGEO</code> will download supplementary files). While the supplementary files have two main types: single count matrix file containing all cells and CellRanger-style outputs (barcode, matrix, feature/gene), users are required to choose the type of supplementary files with <code>supp.type</code>.</p>
<p>With the count matrix, <code>ParseGEO</code> will load the matrix to Seurat automatically. If multiple samples available, users can choose to merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for cellranger output</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE200257.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># for count matrix, no need to specify out.folder, download count matrix to tmp folder</span></span>
<span><span class="va">GSE94820.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span></code></pre></div>
<p><strong>For bulk RNA-seq</strong>, set <code>data.type = "bulk"</code> in <code>ParseGEO</code>, this will return count matrix.</p>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="panglaodb">PanglaoDB<a class="anchor" aria-label="anchor" href="#panglaodb"></a>
</h4>
<p><a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> is a database which contains scRNA-seq datasets from mouse and human. Up to now, it contains <strong>5,586,348 cells</strong> from <strong>1368 datasets (1063 from Mus musculus and 305 from Homo sapiens)</strong>. It has well organized metadata for every dataset, including tissue, protocol, species, number of cells and cell type annotation (computationally identified). Daniel Osorio has developed <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a> to access <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> data, the functions of <code>GEfetch2R</code> here are based on <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a>.</p>
<p>Since <a href="https://panglaodb.se/about.html" class="external-link">PanglaoDB</a> is no longer maintained, <code>GEfetch2R</code> has cached all metadata and cell type composition and use these cached data by default to accelerate, users can access the cached data with <code>PanglaoDBMeta</code> (all metadata) and <code>PanglaoDBComposition</code> (all cell type composition).</p>
<div class="section level5">
<h5 id="summarise-attributes">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to summary attributes of <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use cached metadata</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="extract-metadata-1">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-1"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractPanglaoDBMeta</code> to select interested datasets with specified <strong>species</strong>, <strong>protocol</strong>, <strong>tissue</strong> and <strong>cell number</strong> (The available values of these attributes can be obtained with <code>StatDBAttribute</code>). User can also choose to whether to add cell type annotation to every dataset with <code>show.cell.type</code>.</p>
<p><code>GEfetch2R</code> uses cached metadata and cell type composition by default, users can change this by setting <code>local.data = FALSE</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>,</span>
<span>  show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="extract-cell-type-composition">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractPanglaoDBComposition</code> to extract cell type annotation and composition (use cached data by default to accelerate, users can change this by setting <code>local.data = FALSE</code>).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-matrix-and-load-to-seurat-1">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat-1"></a>
</h5>
<p>After manually check the extracted metadata, <code>GEfetch2R</code> provides <code>ParsePanglaoDB</code> to <strong>download count matrix</strong> and <strong>load the count matrix</strong> to Seurat. With available cell type annotation, uses can filter datasets without specified cell type with <code>cell.type</code>. Users can also include/exclude cells expressing specified genes with <code>include.gene</code>/<code>exclude.gene</code>.</p>
<p>With the count matrix, <code>ParsePanglaoDB</code> will load the matrix to Seurat automatically. If multiple datasets available, users can choose to merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># small test</span></span>
<span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="ucsc-cell-browser">UCSC Cell Browser<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser"></a>
</h4>
<p>The <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a> is a web-based tool that allows scientists to interactively visualize scRNA-seq datasets. It contains <strong>1040 single cell datasets</strong> from <strong>17 different species</strong>. And, it is <strong>organized with the hierarchical structure</strong>, which can help users quickly locate the datasets they are interested in.</p>
<div class="section level5">
<h5 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ShowCBDatasets</code> to show all available datasets. Due to the large number of datasets, <code>ShowCBDatasets</code> enables users to perform <em>lazy load</em> of dataset json files instead of downloading the json files online (time-consuming!!!). This <em>lazy load</em> requires users to provide <code>json.folder</code> to save json files and set <code>lazy = TRUE</code> (for the first time of run, <code>ShowCBDatasets</code> will download current json files to <code>json.folder</code>, for next time of run, with <code>lazy = TRUE</code>, <code>ShowCBDatasets</code> will load the downloaded json files from <code>json.folder</code>.). And, <code>ShowCBDatasets</code> supports updating the local datasets with <code>update = TRUE</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">json.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># first time run, the json files are stored under json.folder</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = json.folder, update = TRUE)</span></span>
<span></span>
<span><span class="co"># second time run, load the downloaded json files</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="va">json.folder</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># always read online</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)</span></span></code></pre></div>
<p>The number of datasets and all available species:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the number of datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># available species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\|parent"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">ucsc.cb.samples</span><span class="op">$</span><span class="va">organisms</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="summarise-attributes-1">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes-1"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to summary attributes of <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"UCSC"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="extract-metadata-2">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-2"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractCBDatasets</code> to filter metadata with <strong>collection</strong>, <strong>sub-collection</strong>, <strong>organ</strong>, <strong>disease status</strong>, <strong>organism</strong>, <strong>project</strong> and <strong>cell number</strong> (The available values of these attributes can be obtained with <code>StatDBAttribute</code> except <strong>cell number</strong>). All attributes except cell number support fuzzy match with <code>fuzzy.match</code>, this is useful when selecting datasets.</p>
<p><code>{r cb_extract, eval=FALSE} hbb.sample.df &lt;- ExtractCBDatasets(all.samples.df = ucsc.cb.samples, organ = c("brain", "blood"), organism = "Human (H. sapiens)", cell.num = c(1000, 2000))</code></p>
<hr>
</div>
<div class="section level5">
<h5 id="extract-cell-type-composition-1">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition-1"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractCBComposition</code> to extract cell type annotation and composition.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span>json.folder <span class="op">=</span> <span class="va">json.folder</span>, sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="load-the-online-datasets-to-seurat">Load the online datasets to Seurat<a class="anchor" aria-label="anchor" href="#load-the-online-datasets-to-seurat"></a>
</h5>
<p>After manually check the extracted metadata, <code>GEfetch2R</code> provides <code>ParseCBDatasets</code> to <strong>load the online count matrix</strong> to Seurat. All the attributes available in <code>ExtractCBDatasets</code> are also same here. Please note that the loading process provided by <code>ParseCBDatasets</code> will load the online count matrix instead of downloading it to local. If multiple datasets available, users can choose to merge the SeuratObject with <code>merge</code>.</p>
<p><code>ParseCBDatasets</code> supports <strong>extracting subset with metadata and gene</strong>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parse the whole datasets</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span>
<span><span class="co"># subset metadata and gene</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span>, obs.value.filter <span class="op">=</span> <span class="st">"Cell.Type == 'MP' &amp; Phase == 'G2M'"</span>,</span>
<span>                                  include.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PAX7"</span>, <span class="st">"MYF5"</span>, <span class="st">"C1QTNF3"</span>, <span class="st">"MYOD1"</span>, <span class="st">"MYOG"</span>, <span class="st">"RASSF4"</span>, <span class="st">"MYH3"</span>, <span class="st">"MYL4"</span>,</span>
<span>                                                    <span class="st">"TNNT3"</span>, <span class="st">"PDGFRA"</span>, <span class="st">"OGN"</span>, <span class="st">"COL3A1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level3">
<h3 id="download-processed-object">Download processed object<a class="anchor" aria-label="anchor" href="#download-processed-object"></a>
</h3>
<p><code>GEfetch2R</code> provides functions for users to download processed single-cell RNA-seq data from <a href="https://zenodo.org/" class="external-link">Zenodo</a>, <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> and <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>, including <code>RDS</code>, <code>RData</code>, <code>h5ad</code>, <code>h5</code>, <code>loom</code> objects.</p>
<p>Until now, the public resources supported and the returned values:</p>
<table class="table">
<colgroup>
<col width="15%">
<col width="29%">
<col width="33%">
<col width="21%">
</colgroup>
<thead><tr class="header">
<th>Resources</th>
<th>URL</th>
<th>Download Type</th>
<th>returned values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Zenodo</td>
<td><a href="https://zenodo.org/" class="external-link uri">https://zenodo.org/</a></td>
<td>count matrix, rds, rdata, h5ad, et al.</td>
<td>SeuratObject(rds) or failed datasets</td>
</tr>
<tr class="even">
<td>CELLxGENE</td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link uri">https://cellxgene.cziscience.com/</a></td>
<td>rds, h5ad</td>
<td>SeuratObject(rds) or failed datasets</td>
</tr>
<tr class="odd">
<td>Human Cell Atlas</td>
<td><a href="https://www.humancellatlas.org/" class="external-link uri">https://www.humancellatlas.org/</a></td>
<td>rds, rdata, h5, h5ad, loom</td>
<td>SeuratObject(rds) or failed projects</td>
</tr>
</tbody>
</table>
<hr>
<div class="section level4">
<h4 id="zenodo">Zenodo<a class="anchor" aria-label="anchor" href="#zenodo"></a>
</h4>
<p><a href="https://zenodo.org/" class="external-link">Zenodo</a> contains various types of processed objects, such as <code>SeuratObject</code> which has been clustered and annotated, <code>AnnData</code> which contains processed results generated by <code>scanpy</code>.</p>
<div class="section level5">
<h5 id="extract-metadata-3">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-3"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractZenodoMeta</code> to extract dataset metadata, including dataset title, description, available files and corresponding md5. Please note that when the dataset is restricted access, the returned dataframe will be empty.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single doi</span></span>
<span><span class="va">zebrafish.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="st">"10.5281/zenodo.7243603"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># vector dois</span></span>
<span><span class="va">multi.dois</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-object-and-load-to-r">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r"></a>
</h5>
<p>After manually check the extracted metadata, users can <strong>download the specified objects</strong> with <code>ParseZenodo</code>. The downloaded objects are controlled by <code>file.ext</code> and <strong>the provided object formats should be in lower case (e.g. rds/rdata/h5ad).</strong></p>
<p>The returned value is a dataframe containing failed objects or a <code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>). If dataframe, users can re-run <code>ParseZenodo</code> by setting <code>doi.df</code> to the returned value.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># download objects</span></span>
<span><span class="va">multi.dois.parse</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span>,</span>
<span>                                file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span><span class="op">)</span>,</span>
<span>                                out.folder <span class="op">=</span> <span class="va">out.folder</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># return SeuratObject</span></span>
<span><span class="va">sinle.doi.parse.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="st">"10.5281/zenodo.8011282"</span>,</span>
<span>                                   file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                   out.folder <span class="op">=</span> <span class="va">out.folder</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h4>
<p>The <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> is a web server contains <strong>910</strong> single-cell datasets, users can explore, download and upload own datasets. The downloaded datasets provided by <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> have two formats: <code>h5ad (AnnData v0.8)</code> and <code>rds (Seurat v4)</code>.</p>
<p><a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> provides an R package (<a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">cellxgene.census</a>) to access the data, but sometimes it’s not timely updated. <code>GEfetch2R</code> also supports users to access <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> via <a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">cellxgene.census</a> (<code>use.census = TRUE</code>).</p>
<div class="section level5">
<h5 id="show-available-datasets-1">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets-1"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ShowCELLxGENEDatasets</code> to extract dataset metadata, including dataset title, description, contact, organism, ethnicity, sex, tissue, disease, assay, suspension type, cell type, et al.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="summarise-attributes-2">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes-2"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to summary attributes of <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span>, <span class="st">"disease"</span><span class="op">)</span>, </span>
<span>                database <span class="op">=</span> <span class="st">"CELLxGENE"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># use cellxgene.census</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"disease"</span>, <span class="st">"tissue"</span>, <span class="st">"cell_type"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"CELLxGENE"</span>, </span>
<span>                use.census <span class="op">=</span> <span class="cn">TRUE</span>, organism <span class="op">=</span> <span class="st">"homo_sapiens"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="extract-metadata-4">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-4"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractCELLxGENEMeta</code> to filter dataset metadata, the available values of attributes can be obtained with <code>StatDBAttribute</code> except <strong>cell number</strong>:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">human.10x.cellxgene.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCELLxGENEMeta.html">ExtractCELLxGENEMeta</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># subset</span></span>
<span><span class="va">cellxgene.down.meta</span> <span class="op">=</span> <span class="va">human.10x.cellxgene.meta</span><span class="op">[</span><span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">==</span> <span class="st">"oligodendrocyte"</span> <span class="op">&amp;</span></span>
<span>                                                 <span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"entorhinal cortex"</span>, <span class="op">]</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-object-and-load-to-r-1">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r-1"></a>
</h5>
<p>After manually check the extracted metadata, users can <strong>download the specified objects</strong> with <code>ParseCELLxGENE</code>. The downloaded objects are controlled by <code>file.ext</code> (choose from <code>"rds"</code> and <code>"h5ad"</code>).</p>
<p>The returned value is a dataframe containing failed objects or a <code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>). If dataframe, users can re-run <code>ParseCELLxGENE</code> by setting <code>meta</code> to the returned value.</p>
<p><strong>When using <code>cellxgene.census</code>, users can subset metadata and gene</strong>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># download objects</span></span>
<span><span class="va">cellxgene.down</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>                                out.folder <span class="op">=</span> <span class="va">out.folder</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># retuen SeuratObject</span></span>
<span><span class="va">cellxgene.down.seu</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    obs.value.filter <span class="op">=</span> <span class="st">"cell_type == 'oligodendrocyte' &amp; disease == 'Alzheimer disease'"</span>,</span>
<span>                                    obs.keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"disease"</span>, <span class="st">"sex"</span>, <span class="st">"suspension_type"</span>, <span class="st">"development_stage"</span><span class="op">)</span>,</span>
<span>                                    out.folder <span class="op">=</span> <span class="va">out.folder</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># use cellxgene.census (support subset, but update is not timely)</span></span>
<span><span class="va">cellxgene.down.census</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span>use.census <span class="op">=</span> <span class="cn">TRUE</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>                                       obs.value.filter <span class="op">=</span> <span class="st">"cell_type == 'B cell' &amp; tissue_general == 'lung' &amp; disease == 'COVID-19'"</span>,</span>
<span>                                       obs.keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"tissue_general"</span>, <span class="st">"disease"</span>, <span class="st">"sex"</span><span class="op">)</span>,</span>
<span>                                       include.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ENSG00000161798'</span>, <span class="st">'ENSG00000188229'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="human-cell-atlas">Human Cell Atlas<a class="anchor" aria-label="anchor" href="#human-cell-atlas"></a>
</h4>
<p>The <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a> aims to map every cell type in the human body, it contains <strong>484 projects</strong>, most of which are from <code>Homo sapiens</code> (also includes projects from <code>Mus musculus</code>, <code>Macaca mulatta</code> and <code>canis lupus familiaris</code>).</p>
<div class="section level5">
<h5 id="show-available-datasets-2">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets-2"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ShowHCAProjects</code> to extract detailed project metadata, including project title, description, organism, sex, organ/organPart, disease, assay, preservation method, sample type, suspension type, cell type, development stage, et al.</p>
<p>There are <strong>484 unique projects</strong> under <em>five different catalogs (<code>dcp29</code>, <code>dcp30</code>, <code>dcp1</code>, <code>lm2</code>, <code>lm3</code>)</em>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.hca.projects</span> <span class="op">=</span> <span class="fu"><a href="reference/ShowHCAProjects.html">ShowHCAProjects</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="summarise-attributes-3">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes-3"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to summary attributes of <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">all.hca.projects</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"HCA"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="extract-metadata-5">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-5"></a>
</h5>
<p><code>GEfetch2R</code> provides <code>ExtractHCAMeta</code> to filter projects metadata, the available values of attributes can be obtained with <code>StatDBAttribute</code> except <strong>cell number</strong>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">hca.human.10x.projects</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractHCAMeta.html">ExtractHCAMeta</a></span><span class="op">(</span></span>
<span>  all.projects.df <span class="op">=</span> <span class="va">all.hca.projects</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="download-object-and-load-to-r-2">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r-2"></a>
</h5>
<p>After manually check the extracted metadata, users can <strong>download the specified objects</strong> with <code>ParseHCA</code>. The downloaded objects are controlled by <code>file.ext</code> (choose from <code>"rds"</code>, <code>"rdata"</code>, <code>"h5"</code>, <code>"h5ad"</code> and <code>"loom"</code>).</p>
<p>The returned value is a dataframe containing failed objects or a <code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>). If dataframe, users can re-run <code>ParseHCA</code> by setting <code>meta</code> to the returned value.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># download objects</span></span>
<span><span class="va">hca.down</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hca.human.10x.projects</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>                    file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h5ad"</span>, <span class="st">"rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># return SeuratObject</span></span>
<span><span class="va">hca.down.seu</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hca.human.10x.projects</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span>,</span>
<span>                        file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h5ad"</span>, <span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level3">
<h3 id="format-conversion">Format conversion<a class="anchor" aria-label="anchor" href="#format-conversion"></a>
</h3>
<p>There are many tools have been developed to process scRNA-seq data, such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">Scanpy</a>, <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a> and <a href="http://cole-trapnell-lab.github.io/monocle-release/" class="external-link">Monocle</a>. These tools have their own object formats, such as <code>Anndata</code> of <code>Scanpy</code>, <code>SeuratObject</code> of <code>Seurat</code>, <code>SingleCellExperiment</code> of <code>scran</code> and <code>CellDataSet</code>/<code>cell_data_set</code> of <code>Monocle2</code>/<code>Monocle3</code>. There are also some file formats designed for large omics datasets, such as <a href="http://loompy.org/" class="external-link">loom</a>. To perform a comprehensive scRNA-seq data analysis, we usually need to combine multiple tools, which means we need to perform object format conversion frequently. To facilitate user analysis of scRNA-seq data, <code>GEfetch2R</code> benchmarked the format conversion tools (<code>Anndata</code> -&gt; <code>SeuratObject</code>, <code>SeuratObject</code> to <code>Anndata</code>, <code>Anndata</code> to <code>SingleCellExperiment</code>, <code>SingleCellExperiment</code> to <code>Anndata</code>), and provides multiple functions to perform format conversion.</p>
<hr>
<div class="section level4">
<h4 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h4>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># pbmc_small</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span> <span class="co"># seger</span></span></code></pre></div>
<p><code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># object</span></span>
<span><span class="va">pbmc_small</span></span></code></pre></div>
<p><code>SingleCellExperiment</code>:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html" class="external-link">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="convert-seuratobject-to-other-objects">Convert SeuratObject to other objects<a class="anchor" aria-label="anchor" href="#convert-seuratobject-to-other-objects"></a>
</h4>
<p>Here, we will convert SeuratObject to <code>SingleCellExperiment</code>, <code>CellDataSet</code>/<code>cell_data_set</code>, <code>Anndata</code>, <code>loom</code>.</p>
<div class="section level5">
<h5 id="seuratobject-to-singlecellexperiment">SeuratObject to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#seuratobject-to-singlecellexperiment"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>Seurat</code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"SCE"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-celldatasetcell_data_set">SeuratObject to CellDataSet/cell_data_set<a class="anchor" aria-label="anchor" href="#seuratobject-to-celldatasetcell_data_set"></a>
</h5>
<p>To <code>CellDataSet</code> (The conversion is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("monocle") # reuqire monocle</span></span>
<span><span class="va">cds.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, to <span class="op">=</span> <span class="st">"CellDataSet"</span><span class="op">)</span></span></code></pre></div>
<p>To <code>cell_data_set</code> (The conversion is performed with functions implemented in <code>SeuratWrappers</code>):</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github('cole-trapnell-lab/monocle3') # reuqire monocle3</span></span>
<span><span class="va">cds3.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"cell_data_set"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-anndata">SeuratObject to AnnData<a class="anchor" aria-label="anchor" href="#seuratobject-to-anndata"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>SeuratObject</code> to <code>Anndata</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept and usability</li>
<li>
<code>sceasy</code> has best performance in running time and disk usage.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SeuratDisk</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, </span>
<span>       assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># sceasy</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, </span>
<span>       assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>       out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-loom">SeuratObject to loom<a class="anchor" aria-label="anchor" href="#seuratobject-to-loom"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>SeuratDisk</code>:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"pbmc_small_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"loom"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="convert-other-objects-to-seuratobject">Convert other objects to SeuratObject<a class="anchor" aria-label="anchor" href="#convert-other-objects-to-seuratobject"></a>
</h4>
<div class="section level5">
<h5 id="singlecellexperiment-to-seuratobject">SingleCellExperiment to SeuratObject<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-seuratobject"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>Seurat</code>:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sce.obj</span>, from <span class="op">=</span> <span class="st">"SCE"</span>, </span>
<span>                            count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>                            assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="celldatasetcell_data_set-to-seuratobject">CellDataSet/cell_data_set to SeuratObject<a class="anchor" aria-label="anchor" href="#celldatasetcell_data_set-to-seuratobject"></a>
</h5>
<p><code>CellDataSet</code> to <code>SeuratObject</code> (The conversion is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds.obj</span>, from <span class="op">=</span> <span class="st">"CellDataSet"</span>, </span>
<span>                            count.assay <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<p><code>cell_data_set</code> to <code>SeuratObject</code> (The conversion is performed with functions implemented in <code>Seurat</code>): <code>{r cds2seu2, eval=FALSE} seu.obj.cds3 &lt;- ImportSeurat(obj = cds3.obj, from = "cell_data_set",                               count.assay = "counts", data.assay = "logcounts",                              assay = "RNA")</code></p>
<hr>
</div>
<div class="section level5">
<h5 id="anndata-to-seuratobject">AnnData to SeuratObject<a class="anchor" aria-label="anchor" href="#anndata-to-seuratobject"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>AnnData</code> to <code>SeuratObject</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept (<strong><code>GEfetch2R</code> integrates <code>scDIOR</code> and <code>SeuratDisk</code> to achieve the best performance in information kept</strong>)</li>
<li>
<code>schard</code> is the best method in terms of usability</li>
<li>
<code>schard</code> and <code>sceasy</code> have comparable performance when cell number below 200k, but <code>sceasy</code> has better performance in scalability</li>
<li>
<code>sceasy</code> has better performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SeuratDisk</span></span>
<span><span class="va">ann.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># sceasy</span></span>
<span><span class="va">ann.sceasy</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>,</span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">ann.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">ann.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, </span>
<span>                     method <span class="op">=</span> <span class="st">"schard"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, use.raw <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># SeuratDisk+scDIOR</span></span>
<span><span class="va">ann.seuscdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"SeuratDisk+scDIOR"</span>, </span>
<span>                        assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="loom-to-seuratobject">loom to SeuratObject<a class="anchor" aria-label="anchor" href="#loom-to-seuratobject"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>SeuratDisk</code> and <code>Seurat</code>:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loom will lose reduction</span></span>
<span><span class="va">seu.obj.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>loom.file <span class="op">=</span> <span class="va">loom.file</span>, from <span class="op">=</span> <span class="st">"loom"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="conversion-between-singlecellexperiment-and-anndata">Conversion between SingleCellExperiment and AnnData<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-anndata"></a>
</h4>
<div class="section level5">
<h5 id="singlecellexperiment-to-anndata">SingleCellExperiment to AnnData<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-anndata"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>SingleCellExperiment</code> to <code>AnnData</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of information kept and running time</li>
<li>
<code>scDIOR</code> is the best method in terms of usability and disk usage</li>
</ul>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sceasy</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"benchmark"</span>, </span>
<span>       slot <span class="op">=</span> <span class="st">"rawcounts"</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span>       </span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">seger.scdior</span> <span class="op">&lt;-</span> <span class="va">seger</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co"># scDIOR does not support varm in rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">seger.scdior</span><span class="op">)</span><span class="op">$</span><span class="va">varm</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger.scdior</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, out.folder <span class="op">=</span> <span class="st">"benchmark"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># zellkonverter</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"zellkonverter"</span>, </span>
<span>       out.folder <span class="op">=</span> <span class="st">"benchmark"</span>, slot <span class="op">=</span> <span class="st">"rawcounts"</span>, </span>
<span>       conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="anndata-to-singlecellexperiment">AnnData to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#anndata-to-singlecellexperiment"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>AnnData</code> to <code>SingleCellExperiment</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of information kept</li>
<li>
<code>schard</code> is the best method in terms of usability and running time</li>
<li>
<code>schard</code> and <code>scDIOR</code> have comparable performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scDIOR</span></span>
<span><span class="va">sce.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># zellkonverter</span></span>
<span><span class="va">sce.zell</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"zellkonverter"</span>,</span>
<span>                   slot <span class="op">=</span> <span class="st">"scale.data"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">sce.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, </span>
<span>                     method <span class="op">=</span> <span class="st">"schard"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="conversion-between-singlecellexperiment-and-loom">Conversion between SingleCellExperiment and loom<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-loom"></a>
</h4>
<p>The conversion is performed with functions implemented in <code>LoomExperiment</code>.</p>
<div class="section level5">
<h5 id="singlecellexperiment-to-loom">SingleCellExperiment to loom<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-loom"></a>
</h5>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.loom first</span></span>
<span><span class="va">seger.loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"seger_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"loom"</span>, sce <span class="op">=</span> <span class="va">seger</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="loom-to-singlecellexperiment">loom to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#loom-to-singlecellexperiment"></a>
</h5>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"loom"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>For any question, feature request or bug report please write an email to <code>songyb0519@gmail.com</code>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the <code>GEfetch2R</code> project is released with a <a href="https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
<p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/showteeth/GEfetch2R/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/showteeth/GEfetch2R/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GEfetch2R</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yabing Song <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
