<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="The goal of scfetch is to access and format scRNA-seq datasets. It can be used to download scRNA-seq
    datasets from widely used public resources, including GEO, Zenodo, CELLxGENE, Human Cell Atlas, PanglaoDB and UCSC Cell Browser.
    And, it can also be used to perform object conversion between SeuratObject, loom,  h5ad, SingleCellExperiment and CellDataSet/cell_data_set.">
<title>Access and Format Single-cell RNA-seq Datasets from Public Resources  • scfetch</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Access and Format Single-cell RNA-seq Datasets from Public Resources ">
<meta property="og:description" content="The goal of scfetch is to access and format scRNA-seq datasets. It can be used to download scRNA-seq
    datasets from widely used public resources, including GEO, Zenodo, CELLxGENE, Human Cell Atlas, PanglaoDB and UCSC Cell Browser.
    And, it can also be used to perform object conversion between SeuratObject, loom,  h5ad, SingleCellExperiment and CellDataSet/cell_data_set.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">

    <a class="navbar-brand me-2" href="index.html">scfetch</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/DownloadRaw.html">Download fastq/bam files</a>
    <a class="dropdown-item" href="articles/DownloadMatrices.html">Download count matrices</a>
    <a class="dropdown-item" href="articles/DownloadObjects.html">Download processed objects</a>
    <a class="dropdown-item" href="articles/ObjectConversion.html">Convert formats between objects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/scfetch">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="scfetch---access-and-format-single-cell-rna-seq-datasets-from-public-resources">scfetch - Access and Format Single-cell RNA-seq Datasets from Public Resources<a class="anchor" aria-label="anchor" href="#scfetch---access-and-format-single-cell-rna-seq-datasets-from-public-resources"></a>
</h1></div>
<p><img src="reference/figures/scfetch.png" align="right" width="200"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>scfetch</code> is designed to accelerate users download and prepare single-cell datasets from public resources. It can be used to:</p>
<ul>
<li>
<strong>Download fastq files</strong> from <code>GEO/SRA</code>, <strong>foramt fastq files</strong> to standard style that can be identified by 10x softwares (e.g. CellRanger).</li>
<li>
<strong>Download bam files</strong> from <code>GEO/SRA</code>, support <strong>downloading original 10x generated bam files (with custom tags) and normal bam files</strong>, and <strong>convert bam files to fastq files</strong>.</li>
<li>Download scRNA-seq <strong>matrix</strong> and <strong>annotation (e.g. cell type)</strong> information from <code>GEO</code>, <code>PanglanDB</code> and <code>UCSC Cell Browser</code>, <strong>load the downnloaded matrix to <code>Seurat</code></strong>.</li>
<li>Download processed objects from <code>Zeenodo</code>, <code>CELLxGENE</code> and <code>Human Cell Atlas</code>.</li>
<li>
<strong>Formats conversion between widely used single cell objects</strong> (<code>SeuratObject</code>, <code>AnnData</code>, <code>SingleCellExperiment</code>, <code>CellDataSet/cell_data_set</code> and <code>loom</code>).</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="framework">Framework<a class="anchor" aria-label="anchor" href="#framework"></a>
</h2>
<div align="center">
<p><img src="reference/figures/scfetch_workflow.png" title="scfetch_framework" alt="scfetch_framework"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of <code>scfetch</code> from <a href="https://github.com/showteeth/scfetch" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"showteeth/scfetch"</span><span class="op">)</span></span></code></pre></div>
<p>For data structures conversion, <code>scfetch</code> requires several python pcakages, you can install with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install python packages</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda loompy anndata</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install anndata loompy</span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="vignette">Vignette<a class="anchor" aria-label="anchor" href="#vignette"></a>
</h3>
<p>Detailed usage is available in <a href="https://showteeth.github.io/scfetch/">website</a>.</p>
<hr>
</div>
<div class="section level3">
<h3 id="function-list">Function list<a class="anchor" aria-label="anchor" href="#function-list"></a>
</h3>
<table class="table">
<thead><tr>
<th>
Type
</th>
<th>
Function
</th>
<th>
Usage
</th>
</tr></thead>
<tbody>
<tr>
<td rowspan="3">
Download and format fastq
</td>
<td>
ExtractRun
</td>
<td>
Extract runs with GEO accession number or GSM number
</td>
</tr>
<tr>
<td>
DownloadSRA
</td>
<td>
Download sra files
</td>
</tr>
<tr>
<td>
SplitSRA
</td>
<td>
Split sra files to fastq files and format to 10x standard style
</td>
</tr>
<tr>
<td rowspan="2">
Download and convert bam
</td>
<td>
DownloadBam
</td>
<td>
Download bam (support 10x original bam)
</td>
</tr>
<tr>
<td>
Bam2Fastq
</td>
<td>
Convert bam files to fastq files
</td>
</tr>
<tr>
<td rowspan="9">
Download matrix and load to Seurat
</td>
<td>
ExtractGEOMeta
</td>
<td>
Extract sample metadata from GEO
</td>
</tr>
<tr>
<td>
ParseGEO
</td>
<td>
Download matrix from GEO and load to Seurat
</td>
</tr>
<tr>
<td>
ExtractPanglaoDBMeta
</td>
<td>
Extract sample metadata from PandlaoDB
</td>
</tr>
<tr>
<td>
ExtractPanglaoDBComposition
</td>
<td>
Extract cell type composition of PanglaoDB datasets
</td>
</tr>
<tr>
<td>
ParsePanglaoDB
</td>
<td>
Download matrix from PandlaoDB and load to Seurat
</td>
</tr>
<tr>
<td>
ShowCBDatasets
</td>
<td>
Show all available datasets in UCSC Cell Browser
</td>
</tr>
<tr>
<td>
ExtractCBDatasets
</td>
<td>
Extract UCSC Cell Browser datasets with attributes
</td>
</tr>
<tr>
<td>
ExtractCBComposition
</td>
<td>
Extract cell type composition of UCSC Cell Browser datasets
</td>
</tr>
<tr>
<td>
ParseCBDatasets
</td>
<td>
Download UCSC Cell Browser datasets and load to Seurat
</td>
</tr>
<tr>
<td rowspan="8">
Download objects
</td>
<td>
ExtractZenodoMeta
</td>
<td>
Extract sample metadata from Zenodo with DOIs
</td>
</tr>
<tr>
<td>
ParseZenodo
</td>
<td>
Download rds/rdata/h5ad/loom from Zenodo with DOIs
</td>
</tr>
<tr>
<td>
ShowCELLxGENEDatasets
</td>
<td>
Show all available datasets in CELLxGENE
</td>
</tr>
<tr>
<td>
ExtractCELLxGENEMeta
</td>
<td>
Extract metadata of CELLxGENE datasets with attributes
</td>
</tr>
<tr>
<td>
ParseCELLxGENE
</td>
<td>
Download rds/h5ad from CELLxGENE
</td>
</tr>
<tr>
<td>
ShowHCAProjects
</td>
<td>
Show all available projects in Human Cell Atlas
</td>
</tr>
<tr>
<td>
ExtractHCAMeta
</td>
<td>
Extract metadata of Human Cell Atlas projects with attributes
</td>
</tr>
<tr>
<td>
ParseHCA
</td>
<td>
Download rds/rdata/h5/h5ad/loom from Human Cell Atlas
</td>
</tr>
<tr>
<td rowspan="4">
Convert between different single-cell objects
</td>
<td>
ExportSeurat
</td>
<td>
Convert SeuratObject to AnnData, SingleCellExperiment, CellDataSet/cell_data_set and loom
</td>
</tr>
<tr>
<td>
ImportSeurat
</td>
<td>
Convert AnnData, SingleCellExperiment, CellDataSet/cell_data_set and loom to SeuratObject
</td>
</tr>
<tr>
<td>
SCEAnnData
</td>
<td>
Convert between SingleCellExperiment and AnnData
</td>
</tr>
<tr>
<td>
SCELoom
</td>
<td>
Convert between SingleCellExperiment and loom
</td>
</tr>
<tr>
<td rowspan="1">
Summarize datasets based on attributes
</td>
<td>
StatDBAttribute
</td>
<td>
Summarize datasets in PandlaoDB, UCSC Cell Browser and CELLxGENE based on attributes
</td>
</tr>
</tbody>
</table>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>For any question, feature request or bug report please write an email to <a href="songyb0519@gmail.com">songyb0519@gmail.com</a>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the <code>scfetch</code> project is released with a <a href="https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
<p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><small><a href="https://opensource.org/license/gpl-3-0/" class="external-link">GPL3</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scfetch</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yabing Song <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>





  </body>
</html>
