<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Access and Format Single-Cell RNA-Seq Datasets from Public Resources • GEfetch2R</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Access and Format Single-Cell RNA-Seq Datasets from Public Resources">
<meta name="description" content="The goal of GEfetch2R is to access and format single-cell RNA-seq datasets. It can be used to download single-cell RNA-seq datasets from widely used public resources, including GEO &lt;https://www.ncbi.nlm.nih.gov/geo/&gt;, Zenodo &lt;https://zenodo.org/&gt;, CELLxGENE &lt;https://cellxgene.cziscience.com/&gt;, Human Cell Atlas &lt;https://www.humancellatlas.org/&gt;, PanglaoDB &lt;https://panglaodb.se/index.html&gt; and UCSC Cell Browser &lt;https://cells.ucsc.edu/&gt;. And, it can also be used to perform object conversion between SeuratObject &lt;https://satijalab.org/seurat/&gt;, loom &lt;http://loompy.org/&gt;, h5ad &lt;https://scanpy.readthedocs.io/en/stable/&gt;, SingleCellExperiment &lt;https://bioconductor.org/packages/release/bioc/html/scran.html&gt;, CellDataSet &lt;http://cole-trapnell-lab.github.io/monocle-release/&gt; and cell_data_set &lt;https://cole-trapnell-lab.github.io/monocle3/&gt;.">
<meta property="og:description" content="The goal of GEfetch2R is to access and format single-cell RNA-seq datasets. It can be used to download single-cell RNA-seq datasets from widely used public resources, including GEO &lt;https://www.ncbi.nlm.nih.gov/geo/&gt;, Zenodo &lt;https://zenodo.org/&gt;, CELLxGENE &lt;https://cellxgene.cziscience.com/&gt;, Human Cell Atlas &lt;https://www.humancellatlas.org/&gt;, PanglaoDB &lt;https://panglaodb.se/index.html&gt; and UCSC Cell Browser &lt;https://cells.ucsc.edu/&gt;. And, it can also be used to perform object conversion between SeuratObject &lt;https://satijalab.org/seurat/&gt;, loom &lt;http://loompy.org/&gt;, h5ad &lt;https://scanpy.readthedocs.io/en/stable/&gt;, SingleCellExperiment &lt;https://bioconductor.org/packages/release/bioc/html/scran.html&gt;, CellDataSet &lt;http://cole-trapnell-lab.github.io/monocle-release/&gt; and cell_data_set &lt;https://cole-trapnell-lab.github.io/monocle3/&gt;.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="gefetch2r---access-single-cellbulk-rna-seq-data-from-public-repositories-and-benchmark-the-subsequent-format-conversion-tools">
<em>GEfetch2R</em> - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools<a class="anchor" aria-label="anchor" href="#gefetch2r---access-single-cellbulk-rna-seq-data-from-public-repositories-and-benchmark-the-subsequent-format-conversion-tools"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="quick-start">
<a href="https://showteeth.github.io/GEfetch2R/articles/QuickStart.html"><mark>Quick start</mark></a><a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h1>
<p><img src="reference/figures/GEfetch2R.png" align="right" width="200"></p>
<p><a href="https://cran.r-project.org/web/licenses/GPLv3" class="external-link"><img src="https://img.shields.io/badge/license-GPLv3-blue.svg" alt="License: GPLv3"></a> <a href="https://github.com/showteeth/GEfetch2R" class="external-link"><img src="https://img.shields.io/github/languages/code-size/showteeth/GEfetch2R.svg" alt="CODE_SIZE"></a></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> is designed to accelerate users’ downloading and preparation of single-cell/bulk RNA-seq datasets from public resources. It can be used to:</p>
<p>For <strong>bulk RNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from <code>SRA/ENA</code> with <strong>GEO accession</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files from <code>SRA/ENA</code>
</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to <code>fastq</code> files</li>
<li><strong>Convert <code>bam</code> to <code>fastq</code> files</strong></li>
<li><strong>Mapping with <code>STAR</code> and load the output to <code>DESeq2</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with <strong>GEO accession</strong>:
<ul>
<li><strong>Generate count matrix from supplementary files</strong></li>
<li><strong>Load the count matrix to <code>DESeq2</code></strong></li>
</ul>
</li>
</ul>
<p>For <strong>scRNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from <code>SRA/ENA</code> with <strong>GEO accession</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files from <code>SRA/ENA</code>
</li>
<li>
<strong>Foramt downloaded <code>fastq</code> files</strong> to standard style that can be identified by 10x softwares (e.g. <code>CellRanger</code>).</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to <code>fastq</code> files</li>
<li><strong>Download original 10x generated <code>bam</code> files (with custom tags)</strong></li>
<li><strong>Convert <code>bam</code> to <code>fastq</code> files (<code>samtools/bamtofastq_linux</code>)</strong></li>
<li><strong>Mapping with <code>STAR/CellRanger</code> and load the output to <code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong>:
<ul>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with <strong>GEO accession</strong>
</li>
<li>Download <strong>count matrix</strong> and <strong>annotation (e.g. cell type)</strong> information from <code>PanglaoDB</code> and <code>UCSC Cell Browser</code> with <strong>key words (filter criteria)</strong>
</li>
<li>
<strong>Extract subset with annotation and gene (<code>PanglaoDB</code> and <code>UCSC Cell Browser</code>)</strong><br>
</li>
<li><strong>Load the count matrix and annotation to <code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>processed object</strong>:
<ul>
<li>Download <strong>processed object</strong> from <code>GEO</code> with <strong>GEO accession</strong>
</li>
<li>Download <strong>processed object</strong> from <code>Zenodo</code> with <strong>DOI</strong>
</li>
<li>Download <strong>processed object</strong> from <code>CELLxGENE</code> and <code>Human Cell Atlas</code> with <strong>key words (filter criteria)</strong>
</li>
<li>
<strong>Parallel</strong> download</li>
<li><strong>Extract subset with annotation and gene (<code>CELLxGENE</code>)</strong></li>
<li><strong>Load the processed object (<code>rds</code>/<code>rdata</code>) to <code>Seurat</code></strong></li>
<li>
<strong>Extract count matrix and metadata from processed object (<code>rds</code>/<code>rdata</code>)</strong>, including <code>Seurat</code>, <code>seurat</code>, <code>SingleCellExperiment</code>, <code>cell_data_set</code>, <code>CellDataSet</code>, <code>DESeqDataSet</code>, <code>DGEList</code>.</li>
</ul>
</li>
<li>
<strong>Format conversion and benchmark related tools</strong>:
<ul>
<li>
<strong>Format conversion between widely used single cell objects</strong> (<code>SeuratObject</code>, <code>AnnData</code>, <code>SingleCellExperiment</code>, <code>CellDataSet/cell_data_set</code> and <code>loom</code>)</li>
<li>
<strong>Benchmark the tools (more than two tools available for the conversion)</strong>:
<ul>
<li>
<code>SeuratObject</code> to <code>AnnData</code>: <code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>
</li>
<li>
<code>SingleCellExperiment</code> to <code>AnnData</code>: <code>sceasy</code>, <code>scDIOR</code>, <code>zellkonverter</code>
</li>
<li>
<code>AnnData</code> to <code>SeuratObject</code>: <code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>, <code>schard</code>
</li>
<li>
<code>AnnData</code> to <code>SingleCellExperiment</code>: <code>schard</code>, <code>scDIOR</code>, <code>zellkonverter</code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="framework">Framework<a class="anchor" aria-label="anchor" href="#framework"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_overview.png" title="GEfetch2R_framework" alt="GEfetch2R_framework"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="manual-installation">Manual installation<a class="anchor" aria-label="anchor" href="#manual-installation"></a>
</h3>
<p>To install <code>GEfetch2R</code>, start <code>R</code> and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install from GitHub</span></span>
<span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"showteeth/GEfetch2R"</span><span class="op">)</span></span></code></pre></div>
<p>There are some conditionally used <code>R</code> packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools") #In case you have not installed it.</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"alexvpickering/GEOfastq"</span><span class="op">)</span> <span class="co"># download fastq</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'tiledbsoma'</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'https://tiledb-inc.r-universe.dev'</span>, <span class="st">'https://cloud.r-project.org'</span><span class="op">)</span><span class="op">)</span> <span class="co"># download from CELLxGENE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'cellxgene.census'</span>, repos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'https://chanzuckerberg.r-universe.dev'</span>, <span class="st">'https://cloud.r-project.org'</span><span class="op">)</span><span class="op">)</span> <span class="co"># download from CELLxGENE</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cellgeni/sceasy"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mojaveazure/seurat-disk"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/seurat-wrappers"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'theislab/zellkonverter@7b118653a471330b3734dcfee60c3537352ecb8d'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'cellgeni/schard'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'JiekaiLab/dior'</span>, upgrade <span class="op">=</span> <span class="st">'never'</span><span class="op">)</span>  <span class="co"># format conversion</span></span></code></pre></div>
<p><strong>For possible issues about installation, please refer <code>INSTALL.md</code>.</strong></p>
<p>For format conversion and downloading <code>fastq</code>/<code>bam</code> files, <code>GEfetch2R</code> requires additional tools, you can install with:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># install additional packages for format conversion</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">pip</span> install diopy</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda loompy anndata </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="ex">pip</span> install anndata loompy</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># install additional packages for downloading fastq/bam files</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda <span class="st">'parallel-fastq-dump'</span> <span class="st">'sra-tools&gt;=3.0.0'</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co"># install bamtofastq, the following installs linux version</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--quiet</span> https://github.com/10XGenomics/bamtofastq/releases/download/v1.4.1/bamtofastq_linux <span class="kw">&amp;&amp;</span> <span class="fu">chmod</span> +x bamtofastq_linux</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co"># install ascp</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> hcc aspera-cli <span class="at">-y</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co"># ascp path: ~/miniconda3/bin/ascp (path/bin/ascp)</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co"># private-key file : ~/miniconda3/etc/asperaweb_id_dsa.openssh (path/etc/asperaweb_id_dsa.openssh)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="docker-image">Docker image<a class="anchor" aria-label="anchor" href="#docker-image"></a>
</h3>
<p>We also provide a <a href="https://hub.docker.com/repository/docker/soyabean/gefetch2r" class="external-link">docker image</a> to use:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># pull the image</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">docker</span> pull soyabean/gefetch2r:1.2</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># run the image</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-p</span> 8888:8787 <span class="at">-e</span> PASSWORD=passwd <span class="at">-e</span> ROOT=TRUE <span class="at">-it</span> soyabean/gefetch2r:1.2</span></code></pre></div>
<p><strong>Notes</strong>:</p>
<ul>
<li>After running the above codes, open browser and enter <code>http://localhost:8888/</code>, the user name is <code>rstudio</code>, the password is <code>passwd</code> (set by <code>-e PASSWORD=passwd</code>)</li>
<li>If port <code>8888</code> is in use, change <code>-p 8888:8787</code>
</li>
<li>The <code>conda.path</code> in <code>ExportSeurat</code> and <code>ImportSeurat</code> can be set <code>/opt/conda</code>.</li>
<li>The <strong>sra-tools</strong> can be found in <code>/opt/sratoolkit.3.0.6-ubuntu64/bin</code>.</li>
<li>The <code>parallel-fastq-dump</code> path: <code>/opt/conda/bin/parallel-fastq-dump</code>.</li>
<li>The <code>bamtofastq_linux</code> path: <code>/opt/bamtofastq_linux</code>.</li>
<li>The <code>samtools</code> path: <code>/opt/conda/bin/samtools</code>.</li>
<li>The <code>STAR</code> and <code>Cell Ranger</code> is not available in the image because customized reference genome is required.</li>
</ul>
<p>Codes used to test the usability of the Docker image: <a href="https://github.com/showteeth/GEfetch2R/blob/main/man/benchmark/Docker_test.R" class="external-link">Docker_test.R</a></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="vignette">Vignette<a class="anchor" aria-label="anchor" href="#vignette"></a>
</h2>
<p>Detailed usage is available in <a href="https://showteeth.github.io/GEfetch2R/">website</a>:</p>
<ul>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadRaw.html" class="external-link">Download raw data (SRA/ENA)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadMatrices.html" class="external-link">Download count matrix and annotaion (GEO, PanglaoDB, UCSC Cell Browser)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadObjects.html" class="external-link">Download processed data (GEO, Zenodo, CELLxGENE, Human Cell Atlas)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/ObjectConversion.html" class="external-link">Format conversion and benchmark (AnnData, SeuratObject, SingleCellExperiment, CellDataSet/cell_data_set, loom)</a></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="comparison-with-similar-tools">Comparison with similar tools<a class="anchor" aria-label="anchor" href="#comparison-with-similar-tools"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_compare.png" title="GEfetch2R_comparison" alt="GEfetch2R_comparison"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="quick-start-1">Quick start<a class="anchor" aria-label="anchor" href="#quick-start-1"></a>
</h2>
<div class="section level3">
<h3 id="check-api">Check API<a class="anchor" aria-label="anchor" href="#check-api"></a>
</h3>
<p>Check the availability of APIs used:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check all databases: "SRA/ENA", "GEO", "PanglaoDB", "UCSC Cell Browser", "Zenodo", "CELLxGENE", "Human Cell Atlas"</span></span>
<span><span class="fu"><a href="reference/CheckAPI.html">CheckAPI</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for a given database</span></span>
<span><span class="fu"><a href="reference/CheckAPI.html">CheckAPI</a></span><span class="op">(</span>database <span class="op">=</span> <span class="st">"GEO"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="sraena-srafastqbam">SRA/ENA (sra/fastq/bam)<a class="anchor" aria-label="anchor" href="#sraena-srafastqbam"></a>
</h3>
<p>Extract all runs, automatically identify the RNA-seq type (10x Genomics scRNA-seq, bulk RNA-seq, Smart-seq2 scRNA-seq/mini-bulk RNA-seq) of each sample, download <code>fastq</code> files from <code>ENA</code>, perform read mapping, and load the results to R:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># mixture of 10x Genomics scRNA-seq and bulk RNA-seq</span></span>
<span><span class="va">GSE305141.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadFastq2R.html">DownloadFastq2R</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE305141"</span>, skip.gsm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM9162729"</span>, <span class="st">"GSM9162725"</span><span class="op">)</span>,</span>
<span>  star.ref <span class="op">=</span> <span class="st">"/path/to/star/ref"</span>, cellranger.ref <span class="op">=</span> <span class="st">"/path/to/cellranger/ref"</span>,</span>
<span>  star.path <span class="op">=</span> <span class="st">"/path/to/STAR"</span>, cellranger.path <span class="op">=</span> <span class="st">"/path/to/cellranger"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># given GSM number</span></span>
<span><span class="va">GSE127942.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadFastq2R.html">DownloadFastq2R</a></span><span class="op">(</span></span>
<span>  gsm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM3656922"</span>, <span class="st">"GSM3656923"</span><span class="op">)</span>, star.ref <span class="op">=</span> <span class="st">"/path/to/ref"</span>,</span>
<span>  star.path <span class="op">=</span> <span class="st">"/path/to/STAR"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>acce</code>: the GEO accession.</li>
<li>
<code>skip.gsm</code>: vector of GSM numbers to skip.</li>
<li>
<code>gsm</code>: the GSM accession (given sample).</li>
<li>
<code>star.ref</code>: path to <code>STAR</code> reference. Used when bulk/Smart-seq2 RNA-seq samples exist.</li>
<li>
<code>star.path</code>: path to <code>STAR</code>, can be detected automatically by <code>Sys.which("STAR")</code>. Used when bulk/Smart-seq2 RNA-seq samples exist.</li>
<li>
<code>cellranger.ref</code>: path to <code>cellranger</code> reference. Used when 10x Genomics scRNA-seq samples exist.</li>
<li>
<code>cellranger.path</code>: path to <code>cellranger</code>, can be detected automatically by <code>Sys.which("cellranger")</code>. Used when 10x Genomics scRNA-seq samples exist.</li>
<li>
<code>out.folder</code>: the output folder, current working directory by default.</li>
<li>
<code>count.col</code>: column contains used count data (<code>2: unstranded (default value)</code>; <code>3: stranded=yes / 1st read strand</code>; <code>4: stranded=reverse/2nd read strand</code>), use when bulk RNA-seq or Smart-seq2 scRNA-seq/mini-bulk RNA-seq.</li>
</ul>
<div class="alert alert-success" role="alert">
<p>The way <code>GEfetch2R</code> automatically identify the RNA-seq type can be find <a href="https://showteeth.github.io/GEfetch2R/articles/DownloadRaw.html#one-step-wrapper">here</a>. Besides, users can specify the RNA-seq type via <code>force.type</code>, choose one from <code>"10x"</code>, <code>"Smart-seq2"</code>, <code>"bulk"</code>.</p>
</div>
<p>The above code equals the following:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract all runs</span></span>
<span><span class="va">GSE127942.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE127942"</span><span class="op">)</span></span>
<span><span class="va">GSE127942.runs</span> <span class="op">&lt;-</span> <span class="va">GSE127942.runs</span><span class="op">[</span><span class="va">GSE127942.runs</span><span class="op">$</span><span class="va">gsm_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM3656922"</span>, <span class="st">"GSM3656923"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download fastq from ENA</span></span>
<span><span class="va">GSE127942.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE127942.runs</span>, out.folder <span class="op">=</span> <span class="st">"/path/to/fastq_out"</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"wget"</span>, <span class="co"># available method: "download.file", "ascp", "wget"</span></span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span>, format.10x <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># 10x-specific</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># read mapping, load to R</span></span>
<span><span class="va">GSE127942.gsms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"/path/to/fastq_out"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM3656922"</span>, <span class="st">"GSM3656923"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GSE127942.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Fastq2R.html">Fastq2R</a></span><span class="op">(</span></span>
<span>  sample.dir <span class="op">=</span> <span class="va">GSE127942.gsms</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"STAR"</span>, st.path <span class="op">=</span> <span class="st">"/path/to/STAR"</span>, ref <span class="op">=</span> <span class="st">"/path/to/STAR.reference"</span>, <span class="co"># STAR reference</span></span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/mapping_out"</span>, count.col <span class="op">=</span> <span class="fl">2</span>, <span class="co"># strand-specific (2: unstranded; 3: stranded=yes; 4: stranded=reverse)</span></span>
<span>  localcores <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE127942.obj is a DESeqDataSet object</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="geo-count-matrix-metadata-processed-object">
<a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO</a> (count matrix, metadata, processed object)<a class="anchor" aria-label="anchor" href="#geo-count-matrix-metadata-processed-object"></a>
</h3>
<div class="section level4">
<h4 id="count-matrix-bulk-rna-seqsmart-seq2">Count matrix (bulk RNA-seq/Smart-seq2)<a class="anchor" aria-label="anchor" href="#count-matrix-bulk-rna-seqsmart-seq2"></a>
</h4>
<p>Supplementary file in <code>csv(.gz)</code>/<code>tsv(.gz)</code>/<code>txt(.gz)</code>/<code>tab(.gz)</code>/<code>xlsx(.gz)</code>/<code>xls(.gz)</code> format or <code>tar(.gz)</code> format (contain <code>csv(.gz)</code>/<code>tsv(.gz)</code>/<code>txt(.gz)</code>/<code>tab(.gz)</code>/<code>xlsx(.gz)</code>/<code>xls(.gz)</code> files):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return SeuratObject</span></span>
<span><span class="va">GSE297431.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE297431"</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>  data.type <span class="op">=</span> <span class="st">"sc"</span>, <span class="co"># scRNA-seq, Smart-seq2 here</span></span>
<span>  load2R <span class="op">=</span> <span class="cn">TRUE</span>, merge <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>down.supp = TRUE</code>: generate count matrix from supplementary file</li>
<li>
<code>supp.idx = 1</code>: the first supplementary file containing the count matrix</li>
<li>
<code>supp.type = "count"</code>: the file containing count matrix is in <code>csv(.gz)</code>/<code>tsv(.gz)</code>/<code>txt(.gz)</code>/<code>tab(.gz)</code>/<code>xlsx(.gz)</code>/<code>xls(.gz)</code> format</li>
<li>
<code>data.type = "sc"</code>: the data type of the dataset, choose from <code>"sc"</code> (single-cell) and <code>"bulk"</code> (bulk)</li>
<li>
<code>load2R = F</code>: return count matrix; <code>load2R = T</code> and <code>data.type = "sc"</code>, return <code>SeuratObject</code>; <code>load2R = T</code> and <code>data.type = "bulk"</code>, return <code>DESeqDataSet</code>
</li>
</ul>
<div class="alert alert-success" role="alert">
<ul>
<li>
<code>ParseGEO</code> is compatible with count matrices generated by <a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html" class="external-link">htseq-count</a>, <a href="https://subread.sourceforge.net/featureCounts.html" class="external-link">featureCounts</a> (contain extra columns: “Chr”, “Start”, “End”, “Strand”, “Length”, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE182219" class="external-link">GSE182219</a>), and <a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a> with <code>--quantMode</code> (contain three count columns: unstranded, 1st read strand, 2nd read strand, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE195839" class="external-link">GSE195839</a>).</li>
<li>
<code>ParseGEO</code> is also compatible with count matrix files containing irregular extra columns, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE268038" class="external-link">GSE268038</a> contains “chromosome”, “start”, “end”, “strand”. Users can specify the extra columns to ignore by setting <code>extra.cols</code> (default: “chr”, “start”, “end”, “strand”, “length”, “width”, “chromosome”, “seqnames”, “seqname”, “chrom”, “chromosome_name”, “seqid”, “stop”)</li>
<li>In general, the rows of the count matrix represent genes, and the columns represent samples. <code>ParseGEO</code> can deal with the transposed count matrix (the number of rows is less than the number of columns) by setting <code>transpose</code> to TRUE.</li>
</ul>
</div>
<hr>
</div>
<div class="section level4">
<h4 id="count-matrix-scrna-seq">Count matrix (scRNA-seq)<a class="anchor" aria-label="anchor" href="#count-matrix-scrna-seq"></a>
</h4>
<p><strong>Supplementary files in <code>h5(.gz)</code> format or composed of <code>barcodes.tsv(.gz)</code>/<code>genes.tsv(.gz)</code>, <code>matrix.mtx(.gz)</code>, <code>features.tsv(.gz)</code> files</strong> (set <code>supp.type = "10xSingle"</code>):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE278892.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE278892"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="st">"10xSingle"</span>, timeout <span class="op">=</span> <span class="fl">36000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/store/count_matrix"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>down.supp = TRUE</code>: generate count matrix from supplementary file</li>
<li>
<code>supp.type = "10xSingle"</code>: the file containing count matrix is in separate files (<code>barcodes.tsv(.gz)</code>/<code>genes.tsv(.gz)</code>, <code>matrix.mtx(.gz)</code>, <code>features.tsv(.gz)</code>) or <code>h5(.gz)</code> file(s)</li>
<li>
<code>load2R = F</code>: return count matrix; <code>load2R = T</code> and <code>data.type = "sc"</code>, return <code>SeuratObject</code>
</li>
</ul>
<p><strong>Supplementary file in <code>tar(.gz)</code> format</strong> (set <code>supp.type = "10x"</code>):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE292908.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE292908"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="st">"10x"</span>, timeout <span class="op">=</span> <span class="fl">36000</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/store/count_matrix"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>down.supp = TRUE</code>: generate count matrix from supplementary file</li>
<li>
<code>supp.idx = 1</code>: the first supplementary file containing the count matrix</li>
<li>
<code>supp.type = "10x"</code>: the file containing count matrix is in <strong><code>tar(.gz)</code> format</strong>. The files in <code>tar(.gz)</code> can be in <code>zip</code>, <code>tar(.gz)</code>, <code>h5(.gz)</code> format, or <strong>separate files</strong> (<code>barcodes.tsv(.gz)</code>/<code>genes.tsv(.gz)</code>, <code>matrix.mtx(.gz)</code>, <code>features.tsv(.gz)</code>).</li>
<li>
<code>load2R = F</code>: return count matrix; <code>load2R = T</code> and <code>data.type = "sc"</code>, return <code>SeuratObject</code>
</li>
</ul>
<div class="alert alert-success" role="alert">
<ul>
<li>
<code>ParseGEO</code> can also load count matrices generated by scRNA-seq platforms other than 10x Genomics, which have a similar output structure to 10x (CellRanger), e.g., <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE300217" class="external-link">SeekOne and MobiDrop</a> and <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE285164" class="external-link">DNBelab C4</a>
</li>
<li>
<code>ParseGEO</code> can handle files with very deep hierarchical structures, e.g. <a href="https://github.com/showteeth/GEfetch2R/issues/16" class="external-link">Compressed files (zip, tar.gz, tar) in downloaded supplemental files (GEO, 10x)</a>
</li>
<li>
<code>ParseGEO</code> can identify sample name before or after the fixed name, e.g. <a href="https://github.com/showteeth/GEfetch2R/issues/17" class="external-link">Sample name is after the fixed name (GEO, 10x)</a>
</li>
</ul>
</div>
<hr>
</div>
<div class="section level4">
<h4 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h4>
<p>The sample metadata can be obtained in two ways:</p>
<ul>
<li>
<strong>user-provided sample metadata when uploading to GEO (applicable to all GEO accessions)</strong>:</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata</span></span>
<span><span class="va">GSE297431.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE297431"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<strong>metadata in supplementary file</strong>:</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE297431.meta.supp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE297431"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">2</span> <span class="co"># specify the index of used supplementary file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="processed-object">Processed object<a class="anchor" aria-label="anchor" href="#processed-object"></a>
</h4>
<p>Supplementary file in <code>rdata(.gz)</code>/<code>rds(.gz)</code>/<code>h5ad(.gz)</code>/<code>loom(.gz)</code> format or <code>tar(.gz)</code> format (contain <code>rdata(.gz)</code>/<code>rds(.gz)</code>/<code>h5ad(.gz)</code>/<code>loom(.gz)</code> files):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return SeuratObject (rds)</span></span>
<span><span class="va">GSE285723.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE285723"</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">T</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/outfoder"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>supp.idx = 1</code>: the first supplementary file containing the processed object.</li>
<li>
<code>file.ext = c("rdata", "rds")</code>: download/keep files in <code>rdata(.gz)</code> and <code>rds(.gz)</code> formats (case-insensitive).</li>
<li>
<code>return.seu = T</code>: load downloaded objects to <code>Seurat</code>.</li>
</ul>
<p>Dissect and extract the <code>RData</code> files:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the object</span></span>
<span><span class="fu"><a href="reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE244572"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span>, <span class="st">"h5ad"</span>, <span class="st">"loom"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># process the object</span></span>
<span><span class="va">GSE244572.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/LoadRData.html">LoadRData</a></span><span class="op">(</span></span>
<span>  rdata <span class="op">=</span> <span class="st">"GSE244572/GSE244572_RPE_CITESeq.RData"</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"seurat"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"cell_data_set"</span>, <span class="st">"CellDataSet"</span>, <span class="st">"DESeqDataSet"</span>, <span class="st">"DGEList"</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span>, return.obj <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Key parameters:</p>
<ul>
<li>
<code>accept.fmt</code>: vector, the format of objects for dissecting and extracting.</li>
<li>
<code>slot</code>: vector, the type of count matrix to pull. <code>'counts'</code>: raw, un-normalized counts, <code>'data'</code>: normalized data, <code>scale.data</code>: z-scored/variance-stabilized data.</li>
<li>
<code>return.obj</code>: logical value, whether to load the available objects in <code>accept.fmt</code> to global environment.</li>
</ul>
<div class="alert alert-success" role="alert">
<p>The way <code>GEfetch2R</code> dissect and extract the <code>RData</code> files can be find <a href="https://showteeth.github.io/GEfetch2R/articles/DownloadObjects.html#process-ddata-files">here</a>.</p>
</div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="panglaodb-count-matrix-cell-type-composition">
<a href="https://panglaodb.se/samples.html" class="external-link">PanglaoDB</a> (count matrix, cell type composition)<a class="anchor" aria-label="anchor" href="#panglaodb-count-matrix-cell-type-composition"></a>
</h3>
<div class="section level4">
<h4 id="given-dataset">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract cell type composition </span></span>
<span><span class="va">lung.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>sra <span class="op">=</span> <span class="st">"SRA570744"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract count matrix and load to Seurat</span></span>
<span><span class="va">lung.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span>sra <span class="op">=</span> <span class="st">"SRA570744"</span>, srs <span class="op">=</span> <span class="st">"SRS2253536"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-samples-based-on-metadata">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarise attributes</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter metadata</span></span>
<span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>,</span>
<span>                                 show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract cell type composition</span></span>
<span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hsa.meta</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download matrix and load to Seurat, small test</span></span>
<span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="ucsc-cell-browser-count-matrix-cell-type-composition">
<a href="https://cells.ucsc.edu/" class="external-link">UCSC Cell Browser</a> (count matrix, cell type composition)<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser-count-matrix-cell-type-composition"></a>
</h3>
<div class="section level4">
<h4 id="given-dataset-1">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset-1"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract cell type composition </span></span>
<span><span class="va">ut.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span>link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-ureter"</span>, <span class="co"># collection</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-testis"</span> <span class="co"># dataset</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract count matrix and load to Seurat</span></span>
<span><span class="va">ut.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-ureter"</span>, <span class="co"># collection</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-testis"</span> <span class="co"># dataset</span></span>
<span><span class="op">)</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>merge = TRUE</code>: whether to merge <code>Seurat</code> list.</li>
</ul>
<hr>
</div>
<div class="section level4">
<h4 id="filter-samples-based-on-metadata-1">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata-1"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first-time run, get all samples and store json to json.folder</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">=</span> <span class="fu"><a href="reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="st">"/path/to/json"</span>, update <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># second-time run, use stored json</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "/path/to/json", update = FALSE)</span></span>
<span></span>
<span><span class="co"># summarise attributes</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"UCSC"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter metadata</span></span>
<span><span class="va">hbb.sample.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCBDatasets.html">ExtractCBDatasets</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, organ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"skeletal muscle"</span><span class="op">)</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Human (H. sapiens)"</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract cell type</span></span>
<span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span></span>
<span>  json.folder <span class="op">=</span> <span class="st">"/path/to/json"</span>,</span>
<span>  meta <span class="op">=</span> <span class="va">hbb.sample.df</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parse the whole datasets</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span>
<span><span class="co"># subset metadata and gene</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">hbb.sample.df</span>, obs.value.filter <span class="op">=</span> <span class="st">"Cell.Type == 'MP' &amp; Phase == 'G2M'"</span>,</span>
<span>  include.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"PAX7"</span>, <span class="st">"MYF5"</span>, <span class="st">"C1QTNF3"</span>, <span class="st">"MYOD1"</span>, <span class="st">"MYOG"</span>, <span class="st">"RASSF4"</span>, <span class="st">"MYH3"</span>, <span class="st">"MYL4"</span>,</span>
<span>    <span class="st">"TNNT3"</span>, <span class="st">"PDGFRA"</span>, <span class="st">"OGN"</span>, <span class="st">"COL3A1"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="zenodo-processed-object">
<a href="https://zenodo.org/" class="external-link">Zenodo</a> (processed object)<a class="anchor" aria-label="anchor" href="#zenodo-processed-object"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract metadata</span></span>
<span><span class="va">multi.dois</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download objects</span></span>
<span><span class="va">multi.dois.parse</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span></span>
<span>  doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span><span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_zenodo"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># return SeuratObject</span></span>
<span><span class="va">sinle.doi.parse.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span></span>
<span>  doi <span class="op">=</span> <span class="st">"10.5281/zenodo.8011282"</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_zenodo"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>return.seu = T</code>: load downloaded objects to <code>Seurat</code>.</li>
</ul>
<p><a href="#processed-object">dissect and extract the <code>RData</code> files</a></p>
<hr>
</div>
<div class="section level3">
<h3 id="cellxgene-processed-object">
<a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> (processed object)<a class="anchor" aria-label="anchor" href="#cellxgene-processed-object"></a>
</h3>
<div class="section level4">
<h4 id="given-dataset-2">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset-2"></a>
</h4>
<p><a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> does not support downloading <code>SeuratObject</code> in <a href="https://cellxgene.cziscience.com/docs/03__Download%20Published%20Data" class="external-link">versions after 2025</a>. <strong>The following code can only download <code>h5ad</code> files</strong>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download h5ad files</span></span>
<span><span class="va">cellxgene.given.h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://cellxgene.cziscience.com/collections/77f9d7e9-5675-49c3-abed-ce02f39eef1b"</span>, <span class="co"># collection</span></span>
<span>    <span class="st">"https://cellxgene.cziscience.com/e/e12eb8a9-5e8b-4b59-90c8-77d29a811c00.cxg/"</span> <span class="co"># dataset</span></span>
<span>  <span class="op">)</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_cellxgene"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-samples-based-on-metadata-2">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata-2"></a>
</h4>
<p>We have downloaded all the <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> datasets in <strong>May 2025</strong> and stored in <a href="https://github.com/showteeth/GEfetch2R/blob/main/man/benchmark/all.cellxgene.datasets.rds" class="external-link">all.cellxgene.datasets.rds</a>. The <code>all.cellxgene.datasets.rds</code> contains the <code>SeuratObject</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the datasets with SeuratObject</span></span>
<span><span class="co"># wget https://github.com/showteeth/GEfetch2R/raw/ff2f19f3b557f90fce5f8bf2f8662cebdfd04298/man/benchmark/all.cellxgene.datasets.rds</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"all.cellxgene.datasets.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># summarise attributes</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span>, <span class="st">"disease"</span><span class="op">)</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"CELLxGENE"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># use cellxgene.census</span></span>
<span><span class="co"># StatDBAttribute(filter = c("disease", "tissue", "cell_type"), database = "CELLxGENE", use.census = TRUE, organism = "homo_sapiens")</span></span>
<span></span>
<span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">human.10x.cellxgene.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractCELLxGENEMeta.html">ExtractCELLxGENEMeta</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset</span></span>
<span><span class="va">cellxgene.down.meta</span> <span class="op">&lt;-</span> <span class="va">human.10x.cellxgene.meta</span><span class="op">[</span><span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">==</span> <span class="st">"oligodendrocyte"</span> <span class="op">&amp;</span></span>
<span>                                                  <span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"entorhinal cortex"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># download objects</span></span>
<span><span class="va">cellxgene.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_cellxgene"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># retuen SeuratObject</span></span>
<span><span class="va">cellxgene.down.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  obs.value.filter <span class="op">=</span> <span class="st">"cell_type == 'oligodendrocyte' &amp; disease == 'Alzheimer disease'"</span>,</span>
<span>  obs.keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"disease"</span>, <span class="st">"sex"</span>, <span class="st">"suspension_type"</span>, <span class="st">"development_stage"</span><span class="op">)</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_cellxgene"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="human-cell-atlas-processed-object">
<a href="https://explore.data.humancellatlas.org/projects" class="external-link">Human Cell Atlas</a> (processed object)<a class="anchor" aria-label="anchor" href="#human-cell-atlas-processed-object"></a>
</h3>
<div class="section level4">
<h4 id="given-dataset-3">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset-3"></a>
</h4>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects</span></span>
<span><span class="va">hca.given.download</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span></span>
<span>  link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://explore.data.humancellatlas.org/projects/902dc043-7091-445c-9442-d72e163b9879"</span>,</span>
<span>    <span class="st">"https://explore.data.humancellatlas.org/projects/cdabcf0b-7602-4abf-9afb-3b410e545703"</span></span>
<span>  <span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>, </span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/download_hca"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><a href="#processed-object">dissect and extract the <code>RData</code> files</a></p>
<hr>
</div>
<div class="section level4">
<h4 id="filter-samples-based-on-metadata-3">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata-3"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.hca.projects</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ShowHCAProjects.html">ShowHCAProjects</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># summarise attributes</span></span>
<span><span class="fu"><a href="reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">all.hca.projects</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"HCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter metadata</span></span>
<span><span class="va">hca.human.10x.projects</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExtractHCAMeta.html">ExtractHCAMeta</a></span><span class="op">(</span></span>
<span>  all.projects.df <span class="op">=</span> <span class="va">all.hca.projects</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># small test</span></span>
<span><span class="va">hca.human.10x.down</span> <span class="op">=</span> <span class="fu"><a href="reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hca.human.10x.projects</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, </span>
<span>                              out.folder <span class="op">=</span> <span class="st">"/path/to/download_hca"</span>, </span>
<span>                              file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h5ad"</span>, <span class="st">"rds"</span><span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span><span class="op">)</span></span></code></pre></div>
<p><a href="#processed-object">dissect and extract the <code>RData</code> files</a></p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="format-conversion">Format conversion<a class="anchor" aria-label="anchor" href="#format-conversion"></a>
</h3>
<p>There are many tools have been developed to process scRNA-seq data, such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">Scanpy</a>, <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a> and <a href="http://cole-trapnell-lab.github.io/monocle-release/" class="external-link">Monocle</a>. These tools have their own object formats, such as <code>Anndata</code> of <code>Scanpy</code>, <code>SeuratObject</code> of <code>Seurat</code>, <code>SingleCellExperiment</code> of <code>scran</code> and <code>CellDataSet</code>/<code>cell_data_set</code> of <code>Monocle2</code>/<code>Monocle3</code>. There are also some file formats designed for large omics datasets, such as <a href="http://loompy.org/" class="external-link">loom</a>. To perform a comprehensive scRNA-seq data analysis, we usually need to combine multiple tools, which means we need to perform object format conversion frequently. To facilitate user analysis of scRNA-seq data, <code>GEfetch2R</code> benchmarked the format conversion tools (<code>Anndata</code> -&gt; <code>SeuratObject</code>, <code>SeuratObject</code> to <code>Anndata</code>, <code>Anndata</code> to <code>SingleCellExperiment</code>, <code>SingleCellExperiment</code> to <code>Anndata</code>), and provides multiple functions to perform format conversion.</p>
<hr>
<div class="section level4">
<h4 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h4>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># pbmc_small</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span> <span class="co"># seger</span></span></code></pre></div>
<p><code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># object</span></span>
<span><span class="va">pbmc_small</span></span></code></pre></div>
<p><code>SingleCellExperiment</code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html" class="external-link">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="convert-seuratobject-to-other-objects">Convert SeuratObject to other objects<a class="anchor" aria-label="anchor" href="#convert-seuratobject-to-other-objects"></a>
</h4>
<p>Here, we will convert SeuratObject to <code>SingleCellExperiment</code>, <code>CellDataSet</code>/<code>cell_data_set</code>, <code>Anndata</code>, <code>loom</code>.</p>
<div class="section level5">
<h5 id="seuratobject-to-singlecellexperiment">SeuratObject to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#seuratobject-to-singlecellexperiment"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>Seurat</code>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"SCE"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-celldatasetcell_data_set">SeuratObject to CellDataSet/cell_data_set<a class="anchor" aria-label="anchor" href="#seuratobject-to-celldatasetcell_data_set"></a>
</h5>
<p>To <code>CellDataSet</code> (The conversion is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("monocle") # reuqire monocle</span></span>
<span><span class="va">cds.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, to <span class="op">=</span> <span class="st">"CellDataSet"</span><span class="op">)</span></span></code></pre></div>
<p>To <code>cell_data_set</code> (The conversion is performed with functions implemented in <code>SeuratWrappers</code>):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github('cole-trapnell-lab/monocle3') # reuqire monocle3</span></span>
<span><span class="va">cds3.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"cell_data_set"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-anndata">SeuratObject to AnnData<a class="anchor" aria-label="anchor" href="#seuratobject-to-anndata"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>SeuratObject</code> to <code>Anndata</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept and usability</li>
<li>
<code>sceasy</code> has best performance in running time and disk usage.</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SeuratDisk</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, </span>
<span>       assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># sceasy</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, </span>
<span>       assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="fu"><a href="reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>       out.folder <span class="op">=</span> <span class="st">"out.folder"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="seuratobject-to-loom">SeuratObject to loom<a class="anchor" aria-label="anchor" href="#seuratobject-to-loom"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>SeuratDisk</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"pbmc_small_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"loom"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="convert-other-objects-to-seuratobject">Convert other objects to SeuratObject<a class="anchor" aria-label="anchor" href="#convert-other-objects-to-seuratobject"></a>
</h4>
<div class="section level5">
<h5 id="singlecellexperiment-to-seuratobject">SingleCellExperiment to SeuratObject<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-seuratobject"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>Seurat</code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sce.obj</span>, from <span class="op">=</span> <span class="st">"SCE"</span>, </span>
<span>                            count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>                            assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="celldatasetcell_data_set-to-seuratobject">CellDataSet/cell_data_set to SeuratObject<a class="anchor" aria-label="anchor" href="#celldatasetcell_data_set-to-seuratobject"></a>
</h5>
<p><code>CellDataSet</code> to <code>SeuratObject</code> (The conversion is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds.obj</span>, from <span class="op">=</span> <span class="st">"CellDataSet"</span>, </span>
<span>                            count.assay <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<p><code>cell_data_set</code> to <code>SeuratObject</code> (The conversion is performed with functions implemented in <code>Seurat</code>): <code>{r cds2seu2, eval=FALSE} seu.obj.cds3 &lt;- ImportSeurat(obj = cds3.obj, from = "cell_data_set",                               count.assay = "counts", data.assay = "logcounts",                              assay = "RNA")</code></p>
<hr>
</div>
<div class="section level5">
<h5 id="anndata-to-seuratobject">AnnData to SeuratObject<a class="anchor" aria-label="anchor" href="#anndata-to-seuratobject"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>AnnData</code> to <code>SeuratObject</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept (<strong><code>GEfetch2R</code> integrates <code>scDIOR</code> and <code>SeuratDisk</code> to achieve the best performance in information kept</strong>)</li>
<li>
<code>schard</code> is the best method in terms of usability</li>
<li>
<code>schard</code> and <code>sceasy</code> have comparable performance when cell number below 200k, but <code>sceasy</code> has better performance in scalability</li>
<li>
<code>sceasy</code> has better performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SeuratDisk</span></span>
<span><span class="va">ann.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># sceasy</span></span>
<span><span class="va">ann.sceasy</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>,</span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">ann.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">ann.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, </span>
<span>                     method <span class="op">=</span> <span class="st">"schard"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, use.raw <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># SeuratDisk+scDIOR</span></span>
<span><span class="va">ann.seuscdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"SeuratDisk+scDIOR"</span>, </span>
<span>                        assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="loom-to-seuratobject">loom to SeuratObject<a class="anchor" aria-label="anchor" href="#loom-to-seuratobject"></a>
</h5>
<p>The conversion is performed with functions implemented in <code>SeuratDisk</code> and <code>Seurat</code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loom will lose reduction</span></span>
<span><span class="va">seu.obj.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>loom.file <span class="op">=</span> <span class="va">loom.file</span>, from <span class="op">=</span> <span class="st">"loom"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="conversion-between-singlecellexperiment-and-anndata">Conversion between SingleCellExperiment and AnnData<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-anndata"></a>
</h4>
<div class="section level5">
<h5 id="singlecellexperiment-to-anndata">SingleCellExperiment to AnnData<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-anndata"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>SingleCellExperiment</code> to <code>AnnData</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of information kept and running time</li>
<li>
<code>scDIOR</code> is the best method in terms of usability and disk usage</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sceasy</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"benchmark"</span>, </span>
<span>       slot <span class="op">=</span> <span class="st">"rawcounts"</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span>       </span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">seger.scdior</span> <span class="op">&lt;-</span> <span class="va">seger</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co"># scDIOR does not support varm in rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">seger.scdior</span><span class="op">)</span><span class="op">$</span><span class="va">varm</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger.scdior</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, out.folder <span class="op">=</span> <span class="st">"benchmark"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># zellkonverter</span></span>
<span><span class="fu"><a href="reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"zellkonverter"</span>, </span>
<span>       out.folder <span class="op">=</span> <span class="st">"benchmark"</span>, slot <span class="op">=</span> <span class="st">"rawcounts"</span>, </span>
<span>       conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="anndata-to-singlecellexperiment">AnnData to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#anndata-to-singlecellexperiment"></a>
</h5>
<p>There are multiple tools available for format conversion from <code>AnnData</code> to <code>SingleCellExperiment</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of information kept</li>
<li>
<code>schard</code> is the best method in terms of usability and running time</li>
<li>
<code>schard</code> and <code>scDIOR</code> have comparable performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scDIOR</span></span>
<span><span class="va">sce.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># zellkonverter</span></span>
<span><span class="va">sce.zell</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"zellkonverter"</span>,</span>
<span>                   slot <span class="op">=</span> <span class="st">"scale.data"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"/path/to/conda"</span><span class="op">)</span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">sce.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span>anndata.file <span class="op">=</span> <span class="st">"pbmc3k.h5ad"</span>, </span>
<span>                     method <span class="op">=</span> <span class="st">"schard"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="conversion-between-singlecellexperiment-and-loom">Conversion between SingleCellExperiment and loom<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-loom"></a>
</h4>
<p>The conversion is performed with functions implemented in <code>LoomExperiment</code>.</p>
<div class="section level5">
<h5 id="singlecellexperiment-to-loom">SingleCellExperiment to loom<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-loom"></a>
</h5>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.loom first</span></span>
<span><span class="va">seger.loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"seger_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"loom"</span>, sce <span class="op">=</span> <span class="va">seger</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level5">
<h5 id="loom-to-singlecellexperiment">loom to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#loom-to-singlecellexperiment"></a>
</h5>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"loom"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>For any question, feature request or bug report please write an email to <code>songyb0519@gmail.com</code>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the <code>GEfetch2R</code> project is released with a <a href="https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
<p><br></p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/showteeth/GEfetch2R/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/showteeth/GEfetch2R/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GEfetch2R</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yabing Song <br><small class="roles"> Author, maintainer </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
