<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scfetch">
<title>DownloadMatrices • scfetch</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadMatrices">
<meta property="og:description" content="scfetch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scfetch</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/DownloadRaw.html">Download fastq/bam files</a>
    <a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrices</a>
    <a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed objects</a>
    <a class="dropdown-item" href="../articles/ObjectConversion.html">Convert formats between objects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/scfetch">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>DownloadMatrices</h1>
            
            <h4 data-toc-skip class="date">2023-07-23</h4>
      
      
      <div class="d-none name"><code>DownloadMatrices.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>scfetch</code> provides functions for users to download
<strong>count matrices</strong> and <strong>annotations</strong>
(e.g. cell type annotation and composition) from GEO and some
single-cell databases (e.g. <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> and <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>).
<code>scfetch</code> also supports loading the downloaded data to
<code>Seurat</code>.</p>
<p>Until now, the public resources supported and the returned
results:</p>
<table class="table">
<thead><tr>
<th>
Resources
</th>
<th>
URL
</th>
<th>
Download Type
</th>
<th>
Returned results
</th>
</tr></thead>
<tbody>
<tr>
<td>
GEO
</td>
<td>
<a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a>
</td>
<td>
count matrix
</td>
<td>
SeuratObject or count matrix for bulk RNA-seq
</td>
</tr>
<tr>
<td>
PanglaoDB
</td>
<td>
<a href="https://panglaodb.se/index.html" class="external-link uri">https://panglaodb.se/index.html</a>
</td>
<td>
count matrix
</td>
<td>
SeuratObject
</td>
</tr>
<tr>
<td>
UCSC Cell Browser
</td>
<td>
<a href="https://cells.ucsc.edu/" class="external-link uri">https://cells.ucsc.edu/</a>
</td>
<td>
count matrix
</td>
<td>
SeuratObject
</td>
</tr>
</tbody>
</table>
<hr>
</div>
<div class="section level2">
<h2 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is
an international public repository that archives and freely distributes
microarray, next-generation sequencing, and other forms of
high-throughput functional genomics data submitted by the research
community.</a> It provides a very convenient way for users to explore
and select interested scRNA-seq datasets.</p>
<div class="section level3">
<h3 id="extract-metadata">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata"></a>
</h3>
<p><code>scfetch</code> provides <code>ExtractGEOMeta</code> to extract
sample metadata, including sample title, source name/tissue,
description, cell type, treatment, paper title, paper abstract,
organism, protocol, data processing methods, et al.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scfetch</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Setting options('download.file.method.GEOquery'='auto')</span></span></code></pre>
<pre><code><span><span class="co">## Setting options('GEOquery.inmemory.gpl'=FALSE)</span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'SeuratDisk':</span></span>
<span><span class="co">##   method            from  </span></span>
<span><span class="co">##   as.sparse.H5Group Seurat</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract metadata of specified platform</span></span>
<span><span class="va">GSE200257.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="st">"GPL24676"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Found 1 file(s)</span></span></code></pre>
<pre><code><span><span class="co">## GSE200257_series_matrix.txt.gz</span></span></code></pre>
<pre><code><span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">9</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (9): ID_REF, GSM6025648, GSM6025649, GSM6025650, GSM6025651, GSM6025652,...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">## File stored at: </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## /var/folders/_4/k4qmvf7s2gx_6789px8n_sxh0000gn/T//RtmpeDiEtS/GPL24676.soft</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span><span class="op">=</span><span class="fl">131072</span><span class="op">*</span><span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata of all platforms</span></span>
<span><span class="va">GSE94820.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Found 2 file(s)</span></span>
<span><span class="co">## GSE94820-GPL15520_series_matrix.txt.gz</span></span>
<span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">651</span><span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (651): ID_REF, GSM2485115, GSM2485116, GSM2485117, GSM2485118, GSM248511...</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.File stored at: </span></span>
<span><span class="co">## /var/folders/_4/k4qmvf7s2gx_6789px8n_sxh0000gn/T//RtmpeDiEtS/GPL15520.soft</span></span>
<span><span class="co">## GSE94820-GPL16791_series_matrix.txt.gz</span></span>
<span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">1735</span><span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (1735): ID_REF, GSM2483594, GSM2483595, GSM2483596, GSM2483597, GSM24835...</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.File stored at: </span></span>
<span><span class="co">## /var/folders/_4/k4qmvf7s2gx_6789px8n_sxh0000gn/T//RtmpeDiEtS/GPL16791.soft</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE94820.meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Platform                  title geo_accession source_name_ch1</span></span>
<span><span class="co">## 1 GPL15520  AXLSIGLEC6_Bin1_S1_S5    GSM2485115            PBMC</span></span>
<span><span class="co">## 2 GPL15520 AXLSIGLEC6_Bin1_S10_S2    GSM2485116            PBMC</span></span>
<span><span class="co">## 3 GPL15520 AXLSIGLEC6_Bin1_S11_S3    GSM2485117            PBMC</span></span>
<span><span class="co">## 4 GPL15520 AXLSIGLEC6_Bin1_S12_S4    GSM2485118            PBMC</span></span>
<span><span class="co">## 5 GPL15520 AXLSIGLEC6_Bin1_S13_S5    GSM2485119            PBMC</span></span>
<span><span class="co">## 6 GPL15520 AXLSIGLEC6_Bin1_S14_S6    GSM2485120            PBMC</span></span>
<span><span class="co">##               description sorted_gate_identity</span></span>
<span><span class="co">## 1 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">## 2 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">## 3 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">## 4 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">## 5 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">## 6 paired-end RNA-seq data      AXLSIGLEC6_Bin1</span></span>
<span><span class="co">##                                                                                             Title</span></span>
<span><span class="co">## 1 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">## 2 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">## 3 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">## 4 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">## 5 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">## 6 Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors</span></span>
<span><span class="co">##                                                 Type     Organism</span></span>
<span><span class="co">## 1 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">## 2 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">## 3 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">## 4 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">## 5 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">## 6 Expression profiling by high throughput sequencing Homo sapiens</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Abstract</span></span>
<span><span class="co">## 1 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">## 2 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">## 3 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">## 4 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">## 5 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">## 6 Peripheral blood mononuclear cells (PBMCs) were isolated from fresh blood using Ficoll-Paque density gradient centrifugation. Single-cell suspensions were stained with different antibody cocktails designed to enrich for particular immune cell populations, which were then single cell sorted in 96-well plates.  Single cell RNA-sequencing libraries were subsequently generated for 2422 single cells using Smart-Seq2 (Picelli et al., Nature Methods, 2014). Cells were sequenced at a depth of 1-2M reads/cell.</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Design</span></span>
<span><span class="co">## 1 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">## 2 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">## 3 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">## 4 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">## 5 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">## 6 The study was divided into two stages: (1) an exploratory phase, where 1140 single human blood dendritic cells and monocytes were profiled and 12 population samples; (2) a deep characterization phase, where an additional 1261 single cells and 9 population samples were profiled as part of follow-up studies.  A total of 2422 single cell and population samples were processed using Smart-Seq2 protocol (Picelli et al., Nature Methods, 2014), which allows for the generation of full-length single cell cDNA, and sequencing libraries were generated using Illumina Nextera XT DNA library preparation kit.\n\nPlease note that [1] the raw data have been submitted to dbGaP (which has controlled access mechanisms at: http://www.ncbi.nlm.nih.gov/gap; phs001294.v1.p1) due to potential privacy concerns. Please contact the submitter or dbGaP to request access to controlled-access datasets.\n\n[2] a few samples (38) were profiled but excluded from the processed data file since they were either a bulk sample (21) or excluded due to QC (17). Therefore, there are 1,140 and 1,244 data columns in two processed data files respectively, corresponding to total 2384 samples described in the records.</span></span>
<span><span class="co">##   SampleCount  Molecule</span></span>
<span><span class="co">## 1         650 polyA RNA</span></span>
<span><span class="co">## 2         650 polyA RNA</span></span>
<span><span class="co">## 3         650 polyA RNA</span></span>
<span><span class="co">## 4         650 polyA RNA</span></span>
<span><span class="co">## 5         650 polyA RNA</span></span>
<span><span class="co">## 6         650 polyA RNA</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                ExtractProtocol</span></span>
<span><span class="co">## 1 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">## 2 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">## 3 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">## 4 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">## 5 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">## 6 Enriched immune cell fractions isolated from healthy blood PBMCs were FACS sorted in 96-well plates (single cell sorted) containing lysis buffer (TCL together with 1% of 2-Mercaptoethanol).. Smart-seq2 (Picelli et al., Nature Methods, 2014). Full-length RNA-sequencing</span></span>
<span><span class="co">##   LibraryStrategy</span></span>
<span><span class="co">## 1         RNA-Seq</span></span>
<span><span class="co">## 2         RNA-Seq</span></span>
<span><span class="co">## 3         RNA-Seq</span></span>
<span><span class="co">## 4         RNA-Seq</span></span>
<span><span class="co">## 5         RNA-Seq</span></span>
<span><span class="co">## 6         RNA-Seq</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  DataProcessing</span></span>
<span><span class="co">## 1 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">## 2 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">## 3 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">## 4 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">## 5 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">## 6 Read were aligned to the UCSC hg19 transcriptomee using Bowtie v0.12.7. Expression levels were quantified using RSEM v1.2.1 (TPM values). UCSC genome table browser was used to map UCSC gene ID (kgID) and the gene name (geneSymbol) for all genes in hg19. If multiple UCSC gene IDs are assigned to the same geneSymbol, the TPM values of all kgIDs that share the same geneSymbol are summed. Genome_build: hg19. Supplementary_files_format_and_content: Text file with tab delimiters</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                  SupplementaryFile</span></span>
<span><span class="co">## 1 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">## 2 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">## 3 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">## 4 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">## 5 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">## 6 ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_DCnMono.discovery.set.submission.txt.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE94nnn/GSE94820/suppl/GSE94820_raw.expMatrix_deeper.characterization.set.submission.txt.gz</span></span>
<span><span class="co">##          Contact               PMID</span></span>
<span><span class="co">## 1 Nir,,Hacohen;  28428369, 30405621</span></span>
<span><span class="co">## 2 Nir,,Hacohen;  28428369, 30405621</span></span>
<span><span class="co">## 3 Nir,,Hacohen;  28428369, 30405621</span></span>
<span><span class="co">## 4 Nir,,Hacohen;  28428369, 30405621</span></span>
<span><span class="co">## 5 Nir,,Hacohen;  28428369, 30405621</span></span>
<span><span class="co">## 6 Nir,,Hacohen;  28428369, 30405621</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="download-matrix-and-load-to-seurat">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat"></a>
</h3>
<p>After manually check the extracted metadata, users can
<strong>download count matrix</strong> and <strong>load the count
matrix</strong> to Seurat with <code>ParseGEO</code>.</p>
<p>For count matrix, <code>ParseGEO</code> supports downloading the
matrix from <strong>supplementary files</strong> and extracting from
<code>ExpressionSet</code>, users can control the source by specifying
<code>down.supp</code> or detecting automatically (<code>ParseGEO</code>
will extract the count matrix from <code>ExpressionSet</code> first, if
the count matrix is NULL or contains non-integer values,
<code>ParseGEO</code> will download supplementary files). While the
supplementary files have two main types: single count matrix file
containing all cells and CellRanger-style outputs (barcode, matrix,
feature/gene), users are required to choose the type of supplementary
files with <code>supp.type</code>.</p>
<p>With the count matrix, <code>ParseGEO</code> will load the matrix to
Seurat automatically. If multiple samples available, users can choose to
merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for cellranger output</span></span>
<span><span class="va">GSE200257.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/dwonload_geo"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># for count matrix, no need to specify out.folder, download count matrix to tmp folder</span></span>
<span><span class="va">GSE94820.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span></code></pre></div>
<p>The structure of downloaded matrix for 10x:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> /Users/soyabean/Desktop/tmp/scdown/dwonload_geo</span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/Users/soyabean/Desktop/tmp/scdown/dwonload_geo</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025652_1</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025653_2</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025654_3</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## └── <span style="color: #0000BB; font-weight: bold;">GSM6025655_4</span></span></span>
<span><span class="co">##     ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">##     ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">##     └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## 5 directories, 12 files</span></span></code></pre>
<p><strong>For bulk RNA-seq</strong>, set
<code>data.type = "bulk"</code> in <code>ParseGEO</code>, this will
return count matrix.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="panglaodb">PanglaoDB<a class="anchor" aria-label="anchor" href="#panglaodb"></a>
</h2>
<p><a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> is a database
which contains scRNA-seq datasets from mouse and human. Up to now, it
contains <strong>5,586,348 cells</strong> from <strong>1368 datasets
(1063 from Mus musculus and 305 from Homo sapiens)</strong>. It has well
organized metadata for every dataset, including tissue, protocol,
species, number of cells and cell type annotation (computationally
identified). Daniel Osorio has developed <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a> to access
<a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> data, the
functions of <code>scfetch</code> here are based on <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a>.</p>
<p>Since <a href="https://panglaodb.se/about.html" class="external-link">PanglaoDB</a> is no
longer maintained, <code>scfetch</code> has cached all metadata and cell
type composition and use these cached data by default to accelerate,
users can access the cached data with <code>PanglaoDBMeta</code> (all
metadata) and <code>PanglaoDBComposition</code> (all cell type
composition).</p>
<div class="section level3">
<h3 id="summary-attributes">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes"></a>
</h3>
<p><code>scfetch</code> provides <code>StatDBAttribute</code> to summary
attributes of <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use cached metadata</span></span>
<span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $species</span></span>
<span><span class="co">##          Value  Num     Key</span></span>
<span><span class="co">## 1 Mus musculus 1063 species</span></span>
<span><span class="co">## 2 Homo sapiens  305 species</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $protocol</span></span>
<span><span class="co">##           Value  Num      Key</span></span>
<span><span class="co">## 1  10x chromium 1046 protocol</span></span>
<span><span class="co">## 2      drop-seq  204 protocol</span></span>
<span><span class="co">## 3 microwell-seq   74 protocol</span></span>
<span><span class="co">## 4    Smart-seq2   26 protocol</span></span>
<span><span class="co">## 5   C1 Fluidigm   16 protocol</span></span>
<span><span class="co">## 6       CEL-seq    1 protocol</span></span>
<span><span class="co">## 7       inDrops    1 protocol</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-metadata-1">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-1"></a>
</h3>
<p><code>scfetch</code> provides <code>ExtractPanglaoDBMeta</code> to
select interested datasets with specified <strong>species</strong>,
<strong>protocol</strong>, <strong>tissue</strong> and <strong>cell
number</strong> (The available values of these attributes can be
obtained with <code>StatDBAttribute</code>). User can also choose to
whether to add cell type annotation to every dataset with
<code>show.cell.type</code>.</p>
<p><code>scfetch</code> uses cached metadata and cell type composition
by default, users can change this by setting
<code>local.data = FALSE</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>, show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         SRA        SRS                             Tissue     Protocol</span></span>
<span><span class="co">## 1 SRA550660 SRS2089635 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 2 SRA550660 SRS2089636 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 3 SRA550660 SRS2089638 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 4 SRA605365 SRS2492922            Nasal airway epithelium 10x chromium</span></span>
<span><span class="co">## 5 SRA608611 SRS2517316                   Lung progenitors 10x chromium</span></span>
<span><span class="co">## 6 SRA608353 SRS2517519           Hepatocellular carcinoma 10x chromium</span></span>
<span><span class="co">##        Species Cells</span></span>
<span><span class="co">## 1 Homo sapiens  1860</span></span>
<span><span class="co">## 2 Homo sapiens  1580</span></span>
<span><span class="co">## 3 Homo sapiens  1818</span></span>
<span><span class="co">## 4 Homo sapiens  1932</span></span>
<span><span class="co">## 5 Homo sapiens  1077</span></span>
<span><span class="co">## 6 Homo sapiens  1230</span></span>
<span><span class="co">##                                                                      CellType</span></span>
<span><span class="co">## 1                                                           Unknown, NK cells</span></span>
<span><span class="co">## 2                              Unknown, T cells, Plasmacytoid dendritic cells</span></span>
<span><span class="co">## 3 Unknown, Gamma delta T cells, Dendritic cells, Plasmacytoid dendritic cells</span></span>
<span><span class="co">## 4       Luminal epithelial cells, Basal cells, Keratinocytes, Ependymal cells</span></span>
<span><span class="co">## 5                                           Unknown, Hepatocytes, Basal cells</span></span>
<span><span class="co">## 6                                        Unknown, Hepatocytes, Foveolar cells</span></span>
<span><span class="co">##   CellNum</span></span>
<span><span class="co">## 1    1860</span></span>
<span><span class="co">## 2    1580</span></span>
<span><span class="co">## 3    1818</span></span>
<span><span class="co">## 4    1932</span></span>
<span><span class="co">## 5    1077</span></span>
<span><span class="co">## 6    1230</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-cell-type-composition">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition"></a>
</h3>
<p><code>scfetch</code> provides
<code>ExtractPanglaoDBComposition</code> to extract cell type annotation
and composition (use cached data by default to accelerate, users can
change this by setting <code>local.data = FALSE</code>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.composition</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           SRA        SRS                        Tissue     Protocol</span></span>
<span><span class="co">## 1.1 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.2 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.3 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.4 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.5 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.6 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">##          Species Cluster Cells Cell Type</span></span>
<span><span class="co">## 1.1 Homo sapiens       0  1572   Unknown</span></span>
<span><span class="co">## 1.2 Homo sapiens       1   563   Unknown</span></span>
<span><span class="co">## 1.3 Homo sapiens       2   280   Unknown</span></span>
<span><span class="co">## 1.4 Homo sapiens       3   270   Unknown</span></span>
<span><span class="co">## 1.5 Homo sapiens       4   220   Unknown</span></span>
<span><span class="co">## 1.6 Homo sapiens       5   192   Unknown</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="download-matrix-and-load-to-seurat-1">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat-1"></a>
</h3>
<p>After manually check the extracted metadata, <code>scfetch</code>
provides <code>ParsePanglaoDB</code> to <strong>download count
matrix</strong> and <strong>load the count matrix</strong> to Seurat.
With available cell type annotation, uses can filter datasets without
specified cell type with <code>cell.type</code>. Users can also
include/exclude cells expressing specified genes with
<code>include.gene</code>/<code>exclude.gene</code>.</p>
<p>With the count matrix, <code>ParsePanglaoDB</code> will load the
matrix to Seurat automatically. If multiple datasets available, users
can choose to merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="ucsc-cell-browser">UCSC Cell Browser<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser"></a>
</h2>
<p>The <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a> is a
web-based tool that allows scientists to interactively visualize
scRNA-seq datasets. It contains <strong>1040 single cell
datasets</strong> from <strong>17 different species</strong>. And, it is
<strong>organized with the hierarchical structure</strong>, which can
help users quickly locate the datasets they are interested in.</p>
<div class="section level3">
<h3 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h3>
<p><code>scfetch</code> provides <code>ShowCBDatasets</code> to show all
available datasets. Due to the large number of datasets,
<code>ShowCBDatasets</code> enables users to perform <em>lazy load</em>
of dataset json files instead of downloading the json files online
(time-consuming!!!). This <em>lazy load</em> requires users to provide
<code>json.folder</code> to save json files and set
<code>lazy = TRUE</code> (for the first time of run,
<code>ShowCBDatasets</code> will download current json files to
<code>json.folder</code>, for next time of run, with
<code>lazy = TRUE</code>, <code>ShowCBDatasets</code> will load the
downloaded json files from <code>json.folder</code>.). And,
<code>ShowCBDatasets</code> supports updating the local datasets with
<code>update = TRUE</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first time run, the json files are stored under json.folder</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "/Users/soyabean/Desktop/tmp/scdown/cell_browser/json", update = TRUE)</span></span>
<span></span>
<span><span class="co"># second time run, load the downloaded json files</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/cell_browser/json"</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Lazy mode is on, load downloaded json from /Users/soyabean/Desktop/tmp/scdown/cell_browser/json</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                               name                 shortLabel</span></span>
<span><span class="co">## 1  adult-brain-vasc/am-endothelial Vasculature of Adult Brain</span></span>
<span><span class="co">## 2       adult-brain-vasc/am-immune Vasculature of Adult Brain</span></span>
<span><span class="co">## 3 adult-brain-vasc/am-perivascular Vasculature of Adult Brain</span></span>
<span><span class="co">## 4     adult-brain-vasc/endothelial Vasculature of Adult Brain</span></span>
<span><span class="co">## 5    adult-brain-vasc/perivascular Vasculature of Adult Brain</span></span>
<span><span class="co">## 6                     adult-testis               Adult Testis</span></span>
<span><span class="co">##                                              subLabel tags body_parts</span></span>
<span><span class="co">## 1  Endothelial - Arteriovenous Malformation &amp; Control           brain</span></span>
<span><span class="co">## 2       Immune - Arteriovenous Malformation &amp; Control           brain</span></span>
<span><span class="co">## 3 Perivascular - Arteriovenous Malformation &amp; Control           brain</span></span>
<span><span class="co">## 4                             Adult Brain Endothelial           brain</span></span>
<span><span class="co">## 5                            Adult Brain Perivascular           brain</span></span>
<span><span class="co">## 6                                                &lt;NA&gt;          testis</span></span>
<span><span class="co">##         diseases                 organisms projects life_stages domains sources</span></span>
<span><span class="co">## 1 Healthy|parent Human (H. sapiens)|parent                           NA      NA</span></span>
<span><span class="co">## 2 Healthy|parent Human (H. sapiens)|parent                           NA      NA</span></span>
<span><span class="co">## 3 Healthy|parent Human (H. sapiens)|parent                           NA      NA</span></span>
<span><span class="co">## 4 Healthy|parent Human (H. sapiens)|parent                           NA      NA</span></span>
<span><span class="co">## 5 Healthy|parent Human (H. sapiens)|parent                           NA      NA</span></span>
<span><span class="co">## 6        Healthy        Human (H. sapiens)                                     </span></span>
<span><span class="co">##   sampleCount assays            matrix         barcode         feature</span></span>
<span><span class="co">## 1        9541        exprMatrix.tsv.gz                                </span></span>
<span><span class="co">## 2       55255            matrix.mtx.gz barcodes.tsv.gz features.tsv.gz</span></span>
<span><span class="co">## 3      101317            matrix.mtx.gz barcodes.tsv.gz features.tsv.gz</span></span>
<span><span class="co">## 4        5018        exprMatrix.tsv.gz                                </span></span>
<span><span class="co">## 5       49553        exprMatrix.tsv.gz                                </span></span>
<span><span class="co">## 6        6199        exprMatrix.tsv.gz                                </span></span>
<span><span class="co">##   matrixType                                               title</span></span>
<span><span class="co">## 1     matrix  Arteriovenous malformation and control endothelial</span></span>
<span><span class="co">## 2        10x       Arteriovenous malformation and control immune</span></span>
<span><span class="co">## 3        10x Arteriovenous malformation and control perivascular</span></span>
<span><span class="co">## 4     matrix                  Adult Brain Endothelial Cell Types</span></span>
<span><span class="co">## 5     matrix                 Adult Brain Perivascular Cell Types</span></span>
<span><span class="co">## 6     matrix   The adult human testis transcriptional cell atlas</span></span>
<span><span class="co">##                                                                          paper</span></span>
<span><span class="co">## 1                                                                             </span></span>
<span><span class="co">## 2                                                                             </span></span>
<span><span class="co">## 3                                                                             </span></span>
<span><span class="co">## 4                                                                             </span></span>
<span><span class="co">## 5                                                                             </span></span>
<span><span class="co">## 6 https://www.nature.com/articles/s41422-018-0099-2 Guo et al. 2018. Cell Res.</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   abstract</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \nArteriovenous malformation and control endothelial cell types coembedded with\ntheir respective cell types: nidus, arterial, venous, and venules.\n</span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \nArteriovenous malformation and control immune cell types coembedding of myeloid\nand lymphoid cells.\n</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \nArteriovenous malformation and control perivascular cell types coembedded with\ntheir respective cell types: smooth muscle cells, pericytes, fibroblasts, and\nfibromyocytes.\n</span></span>
<span><span class="co">## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \nAdult brain endothelial cell types broken down into four broad cell types:\ncapillary, arterial, venous, and venules.\n</span></span>
<span><span class="co">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \nAdult brain perivascular cell types broken down into four broad cell\ntypes: smooth muscle cells, pericytes, fibroblasts, and fibromyocytes.\n</span></span>
<span><span class="co">## 6 \n&lt;p&gt;\nFrom &lt;a href="https://www.nature.com/articles/s41422-018-0099-2"\ntarget="_blank"&gt;Guo et al&lt;/a&gt;:\n&lt;/p&gt;\n\n&lt;p&gt;\nHuman adult spermatogenesis balances spermatogonial stem cell (SSC)\nself-renewal and differentiation, alongside complex germ cell-niche\ninteractions, to ensure long-term fertility and faithful genome propagation.\nHere, we performed single-cell RNA sequencing of ~6500 testicular cells from\nyoung adults. We found five niche/somatic cell types (Leydig, myoid, Sertoli,\nendothelial, macrophage), and observed germline-niche interactions and key\nhuman-mouse differences. Spermatogenesis, including meiosis, was reconstructed\ncomputationally, revealing sequential coding, non-coding, and repeat-element\ntranscriptional signatures. Interestingly, we identified five discrete\ntranscriptional/developmental spermatogonial states, including a novel early\nSSC state, termed State 0. Epigenetic features and nascent transcription\nanalyses suggested developmental plasticity within spermatogonial States. To\nunderstand the origin of State 0, we profiled testicular cells from infants,\nand identified distinct similarities between adult State 0 and infant SSCs.\nOverall, our datasets describe key transcriptional and epigenetic signatures of\nthe normal adult human testis, and provide new insights into germ cell\ndevelopmental transitions and plasticity.\n&lt;/p&gt;\n</span></span>
<span><span class="co">##   unit                    coords</span></span>
<span><span class="co">## 1      Seurat_umap.coords.tsv.gz</span></span>
<span><span class="co">## 2      Seurat_umap.coords.tsv.gz</span></span>
<span><span class="co">## 3      Seurat_umap.coords.tsv.gz</span></span>
<span><span class="co">## 4      Seurat_umap.coords.tsv.gz</span></span>
<span><span class="co">## 5             UMAP.coords.tsv.gz</span></span>
<span><span class="co">## 6          umap_hm.coords.tsv.gz</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  methods</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       </span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       </span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       </span></span>
<span><span class="co">## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       </span></span>
<span><span class="co">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       </span></span>
<span><span class="co">## 6 \n&lt;p&gt;\nDataset was imported from the h5ad file available on the &lt;a href="https://www.covid19cellatlas.org/"\ntarget="_blank"&gt; COVID-19 Cell Atlas website&lt;/a&gt; using the UCSC Cell Browser utility\n&lt;code&gt;cbImportScanpy&lt;/code&gt;\n&lt;/p&gt;\n\n&lt;section&gt;Single cell RNA-seq performance, library preparation and sequencing&lt;/section&gt;\n&lt;p&gt;\nscRNA-Seq was performed using the 10× Genomics system. Briefly, each experiment\ncaptured ~1500 single cells, in order to obtain ~0.8% multiplex rate. Cells\nwere diluted following manufacturer recommendations, and mixed with 33.8 µL of\ntotal mixed buffer before being loaded into 10× Chromium Controller using\nChromium Single Cell 3’ v2 reagents. Each sequencing library was prepared\nfollowing the manufacturer’s instructions, with 13 cycles used for cDNA\namplification. Then ~100 ng of cDNA were used for library amplification by 12\ncycles. The resulting libraries were then sequenced on a 26 × 100 cycle\npaired-end run on an Illumina HiSeq 2500 instrument.\n&lt;/p&gt;\n\n&lt;section&gt;Process of single cell RNA-seq data&lt;/section&gt;\n&lt;p&gt;\nRaw sequencing data were demultiplexed using the mkfastq application (Cell\nRanger v1.2.1). Three types of fastq files were generated: I1 contains 8 bp\nsample index; R1 contains 26 bp (10 bp cell-BC + 16 bp UMI) index and R2\ncontains 100 bp cDNA sequence. Fastq files were then run with the cellranger\ncount application (Cell Ranger v1.2.1) using default settings, to perform\nalignment (using STAR v2.5.4a), filtering and cellular barcode and UMI\ncounting. The UMI count tables of each cellular barcode were used for further\nanalysis.\n&lt;/p&gt;\n\n&lt;section&gt;Cell type identification and clustering analysis using Seurat program&lt;/section&gt;\n&lt;p&gt;\nThe Seurat program (http://satijalab.org/seurat/, R package, v.2.0.0) was\nfirstly applied for analysis of RNA-Sequencing data. To start with, UMI count\ntables from each replicates and donors were loaded into R using Read10X\nfunction, and Seurat objects were built from each experiment. Each experiment\nwas filtered and normalized with default settings. Specifically, cells were\nretained only when they had greater than 500 genes expressed, and less than 20%\nreads mapped to mitochondrial genome. We first ran t-SNE and the clustering\nanalysis for each replicate, which resulted in similar t-SNE map. Next, to\nminimize variation between technical replicates, we normalized and combined\ntechnical replicates from the same donor using the 10× Genomics built-in\napplication from Cell Ranger “cellrange aggr”. Data matrices from different\ndonors were then loaded into R using Seurat. Next, cells were normalized to the\ntotal UMI read count as well as mitochondrial read percentage, as instructed in\nthe manufacturer’s manual (http://satijalab.org/seurat/). Seurat objects\n(matrices from different donors) were then combined using RunCCA function.\nt-SNE and clustering analyses were then performed on the combined dataset using\nthe top 5000 highly variable genes and PCs 1–15, which showed most significant\np-values. Given the low number of Sertoli cells (underrepresented due to size\nfiltering), the initial clustering analysis did not identify them as a separate\ncluster. We performed deeper clustering of somatic cells, identified the\nSertoli cell cluster, and projected it back to the overall clusters, which\nresulted in 13 discrete cell clusters. Correlation of different replicates was\ncalculated based on average expression (normalized UMIs by Seurat) in each\nexperiment.\n&lt;/p&gt;\n\n&lt;p&gt;\nSee the source paper &lt;a href="https://www.nature.com/articles/s41422-018-0099-2"\ntarget="_blank"&gt;Guo et al. 2018. Cell Res.&lt;/a&gt; for more details.\n&lt;/p&gt;\n</span></span>
<span><span class="co">##         geo</span></span>
<span><span class="co">## 1          </span></span>
<span><span class="co">## 2          </span></span>
<span><span class="co">## 3          </span></span>
<span><span class="co">## 4          </span></span>
<span><span class="co">## 5          </span></span>
<span><span class="co">## 6 GSE120508</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># always read online</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)</span></span></code></pre></div>
<p>The number of datasets and all available species:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the number of datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1040</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># available species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\|parent"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">ucsc.cb.samples</span><span class="op">$</span><span class="va">organisms</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Human (H. sapiens)"          "Mouse (M. musculus)"        </span></span>
<span><span class="co">##  [3] "Rhesus macaque (M. mulatta)" "Dog (C. familiaris)"        </span></span>
<span><span class="co">##  [5] "Human (H. Sapiens)"          "C. intestinalis"            </span></span>
<span><span class="co">##  [7] "C. robusta"                  "Zebrafish (D. rerio)"       </span></span>
<span><span class="co">##  [9] "Fruit fly (D. melanogaster)" "Hydra vulgaris"             </span></span>
<span><span class="co">## [11] "Capitella teleta"            "Spongilla lacustris"        </span></span>
<span><span class="co">## [13] "X. tropicalis"               "Chimp (P. troglodytes)"     </span></span>
<span><span class="co">## [15] "Bonobo (P. paniscus)"        "S. mansoni"                 </span></span>
<span><span class="co">## [17] "Sea urchin (S. purpuratus)"  "Human-Mouse Xenograft"</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="summary-attributes-1">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes-1"></a>
</h3>
<p><code>scfetch</code> provides <code>StatDBAttribute</code> to summary
attributes of <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell
Browser</a>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"UCSC"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $organism</span></span>
<span><span class="co">##                          Value Num      Key</span></span>
<span><span class="co">## 1           human (h. sapiens) 525 organism</span></span>
<span><span class="co">## 2          mouse (m. musculus) 196 organism</span></span>
<span><span class="co">## 3  fruit fly (d. melanogaster)  32 organism</span></span>
<span><span class="co">## 4  rhesus macaque (m. mulatta)  25 organism</span></span>
<span><span class="co">## 5             capitella teleta  18 organism</span></span>
<span><span class="co">## 6               hydra vulgaris  18 organism</span></span>
<span><span class="co">## 7          spongilla lacustris  18 organism</span></span>
<span><span class="co">## 8         zebrafish (d. rerio)  10 organism</span></span>
<span><span class="co">## 9              c. intestinalis   9 organism</span></span>
<span><span class="co">## 10      chimp (p. troglodytes)   8 organism</span></span>
<span><span class="co">## 11         dog (c. familiaris)   6 organism</span></span>
<span><span class="co">## 12                  c. robusta   5 organism</span></span>
<span><span class="co">## 13        bonobo (p. paniscus)   3 organism</span></span>
<span><span class="co">## 14  sea urchin (s. purpuratus)   3 organism</span></span>
<span><span class="co">## 15       human-mouse xenograft   1 organism</span></span>
<span><span class="co">## 16                  s. mansoni   1 organism</span></span>
<span><span class="co">## 17               x. tropicalis   1 organism</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $organ</span></span>
<span><span class="co">##                              Value Num   Key</span></span>
<span><span class="co">## 1                            brain 175 organ</span></span>
<span><span class="co">## 2                              eye 136 organ</span></span>
<span><span class="co">## 3                           retina 133 organ</span></span>
<span><span class="co">## 4                             lung  72 organ</span></span>
<span><span class="co">## 5                           muscle  44 organ</span></span>
<span><span class="co">## 6                            blood  42 organ</span></span>
<span><span class="co">## 7                            heart  36 organ</span></span>
<span><span class="co">## 8                  skeletal muscle  35 organ</span></span>
<span><span class="co">## 9                         pancreas  24 organ</span></span>
<span><span class="co">## 10                          thymus  23 organ</span></span>
<span><span class="co">## 11                          kidney  22 organ</span></span>
<span><span class="co">## 12                          immune  21 organ</span></span>
<span><span class="co">## 13                     bone marrow  20 organ</span></span>
<span><span class="co">## 14                            skin  20 organ</span></span>
<span><span class="co">## 15                             gut  19 organ</span></span>
<span><span class="co">## 16                           liver  19 organ</span></span>
<span><span class="co">## 17                  whole organism  18 organ</span></span>
<span><span class="co">## 18                          embryo  17 organ</span></span>
<span><span class="co">## 19                           ovary  16 organ</span></span>
<span><span class="co">## 20                          spleen  15 organ</span></span>
<span><span class="co">## 21                peripheral blood  13 organ</span></span>
<span><span class="co">## 22                             all  12 organ</span></span>
<span><span class="co">## 23                           colon  11 organ</span></span>
<span><span class="co">## 24                           nasal  11 organ</span></span>
<span><span class="co">## 25                           tumor  11 organ</span></span>
<span><span class="co">## 26                          testis  10 organ</span></span>
<span><span class="co">## 27                        organoid   9 organ</span></span>
<span><span class="co">## 28                          cortex   8 organ</span></span>
<span><span class="co">## 29                           fetal   8 organ</span></span>
<span><span class="co">## 30                     hippocampus   7 organ</span></span>
<span><span class="co">## 31                       intestine   7 organ</span></span>
<span><span class="co">## 32                 large intestine   7 organ</span></span>
<span><span class="co">## 33                      lymph node   7 organ</span></span>
<span><span class="co">## 34                 small intestine   6 organ</span></span>
<span><span class="co">## 35                          airway   5 organ</span></span>
<span><span class="co">## 36                          breast   5 organ</span></span>
<span><span class="co">## 37       leptomeningeal metastasis   5 organ</span></span>
<span><span class="co">## 38                        placenta   5 organ</span></span>
<span><span class="co">## 39              respiratory system   5 organ</span></span>
<span><span class="co">## 40                     spinal cord   5 organ</span></span>
<span><span class="co">## 41                         stomach   5 organ</span></span>
<span><span class="co">## 42                         trachea   5 organ</span></span>
<span><span class="co">## 43                          ureter   5 organ</span></span>
<span><span class="co">## 44                            balf   4 organ</span></span>
<span><span class="co">## 45                         bladder   4 organ</span></span>
<span><span class="co">## 46                      epithelium   4 organ</span></span>
<span><span class="co">## 47                       esophagus   4 organ</span></span>
<span><span class="co">## 48                   mammary gland   4 organ</span></span>
<span><span class="co">## 49                        striatum   4 organ</span></span>
<span><span class="co">## 50                          tongue   4 organ</span></span>
<span><span class="co">## 51                         adrenal   3 organ</span></span>
<span><span class="co">## 52                      cerebellum   3 organ</span></span>
<span><span class="co">## 53                  early develop.   3 organ</span></span>
<span><span class="co">## 54                     oral cavity   3 organ</span></span>
<span><span class="co">## 55                        prostate   3 organ</span></span>
<span><span class="co">## 56                  salivary gland   3 organ</span></span>
<span><span class="co">## 57                       cell line   2 organ</span></span>
<span><span class="co">## 58                        cerebrum   2 organ</span></span>
<span><span class="co">## 59                      cord blood   2 organ</span></span>
<span><span class="co">## 60                         decidua   2 organ</span></span>
<span><span class="co">## 61                        ectoderm   2 organ</span></span>
<span><span class="co">## 62                     endothelial   2 organ</span></span>
<span><span class="co">## 63                      epithelial   2 organ</span></span>
<span><span class="co">## 64                             fat   2 organ</span></span>
<span><span class="co">## 65                     fetal liver   2 organ</span></span>
<span><span class="co">## 66                       forebrain   2 organ</span></span>
<span><span class="co">## 67                         gingiva   2 organ</span></span>
<span><span class="co">## 68                           ileum   2 organ</span></span>
<span><span class="co">## 69                            limb   2 organ</span></span>
<span><span class="co">## 70                     nasopharynx   2 organ</span></span>
<span><span class="co">## 71                            nose   2 organ</span></span>
<span><span class="co">## 72                      oesophagus   2 organ</span></span>
<span><span class="co">## 73                placenta/decidua   2 organ</span></span>
<span><span class="co">## 74                          rectum   2 organ</span></span>
<span><span class="co">## 75                           teeth   2 organ</span></span>
<span><span class="co">## 76                         antenna   1 organ</span></span>
<span><span class="co">## 77                            body   1 organ</span></span>
<span><span class="co">## 78                       body wall   1 organ</span></span>
<span><span class="co">## 79                            bone   1 organ</span></span>
<span><span class="co">## 80                     bone marroe   1 organ</span></span>
<span><span class="co">## 81            brown adipose tissue   1 organ</span></span>
<span><span class="co">## 82                    cell culture   1 organ</span></span>
<span><span class="co">## 83                          cornea   1 organ</span></span>
<span><span class="co">## 84                           cotex   1 organ</span></span>
<span><span class="co">## 85                       diaphragm   1 organ</span></span>
<span><span class="co">## 86  dorsolateral prefrontal cortex   1 organ</span></span>
<span><span class="co">## 87                     endometrium   1 organ</span></span>
<span><span class="co">## 88          enteric nervous system   1 organ</span></span>
<span><span class="co">## 89                       epidermis   1 organ</span></span>
<span><span class="co">## 90                         fatbody   1 organ</span></span>
<span><span class="co">## 91                           fovea   1 organ</span></span>
<span><span class="co">## 92                     gallbladder   1 organ</span></span>
<span><span class="co">## 93          gonadal adipose tissue   1 organ</span></span>
<span><span class="co">## 94                         haltere   1 organ</span></span>
<span><span class="co">## 95                            head   1 organ</span></span>
<span><span class="co">## 96                             leg   1 organ</span></span>
<span><span class="co">## 97        male reproductive glands   1 organ</span></span>
<span><span class="co">## 98               malpighian-tubule   1 organ</span></span>
<span><span class="co">## 99       mesechymal adipost tissue   1 organ</span></span>
<span><span class="co">## 100                   nasal mucosa   1 organ</span></span>
<span><span class="co">## 101                      neocortex   1 organ</span></span>
<span><span class="co">## 102                   neural crest   1 organ</span></span>
<span><span class="co">## 103                       oenocyte   1 organ</span></span>
<span><span class="co">## 104              peritoneal cavity   1 organ</span></span>
<span><span class="co">## 105  proboscis and maxillary palps   1 organ</span></span>
<span><span class="co">## 106                        stromal   1 organ</span></span>
<span><span class="co">## 107    subcutaneous adipose tissue   1 organ</span></span>
<span><span class="co">## 108                         uterus   1 organ</span></span>
<span><span class="co">## 109                    vasculature   1 organ</span></span>
<span><span class="co">## 110                   whole embryo   1 organ</span></span>
<span><span class="co">## 111                           wing   1 organ</span></span>
<span><span class="co">## 112                       yolk sac   1 organ</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-metadata-2">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-2"></a>
</h3>
<p><code>scfetch</code> provides <code>ExtractCBDatasets</code> to
filter metadata with <strong>collection</strong>,
<strong>sub-collection</strong>, <strong>organ</strong>, <strong>disease
status</strong>, <strong>organism</strong>, <strong>project</strong> and
<strong>cell number</strong> (The available values of these attributes
can be obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>). All attributes except cell number support fuzzy match
with <code>fuzzy.match</code>, this is useful when selecting
datasets.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBDatasets.html">ExtractCBDatasets</a></span><span class="op">(</span>all.samples.df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, organ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"blood"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Human (H. sapiens)"</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Use all shortLabel as input!</span></span></code></pre>
<pre><code><span><span class="co">## Use all subLabel as input!</span></span></code></pre>
<pre><code><span><span class="co">## Use all diseases as input!</span></span></code></pre>
<pre><code><span><span class="co">## Use all projects as input!</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                              name</span></span>
<span><span class="co">## 1   allen-celltypes/comparative-thalmus/human-lgn</span></span>
<span><span class="co">## 2 allen-celltypes/comparative-thalmus/macaque-lgn</span></span>
<span><span class="co">## 3   allen-celltypes/comparative-thalmus/mouse-lgd</span></span>
<span><span class="co">## 4                      lepto-metastasis/patient-d</span></span>
<span><span class="co">##                                             shortLabel</span></span>
<span><span class="co">## 1                 Allen Brain Map: Cell Types Database</span></span>
<span><span class="co">## 2                 Allen Brain Map: Cell Types Database</span></span>
<span><span class="co">## 3                 Allen Brain Map: Cell Types Database</span></span>
<span><span class="co">## 4 Single-cell atlas of human leptomeningeal metastasis</span></span>
<span><span class="co">##                            subLabel tags</span></span>
<span><span class="co">## 1                         Human LGN     </span></span>
<span><span class="co">## 2                       Macaque LGN     </span></span>
<span><span class="co">## 3                         Mouse LGd     </span></span>
<span><span class="co">## 4 Patient D with Lung Primary Tumor     </span></span>
<span><span class="co">##                                             body_parts</span></span>
<span><span class="co">## 1                                         brain|parent</span></span>
<span><span class="co">## 2                                         brain|parent</span></span>
<span><span class="co">## 3                                         brain|parent</span></span>
<span><span class="co">## 4 brain, spinal cord, tumor, leptomeningeal metastasis</span></span>
<span><span class="co">##                         diseases</span></span>
<span><span class="co">## 1                 Healthy|parent</span></span>
<span><span class="co">## 2                 Healthy|parent</span></span>
<span><span class="co">## 3                 Healthy|parent</span></span>
<span><span class="co">## 4 Leptomeningeal Melanoma|parent</span></span>
<span><span class="co">##                                                                     organisms</span></span>
<span><span class="co">## 1 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 2 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 3 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 4                                                   Human (H. sapiens)|parent</span></span>
<span><span class="co">##                   projects life_stages domains sources sampleCount assays</span></span>
<span><span class="co">## 1 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 2 Allen Brain Atlas|parent                  NA      NA        1092       </span></span>
<span><span class="co">## 3 Allen Brain Atlas|parent                  NA      NA        1996       </span></span>
<span><span class="co">## 4                                           NA      NA        1682       </span></span>
<span><span class="co">##              matrix barcode feature matrixType</span></span>
<span><span class="co">## 1 exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 2 exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 3 exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 4 exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">##                                                                title paper</span></span>
<span><span class="co">## 1                             Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 2                           Macaque Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 3 Comparative Thalamus - Mouse Dorsolateral Geniculate Complex (LGd)      </span></span>
<span><span class="co">## 4                                  Patient D with Lung Primary Tumor      </span></span>
<span><span class="co">##                                                                                                           abstract</span></span>
<span><span class="co">## 1                                                             This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 2                                                           This dataset covers 1,092 nuclei from macaque samples.</span></span>
<span><span class="co">## 3                                                              This dataset covers 1,996 cells from mouse samples.</span></span>
<span><span class="co">## 4 CSF cell fraction isolated from Patient D (primary tumor in lung) with newly diagnosed leptomeningeal metastasis</span></span>
<span><span class="co">##   unit                                                            coords</span></span>
<span><span class="co">## 1      UMAP.coords.tsv.gz, ForceAtlas2.coords.tsv.gz, tSNE.coords.tsv.gz</span></span>
<span><span class="co">## 2      UMAP.coords.tsv.gz, ForceAtlas2.coords.tsv.gz, tSNE.coords.tsv.gz</span></span>
<span><span class="co">## 3      UMAP.coords.tsv.gz, ForceAtlas2.coords.tsv.gz, tSNE.coords.tsv.gz</span></span>
<span><span class="co">## 4                                                     UMAP.coords.tsv.gz</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 methods</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 4 &lt;section&gt;Human CSF Single-Cell Transcriptomic Analysis&lt;/section&gt;\n&lt;p&gt;Single-cell and bulk RNA-sequencing data have been deposited to NCBI GEO as\nGSE150681 SuperSeries. CSF, collected with informed consent from patients under\nprotocol IRB 13-039, was processed to isolate the sample into cell-free CSF and the cellular contents of the CSF: The whole CSF sample was centrifuged at 600 x g for 5 minutes without brake at 4 ºC to pellet the cells, and the supernatant was saved as cellfree CSF. The pellet was resuspended, washed with PBS supplemented with 0.4% BSA twice and processed immediately. The cells were manually counted with a hematocytometer. scRNA-Seq was performed with 10X genomics system using Chromium Single Cell 3' Library and Gel Bead Kit V2 (catalog no. 120234). Briefly,8,700 cells (viability 70-80%) were processed per sample, targeting recovery of ~5,000\ncells with 3.9% multiplet rate. In cases, where cell count was too low to target 5,000\ncells, maximum volume (34 µl) was loaded in the microfluidic droplet generation device.\nAfter reverse transcription reaction emulsions were broken, barcoded cDNA was purified\nwith DynaBeads, followed by 12-cycles of PCR amplification. The resulting amplified\nbarcoded-cDNA library was fragmented to ~400-600 bp, ligated to sequencing adapter\nand PCR-amplified to obtain sufficient amount of material for next-generation\nsequencing. The final libraries were sequenced on an Illumina NovaSeq 6000 system\n(Read 1 - 28 cycles, Index Read  8 cycles, and Read 2 - 96 cycles).\n&lt;/p&gt; \n&lt;section&gt;Annotation of cell types&lt;/section&gt;\n&lt;p&gt;Cell quality control steps described earlier are quite permissive and sometimes fail to\neliminate all low-quality cells. These cells typically do not express any particular\ngenes and frequently have low library size. In our dataset, after careful consideration,\ncluster 10 was identified as low-quality, dying cells and ultimately eliminated.\nMoreover, cluster 5 and 17 exhibited both T cell and myeloid lineage phenotypic\nprofiles and were eliminated from the dataset. To facilitate annotation of remaining 15\nclusters identified by PhenoGraph, we have examined MAGIC imputed gene expression\nof known marker genes (Figure S6A) and MAST derived differentially expressed genes\nbetween PhenoGraph clusters. Markers used to identify major cell types included MS4A1\nand CD79A (B cells), IL3RA and CLEC4C (plasmacytoid DC), EPCAM and KRT18\n(Cancer cells), CD3D and CD8A (CD8 T cells), CD3D, IL7R and CD4 (CD4 T cells),\nGNLY, NKG7, KLRB1 and NCAM1 (NK cells), CD14, CD68 and CST3\n(Macrophages), FCGR3A, LYZ and CST3 (Monocytes), CST3 and FCER1A (conventional DC).&lt;/p&gt;\n\n\n</span></span>
<span><span class="co">##   geo</span></span>
<span><span class="co">## 1    </span></span>
<span><span class="co">## 2    </span></span>
<span><span class="co">## 3    </span></span>
<span><span class="co">## 4</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-cell-type-composition-1">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition-1"></a>
</h3>
<p><code>scfetch</code> provides <code>ExtractCBComposition</code> to
extract cell type annotation and composition.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span>json.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/cell_browser/json"</span>, sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hbb.sample.ct</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                             shortLabel  subLabel           CellType Num tags</span></span>
<span><span class="co">## 1 Allen Brain Map: Cell Types Database Human LGN      LGN Exc BTNL9 908     </span></span>
<span><span class="co">## 2 Allen Brain Map: Cell Types Database Human LGN          Oligo MAG 188     </span></span>
<span><span class="co">## 3 Allen Brain Map: Cell Types Database Human LGN LGN Exc PRKCG BCHE 102     </span></span>
<span><span class="co">## 4 Allen Brain Map: Cell Types Database Human LGN      LGN Inh CTXN3  96     </span></span>
<span><span class="co">## 5 Allen Brain Map: Cell Types Database Human LGN      LGN Inh LAMP5  72     </span></span>
<span><span class="co">## 6 Allen Brain Map: Cell Types Database Human LGN      LGN Inh NTRK1  42     </span></span>
<span><span class="co">##     body_parts       diseases</span></span>
<span><span class="co">## 1 brain|parent Healthy|parent</span></span>
<span><span class="co">## 2 brain|parent Healthy|parent</span></span>
<span><span class="co">## 3 brain|parent Healthy|parent</span></span>
<span><span class="co">## 4 brain|parent Healthy|parent</span></span>
<span><span class="co">## 5 brain|parent Healthy|parent</span></span>
<span><span class="co">## 6 brain|parent Healthy|parent</span></span>
<span><span class="co">##                                                                     organisms</span></span>
<span><span class="co">## 1 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 2 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 3 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 4 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 5 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">## 6 Human (H. sapiens), Mouse (M. musculus), Rhesus macaque (M. mulatta)|parent</span></span>
<span><span class="co">##                   projects life_stages domains sources sampleCount assays</span></span>
<span><span class="co">## 1 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 2 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 3 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 4 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 5 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">## 6 Allen Brain Atlas|parent                  NA      NA        1576       </span></span>
<span><span class="co">##                                    title paper</span></span>
<span><span class="co">## 1 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 2 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 3 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 4 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 5 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">## 6 Human Lateral Geniculate Nucleus (LGN)      </span></span>
<span><span class="co">##                                               abstract</span></span>
<span><span class="co">## 1 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 2 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 3 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 4 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 5 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">## 6 This dataset covers 1,576 nuclei from human samples.</span></span>
<span><span class="co">##                                                                                                                                                                                                               methods</span></span>
<span><span class="co">## 1 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 2 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 3 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 4 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 5 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">## 6 See the Allen Brain Atlas\n&lt;a href="http://help.brain-map.org/display/celltypes/Documentation" traget="_blank"\n&gt;Documentation&lt;/a&gt; for details about how the different aspects of this project\nwere carried out.\n</span></span>
<span><span class="co">##   geo</span></span>
<span><span class="co">## 1    </span></span>
<span><span class="co">## 2    </span></span>
<span><span class="co">## 3    </span></span>
<span><span class="co">## 4    </span></span>
<span><span class="co">## 5    </span></span>
<span><span class="co">## 6</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="load-the-online-datasets-to-seurat">Load the online datasets to Seurat<a class="anchor" aria-label="anchor" href="#load-the-online-datasets-to-seurat"></a>
</h3>
<p>After manually check the extracted metadata, <code>scfetch</code>
provides <code>ParseCBDatasets</code> to <strong>load the online count
matrix</strong> to Seurat. All the attributes available in
<code>ExtractCBDatasets</code> are also same here. Please note that the
loading process provided by <code>ParseCBDatasets</code> will load the
online count matrix instead of downloading it to local. If multiple
datasets available, users can choose to merge the SeuratObject with
<code>merge</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
