<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scfetch">
<title>DownloadRaw • scfetch</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadRaw">
<meta property="og:description" content="scfetch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scfetch</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/DownloadRaw.html">Download fastq/bam files</a>
    <a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrices</a>
    <a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed objects</a>
    <a class="dropdown-item" href="../articles/ObjectConversion.html">Convert formats between objects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/scfetch">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>DownloadRaw</h1>
            
      
      
      <div class="d-none name"><code>DownloadRaw.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>A common situation is that we need to use a unified software version
(e.g. CellRanger) to obtain the count matrix, in order to better
integrate and compare multiple datasets. Here, we will use
<code>scfetch</code> to download fastq and bam files. With bam files,
<code>scfetch</code> also provides function for user to convert the bam
to fastq.</p>
<hr>
</div>
<div class="section level2">
<h2 id="download-fastq">Download fastq<a class="anchor" aria-label="anchor" href="#download-fastq"></a>
</h2>
<div class="section level3">
<h3 id="prepare-run-number">Prepare run number<a class="anchor" aria-label="anchor" href="#prepare-run-number"></a>
</h3>
<p>For fastq files stored in SRA, <code>scfetch</code> can extract
sample information and run number with GEO accession number or users can
also provide a dataframe contains the run number of interested
samples.</p>
<p>Extract all samples under <code>GSE130636</code> and the platform is
<code>GPL20301</code> (use <code>platform = NULL</code> for all
platforms):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scfetch</span><span class="op">)</span></span>
<span><span class="co">#&gt; Setting options('download.file.method.GEOquery'='auto')</span></span>
<span><span class="co">#&gt; Setting options('GEOquery.inmemory.gpl'=FALSE)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'SeuratDisk':</span></span>
<span><span class="co">#&gt;   method            from  </span></span>
<span><span class="co">#&gt;   as.sparse.H5Group Seurat</span></span>
<span></span>
<span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE130636"</span>, platform <span class="op">=</span> <span class="st">"GPL20301"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Extract all GSM with acce: GSE130636 and platform: GPL20301</span></span>
<span><span class="co">#&gt; Found 1 file(s)</span></span>
<span><span class="co">#&gt; GSE130636_series_matrix.txt.gz</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">7</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (7): ID_REF, GSM3745992, GSM3745993, GSM3745994, GSM3745995, GSM3745996,...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt; File stored at: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; /var/folders/_4/k4qmvf7s2gx_6789px8n_sxh0000gn/T//RtmpXDGHCE/GPL20301.soft</span></span>
<span><span class="co">#&gt; 6 GSMs to process</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE130636.runs</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    title   gsm_name experiment        run</span></span>
<span><span class="co">#&gt; SRR9004325 Fovea donor 1 GSM3745992 SRX5783051 SRR9004325</span></span>
<span><span class="co">#&gt; SRR9004326 Fovea donor 1 GSM3745992 SRX5783051 SRR9004326</span></span>
<span><span class="co">#&gt; SRR9004327 Fovea donor 1 GSM3745992 SRX5783051 SRR9004327</span></span>
<span><span class="co">#&gt; SRR9004328 Fovea donor 1 GSM3745992 SRX5783051 SRR9004328</span></span>
<span><span class="co">#&gt; SRR9004329 Fovea donor 1 GSM3745992 SRX5783051 SRR9004329</span></span>
<span><span class="co">#&gt; SRR9004330 Fovea donor 1 GSM3745992 SRX5783051 SRR9004330</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-sra">Download sra<a class="anchor" aria-label="anchor" href="#download-sra"></a>
</h3>
<p>With the dataframe contains gsm and run number, <code>scfetch</code>
will download sra files using <code>prefetch</code>. The returned result
is a dataframe contains failed runs. If not <code>NULL</code>, users can
re-run <code>DownloadSRA</code> by setting <code>gsm.df</code> to the
returned result.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="va">GSE130636.runs</span><span class="op">[</span><span class="va">GSE130636.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR9004346"</span>, <span class="st">"SRR9004351"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>,</span>
<span>  prefetch.path <span class="op">=</span> <span class="st">"/Users/soyabean/software/sratoolkit.3.0.6-mac64/bin/prefetch"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/download_fastq"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be:
<code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level3">
<h3 id="split-fastq">Split fastq<a class="anchor" aria-label="anchor" href="#split-fastq"></a>
</h3>
<p>After obtaining the sra files, <code>scfetch</code> provides function
<code>SplitSRA</code> to split sra files to fastq files using
<code>parallel-fastq-dump</code> (parallel, fastest and gzip output),
<code>fasterq-dump</code> (parallel, fast but unzipped output) and
<code>fastq-dump</code> (slowest and gzip output).</p>
<p>For fastqs generated with 10x Genomics, <code>SplitSRA</code> can
identify read1, read2 and index files and format the read1 and read2 to
10x required format (<code>sample1_S1_L001_R1_001.fastq.gz</code> and
<code>sample1_S1_L001_R2_001.fastq.gz</code>). In detail, the file with
read length 26 or 28 is considered as read1, the files with read length
8 or 10 are considered as index files and the remain file is considered
as read2. The read length rules is from <a href="https://www.10xgenomics.com/cn/support/single-cell-gene-expression/documentation/steps/sequencing/sequencing-requirements-for-single-cell-3" class="external-link">Sequencing
Requirements for Single Cell 3’</a> and <a href="https://www.10xgenomics.com/cn/support/single-cell-immune-profiling/documentation/steps/sequencing/sequencing-requirements-for-single-cell-v-d-j" class="external-link">Sequencing
Requirements for Single Cell V(D)J</a>.</p>
<p>The returned result is a vector of failed sra files. If not
<code>NULL</code>, users can re-run <code>SplitSRA</code> by setting
<code>sra.path</code> to the returned result.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parallel-fastq-dump requires sratools.path</span></span>
<span><span class="va">GSE130636.split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SplitSRA.html">SplitSRA</a></span><span class="op">(</span></span>
<span>  sra.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/download_fastq"</span>,</span>
<span>  fastq.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  split.cmd.path <span class="op">=</span> <span class="st">"/Applications/anaconda3/bin/parallel-fastq-dump"</span>,</span>
<span>  sratools.path <span class="op">=</span> <span class="st">"/usr/local/bin"</span>, split.cmd.threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The final <code>out.folder</code> structure will be:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> /Users/soyabean/Desktop/tmp/scdown/download_fastq</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [01;34m/Users/soyabean/Desktop/tmp/scdown/download_fastq [0m</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; └──  [01;34mGSM3745993 [0m</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ├──  [01;34mSRR9004346 [0m</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     │   ├──  [00mSRR9004346.sra [0m</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     │   ├──  [01;31mSRR9004346_1.fastq.gz [0m</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     │   ├──  [01;31mSRR9004346_2.fastq.gz [0m</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     │   ├──  [01;31mSRR9004346_S1_L001_R1_001.fastq.gz [0m</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     │   └──  [01;31mSRR9004346_S1_L001_R2_001.fastq.gz [0m</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     └──  [01;34mSRR9004351 [0m</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         ├──  [00mSRR9004351.sra [0m</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         ├──  [01;31mSRR9004351_1.fastq.gz [0m</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         ├──  [01;31mSRR9004351_2.fastq.gz [0m</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         ├──  [01;31mSRR9004351_S1_L001_R1_001.fastq.gz [0m</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         └──  [01;31mSRR9004351_S1_L001_R2_001.fastq.gz [0m</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 directories, 10 files</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="download-bam">Download bam<a class="anchor" aria-label="anchor" href="#download-bam"></a>
</h2>
<div class="section level3">
<h3 id="prepare-run-number-1">Prepare run number<a class="anchor" aria-label="anchor" href="#prepare-run-number-1"></a>
</h3>
<p><code>scfetch</code> can extract sample information and run number
with GEO accession number or users can also provide a dataframe contains
the run number of interested samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE138266"</span>, platform <span class="op">=</span> <span class="st">"GPL18573"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Extract all GSM with acce: GSE138266 and platform: GPL18573</span></span>
<span><span class="co">#&gt; Found 1 file(s)</span></span>
<span><span class="co">#&gt; GSE138266_series_matrix.txt.gz</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">0</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">23</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (23): ID_REF, GSM4104122, GSM4104123, GSM4104124, GSM4104125, GSM4104126...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt; File stored at: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; /var/folders/_4/k4qmvf7s2gx_6789px8n_sxh0000gn/T//RtmpXDGHCE/GPL18573.soft</span></span>
<span><span class="co">#&gt; 22 GSMs to process</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE138266.runs</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   title   gsm_name experiment         run</span></span>
<span><span class="co">#&gt; SRR10211551 MS19270_CSF GSM4104122 SRX6931239 SRR10211551</span></span>
<span><span class="co">#&gt; SRR10211552 MS58637_CSF GSM4104123 SRX6931240 SRR10211552</span></span>
<span><span class="co">#&gt; SRR10211553 MS71658_CSF GSM4104124 SRX6931241 SRR10211553</span></span>
<span><span class="co">#&gt; SRR10211554 MS49131_CSF GSM4104125 SRX6931242 SRR10211554</span></span>
<span><span class="co">#&gt; SRR10211555 MS60249_CSF GSM4104126 SRX6931243 SRR10211555</span></span>
<span><span class="co">#&gt; SRR10211556 MS74594_CSF GSM4104127 SRX6931244 SRR10211556</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-bam-1">Download bam<a class="anchor" aria-label="anchor" href="#download-bam-1"></a>
</h3>
<p>With the dataframe contains gsm and run number, <code>scfetch</code>
provides <code>DownloadBam</code> to download bam files using
<code>prefetch</code>. It suooorts 10x generated bam files and normal
bam files.</p>
<ul>
<li>10x generated bam: While bam files generated from 10x softwares
(e.g. CellRanger) contain custom tags which are not kept when using
default parameters of <code>prefetch</code>, <code>scfetch</code> adds
<code>--type TenX</code> to make sure the downloaded bam files contain
these tags.</li>
<li>normal bam: For normal bam files, <code>DownloadBam</code> will
download sra files first and then convert sra files to bam files with
<code>sam-dump</code>. After testing the efficiency of
<code>prefetch</code> + <code>sam-dump</code> and <code>sam-dump</code>,
the former is much faster than the latter (52G sra and 72G bam
files):</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # use prefetch to download sra file</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># prefetch -X 60G SRR1976036</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    117m26.334s</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    16m42.062s</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 3m28.295s</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # use sam-dump to convert sra to bam</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036.sra | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    536m2.721s</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    749m41.421s</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 20m49.069s</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># use sam-dump to download bam directly</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036 | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co"># # more than 36hrs only get ~3G bam files, too slow</span></span></code></pre></div>
<p>The returned result is a dataframe containing failed runs (either
failed to download sra files or failed to convert to bam files for
normal bam; failed to download bam files for 10x generated bam). If not
<code>NULL</code>, users can re-run <code>DownloadBam</code> by setting
<code>gsm.df</code> to the returned result. The following is an example
to download 10x generated bam file:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="va">GSE138266.runs</span><span class="op">[</span><span class="va">GSE138266.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR10211566"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>,</span>
<span>  prefetch.path <span class="op">=</span> <span class="st">"/Users/soyabean/software/sratoolkit.3.0.6-mac64/bin/prefetch"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/download_bam"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE138266.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be:
<code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level3">
<h3 id="convert-bam-to-fastq">Convert bam to fastq<a class="anchor" aria-label="anchor" href="#convert-bam-to-fastq"></a>
</h3>
<p>With downloaded bam files, <code>scfetch</code> provides function
<code>Bam2Fastq</code> to convert bam files to fastq files. For bam
files generated from 10x softwares, <code>Bam2Fastq</code> utilizes
<code>bamtofastq</code> tool developed by 10x Genomics.</p>
<p>The returned result is a vector of bam files failed to convert to
fastq files. If not <code>NULL</code>, users can re-run
<code>Bam2Fastq</code> by setting <code>bam.path</code> to the returned
result.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Bam2Fastq.html">Bam2Fastq</a></span><span class="op">(</span></span>
<span>  bam.folder <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/download_bam"</span>,</span>
<span>  bamtofastq.path <span class="op">=</span> <span class="st">"/Users/soyabean/software/bamtofastq_macos"</span>,</span>
<span>  bamtofastq.paras <span class="op">=</span> <span class="st">"--nthreads 4"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The final <code>out.folder</code> structure will be:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> /Users/soyabean/Desktop/tmp/scdown/download_bam</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [01;34m/Users/soyabean/Desktop/tmp/scdown/download_bam [0m</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; └──  [01;34mGSM4104137 [0m</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     └──  [01;34mSRR10211566 [0m</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         ├──  [01;34mbam2fastq [0m</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │   └──  [01;34mMS60249_PBMC_2_0_MissingLibrary_1_H72VGBGX2 [0m</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_I1_001.fastq.gz [0m</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_I1_002.fastq.gz [0m</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_R1_001.fastq.gz [0m</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_R1_002.fastq.gz [0m</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_R2_001.fastq.gz [0m</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L001_R2_002.fastq.gz [0m</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_I1_001.fastq.gz [0m</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_I1_002.fastq.gz [0m</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_R1_001.fastq.gz [0m</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_R1_002.fastq.gz [0m</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_R2_001.fastq.gz [0m</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L002_R2_002.fastq.gz [0m</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_I1_001.fastq.gz [0m</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_I1_002.fastq.gz [0m</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_R1_001.fastq.gz [0m</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_R1_002.fastq.gz [0m</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_R2_001.fastq.gz [0m</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L003_R2_002.fastq.gz [0m</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L004_I1_001.fastq.gz [0m</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L004_I1_002.fastq.gz [0m</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L004_R1_001.fastq.gz [0m</span></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L004_R1_002.fastq.gz [0m</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       ├──  [01;31mbamtofastq_S1_L004_R2_001.fastq.gz [0m</span></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         │       └──  [01;31mbamtofastq_S1_L004_R2_002.fastq.gz [0m</span></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         └──  [00mbamfiles_MS60249_PBMC_possorted_genome_bam.bam [0m</span></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5 directories, 25 files</span></span></code></pre></div>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
