<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>DownloadRaw • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadRaw">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DownloadRaw</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/DownloadRaw.Rmd" class="external-link"><code>vignettes/DownloadRaw.Rmd</code></a></small>
      <div class="d-none name"><code>DownloadRaw.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>A common situation is that we need to use a unified software version
(e.g. CellRanger) to obtain the count matrix, in order to better
integrate and compare multiple datasets. Here, we will use
<code>GEfetch2R</code> to download raw data
(<code>sra/fastq/bam</code>). With <code>bam</code> files,
<code>GEfetch2R</code> also provides function for user to convert the
<code>bam</code> to <code>fastq</code> files.</p>
<blockquote>
<p><code>GEfetch2R</code> supports downloading raw data
(<code>sra/fastq/bam</code>) from <code>SRA</code> and <code>ENA</code>
with <strong>GEO accessions</strong>. In general, downloading raw data
from ENA is much faster than SRA, because of <code>ascp</code> and
<strong>parallel</strong> support.</p>
</blockquote>
<hr>
</div>
<div class="section level2">
<h2 id="check-api">Check API<a class="anchor" aria-label="anchor" href="#check-api"></a>
</h2>
<p>Check the availability of APIs used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/CheckAPI.html">CheckAPI</a></span><span class="op">(</span>database <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GEO"</span>, <span class="st">"SRA/ENA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># 1 GSMs to process</span></span>
<span><span class="co"># The API to access the SRA accession is OK!</span></span>
<span><span class="co"># The API to access ENA download links is OK!</span></span>
<span><span class="co"># start checking APIs to access GEO!</span></span>
<span><span class="co"># trying URL 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE302nnn/GSE302912/matrix/GSE302912_series_matrix.txt.gz'</span></span>
<span><span class="co"># Content type 'application/x-gzip' length 2642 bytes</span></span>
<span><span class="co"># ==================================================</span></span>
<span><span class="co"># downloaded 2642 bytes</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># trying URL 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE302nnn/GSE302912/suppl//GSE302912_counts.csv.gz?tool=geoquery'</span></span>
<span><span class="co"># Content type 'application/x-gzip' length 332706 bytes (324 KB)</span></span>
<span><span class="co"># ==================================================</span></span>
<span><span class="co"># downloaded 324 KB</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># The API to access supplementary files is OK!</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="one-step-wrapper">One-step wrapper<a class="anchor" aria-label="anchor" href="#one-step-wrapper"></a>
</h2>
<p>With a given <code>GEO accession</code> or
<code>GSM accession</code>, <code>GEfetch2R</code> provides a one-step
function <code>DownloadFastq2R</code> to <strong>extract all
runs</strong>, <strong>automatically identify the RNA-seq type (10x
Genomics scRNA-seq, bulk RNA-seq, Smart-seq2 scRNA-seq/mini-bulk
RNA-seq) of each run</strong>, <strong>download <code>fastq</code> files
from <code>ENA</code></strong>, <strong>perform read mapping (merge
multiple runs of a sample)</strong>, and <strong>load the results to
R</strong>.</p>
<p>The way <code>DownloadFastq2R</code> automatically identify the
RNA-seq type:</p>
<ul>
<li>filter the <code>library_strategy</code> field
(<code>"rna-seq"</code>, <code>"scrna-seq"</code>,
<code>"snrna-seq"</code>) to <strong>remove non-RNA-seq
runs</strong>.</li>
<li>if <code>library_source</code> field is
<code>"transcriptomic single cell"</code> or
<code>library_strategy</code> field is in
<code>c("scrna-seq", "snrna-seq")</code> or key words
(<code>"scRNA", "snRNA", "single.cell", "single.nuclei", "single.nucleus", "singlecell", "singlenuclei", "singlenucleus", "10X.Genomics", "10XGenomics", "Smart.seq2", "Smartseq2"</code>)
appear in <code>title|source_name|characteristics|description</code>
fields, the run is classified as <strong>scRNA-seq</strong>.
<ul>
<li>if key words
(<code>"10X.Genomics", "Cell.Ranger", "10XGenomics", "CellRanger"</code>)
appear in any field, the run is classified as <strong>10x Genomics
scRNA-seq</strong>.</li>
<li>if key words (<code>"Smart.seq2", "Smartseq2"</code>) appear in any
field, the run is classified as <strong>Smart-seq2 scRNA-seq/mini bulk
RNA-seq</strong>.</li>
<li>
<strong>else the run will skip subsequent processing
steps</strong>.</li>
</ul>
</li>
<li>
<strong>else the run is classified as bulk RNA-seq</strong>.</li>
</ul>
<p>If the automatic identification failed, users can specify the RNA-seq
type via <code>force.type</code>, choose one from <code>"10x"</code>,
<code>"Smart-seq2"</code>, <code>"bulk"</code>.</p>
<p>With a given <code>GEO accession</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">GEfetch2R</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSE305141.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadFastq2R.html">DownloadFastq2R</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE305141"</span>, skip.gsm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM9162729"</span>, <span class="st">"GSM9162725"</span><span class="op">)</span>, star.ref <span class="op">=</span> <span class="st">"~/Mus_musculus/star"</span>,</span>
<span>  cellranger.ref <span class="op">=</span> <span class="st">"~/refdata-cellranger-mm10-3.0.0"</span>,</span>
<span>  star.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/STAR"</span>,</span>
<span>  cellranger.path <span class="op">=</span> <span class="st">"~/cellranger-3.0.2/cellranger"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/fastq2r"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE305141.list</span></span>
<span><span class="co"># $bulk.rna</span></span>
<span><span class="co"># class: DESeqDataSet</span></span>
<span><span class="co"># dim: 55573 11</span></span>
<span><span class="co"># metadata(1): version</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(55573): ENSMUSG00000000001 ENSMUSG00000000003 ...</span></span>
<span><span class="co">#   ENSMUSG00000118487 ENSMUSG00000118488</span></span>
<span><span class="co"># rowData names(0):</span></span>
<span><span class="co"># colnames(11): GSM9162716 GSM9162717 ... GSM9162726 GSM9162727</span></span>
<span><span class="co"># colData names(1): condition</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $scrna.10x</span></span>
<span><span class="co"># $scrna.10x$`/home/songyabing/Gaolab/07_dev/gefetch2r/doc/fastq2r/mapping/10x/GSM9162728/outs/filtered_feature_bc_matrix`</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 31053 features across 6774 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (31053 features, 0 variable features)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $scrna.10x$`/home/songyabing/Gaolab/07_dev/gefetch2r/doc/fastq2r/mapping/10x/GSM9162730/outs/filtered_feature_bc_matrix`</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 31053 features across 9344 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (31053 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree -L 3 ~/gefetch2r/doc/fastq2r</span></span>
<span><span class="co"># ~/gefetch2r/doc/fastq2r</span></span>
<span><span class="co"># ├── fastq</span></span>
<span><span class="co"># │   ├── 10x</span></span>
<span><span class="co"># │   │   ├── GSM9162728</span></span>
<span><span class="co"># │   │   └── GSM9162730</span></span>
<span><span class="co"># │   └── bulkRNAseq</span></span>
<span><span class="co"># │       ├── GSM9162716</span></span>
<span><span class="co"># │       ├── GSM9162717</span></span>
<span><span class="co"># │       ├── GSM9162718</span></span>
<span><span class="co"># │       ├── GSM9162719</span></span>
<span><span class="co"># │       ├── GSM9162720</span></span>
<span><span class="co"># │       ├── GSM9162721</span></span>
<span><span class="co"># │       ├── GSM9162722</span></span>
<span><span class="co"># │       ├── GSM9162723</span></span>
<span><span class="co"># │       ├── GSM9162724</span></span>
<span><span class="co"># │       ├── GSM9162726</span></span>
<span><span class="co"># │       └── GSM9162727</span></span>
<span><span class="co"># └── mapping</span></span>
<span><span class="co">#     ├── 10x</span></span>
<span><span class="co">#     │   ├── GSM9162728</span></span>
<span><span class="co">#     │   └── GSM9162730</span></span>
<span><span class="co">#     └── bulkRNAseq</span></span>
<span><span class="co">#         ├── GSM9162716</span></span>
<span><span class="co">#         ├── GSM9162717</span></span>
<span><span class="co">#         ├── GSM9162718</span></span>
<span><span class="co">#         ├── GSM9162719</span></span>
<span><span class="co">#         ├── GSM9162720</span></span>
<span><span class="co">#         ├── GSM9162721</span></span>
<span><span class="co">#         ├── GSM9162722</span></span>
<span><span class="co">#         ├── GSM9162723</span></span>
<span><span class="co">#         ├── GSM9162724</span></span>
<span><span class="co">#         ├── GSM9162726</span></span>
<span><span class="co">#         └── GSM9162727</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 32 directories, 0 files</span></span></code></pre></div>
<p>With a given <code>GSM accession</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE127942.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadFastq2R.html">DownloadFastq2R</a></span><span class="op">(</span></span>
<span>  gsm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM3656922"</span>, <span class="st">"GSM3656923"</span><span class="op">)</span>, star.ref <span class="op">=</span> <span class="st">"~/Mus_musculus/star"</span>,</span>
<span>  star.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/STAR"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/fastq2r2"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE127942.list</span></span>
<span><span class="co"># $bulk.rna</span></span>
<span><span class="co"># class: DESeqDataSet</span></span>
<span><span class="co"># dim: 55573 2</span></span>
<span><span class="co"># metadata(1): version</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(55573): ENSMUSG00000000001 ENSMUSG00000000003 ...</span></span>
<span><span class="co">#   ENSMUSG00000118487 ENSMUSG00000118488</span></span>
<span><span class="co"># rowData names(0):</span></span>
<span><span class="co"># colnames(2): GSM3656922 GSM3656923</span></span>
<span><span class="co"># colData names(1): condition</span></span>
<span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/fastq2r2</span></span>
<span><span class="co"># ~/gefetch2r/doc/fastq2r2</span></span>
<span><span class="co"># ├── fastq</span></span>
<span><span class="co"># │   └── bulkRNAseq</span></span>
<span><span class="co"># │       ├── GSM3656922</span></span>
<span><span class="co"># │       │   ├── SRR8690244</span></span>
<span><span class="co"># │       │   │   ├── SRR8690244_1.fastq.gz</span></span>
<span><span class="co"># │       │   │   └── SRR8690244_2.fastq.gz</span></span>
<span><span class="co"># │       │   └── SRR8690245</span></span>
<span><span class="co"># │       │       ├── SRR8690245_1.fastq.gz</span></span>
<span><span class="co"># │       │       └── SRR8690245_2.fastq.gz</span></span>
<span><span class="co"># │       └── GSM3656923</span></span>
<span><span class="co"># │           ├── SRR8690246</span></span>
<span><span class="co"># │           │   ├── SRR8690246_1.fastq.gz</span></span>
<span><span class="co"># │           │   └── SRR8690246_2.fastq.gz</span></span>
<span><span class="co"># │           └── SRR8690247</span></span>
<span><span class="co"># │               ├── SRR8690247_1.fastq.gz</span></span>
<span><span class="co"># │               └── SRR8690247_2.fastq.gz</span></span>
<span><span class="co"># └── mapping</span></span>
<span><span class="co">#     └── bulkRNAseq</span></span>
<span><span class="co">#         ├── GSM3656922</span></span>
<span><span class="co">#         │   ├── Aligned.sortedByCoord.out.bam</span></span>
<span><span class="co">#         │   ├── GSM3656922.txt</span></span>
<span><span class="co">#         │   ├── Log.final.out</span></span>
<span><span class="co">#         │   ├── Log.out</span></span>
<span><span class="co">#         │   ├── Log.progress.out</span></span>
<span><span class="co">#         │   ├── SJ.out.tab</span></span>
<span><span class="co">#         │   └── _STARtmp</span></span>
<span><span class="co">#         │       └── BAMsort</span></span>
<span><span class="co">#         │           ├── 0</span></span>
<span><span class="co">#         │           ├── 1</span></span>
<span><span class="co">#         │           ├── 2</span></span>
<span><span class="co">#         │           └── 3</span></span>
<span><span class="co">#         └── GSM3656923</span></span>
<span><span class="co">#             ├── Aligned.sortedByCoord.out.bam</span></span>
<span><span class="co">#             ├── GSM3656923.txt</span></span>
<span><span class="co">#             ├── Log.final.out</span></span>
<span><span class="co">#             ├── Log.out</span></span>
<span><span class="co">#             ├── Log.progress.out</span></span>
<span><span class="co">#             ├── SJ.out.tab</span></span>
<span><span class="co">#             └── _STARtmp</span></span>
<span><span class="co">#                 └── BAMsort</span></span>
<span><span class="co">#                     ├── 0</span></span>
<span><span class="co">#                     ├── 1</span></span>
<span><span class="co">#                     ├── 2</span></span>
<span><span class="co">#                     └── 3</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 24 directories, 20 files</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="download-sra">Download sra<a class="anchor" aria-label="anchor" href="#download-sra"></a>
</h2>
<div class="section level3">
<h3 id="extract-all-samples-runs">Extract all samples (runs)<a class="anchor" aria-label="anchor" href="#extract-all-samples-runs"></a>
</h3>
<p>For <code>fastq</code> files stored in <code>SRA/ENA</code>,
<code>GEfetch2R</code> can extract sample information and run number
with <strong>GEO accessions</strong> or users can also <strong>provide a
dataframe containing the run number of interested samples</strong>.</p>
<p>Extract all samples under <code>GSE130636</code> and the platform is
<code>GPL20301</code> (use <code>platform = NULL</code> for all
platforms):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE130636"</span>, platform <span class="op">=</span> <span class="st">"GPL20301"</span><span class="op">)</span></span>
<span><span class="co"># a small test</span></span>
<span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="va">GSE130636.runs</span><span class="op">[</span><span class="va">GSE130636.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR9004346"</span>, <span class="st">"SRR9004351"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">GSE130636.runs</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"run"</span>, <span class="st">"experiment"</span>, <span class="st">"gsm_name"</span>, <span class="st">"title"</span>, <span class="st">"geo_accession"</span>, <span class="st">"source_name_ch1"</span>, <span class="st">"organism_ch1"</span>, <span class="st">"characteristics_ch1"</span>, <span class="st">"characteristics_ch1.1"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                   run experiment   gsm_name         title geo_accession source_name_ch1</span></span>
<span><span class="co"># SRR9004346 SRR9004346 SRX5783052 GSM3745993 Fovea donor 2    GSM3745993          Retina</span></span>
<span><span class="co"># SRR9004351 SRR9004351 SRX5783052 GSM3745993 Fovea donor 2    GSM3745993          Retina</span></span>
<span><span class="co">#            organism_ch1 characteristics_ch1 characteristics_ch1.1</span></span>
<span><span class="co"># SRR9004346 Homo sapiens     location: Fovea        donor: Donor 2</span></span>
<span><span class="co"># SRR9004351 Homo sapiens     location: Fovea        donor: Donor 2</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-sra-1">Download sra<a class="anchor" aria-label="anchor" href="#download-sra-1"></a>
</h3>
<p>With the dataframe contains gsm and run number,
<code>GEfetch2R</code> will use <code>prefetch</code> to download
<code>sra</code> files from <code>SRA</code> or using
<code>ascp/download.file</code> to download <code>sra</code> files from
<code>ENA</code>. The returned value is a dataframe contains failed
runs. If not <code>NULL</code>, users can re-run
<code>DownloadSRA</code> by setting <code>gsm.df</code> to the returned
value.</p>
<p>Download from <code>SRA</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>,</span>
<span>  prefetch.path <span class="op">=</span> <span class="st">"~/data/software/sratoolkit.2.11.2-centos_linux64/bin/prefetch"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_sra/prefetch"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span>
<span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_sra/prefetch</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_sra/prefetch</span></span>
<span><span class="co"># └── GSM3745993</span></span>
<span><span class="co">#     ├── SRR9004346</span></span>
<span><span class="co">#     │   └── SRR9004346.sra</span></span>
<span><span class="co">#     └── SRR9004351</span></span>
<span><span class="co">#         └── SRR9004351.sra</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 3 directories, 2 files</span></span></code></pre></div>
<p>Download from <code>ENA</code> (<strong>parallel</strong>):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download from ENA using download.file</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, download.method <span class="op">=</span> <span class="st">"download.file"</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">3600</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_sra/download_file"</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download from ENA using ascp</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, download.method <span class="op">=</span> <span class="st">"ascp"</span>,</span>
<span>  ascp.path <span class="op">=</span> <span class="st">"~/.aspera/connect/bin/ascp"</span>, max.rate <span class="op">=</span> <span class="st">"300m"</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_sra/ascp"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download from ENA using wget</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, download.method <span class="op">=</span> <span class="st">"wget"</span>,</span>
<span>  wget.path <span class="op">=</span> <span class="st">"/usr/bin/wget"</span>, timeout <span class="op">=</span> <span class="fl">3600</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_sra/wget"</span>, rename <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="download-fastq">Download fastq<a class="anchor" aria-label="anchor" href="#download-fastq"></a>
</h2>
<div class="section level3">
<h3 id="split-sra-to-generate-fastq">Split sra to generate fastq<a class="anchor" aria-label="anchor" href="#split-sra-to-generate-fastq"></a>
</h3>
<p>After obtaining the <code>sra</code> files, <code>GEfetch2R</code>
provides function <code>SplitSRA</code> to split <code>sra</code> files
to fastq files using <code>parallel-fastq-dump</code> (<strong>parallel,
fastest and gzip output</strong>), <code>fasterq-dump</code>
(<strong>parallel, fast but unzipped output</strong>) and
<code>fastq-dump</code> (<strong>slowest and gzip output</strong>).</p>
<p>For fastqs generated with 10x Genomics, <code>SplitSRA</code> can
identify read1, read2 and index files and format the read1 and read2 to
10x required format (<code>sample1_S1_L001_R1_001.fastq.gz</code> and
<code>sample1_S1_L001_R2_001.fastq.gz</code>). In detail, the file with
read length 26 or 28 is considered as read1, the files with read length
8 or 10 are considered as index files and the remain file is considered
as read2. The read length rules is from <a href="https://www.10xgenomics.com/cn/support/single-cell-gene-expression/documentation/steps/sequencing/sequencing-requirements-for-single-cell-3" class="external-link">Sequencing
Requirements for Single Cell 3’</a> and <a href="https://www.10xgenomics.com/cn/support/single-cell-immune-profiling/documentation/steps/sequencing/sequencing-requirements-for-single-cell-v-d-j" class="external-link">Sequencing
Requirements for Single Cell V(D)J</a>.</p>
<p>The returned value is a vector of failed <code>sra</code> files. If
not <code>NULL</code>, users can re-run <code>SplitSRA</code> by setting
<code>sra.path</code> to the returned value.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parallel-fastq-dump requires sratools.path</span></span>
<span><span class="va">GSE130636.split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SplitSRA.html">SplitSRA</a></span><span class="op">(</span></span>
<span>  sra.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_sra/prefetch"</span>,</span>
<span>  fastq.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  split.cmd.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/parallel-fastq-dump"</span>,</span>
<span>  sratools.path <span class="op">=</span> <span class="st">"~/data/software/sratoolkit.2.11.2-centos_linux64/bin"</span>, split.cmd.paras <span class="op">=</span> <span class="st">"--gzip"</span>,</span>
<span>  split.cmd.threads <span class="op">=</span> <span class="fl">4</span>, format.10x <span class="op">=</span> <span class="cn">TRUE</span>, remove.raw <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># remove the unformatted fastq files, used when format.10x = TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_sra/prefetch</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_sra/prefetch</span></span>
<span><span class="co"># └── GSM3745993</span></span>
<span><span class="co">#     ├── SRR9004346</span></span>
<span><span class="co">#     │   ├── SRR9004346_S1_L001_R1_001.fastq.gz</span></span>
<span><span class="co">#     │   ├── SRR9004346_S1_L001_R2_001.fastq.gz</span></span>
<span><span class="co">#     │   └── SRR9004346.sra</span></span>
<span><span class="co">#     └── SRR9004351</span></span>
<span><span class="co">#         ├── SRR9004351_S1_L001_R1_001.fastq.gz</span></span>
<span><span class="co">#         ├── SRR9004351_S1_L001_R2_001.fastq.gz</span></span>
<span><span class="co">#         └── SRR9004351.sra</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 3 directories, 6 files</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-fastq-directly-from-ena">Download fastq directly from ENA<a class="anchor" aria-label="anchor" href="#download-fastq-directly-from-ena"></a>
</h3>
<p>Alternatively, <code>GEfetch2R</code> provides function
<code>DownloadFastq</code> to download <code>fastq</code> files directly
from <code>ENA</code> (<strong>parallel, faster than the above
method</strong>). The returned value is a dataframe contains failed
runs. If not <code>NULL</code>, users can re-run
<code>DownloadFastq</code> by setting <code>gsm.df</code> to the
returned value.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use download.file</span></span>
<span><span class="va">GSE130636.down.fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_fastq/download_file"</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"download.file"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span>, format.10x <span class="op">=</span> <span class="cn">TRUE</span>, remove.raw <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># remove the unformatted fastq files, used when format.10x = TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_fastq/download_file</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_fastq/download_file</span></span>
<span><span class="co"># └── GSM3745993</span></span>
<span><span class="co">#     ├── SRR9004346</span></span>
<span><span class="co">#     │   ├── SRR9004346_S1_L001_R1_001.fastq.gz</span></span>
<span><span class="co">#     │   └── SRR9004346_S1_L001_R2_001.fastq.gz</span></span>
<span><span class="co">#     └── SRR9004351</span></span>
<span><span class="co">#         ├── SRR9004351_S1_L001_R1_001.fastq.gz</span></span>
<span><span class="co">#         └── SRR9004351_S1_L001_R2_001.fastq.gz</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 3 directories, 4 files</span></span>
<span></span>
<span><span class="co"># use ascp</span></span>
<span><span class="va">GSE130636.down.fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_fastq/ascp"</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"ascp"</span>, ascp.path <span class="op">=</span> <span class="st">"~/.aspera/connect/bin/ascp"</span>, max.rate <span class="op">=</span> <span class="st">"300m"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span>, format.10x <span class="op">=</span> <span class="cn">TRUE</span>, remove.raw <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># use wget</span></span>
<span><span class="va">GSE130636.down.fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadFastq.html">DownloadFastq</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_fastq/wget"</span>,</span>
<span>  download.method <span class="op">=</span> <span class="st">"wget"</span>, wget.path <span class="op">=</span> <span class="st">"/usr/bin/wget"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span>, format.10x <span class="op">=</span> <span class="cn">TRUE</span>, remove.raw <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="download-bam">Download bam<a class="anchor" aria-label="anchor" href="#download-bam"></a>
</h2>
<div class="section level3">
<h3 id="extract-all-samples-runs-1">Extract all samples (runs)<a class="anchor" aria-label="anchor" href="#extract-all-samples-runs-1"></a>
</h3>
<p><code>GEfetch2R</code> can extract sample information and run number
with <strong>GEO accessions</strong> or users can also <strong>provide a
dataframe containing the run number of interested samples</strong>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE138266"</span>, platform <span class="op">=</span> <span class="st">"GPL18573"</span><span class="op">)</span></span>
<span><span class="va">GSE138266.runs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"run"</span>, <span class="st">"experiment"</span>, <span class="st">"gsm_name"</span>, <span class="st">"title"</span>, <span class="st">"geo_accession"</span>, <span class="st">"source_name_ch1"</span>, <span class="st">"organism_ch1"</span>, <span class="st">"characteristics_ch1"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                     run experiment   gsm_name       title geo_accession     source_name_ch1</span></span>
<span><span class="co"># SRR10211551 SRR10211551 SRX6931239 GSM4104122 MS19270_CSF    GSM4104122 Cerebrospinal fluid</span></span>
<span><span class="co"># SRR10211552 SRR10211552 SRX6931240 GSM4104123 MS58637_CSF    GSM4104123 Cerebrospinal fluid</span></span>
<span><span class="co"># SRR10211553 SRR10211553 SRX6931241 GSM4104124 MS71658_CSF    GSM4104124 Cerebrospinal fluid</span></span>
<span><span class="co">#             organism_ch1                   characteristics_ch1</span></span>
<span><span class="co"># SRR10211551 Homo sapiens disease condition: Multiple Sclerosis</span></span>
<span><span class="co"># SRR10211552 Homo sapiens disease condition: Multiple Sclerosis</span></span>
<span><span class="co"># SRR10211553 Homo sapiens disease condition: Multiple Sclerosis</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-bam-from-sra">Download bam from SRA<a class="anchor" aria-label="anchor" href="#download-bam-from-sra"></a>
</h3>
<p>With the dataframe contains gsm and run number,
<code>GEfetch2R</code> provides <code>DownloadBam</code> to download
<code>bam</code> files using <code>prefetch</code>. It supports 10x
generated <code>bam</code> files and normal <code>bam</code> files.</p>
<ul>
<li>
<strong>10x generated bam</strong>: While <code>bam</code> files
generated from 10x softwares (e.g. CellRanger) contain custom tags which
are not kept when using default parameters of <code>prefetch</code>,
<code>GEfetch2R</code> adds <code>--type TenX</code> to make sure the
downloaded <code>bam</code> files contain these tags.</li>
<li>
<strong>normal bam</strong>: For normal bam files,
<code>DownloadBam</code> will download <code>sra</code> files first and
then convert <code>sra</code> files to <code>bam</code> files with
<code>sam-dump</code>. After testing the efficiency of
<code>prefetch</code> + <code>sam-dump</code> and <code>sam-dump</code>,
the former is much faster than the latter (52G <code>sra</code> and 72G
<code>bam</code> files):</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># # use prefetch to download sra file</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co"># prefetch -X 60G SRR1976036</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># # real    117m26.334s</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># # user    16m42.062s</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># # sys 3m28.295s</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># # use sam-dump to convert sra to bam</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036.sra | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co"># # real    536m2.721s</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co"># # user    749m41.421s</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co"># # sys 20m49.069s</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co"># use sam-dump to download bam directly</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036 | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co"># # more than 36hrs only get ~3G bam files, too slow</span></span></code></pre></div>
<p>The returned value is a dataframe containing failed runs (either
failed to download <code>sra</code> files or failed to convert to
<code>bam</code> files for normal <code>bam</code>; failed to download
<code>bam</code> files for 10x generated <code>bam</code>). If not
<code>NULL</code>, users can re-run <code>DownloadBam</code> by setting
<code>gsm.df</code> to the returned value. The following is an example
to download 10x generated <code>bam</code> file:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="va">GSE138266.runs</span><span class="op">[</span><span class="va">GSE138266.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR10211566"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download (slow, use wget instead)</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>,</span>
<span>  prefetch.path <span class="op">=</span> <span class="st">"~/data/software/sratoolkit.2.11.2-centos_linux64/bin/prefetch"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_bam"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE138266.down is null or dataframe contains failed runs</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-bam-from-ena">Download bam from ENA<a class="anchor" aria-label="anchor" href="#download-bam-from-ena"></a>
</h3>
<p>The returned value is a dataframe containing failed runs. If not
<code>NULL</code>, users can re-run <code>DownloadBam</code> by setting
<code>gsm.df</code> to the returned value. The following is an example
to download 10x generated <code>bam</code> file from
<code>ENA</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download.file</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>, download.method <span class="op">=</span> <span class="st">"download.file"</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">3600</span>, rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_bam/download_file"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ascp</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>, download.method <span class="op">=</span> <span class="st">"ascp"</span>,</span>
<span>  ascp.path <span class="op">=</span> <span class="st">"~/.aspera/connect/bin/ascp"</span>, max.rate <span class="op">=</span> <span class="st">"300m"</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_bam/ascp"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># wget</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>, download.method <span class="op">=</span> <span class="st">"wget"</span>,</span>
<span>  wget.path <span class="op">=</span> <span class="st">"/usr/bin/wget"</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  rename <span class="op">=</span> <span class="cn">TRUE</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_bam/wget"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>, use.cores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_bam/wget</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_bam/wget</span></span>
<span><span class="co"># └── GSM4104137</span></span>
<span><span class="co">#     └── SRR10211566</span></span>
<span><span class="co">#         ├── bamfiles_MS60249_PBMC_possorted_genome_bam_1.bam</span></span>
<span><span class="co">#         └── bamfiles_MS60249_PBMC_possorted_genome_bam_1.bam.bai</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 2 directories, 2 files</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="convert-bam-to-fastq">Convert bam to fastq<a class="anchor" aria-label="anchor" href="#convert-bam-to-fastq"></a>
</h3>
<p>With downloaded <code>bam</code> files, <code>GEfetch2R</code>
provides function <code>Bam2Fastq</code> to convert <code>bam</code>
files to <code>fastq</code> files. For <code>bam</code> files generated
from 10x softwares, <code>Bam2Fastq</code> utilizes
<code>bamtofastq</code> tool developed by 10x Genomics, otherwise,
<code>samtools</code> is utilized.</p>
<p>The returned value is a vector of <code>bam</code> files failed to
convert to <code>fastq</code> files. If not <code>NULL</code>, users can
re-run <code>Bam2Fastq</code> by setting <code>bam.path</code> to the
returned value.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Bam2Fastq.html">Bam2Fastq</a></span><span class="op">(</span></span>
<span>  bam.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_bam/wget"</span>,</span>
<span>  bam.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  bamtofastq.path <span class="op">=</span> <span class="st">"~/data/software/bamtofastq_linux"</span>,</span>
<span>  bamtofastq.paras <span class="op">=</span> <span class="st">"--nthreads 4"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_bam/wget</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_bam/wget</span></span>
<span><span class="co"># └── GSM4104137</span></span>
<span><span class="co">#     └── SRR10211566</span></span>
<span><span class="co">#         ├── bam2fastq                                                  # generated by Bam2Fastq, can remove</span></span>
<span><span class="co">#         │   └── MS60249_PBMC_2_0_MissingLibrary_1_H72VGBGX2</span></span>
<span><span class="co">#         ├── bamfiles_MS60249_PBMC_possorted_genome_bam_1.bam</span></span>
<span><span class="co">#         ├── bamfiles_MS60249_PBMC_possorted_genome_bam_1.bam.bai</span></span>
<span><span class="co">#         ├── SRR10211566_S1_L001_R1_001.fastq.gz                        # generated by Bam2Fastq</span></span>
<span><span class="co">#         └── SRR10211566_S1_L001_R2_001.fastq.gz                        # generated by Bam2Fastq</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 4 directories, 4 files</span></span>
<span></span>
<span><span class="co"># bulk RNA-seq</span></span>
<span><span class="va">GSE127942.convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Bam2Fastq.html">Bam2Fastq</a></span><span class="op">(</span></span>
<span>  bam.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/fastq2r2/mapping/bulkRNAseq"</span>,</span>
<span>  bam.type <span class="op">=</span> <span class="st">"other"</span>,</span>
<span>  bamtofastq.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/samtools"</span>,</span>
<span>  bamtofastq.paras <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/fastq2r2/mapping/bulkRNAseq</span></span>
<span><span class="co"># ~/gefetch2r/doc/fastq2r2/mapping/bulkRNAseq</span></span>
<span><span class="co"># ├── GSM3656922</span></span>
<span><span class="co"># │   ├── Aligned.sortedByCoord.out.bam</span></span>
<span><span class="co"># │   ├── Aligned.sortedByCoord.out.sortname.bam               # generate by Bam2Fastq</span></span>
<span><span class="co"># │   ├── bam2fastq                                            # generate by Bam2Fastq</span></span>
<span><span class="co"># │   │   ├── Aligned.sortedByCoord.out.sortname_1.fastq.gz    # generate by Bam2Fastq</span></span>
<span><span class="co"># │   │   └── Aligned.sortedByCoord.out.sortname_2.fastq.gz    # generate by Bam2Fastq</span></span>
<span><span class="co"># │   └── GSM3656922.txt</span></span>
<span><span class="co"># │   ......</span></span>
<span><span class="co"># └── GSM3656923</span></span>
<span><span class="co">#     ├── Aligned.sortedByCoord.out.bam</span></span>
<span><span class="co">#     ├── Aligned.sortedByCoord.out.sortname.bam               # generate by Bam2Fastq</span></span>
<span><span class="co">#     ├── bam2fastq                                            # generate by Bam2Fastq</span></span>
<span><span class="co">#     │   ├── Aligned.sortedByCoord.out.sortname_1.fastq.gz    # generate by Bam2Fastq</span></span>
<span><span class="co">#     │   └── Aligned.sortedByCoord.out.sortname_2.fastq.gz    # generate by Bam2Fastq</span></span>
<span><span class="co">#     └── GSM3656923.txt</span></span>
<span><span class="co">#     ......</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 16 directories, 18 files</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="load-fastq-to-r">Load fastq to R<a class="anchor" aria-label="anchor" href="#load-fastq-to-r"></a>
</h2>
<p>With downloaded/converted <code>fastq</code> files,
<code>GEfetch2R</code> provides function <code>Fastq2R</code> to align
them to reference genome with <a href="https://www.10xgenomics.com/cn/support/software/cell-ranger/latest/tutorials/cr-tutorial-ct" class="external-link">CellRanger</a>
(10x-generated <code>fastq</code> files) or <a href="https://github.com/alexdobin/STAR/tree/master" class="external-link">STAR</a>
(Smart-seq2 or bulk RNA-seq data), and load the output to
<code>Seurat</code> (10x-generated <code>fastq</code> files) or
<code>DESEq2</code> (Smart-seq2 or bulk RNA-seq data).</p>
<p>Here, we use the downloaded <code>fastq</code> files as an example.
There are two runs (<code>SRR9004346</code> and <code>SRR9004351</code>)
corresponding to sample name <code>GSM3745993</code>. When running
<code>CellRanger</code>, we will process <code>SRR9004346</code> and
<code>SRR9004351</code> as a single merged sample by specifying
<code>--sample=SRR9004346,SRR9004351</code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run CellRanger (10x Genomics)</span></span>
<span><span class="co"># the sample.dir corresponding to sra.folder/GSMXXXX (SplitSRA) or out.folder/GSMXXXX (DownloadFastq)</span></span>
<span><span class="va">GSE130636.gsms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"~/gefetch2r/doc/download_fastq/wget"</span>, <span class="st">"GSM3745993"</span><span class="op">)</span></span>
<span><span class="va">GSE130636.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Fastq2R.html">Fastq2R</a></span><span class="op">(</span></span>
<span>  sample.dir <span class="op">=</span> <span class="va">GSE130636.gsms</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"~/refdata-cellranger-GRCh38-3.0.0"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"CellRanger"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_fastq/cellranger"</span>,</span>
<span>  st.path <span class="op">=</span> <span class="st">"~/cellranger-3.0.2/cellranger"</span>,</span>
<span>  st.paras <span class="op">=</span> <span class="st">"--chemistry=auto --jobmode=local"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE130636.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 33538 features across 956 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (33538 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># # The output structure</span></span>
<span><span class="co"># tree -L 2 ~/gefetch2r/doc/download_fastq/cellranger</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_fastq/cellranger</span></span>
<span><span class="co"># └── GSM3745993</span></span>
<span><span class="co">#     ├── _cmdline</span></span>
<span><span class="co">#     ├── _filelist</span></span>
<span><span class="co">#     ├── _finalstate</span></span>
<span><span class="co">#     ├── GSM3745993.mri.tgz</span></span>
<span><span class="co">#     ├── _invocation</span></span>
<span><span class="co">#     ├── _jobmode</span></span>
<span><span class="co">#     ├── _log</span></span>
<span><span class="co">#     ├── _mrosource</span></span>
<span><span class="co">#     ├── outs</span></span>
<span><span class="co">#     ├── _perf</span></span>
<span><span class="co">#     ├── SC_RNA_COUNTER_CS</span></span>
<span><span class="co">#     ├── _sitecheck</span></span>
<span><span class="co">#     ├── _tags</span></span>
<span><span class="co">#     ├── _timestamp</span></span>
<span><span class="co">#     ├── _uuid</span></span>
<span><span class="co">#     ├── _vdrkill</span></span>
<span><span class="co">#     └── _versions</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 3 directories, 15 files</span></span>
<span></span>
<span></span>
<span><span class="co"># run STAR (Smart-seq2 or bulk RNA-seq)</span></span>
<span><span class="va">GSE127942.gsms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"~/gefetch2r/doc/fastq2r2/fastq/bulkRNAseq"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM3656922"</span>, <span class="st">"GSM3656923"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GSE127942.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Fastq2R.html">Fastq2R</a></span><span class="op">(</span></span>
<span>  sample.dir <span class="op">=</span> <span class="va">GSE127942.gsms</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"~/Mus_musculus/star"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"STAR"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_fastq/STAR"</span>,</span>
<span>  st.path <span class="op">=</span> <span class="st">"~/anaconda3/bin/STAR"</span>,</span>
<span>  st.paras <span class="op">=</span> <span class="st">"--outBAMsortingThreadN 4 --twopassMode None"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE127942.obj</span></span>
<span><span class="co"># class: DESeqDataSet</span></span>
<span><span class="co"># dim: 55573 2</span></span>
<span><span class="co"># metadata(1): version</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(55573): ENSMUSG00000000001 ENSMUSG00000000003 ...</span></span>
<span><span class="co">#   ENSMUSG00000118487 ENSMUSG00000118488</span></span>
<span><span class="co"># rowData names(0):</span></span>
<span><span class="co"># colnames(2): GSM3656922 GSM3656923</span></span>
<span><span class="co"># colData names(1): condition</span></span></code></pre></div>
<hr>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
