<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>DownloadMatrices • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadMatrices">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DownloadMatrices</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/DownloadMatrices.Rmd" class="external-link"><code>vignettes/DownloadMatrices.Rmd</code></a></small>
      <div class="d-none name"><code>DownloadMatrices.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> provides functions for users to download
<strong>count matrices</strong> and <strong>annotations</strong>
(e.g. cell type annotation and composition) from GEO and some
single-cell databases (e.g. <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> and <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>).
<code>GEfetch2R</code> also supports loading the downloaded data to
<code>Seurat</code>.</p>
<p>Until now, the public resources supported and the returned
values:</p>
<table class="table">
<thead><tr>
<th>
Resources
</th>
<th>
URL
</th>
<th>
Download Type
</th>
<th>
Returned values
</th>
</tr></thead>
<tbody>
<tr>
<td>
GEO
</td>
<td>
<a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a>
</td>
<td>
count matrix
</td>
<td>
SeuratObject (scRNA-seq) or DESeqDataSet (bulk RNA-seq)
</td>
</tr>
<tr>
<td>
PanglaoDB
</td>
<td>
<a href="https://panglaodb.se/index.html" class="external-link uri">https://panglaodb.se/index.html</a>
</td>
<td>
count matrix and annotation
</td>
<td>
SeuratObject
</td>
</tr>
<tr>
<td>
UCSC Cell Browser
</td>
<td>
<a href="https://cells.ucsc.edu/" class="external-link uri">https://cells.ucsc.edu/</a>
</td>
<td>
count matrix and annotation
</td>
<td>
SeuratObject
</td>
</tr>
</tbody>
</table>
<hr>
</div>
<div class="section level2">
<h2 id="check-api">Check API<a class="anchor" aria-label="anchor" href="#check-api"></a>
</h2>
<p>Check the availability of APIs used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/CheckAPI.html">CheckAPI</a></span><span class="op">(</span>database <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GEO"</span>, <span class="st">"PanglaoDB"</span>, <span class="st">"UCSC Cell Browser"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># start checking APIs to access GEO!</span></span>
<span><span class="co"># The API to access the GEO object is OK!</span></span>
<span><span class="co"># trying URL 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE302nnn/GSE302912/suppl//GSE302912_counts.csv.gz?tool=geoquery'</span></span>
<span><span class="co"># Content type 'application/x-gzip' length 332706 bytes (324 KB)</span></span>
<span><span class="co"># ==================================================</span></span>
<span><span class="co"># downloaded 324 KB</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># The API to access supplementary files is OK!</span></span>
<span><span class="co"># start checking APIs to access PanglaoDB!</span></span>
<span><span class="co"># The API to access all available samples is OK!</span></span>
<span><span class="co"># Processing given dataset(s): SRA429320, SRS1467249</span></span>
<span><span class="co"># The API to access cell type composition is OK!</span></span>
<span><span class="co">#   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=01m 00s</span></span>
<span><span class="co"># start checking APIs to access UCSC Cell Browser!</span></span>
<span><span class="co"># The API to access all available projects is OK!</span></span>
<span><span class="co"># The API to access detailed information of a given dataset is OK!</span></span>
<span><span class="co"># The API to access the available data of a given dataset is OK!</span></span>
<span><span class="co"># The API to access available files is OK!</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is
an international public repository that archives and freely distributes
microarray, next-generation sequencing, and other forms of
high-throughput functional genomics data submitted by the research
community.</a> It provides a very convenient way for users to explore
and select interested scRNA-seq datasets.</p>
<div class="section level3">
<h3 id="extract-metadata-optional">Extract metadata (optional)<a class="anchor" aria-label="anchor" href="#extract-metadata-optional"></a>
</h3>
<p><code>ExtractGEOMeta</code> provides two ways to extract sample
metadata:</p>
<ul>
<li>
<strong>user-provided sample metadata when uploading to GEO
(applicable to all GEO accessions)</strong>, including sample title,
source name/tissue, description, cell type, treatment, paper title,
paper abstract, organism, protocol, data processing methods, et al:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">GEfetch2R</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata</span></span>
<span><span class="va">GSE297431.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE297431"</span><span class="op">)</span></span>
<span><span class="va">GSE297431.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"title"</span>, <span class="st">"geo_accession"</span>, <span class="st">"source_name_ch1"</span>, <span class="st">"description"</span>, <span class="st">"cell type"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                        title geo_accession source_name_ch1                       description</span></span>
<span><span class="co"># 1  MOS_mutant_M3_replicate_4    GSM8991251          Oocyte   Library name: Plate1_mut_A1_S70</span></span>
<span><span class="co"># 2  MOS_mutant_M3_replicate_5    GSM8991252          Oocyte Library name: Plate1_mut_A11_S135</span></span>
<span><span class="co"># 3 MOS_mutant_M1_replicate_10    GSM8991253          Oocyte Library name: Plate1_mut_A12_S141</span></span>
<span><span class="co">#   cell type</span></span>
<span><span class="co"># 1    Oocyte</span></span>
<span><span class="co"># 2    Oocyte</span></span>
<span><span class="co"># 3    Oocyte</span></span></code></pre></div>
<ul>
<li>
<strong>metadata in supplementary file</strong>:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE297431.meta.supp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE297431"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">2</span> <span class="co"># specify the index of used supplementary file</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE297431.meta.supp</span><span class="op">)</span></span>
<span><span class="co">#             Sample_ID  Batch  Plate   Type Growths Class</span></span>
<span><span class="co"># 1   Plate1_mut_A1_S70 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 2 Plate1_mut_A11_S135 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 3 Plate1_mut_A12_S141 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 4   Plate1_mut_A2_S77 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 5   Plate1_mut_A3_S84 batch2 Plate1 mutant       2    M2</span></span>
<span><span class="co"># 6   Plate1_mut_A4_S91 batch2 Plate1 mutant       2    M2</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-matrix-and-load-to-seuratdeseq2">Download matrix and load to Seurat/DESeq2<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seuratdeseq2"></a>
</h3>
<p>After manually check the extracted metadata, users can
<strong>download count matrix</strong> and <strong>load the count
matrix</strong> to <code>Seurat</code>/<code>DESeq2</code> with
<code>ParseGEO</code>.</p>
<p>For count matrix, <code>ParseGEO</code> supports downloading the
matrix from <strong>supplementary files</strong> and extracting from
<code>ExpressionSet</code>, users can control the source by specifying
<code>down.supp</code> or detecting automatically (<code>ParseGEO</code>
will extract the count matrix from <code>ExpressionSet</code> first, if
the count matrix is <code>NULL</code> or contains non-integer values,
<code>ParseGEO</code> will download supplementary files).</p>
<div class="section level4">
<h4 id="smart-seq2-scrna-seq-or-bulk-rna-seq">Smart-seq2 scRNA-seq or Bulk RNA-seq<a class="anchor" aria-label="anchor" href="#smart-seq2-scrna-seq-or-bulk-rna-seq"></a>
</h4>
<p>For Smart-seq2 scRNA-seq or bulk RNA-seq, the supplementary files
have two formats:</p>
<ul>
<li>single file
(<code>csv(.gz)</code>/<code>tsv(.gz)</code>/<code>txt(.gz)</code>/<code>tab(.gz)</code>/<code>xlsx(.gz)</code>/<code>xls(.gz)</code>)
contain the count matrix of all samples, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE297431" class="external-link">GSE297431-Smart-seq2</a>
</li>
<li>(gzip) archive (contain
<code>csv(.gz)</code>/<code>tsv(.gz)</code>/<code>txt(.gz)</code>/<code>tab(.gz)</code>/<code>xlsx(.gz)</code>/<code>xls(.gz)</code>
files) contain the count matrix of every sample, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE241004" class="external-link">GSE241004</a>
</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single file (Smart-seq2)</span></span>
<span><span class="va">GSE297431.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE297431"</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE297431.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 21564 features across 107 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (21564 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># gzip archive of files (Smart-seq2)</span></span>
<span><span class="va">GSE241004.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE241004"</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE241004.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 61911 features across 30 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (61911 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># gzip archive of files (bulk RNA-seq)</span></span>
<span><span class="va">GSE223226.dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE223226"</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>  data.type <span class="op">=</span> <span class="st">"bulk"</span> <span class="co"># set data.type = "bulk", return DESeqDataSet</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE223226.dds</span></span>
<span><span class="co"># class: DESeqDataSet</span></span>
<span><span class="co"># dim: 32525 20</span></span>
<span><span class="co"># metadata(1): version</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(32525): __alignment_not_unique __ambiguous ... ENSDARG00000117826</span></span>
<span><span class="co">#   ENSDARG00000117827</span></span>
<span><span class="co"># rowData names(0):</span></span>
<span><span class="co"># colnames(20): GSM6943634.x GSM6943634.y ... GSM6943643.x GSM6943643.y</span></span>
<span><span class="co"># colData names(1): condition</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="scrna-seq">scRNA-seq<a class="anchor" aria-label="anchor" href="#scrna-seq"></a>
</h4>
<p>For 10x Genomics (or similiar) scRNA-seq, the supplementary files
have two formats:</p>
<ul>
<li>
<code>h5(.gz)</code> file(s) or
<code>barcodes.tsv(.gz)</code>/<code>genes.tsv(.gz)</code>,
<code>matrix.mtx(.gz)</code>, <code>features.tsv(.gz)</code> files
contain the count matrix, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE278892" class="external-link">GSE278892</a>
</li>
<li>archive (<code>tar(.gz)</code>), the files in archive can be in
<code>zip</code>, <code>tar(.gz)</code>, <code>h5(.gz)</code> format, or
separate files
(<code>barcodes.tsv(.gz)</code>/<code>genes.tsv(.gz)</code>,
<code>matrix.mtx(.gz)</code>, <code>features.tsv(.gz)</code>), e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE271836" class="external-link">GSE271836-h5</a>,
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE292908" class="external-link">GSE292908-separate
files</a>, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE253859" class="external-link">GSE253859-zip</a>,
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE274058" class="external-link">GSE274058-tar.gz</a>
</li>
</ul>
<p>With the count matrix, <code>ParseGEO</code> will load the matrix to
<code>Seurat</code> automatically. If multiple samples available, users
can choose to merge the <code>SeuratObject</code> with
<code>merge</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># separate files</span></span>
<span><span class="va">GSE278892.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE278892"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="st">"10xSingle"</span>, timeout <span class="op">=</span> <span class="fl">36000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_geo"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE278892.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 22548 features across 6079 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (22548 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># archive of files</span></span>
<span><span class="va">GSE292908.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE292908"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="st">"10x"</span>, timeout <span class="op">=</span> <span class="fl">36000</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, <span class="co"># specify the index of used supplementary file</span></span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_geo"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE292908.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 32285 features across 16062 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (32285 features, 0 variable features)</span></span>
<span></span>
<span><span class="co"># # The structure of downloaded count matrix for 10x</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_geo</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_geo</span></span>
<span><span class="co"># ├── GSE278892</span></span>
<span><span class="co"># │   └── GSE278892_BFPRFP</span></span>
<span><span class="co"># │       ├── barcodes.tsv.gz</span></span>
<span><span class="co"># │       ├── features.tsv.gz</span></span>
<span><span class="co"># │       └── matrix.mtx.gz</span></span>
<span><span class="co"># └── GSE292908</span></span>
<span><span class="co">#     ├── GSM8869153_2151R_CTX</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869154_2151R_CTX_SNDX_ms6352</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869155_2151R_IgG</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869156_2151R_SNDX_ms6352</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869157_T12_CTX</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869158_T12_CTX_SNDX_ms6352</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     ├── GSM8869159_T12_IgG</span></span>
<span><span class="co">#     │   ├── barcodes.tsv.gz</span></span>
<span><span class="co">#     │   ├── features.tsv.gz</span></span>
<span><span class="co">#     │   └── matrix.mtx.gz</span></span>
<span><span class="co">#     └── GSM8869160_T12_SNDX_ms6352</span></span>
<span><span class="co">#         ├── barcodes.tsv.gz</span></span>
<span><span class="co">#         ├── features.tsv.gz</span></span>
<span><span class="co">#         └── matrix.mtx.gz</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 11 directories, 27 files</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="panglaodb">PanglaoDB<a class="anchor" aria-label="anchor" href="#panglaodb"></a>
</h2>
<p><a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> is a database
which contains scRNA-seq datasets from mouse and human. Up to now, it
contains <strong>5,586,348 cells</strong> from <strong>1368 datasets
(1063 from Mus musculus and 305 from Homo sapiens)</strong>. It has well
organized metadata for every dataset, including tissue, protocol,
species, number of cells and cell type annotation (computationally
identified). Daniel Osorio has developed <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a> to access
<a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> data, the
functions of <code>GEfetch2R</code> here are based on <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a>.</p>
<p>Since <a href="https://panglaodb.se/about.html" class="external-link">PanglaoDB</a> is no
longer maintained, <code>GEfetch2R</code> has cached all metadata and
cell type composition and use these cached data by default to
accelerate, users can access the cached data with
<code>PanglaoDBMeta</code> (all metadata) and
<code>PanglaoDBComposition</code> (all cell type composition).</p>
<div class="section level3">
<h3 id="given-dataset">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset"></a>
</h3>
<p>With <code>SRA</code> or <code>SRS</code> accession, user can access
the <strong>cell type composition</strong> and <strong>count
matrix</strong>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cell type composition</span></span>
<span><span class="va">lung.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>sra <span class="op">=</span> <span class="st">"SRA570744"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lung.composition</span><span class="op">)</span></span>
<span><span class="co">#           SRA        SRS          Tissue     Protocol      Species Cluster Cells</span></span>
<span><span class="co"># 2.1 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       0  1606</span></span>
<span><span class="co"># 2.2 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       1   874</span></span>
<span><span class="co"># 2.3 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       2   762</span></span>
<span><span class="co"># 2.4 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       3   627</span></span>
<span><span class="co"># 2.5 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       4   400</span></span>
<span><span class="co"># 2.6 SRA570744 SRS2253536 Lung mesenchyme 10x chromium Mus musculus       5    74</span></span>
<span><span class="co">#               Cell Type</span></span>
<span><span class="co"># 2.1         Fibroblasts</span></span>
<span><span class="co"># 2.2         Fibroblasts</span></span>
<span><span class="co"># 2.3 Smooth muscle cells</span></span>
<span><span class="co"># 2.4         Fibroblasts</span></span>
<span><span class="co"># 2.5 Smooth muscle cells</span></span>
<span><span class="co"># 2.6   Mesothelial cells、</span></span>
<span></span>
<span><span class="co"># count matrix</span></span>
<span><span class="va">lung.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span>sra <span class="op">=</span> <span class="st">"SRA570744"</span>, srs <span class="op">=</span> <span class="st">"SRS2253536"</span><span class="op">)</span></span>
<span><span class="va">lung.seu</span></span>
<span><span class="co"># $SRS2253536</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 23411 features across 4374 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (23411 features, 0 variable features)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="filter-samples-based-on-metadata">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata"></a>
</h3>
<div class="section level4">
<h4 id="summary-attributes">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use cached metadata</span></span>
<span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span>
<span><span class="co"># $species</span></span>
<span><span class="co">#          Value  Num     Key</span></span>
<span><span class="co"># 1 Mus musculus 1063 species</span></span>
<span><span class="co"># 2 Homo sapiens  305 species</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $protocol</span></span>
<span><span class="co">#           Value  Num      Key</span></span>
<span><span class="co"># 1  10x chromium 1046 protocol</span></span>
<span><span class="co"># 2      drop-seq  204 protocol</span></span>
<span><span class="co"># 3 microwell-seq   74 protocol</span></span>
<span><span class="co"># 4    Smart-seq2   26 protocol</span></span>
<span><span class="co"># 5   C1 Fluidigm   16 protocol</span></span>
<span><span class="co"># 6       CEL-seq    1 protocol</span></span>
<span><span class="co"># 7       inDrops    1 protocol</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-metadata">Filter metadata<a class="anchor" aria-label="anchor" href="#filter-metadata"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractPanglaoDBMeta</code> to
filter interested datasets with specified <strong>species</strong>,
<strong>protocol</strong>, <strong>tissue</strong> and <strong>cell
number</strong> (The available values of these attributes can be
obtained with <code>StatDBAttribute</code>). User can also choose to
whether to add cell type annotation to every dataset with
<code>show.cell.type</code>.</p>
<p><code>GEfetch2R</code> uses cached metadata and cell type composition
by default, users can change this by setting
<code>local.data = FALSE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>,</span>
<span>  show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">)</span></span>
<span><span class="co">#         SRA        SRS                             Tissue     Protocol      Species Cells</span></span>
<span><span class="co"># 1 SRA550660 SRS2089635 Peripheral blood mononuclear cells 10x chromium Homo sapiens  1860</span></span>
<span><span class="co"># 2 SRA550660 SRS2089636 Peripheral blood mononuclear cells 10x chromium Homo sapiens  1580</span></span>
<span><span class="co"># 3 SRA550660 SRS2089638 Peripheral blood mononuclear cells 10x chromium Homo sapiens  1818</span></span>
<span><span class="co"># 4 SRA605365 SRS2492922            Nasal airway epithelium 10x chromium Homo sapiens  1932</span></span>
<span><span class="co"># 5 SRA608611 SRS2517316                   Lung progenitors 10x chromium Homo sapiens  1077</span></span>
<span><span class="co"># 6 SRA608353 SRS2517519           Hepatocellular carcinoma 10x chromium Homo sapiens  1230</span></span>
<span><span class="co">#                                                                      CellType CellNum</span></span>
<span><span class="co"># 1                                                           Unknown, NK cells    1860</span></span>
<span><span class="co"># 2                              Unknown, T cells, Plasmacytoid dendritic cells    1580</span></span>
<span><span class="co"># 3 Unknown, Gamma delta T cells, Dendritic cells, Plasmacytoid dendritic cells    1818</span></span>
<span><span class="co"># 4       Luminal epithelial cells, Basal cells, Keratinocytes, Ependymal cells    1932</span></span>
<span><span class="co"># 5                                           Unknown, Hepatocytes, Basal cells    1077</span></span>
<span><span class="co"># 6                                        Unknown, Hepatocytes, Foveolar cells    1230</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-cell-type-composition">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition"></a>
</h4>
<p><code>GEfetch2R</code> provides
<code>ExtractPanglaoDBComposition</code> to extract cell type annotation
and composition (use cached data by default to accelerate, users can
change this by setting <code>local.data = FALSE</code>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">hsa.meta</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.composition</span><span class="op">)</span></span>
<span><span class="co">#            SRA        SRS           Tissue     Protocol      Species Cluster Cells   Cell Type</span></span>
<span><span class="co"># 13.1 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       0   214     Unknown</span></span>
<span><span class="co"># 13.2 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       1   210 Hepatocytes</span></span>
<span><span class="co"># 13.3 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       2   175 Basal cells</span></span>
<span><span class="co"># 13.4 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       3   121 Basal cells</span></span>
<span><span class="co"># 13.5 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       4    81     Unknown</span></span>
<span><span class="co"># 13.6 SRA608611 SRS2517316 Lung progenitors 10x chromium Homo sapiens       5    80     Unknown</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-matrix-and-load-to-seurat">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat"></a>
</h4>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParsePanglaoDB</code> to <strong>download count
matrix</strong> and <strong>load the count matrix</strong> to
<code>Seurat</code>. With available cell type annotation, uses can
filter datasets without specified cell type with <code>cell.type</code>.
Users can also include/exclude cells expressing specified genes with
<code>include.gene</code>/<code>exclude.gene</code>.</p>
<p>With the count matrix, <code>ParsePanglaoDB</code> will load the
matrix to <code>Seurat</code> automatically. If multiple datasets
available, users can choose to merge the <code>SeuratObject</code> with
<code>merge</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">hsa.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 25917 features across 4996 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (25917 features, 0 variable features)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="ucsc-cell-browser">UCSC Cell Browser<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser"></a>
</h2>
<p>The <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a> is a
web-based tool that allows scientists to interactively visualize
scRNA-seq datasets. It contains <strong>1427 single cell
datasets</strong> from <strong>37 different species</strong>. And, it is
<strong>organized with the hierarchical structure</strong>, which can
help users quickly locate the datasets they are interested in.</p>
<div class="section level3">
<h3 id="given-dataset-1">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset-1"></a>
</h3>
<p>With the collection or dataset link(s), user can access the
<strong>cell type composition</strong> and <strong>count
matrix</strong>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract cell type composition</span></span>
<span><span class="va">ut.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span>link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-ureter"</span>, <span class="co"># collection</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-testis"</span> <span class="co"># dataset</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ut.sample.ct</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"title"</span>, <span class="st">"CellType"</span>, <span class="st">"Num"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                                               title                    CellType  Num</span></span>
<span><span class="co"># 1 The adult human testis transcriptional cell atlas                       Sperm 1885</span></span>
<span><span class="co"># 2 The adult human testis transcriptional cell atlas        Elongated Spermatids  775</span></span>
<span><span class="co"># 3 The adult human testis transcriptional cell atlas                Leydig cells  605</span></span>
<span><span class="co"># 4 The adult human testis transcriptional cell atlas Early Primary Spermatocytes  557</span></span>
<span><span class="co"># 5 The adult human testis transcriptional cell atlas            Round Spermatids  444</span></span>
<span></span>
<span><span class="co"># extract count matrix and load to Seurat</span></span>
<span><span class="va">ut.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-ureter"</span>, <span class="co"># collection</span></span>
<span>  <span class="st">"https://cells.ucsc.edu/?ds=adult-testis"</span> <span class="co"># dataset</span></span>
<span><span class="op">)</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ut.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 43370 features across 79053 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (43370 features, 0 variable features)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="filter-samples-based-on-metadata-1">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata-1"></a>
</h3>
<div class="section level4">
<h4 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ShowCBDatasets</code> to show
all available datasets. Due to the large number of datasets,
<code>ShowCBDatasets</code> enables users to perform <em>lazy load</em>
of dataset json files instead of downloading the json files online
(time-consuming!!!). This <em>lazy load</em> requires users to provide
<code>json.folder</code> to save json files and set
<code>lazy = TRUE</code> (for the first time of run,
<code>ShowCBDatasets</code> will download current json files to
<code>json.folder</code>, for next time of run, with
<code>lazy = TRUE</code>, <code>ShowCBDatasets</code> will load the
downloaded json files from <code>json.folder</code>.). And,
<code>ShowCBDatasets</code> supports updating the local datasets with
<code>update = TRUE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first time run, the json files are stored under json.folder</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "~/gefetch2r/doc/cell_browser/json", update = TRUE)</span></span>
<span></span>
<span><span class="co"># second time run, load the downloaded json files</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/cell_browser/json"</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># always read online</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)</span></span></code></pre></div>
<p>The number of datasets and all available species:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the number of datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span>
<span><span class="co"># 1427</span></span>
<span></span>
<span><span class="co"># available species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\|parent"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">ucsc.cb.samples</span><span class="op">$</span><span class="va">organisms</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  [1] "human (h. sapiens)"                      "mouse (m. musculus)"</span></span>
<span><span class="co">#  [3] "rhesus macaque (m. mulatta)"             "chimp (p. troglodytes)"</span></span>
<span><span class="co">#  [5] "brine shrimp (a. franciscana)"           "canis lupus familiaris"</span></span>
<span><span class="co">#  [7] "dog (c. familiaris)"                     "canis familiaris"</span></span>
<span><span class="co">#  [9] "sea squirt (c. intestinalis)"            "sea squirt (c. robusta)"</span></span>
<span><span class="co"># [11] "macaque (m. fascicularis)"               "rabbit (o. cuniculus)"</span></span>
<span><span class="co"># [13] "rat (r. norvegicus)"                     "pig (s. scrofa)"</span></span>
<span><span class="co"># [15] "ferret (m. putorius furo)"               "opossum (m. domestica)"</span></span>
<span><span class="co"># [17] "sugar glider (p. breviceps)"             "zebrafish (d. rerio)"</span></span>
<span><span class="co"># [19] "fruit fly (d. melanogaster)"             "horse (e. caballus)"</span></span>
<span><span class="co"># [21] "freshwater hydra (h. vulgaris)"          "capitellid worm (c. teleta)"</span></span>
<span><span class="co"># [23] "freshwater sponge (s. lacustris)"        "colonial hydroid (h. symbiolongicarpus)"</span></span>
<span><span class="co"># [25] "western clawed frog (x. tropicalis)"     "marmoset (c. jacchus)"</span></span>
<span><span class="co"># [27] "mosquito (a. aegypti)"                   "pacific oyster (crassostrea gigas)"</span></span>
<span><span class="co"># [29] "freshwater worm (p. leidyi)"             "bonobo (p. paniscus)"</span></span>
<span><span class="co"># [31] "blood fluke (s. mansoni)"                "starlet sea anemone (n. vectensis)"</span></span>
<span><span class="co"># [33] "nematostella vectensis"                  "sea urchin (s. purpuratus)"</span></span>
<span><span class="co"># [35] "mouse lemur (t. microcebus)"             "human-mouse xenograft"</span></span>
<span><span class="co"># [37] "african clawed frog (x. laevis)"</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="summary-attributes-1">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell
Browser</a>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"UCSC"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 6 × 3</span></span>
<span><span class="co"># # Groups:   organisms [2]</span></span>
<span><span class="co">#   organisms           body_parts   Num</span></span>
<span><span class="co">#   &lt;chr&gt;               &lt;chr&gt;      &lt;int&gt;</span></span>
<span><span class="co"># 1 human (h. sapiens)  eye          350</span></span>
<span><span class="co"># 2 human (h. sapiens)  retina       348</span></span>
<span><span class="co"># 3 human (h. sapiens)  brain        198</span></span>
<span><span class="co"># 4 mouse (m. musculus) brain        107</span></span>
<span><span class="co"># 5 human (h. sapiens)  lung          72</span></span>
<span><span class="co"># 6 human (h. sapiens)  cortex        50</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-metadata-1">Filter metadata<a class="anchor" aria-label="anchor" href="#filter-metadata-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCBDatasets</code> to
filter metadata with <strong>collection</strong>,
<strong>sub-collection</strong>, <strong>organ</strong>, <strong>disease
status</strong>, <strong>organism</strong>, <strong>project</strong> and
<strong>cell number</strong> (The available values of these attributes
can be obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>). All attributes except cell number support fuzzy match
with <code>fuzzy.match</code>, this is useful when selecting
datasets.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBDatasets.html">ExtractCBDatasets</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, organ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"skeletal muscle"</span><span class="op">)</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Human (H. sapiens)"</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hbb.sample.df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"title"</span>, <span class="st">"body_parts"</span>, <span class="st">"diseases"</span>, <span class="st">"organisms"</span>, <span class="st">"sampleCount"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                                title                     body_parts       diseases</span></span>
<span><span class="co"># 1 Embryonic Week 7-8 Myogenic Subset muscle, skeletal muscle|parent Healthy|parent</span></span>
<span><span class="co"># 2   Fetal Week 12-14 Hindlimb Muscle muscle, skeletal muscle|parent Healthy|parent</span></span>
<span><span class="co"># 3         HX Protocol Week 4 Culture muscle, skeletal muscle|parent Healthy|parent</span></span>
<span><span class="co"># 4 HX Protocol Week 6 Myogenic Subset muscle, skeletal muscle|parent Healthy|parent</span></span>
<span><span class="co"># 5 HX Protocol Week 8 Myogenic Subset muscle, skeletal muscle|parent Healthy|parent</span></span>
<span><span class="co">#                   organisms sampleCount</span></span>
<span><span class="co"># 1 Human (H. sapiens)|parent        1448</span></span>
<span><span class="co"># 2 Human (H. sapiens)|parent        1545</span></span>
<span><span class="co"># 3 Human (H. sapiens)|parent        1562</span></span>
<span><span class="co"># 4 Human (H. sapiens)|parent        1598</span></span>
<span><span class="co"># 5 Human (H. sapiens)|parent        1350</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-cell-type-composition-1">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCBComposition</code> to
extract cell type annotation and composition.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span></span>
<span>  json.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/cell_browser/json"</span>,</span>
<span>  meta <span class="op">=</span> <span class="va">hbb.sample.df</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hbb.sample.ct</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"title"</span>, <span class="st">"CellType"</span>, <span class="st">"Num"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#                                title  CellType Num</span></span>
<span><span class="co"># 1 Embryonic Week 7-8 Myogenic Subset        MP 785</span></span>
<span><span class="co"># 2 Embryonic Week 7-8 Myogenic Subset        MB 303</span></span>
<span><span class="co"># 3 Embryonic Week 7-8 Myogenic Subset SkM.Mesen 264</span></span>
<span><span class="co"># 4 Embryonic Week 7-8 Myogenic Subset        MC  96</span></span>
<span><span class="co"># 5   Fetal Week 12-14 Hindlimb Muscle       MSC 822</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="load-the-online-datasets-to-seurat">Load the online datasets to Seurat<a class="anchor" aria-label="anchor" href="#load-the-online-datasets-to-seurat"></a>
</h4>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParseCBDatasets</code> to <strong>load the online count
matrix</strong> to <code>Seurat</code>. All the attributes available in
<code>ExtractCBDatasets</code> are also same here. Please note that the
loading process provided by <code>ParseCBDatasets</code> will load the
online count matrix instead of downloading it to local. If multiple
datasets available, users can choose to merge the
<code>SeuratObject</code> with <code>merge</code>.</p>
<p><code>ParseCBDatasets</code> supports <strong>extracting subset with
metadata and gene</strong>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parse the whole datasets</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>meta <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span>
<span><span class="co"># subset metadata and gene</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">hbb.sample.df</span>, obs.value.filter <span class="op">=</span> <span class="st">"Cell.Type == 'MP' &amp; Phase == 'G2M'"</span>,</span>
<span>  include.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"PAX7"</span>, <span class="st">"MYF5"</span>, <span class="st">"C1QTNF3"</span>, <span class="st">"MYOD1"</span>, <span class="st">"MYOG"</span>, <span class="st">"RASSF4"</span>, <span class="st">"MYH3"</span>, <span class="st">"MYL4"</span>,</span>
<span>    <span class="st">"TNNT3"</span>, <span class="st">"PDGFRA"</span>, <span class="st">"OGN"</span>, <span class="st">"COL3A1"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hbb.sample.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 14 features across 5684 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (14 features, 0 variable features)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
