<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>DownloadObjects • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadObjects">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DownloadObjects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/DownloadObjects.Rmd" class="external-link"><code>vignettes/DownloadObjects.Rmd</code></a></small>
      <div class="d-none name"><code>DownloadObjects.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> provides functions for users to download
processed single-cell RNA-seq data from <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO</a>, <a href="https://zenodo.org/" class="external-link">Zenodo</a>, <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> and <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>, including
files in <code>RDS</code>, <code>RData</code>, <code>h5ad</code>,
<code>h5</code>, <code>loom</code> formats.</p>
<p>Until now, the public resources supported and the returned
values:</p>
<table class="table">
<thead><tr>
<th>
Resources
</th>
<th>
URL
</th>
<th>
Download Type
</th>
<th>
Returned values
</th>
</tr></thead>
<tbody>
<tr>
<td>
GEO
</td>
<td>
<a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a>
</td>
<td>
rds, rdata, h5ad, loom
</td>
<td>
SeuratObject(rds) or failed datasets
</td>
</tr>
<tr>
<td>
Zenodo
</td>
<td>
<a href="https://zenodo.org/" class="external-link uri">https://zenodo.org/</a>
</td>
<td>
count matrix, rds, rdata, h5ad, et al.
</td>
<td>
SeuratObject(rds) or failed datasets
</td>
</tr>
<tr>
<td>
CELLxGENE
</td>
<td>
<a href="https://cellxgene.cziscience.com/" class="external-link uri">https://cellxgene.cziscience.com/</a>
</td>
<td>
rds, h5ad
</td>
<td>
SeuratObject(rds) or failed datasets
</td>
</tr>
<tr>
<td>
Human Cell Atlas
</td>
<td>
<a href="https://www.humancellatlas.org/" class="external-link uri">https://www.humancellatlas.org/</a>
</td>
<td>
rds, rdata, h5, h5ad, loom
</td>
<td>
SeuratObject(rds) or failed projects
</td>
</tr>
</tbody>
</table>
<hr>
</div>
<div class="section level2">
<h2 id="check-api">Check API<a class="anchor" aria-label="anchor" href="#check-api"></a>
</h2>
<p>Check the availability of APIs used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/CheckAPI.html">CheckAPI</a></span><span class="op">(</span>database <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GEO"</span>, <span class="st">"Zenodo"</span>, <span class="st">"CELLxGENE"</span>, <span class="st">"Human Cell Atlas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># start checking APIs to access GEO!</span></span>
<span><span class="co"># The API to access the GEO object is OK!</span></span>
<span><span class="co"># trying URL 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE302nnn/GSE302912/suppl//GSE302912_counts.csv.gz?tool=geoquery'</span></span>
<span><span class="co"># Content type 'application/x-gzip' length 332706 bytes (324 KB)</span></span>
<span><span class="co"># ==================================================</span></span>
<span><span class="co"># downloaded 324 KB</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># The API to access supplementary files is OK!</span></span>
<span><span class="co"># start checking APIs to access Zenodo!</span></span>
<span><span class="co"># The API to access detailed information of a given doi is OK!</span></span>
<span><span class="co"># The API to access available files is OK!</span></span>
<span><span class="co"># start checking APIs to access CELLxGENE!</span></span>
<span><span class="co"># The API to access all available collections is OK!</span></span>
<span><span class="co"># The API to access detailed information of a given collection is OK!</span></span>
<span><span class="co"># The API to access available files is OK!</span></span>
<span><span class="co"># The API to access detailed information of a given dataset is OK!</span></span>
<span><span class="co"># start checking APIs to access Human Cell Atlas!</span></span>
<span><span class="co"># The API to access all available catalogs is OK!</span></span>
<span><span class="co"># The API to access all available projects is OK!</span></span>
<span><span class="co"># The API to access available files is OK!</span></span>
<span><span class="co"># The API to access detailed information of a given project is OK!</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is
an international public repository that archives and freely distributes
microarray, next-generation sequencing, and other forms of
high-throughput functional genomics data submitted by the research
community.</a> It provides a very convenient way for users to explore
and select interested scRNA-seq datasets. <strong>Nowadays, in addition
to the count matrix, GEO includes processed objects uploaded by users as
supplementary files</strong>.</p>
<div class="section level3">
<h3 id="extract-metadata">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata"></a>
</h3>
<p><code>ExtractGEOMeta</code> provides two ways to extract sample
metadata:</p>
<ul>
<li>
<strong>user-provided sample metadata when uploading to GEO
(applicable to all GEO accessions)</strong>, including sample title,
source name/tissue, description, cell type, treatment, paper title,
paper abstract, organism, protocol, data processing methods, et al:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">GEfetch2R</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata</span></span>
<span><span class="va">GSE285723.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE285723"</span><span class="op">)</span></span>
<span><span class="va">GSE285723.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"title"</span>, <span class="st">"geo_accession"</span>, <span class="st">"source_name_ch1"</span>, <span class="st">"description"</span>, <span class="st">"cell type"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#     title geo_accession source_name_ch1           description  cell type</span></span>
<span><span class="co"># 1   DRA-F    GSM8707297            Lung   Library name: DRA-F Homogenate</span></span>
<span><span class="co"># 2   DRA-M    GSM8707298            Lung   Library name: DRA-M Homogenate</span></span>
<span><span class="co"># 3 DRAO3-F    GSM8707299            Lung Library name: DRAO3-F Homogenate</span></span></code></pre></div>
<ul>
<li>
<strong>metadata in supplementary file</strong>:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example</span></span>
<span><span class="va">GSE297431.meta.supp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE297431"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">2</span> <span class="co"># specify the index of used supplementary file</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE297431.meta.supp</span><span class="op">)</span></span>
<span><span class="co">#             Sample_ID  Batch  Plate   Type Growths Class</span></span>
<span><span class="co"># 1   Plate1_mut_A1_S70 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 2 Plate1_mut_A11_S135 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 3 Plate1_mut_A12_S141 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 4   Plate1_mut_A2_S77 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 5   Plate1_mut_A3_S84 batch2 Plate1 mutant       2    M2</span></span>
<span><span class="co"># 6   Plate1_mut_A4_S91 batch2 Plate1 mutant       2    M2</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-object-and-load-to-r">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r"></a>
</h3>
<p>After downloading the metadata, users can <strong>download the
specified objects</strong> with <code>ParseGEOProcessed</code>. The
format of downloaded objects are controlled by <code>file.ext</code>
(choose from <code>"rds"</code>, <code>"rdata"</code>,
<code>"h5ad"</code> and <code>"loom"</code>).</p>
<p>The processed objects in the supplementary files are in two
forms:</p>
<ul>
<li>single file
(<code>rds(.gz)</code>/<code>rdata(.gz)</code>/<code>h5ad(.gz)</code>/<code>loom(.gz)</code>)
contain the processed object, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE285723" class="external-link">GSE285723-RDS.gz</a>
</li>
<li>(gzip) archive (contain
<code>rds(.gz)</code>/<code>rdata(.gz)</code>/<code>h5ad(.gz)</code>/<code>loom(.gz)</code>
files) contain the processed objects, e.g. <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE298041" class="external-link">GSE298041-RDS
in tar.gz</a>
</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return SeuratObject</span></span>
<span><span class="va">GSE285723.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE285723"</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">T</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_geoObj"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">GSE285723.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 56055 features across 49806 samples within 2 assays</span></span>
<span><span class="co"># Active assay: SCT (23770 features, 3000 variable features)</span></span>
<span><span class="co">#  1 other assay present: RNA</span></span>
<span><span class="co">#  4 dimensional reductions calculated: pca, harmony, umap, tsne</span></span>
<span></span>
<span><span class="co"># download h5ad objects</span></span>
<span><span class="va">GSE311813.h5ad.log</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE311813"</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h5ad"</span><span class="op">)</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_geoObj"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # The structure of downloaded objects</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_geoObj</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_geoObj</span></span>
<span><span class="co"># ├── GSE285723</span></span>
<span><span class="co"># │   └── GSE285723_Final_Ballinger.RDS</span></span>
<span><span class="co"># └── GSE311813</span></span>
<span><span class="co">#     └── GSM9332642_merged_raw.ARS.vivo.clean.labelled.h5ad</span></span>
<span></span>
<span><span class="co"># 2 directories, 2 files</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="zenodo">Zenodo<a class="anchor" aria-label="anchor" href="#zenodo"></a>
</h2>
<p><a href="https://zenodo.org/" class="external-link">Zenodo</a> contains various types of
processed objects, such as <code>SeuratObject</code> which has been
clustered and annotated, <code>AnnData</code> which contains processed
results generated by <code>scanpy</code>.</p>
<div class="section level3">
<h3 id="extract-metadata-1">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-1"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ExtractZenodoMeta</code> to
extract dataset metadata, including dataset title, description,
available files and corresponding md5. Please note that when the dataset
is restricted access, the returned dataframe will be empty.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single doi</span></span>
<span><span class="va">zebrafish.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="st">"10.5281/zenodo.7243603"</span><span class="op">)</span></span>
<span><span class="va">zebrafish.df</span></span>
<span><span class="co">#                              title</span></span>
<span><span class="co"># 1 zebrafish scRNA data set objects</span></span>
<span><span class="co"># 2 zebrafish scRNA data set objects</span></span>
<span><span class="co">#                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                description</span></span>
<span><span class="co"># 1 &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Farrel et al. 2018, Wagner et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co"># 2 &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Farrel et al. 2018, Wagner et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co">#                                                                         url             filename</span></span>
<span><span class="co"># 1  https://zenodo.org/api/records/7243603/files/zebrafish_data.h5ad/content  zebrafish_data.h5ad</span></span>
<span><span class="co"># 2 https://zenodo.org/api/records/7243603/files/zebrafish_data.RData/content zebrafish_data.RData</span></span>
<span><span class="co">#                                md5   license</span></span>
<span><span class="co"># 1 124f2229128918b411a7dc7931558f97 cc-by-4.0</span></span>
<span><span class="co"># 2 a08c3ebd285b370fcf34cf2f8f9bdb59 cc-by-4.0</span></span>
<span></span>
<span></span>
<span><span class="co"># vector dois</span></span>
<span><span class="va">multi.dois</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">multi.dois</span></span>
<span><span class="co">#                              title</span></span>
<span><span class="co"># 1 zebrafish scRNA data set objects</span></span>
<span><span class="co"># 2 zebrafish scRNA data set objects</span></span>
<span><span class="co"># 3      frog scRNA data set objects</span></span>
<span><span class="co"># 4      frog scRNA data set objects</span></span>
<span><span class="co">#                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                description</span></span>
<span><span class="co"># 1 &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Farrel et al. 2018, Wagner et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co"># 2 &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Farrel et al. 2018, Wagner et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co"># 3                     &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Briggs et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co"># 4                     &lt;p&gt;Combined and converted scRNA data from http://tome.gs.washington.edu/ (Qiu et al. 2022), see a detailed description of the study here: https://www.nature.com/articles/s41588-022-01018-x&lt;/p&gt;\n\n&lt;p&gt;Data were downloaded from http://tome.gs.washington.edu/ as R rds files, combined into a single Seurat object and converted into loom and AnnData (h5ad) files to be able to analyse with e.g. python scanpy package.&lt;/p&gt;\n\n&lt;p&gt;If you use this data, please cite Briggs et al. 2018 and Qiu et al. 2022.&lt;/p&gt;</span></span>
<span><span class="co">#                                                                         url             filename</span></span>
<span><span class="co"># 1  https://zenodo.org/api/records/7243603/files/zebrafish_data.h5ad/content  zebrafish_data.h5ad</span></span>
<span><span class="co"># 2 https://zenodo.org/api/records/7243603/files/zebrafish_data.RData/content zebrafish_data.RData</span></span>
<span><span class="co"># 3       https://zenodo.org/api/records/7244441/files/frog_data.h5ad/content       frog_data.h5ad</span></span>
<span><span class="co"># 4      https://zenodo.org/api/records/7244441/files/frog_data.RData/content      frog_data.RData</span></span>
<span><span class="co">#                                md5   license</span></span>
<span><span class="co"># 1 124f2229128918b411a7dc7931558f97 cc-by-4.0</span></span>
<span><span class="co"># 2 a08c3ebd285b370fcf34cf2f8f9bdb59 cc-by-4.0</span></span>
<span><span class="co"># 3 7be7d6ff024ab2c8579b4d0edb2428e3 cc-by-4.0</span></span>
<span><span class="co"># 4 c80f46320c0cff9e341bed195f12c3b1 cc-by-4.0</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-object-and-load-to-r-1">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r-1"></a>
</h3>
<p>After manually check the extracted metadata, users can
<strong>download the specified objects</strong> with
<code>ParseZenodo</code>. The format of downloaded objects are
controlled by <code>file.ext</code> and <strong>the provided object
formats should be in lower case (e.g. rds/rdata/h5ad).</strong></p>
<p>The returned value is a dataframe containing failed objects or a
<code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>
and <code>return.seu = TRUE</code>). If dataframe, users can re-run
<code>ParseZenodo</code> by setting <code>doi.df</code> to the returned
value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects</span></span>
<span><span class="va">multi.dois.parse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span></span>
<span>  doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span><span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_zenodo"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># return SeuratObject</span></span>
<span><span class="va">sinle.doi.parse.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span></span>
<span>  doi <span class="op">=</span> <span class="st">"10.5281/zenodo.8011282"</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rds"</span><span class="op">)</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_zenodo"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sinle.doi.parse.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 19594 features across 9219 samples within 2 assays</span></span>
<span><span class="co"># Active assay: RNA (17594 features, 0 variable features)</span></span>
<span><span class="co">#  1 other assay present: integrated</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, umap</span></span>
<span></span>
<span><span class="co"># # The structure of downloaded objects</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_zenodo</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_zenodo</span></span>
<span><span class="co"># ├── frog_data.RData</span></span>
<span><span class="co"># ├── PyMTM_immune_scRNA.rds</span></span>
<span><span class="co"># └── zebrafish_data.RData</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h2>
<p>The <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> is a
web server contains <strong>2043</strong> single-cell datasets, users
can explore, download and upload own datasets. The downloaded datasets
provided by <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a>
have two formats: <code>h5ad (AnnData v0.8)</code> and
<code>rds (Seurat v4)</code>.</p>
<p><a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> does not
support downloading <code>SeuratObject</code> in <a href="https://cellxgene.cziscience.com/docs/03__Download%20Published%20Data" class="external-link">versions
after 2025</a>.Fortunately, we have downloaded all the <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> datasets in
<strong>May 2025</strong> and stored in <a href="https://github.com/showteeth/GEfetch2R/blob/main/man/benchmark/all.cellxgene.datasets.rds" class="external-link">all.cellxgene.datasets.rds</a>.
The <code>all.cellxgene.datasets.rds</code> contains the
<code>SeuratObject</code> for downloading. However, this does not apply
to the case of a given dataset.</p>
<p><a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> provides an
R package (<a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">cellxgene.census</a>)
to access the data, but sometimes it’s not timely updated.
<code>GEfetch2R</code> also supports users to access <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> via <a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">cellxgene.census</a>
(<code>use.census = TRUE</code>).</p>
<div class="section level3">
<h3 id="given-dataset">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset"></a>
</h3>
<p>With the collection or dataset link(s), user can download the
<code>h5ad</code> objects:</p>
<p>With the collection or dataset link(s), users can <strong>download
the <code>h5ad</code> objects</strong> with
<code>ParseCELLxGENE</code>.</p>
<p>The returned value is <code>NULL</code> or a dataframe containing
failed objects. If dataframe, users can re-run
<code>ParseCELLxGENE</code> by setting <code>meta</code> to the returned
value.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellxgene.given.h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://cellxgene.cziscience.com/collections/77f9d7e9-5675-49c3-abed-ce02f39eef1b"</span>, <span class="co"># collection</span></span>
<span>    <span class="st">"https://cellxgene.cziscience.com/e/e12eb8a9-5e8b-4b59-90c8-77d29a811c00.cxg/"</span> <span class="co"># dataset</span></span>
<span>  <span class="op">)</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_cellxgene"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># The structure of downloaded objects</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_cellxgene</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_cellxgene</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.B.and.Plasma.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.CD4.T.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.CD8.T.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.DCs.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.Monocytes.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.NK.cells.and.ILCs.h5ad</span></span>
<span><span class="co"># └── Human.Immune.Health.Atlas.Other.cells.h5ad</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 0 directories, 8 files</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="filter-samples-based-on-metadata">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata"></a>
</h3>
<div class="section level4">
<h4 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ShowCELLxGENEDatasets</code> to
extract dataset metadata, including dataset title, description, contact,
organism, ethnicity, sex, tissue, disease, assay, suspension type, cell
type, et al.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">all.cellxgene.datasets</span><span class="op">)</span></span>
<span><span class="co"># [1] 2043</span></span></code></pre></div>
<p>Load datasets with <code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the datasets with SeuratObject</span></span>
<span><span class="co"># wget https://github.com/showteeth/GEfetch2R/raw/ff2f19f3b557f90fce5f8bf2f8662cebdfd04298/man/benchmark/all.cellxgene.datasets.rds</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"all.cellxgene.datasets.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">all.cellxgene.datasets</span><span class="op">)</span></span>
<span><span class="co"># [1] 1320</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="summary-attributes">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span>, <span class="st">"disease"</span><span class="op">)</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"CELLxGENE"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 280 × 4</span></span>
<span><span class="co"># # Groups:   organism, sex [18]</span></span>
<span><span class="co">#    organism     sex     disease    Num</span></span>
<span><span class="co">#    &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt;    &lt;int&gt;</span></span>
<span><span class="co">#  1 homo sapiens male    normal     687</span></span>
<span><span class="co">#  2 homo sapiens female  normal     535</span></span>
<span><span class="co">#  3 mus musculus male    normal     178</span></span>
<span><span class="co">#  4 mus musculus female  normal     125</span></span>
<span><span class="co">#  5 mus musculus unknown normal     100</span></span>
<span><span class="co">#  6 homo sapiens unknown normal      79</span></span>
<span><span class="co">#  7 homo sapiens female  covid-19    50</span></span>
<span><span class="co">#  8 homo sapiens female  dementia    50</span></span>
<span><span class="co">#  9 homo sapiens male    dementia    50</span></span>
<span><span class="co"># 10 homo sapiens male    covid-19    38</span></span>
<span><span class="co"># # ℹ 270 more rows</span></span>
<span><span class="co"># # ℹ Use `print(n = ...)` to see more rows</span></span>
<span></span>
<span><span class="co"># # use cellxgene.census</span></span>
<span><span class="co"># StatDBAttribute(filter = c("disease", "tissue", "cell_type"), database = "CELLxGENE",</span></span>
<span><span class="co">#                 use.census = TRUE, organism = "homo_sapiens")</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-metadata">Filter metadata<a class="anchor" aria-label="anchor" href="#filter-metadata"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCELLxGENEMeta</code> to
filter dataset metadata, the available values of attributes can be
obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">human.10x.cellxgene.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCELLxGENEMeta.html">ExtractCELLxGENEMeta</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">human.10x.cellxgene.meta</span><span class="op">)</span></span>
<span><span class="co"># [1] 627</span></span>
<span></span>
<span><span class="co"># subset</span></span>
<span><span class="va">cellxgene.down.meta</span> <span class="op">&lt;-</span> <span class="va">human.10x.cellxgene.meta</span><span class="op">[</span><span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">==</span> <span class="st">"oligodendrocyte"</span> <span class="op">&amp;</span></span>
<span>  <span class="va">human.10x.cellxgene.meta</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"entorhinal cortex"</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cellxgene.down.meta</span><span class="op">)</span></span>
<span><span class="co"># [1] 1</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-object-and-load-to-r-2">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r-2"></a>
</h4>
<p>After manually check the extracted metadata, users can
<strong>download the specified objects</strong> with
<code>ParseCELLxGENE</code>. The downloaded objects are controlled by
<code>file.ext</code> (choose from <code>"rds"</code> and
<code>"h5ad"</code>).</p>
<p>The returned value is a dataframe containing failed objects or a
<code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>
and <code>return.seu = TRUE</code>). If dataframe, users can re-run
<code>ParseCELLxGENE</code> by setting <code>meta</code> to the returned
value.</p>
<p><strong>When using <code>cellxgene.census</code>, users can subset
metadata and gene</strong>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects</span></span>
<span><span class="va">cellxgene.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_cellxgene"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cellxgene.down</span></span>
<span><span class="co"># NULL</span></span>
<span></span>
<span><span class="co"># retuen SeuratObject</span></span>
<span><span class="va">cellxgene.down.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">cellxgene.down.meta</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  obs.value.filter <span class="op">=</span> <span class="st">"cell_type == 'oligodendrocyte' &amp; disease == 'Alzheimer disease'"</span>,</span>
<span>  obs.keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"disease"</span>, <span class="st">"sex"</span>, <span class="st">"suspension_type"</span>, <span class="st">"development_stage"</span><span class="op">)</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_cellxgene"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cellxgene.down.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 32743 features across 6873 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (32743 features, 0 variable features)</span></span>
<span><span class="co">#  3 dimensional reductions calculated: cca, cca.aligned, tsne</span></span>
<span></span>
<span><span class="co"># # The structure of downloaded objects</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_cellxgene</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_cellxgene</span></span>
<span><span class="co"># ├── all.cellxgene.datasets.rds</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.B.and.Plasma.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.CD4.T.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.CD8.T.cells.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.DCs.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.Monocytes.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.NK.cells.and.ILCs.h5ad</span></span>
<span><span class="co"># ├── Human.Immune.Health.Atlas.Other.cells.h5ad</span></span>
<span><span class="co"># └── Molecular.characterization.of.selectively.vulnerable.neurons.in.Alzheimer.s.Disease..EC.oligodendrocyte.rds</span></span>
<span></span>
<span><span class="co"># 0 directories, 10 files</span></span>
<span></span>
<span><span class="co"># # use cellxgene.census (support subset, but update is not timely)</span></span>
<span><span class="co"># cellxgene.down.census &lt;- ParseCELLxGENE(</span></span>
<span><span class="co">#   use.census = TRUE, organism = "Homo sapiens",</span></span>
<span><span class="co">#   obs.value.filter = "cell_type == 'B cell' &amp; tissue_general == 'lung' &amp; disease == 'COVID-19'",</span></span>
<span><span class="co">#   obs.keys = c("cell_type", "tissue_general", "disease", "sex"),</span></span>
<span><span class="co">#   include.genes = c("ENSG00000161798", "ENSG00000188229")</span></span>
<span><span class="co"># )</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="human-cell-atlas">Human Cell Atlas<a class="anchor" aria-label="anchor" href="#human-cell-atlas"></a>
</h2>
<p>The <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>
aims to map every cell type in the human body, it contains <strong>546
projects</strong>, most of which are from <code>Homo sapiens</code>
(also includes projects from <code>Mus musculus</code>,
<code>Macaca mulatta</code> and
<code>canis lupus familiaris</code>).</p>
<div class="section level3">
<h3 id="given-dataset-1">Given dataset<a class="anchor" aria-label="anchor" href="#given-dataset-1"></a>
</h3>
<p>With the dataset link(s), users can <strong>download the processed
objects</strong> with <code>ParseHCA</code>. The format of downloaded
objects are controlled by <code>file.ext</code> (choose from
<code>"tsv"</code>, <code>"rds"</code>, <code>"rdata"</code>,
<code>"h5"</code>, <code>"h5ad"</code> and <code>"loom"</code>).</p>
<p>The returned value is a dataframe containing failed objects or a
<code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>
and <code>return.seu = TRUE</code>). If dataframe, users can re-run
<code>ParseHCA</code> by setting <code>meta</code> to the returned
value.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects</span></span>
<span><span class="va">hca.given.download</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span></span>
<span>  link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://explore.data.humancellatlas.org/projects/902dc043-7091-445c-9442-d72e163b9879"</span>,</span>
<span>    <span class="st">"https://explore.data.humancellatlas.org/projects/cdabcf0b-7602-4abf-9afb-3b410e545703"</span></span>
<span>  <span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_hca"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # The structure of downloaded objects</span></span>
<span><span class="co"># tree ~/gefetch2r/doc/download_hca</span></span>
<span><span class="co"># ~/gefetch2r/doc/download_hca</span></span>
<span><span class="co"># ├── COMBAT2022.h5ad</span></span>
<span><span class="co"># └── seurat_object_hca_as_harmonized_AS_SP_nuc_refined_cells.rds.gz</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 0 directories, 2 files</span></span>
<span></span>
<span><span class="co"># retuen SeuratObject</span></span>
<span><span class="va">hypertrophic.heart.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span></span>
<span>  link <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://explore.data.humancellatlas.org/projects/902dc043-7091-445c-9442-d72e163b9879"</span></span>
<span>  <span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span>, return.seu <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_hca"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="filter-samples-based-on-metadata-1">Filter samples based on metadata<a class="anchor" aria-label="anchor" href="#filter-samples-based-on-metadata-1"></a>
</h3>
<div class="section level4">
<h4 id="show-available-datasets-1">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ShowHCAProjects</code> to
extract detailed project metadata, including project title, description,
organism, sex, organ/organPart, disease, assay, preservation method,
sample type, suspension type, cell type, development stage, et al.</p>
<p>There are <strong>484 unique projects</strong> under <em>five
different catalogs (<code>dcp29</code>, <code>dcp30</code>,
<code>dcp1</code>, <code>lm2</code>, <code>lm3</code>)</em>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.hca.projects</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowHCAProjects.html">ShowHCAProjects</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">all.hca.projects</span><span class="op">)</span></span>
<span><span class="co"># [1] 546</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="summary-attributes-1">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://www.humancellatlas.org/" class="external-link">Human
Cell Atlas</a>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">all.hca.projects</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"HCA"</span><span class="op">)</span></span>
<span><span class="co"># $organism</span></span>
<span><span class="co">#                    Value Num      Key</span></span>
<span><span class="co"># 1           homo sapiens 520 organism</span></span>
<span><span class="co"># 2           mus musculus  58 organism</span></span>
<span><span class="co"># 3 canis lupus familiaris   1 organism</span></span>
<span><span class="co"># 4         macaca mulatta   1 organism</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $sex</span></span>
<span><span class="co">#     Value Num Key</span></span>
<span><span class="co"># 1  female 405 sex</span></span>
<span><span class="co"># 2    male 392 sex</span></span>
<span><span class="co"># 3 unknown 164 sex</span></span>
<span><span class="co"># 4   mixed   6 sex</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="filter-metadata-1">Filter metadata<a class="anchor" aria-label="anchor" href="#filter-metadata-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractHCAMeta</code> to filter
projects metadata, the available values of attributes can be obtained
with <code>StatDBAttribute</code> except <strong>cell
number</strong>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">hca.human.10x.projects</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractHCAMeta.html">ExtractHCAMeta</a></span><span class="op">(</span></span>
<span>  all.projects.df <span class="op">=</span> <span class="va">all.hca.projects</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">hca.human.10x.projects</span><span class="op">)</span></span>
<span><span class="co"># [1] 251</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-object-and-load-to-r-3">Download object and load to R<a class="anchor" aria-label="anchor" href="#download-object-and-load-to-r-3"></a>
</h4>
<p>After manually check the extracted metadata, users can
<strong>download the specified objects</strong> with
<code>ParseHCA</code>. The format of downloaded objects are controlled
by <code>file.ext</code> (choose from <code>"tsv"</code>,
<code>"rds"</code>, <code>"rdata"</code>, <code>"h5"</code>,
<code>"h5ad"</code> and <code>"loom"</code>).</p>
<p>The returned value is a dataframe containing failed objects or a
<code>SeuratObject</code> (if <code>file.ext</code> is <code>rds</code>
and <code>return.seu = TRUE</code>). If dataframe, users can re-run
<code>ParseHCA</code> by setting <code>meta</code> to the returned
value.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects</span></span>
<span><span class="va">hca.human.10x.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseHCA.html">ParseHCA</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">hca.human.10x.projects</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/download_hca"</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h5ad"</span>, <span class="st">"rds"</span><span class="op">)</span>, timeout <span class="op">=</span> <span class="fl">36000000</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># file downloaded</span></span>
<span><span class="co"># 4f30b962-d49b-4624-a233-64f048cf8632_b61a921b-7fa3-4b42-b455-aaaf32447920.h5ad</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="process-rdata-files">Process RData files<a class="anchor" aria-label="anchor" href="#process-rdata-files"></a>
</h2>
<p>As illustrated before, the downloaded <code>rds</code> file
containing <code>SeuratObject</code> will be automatically loaded into
R. Here, <code>GEfetch2R</code> provides <code>LoadRData</code> to
dissect and extract the <code>RData</code> files.</p>
<p><code>LoadRData</code> process the objects in <code>RData</code>
files by two ways:</p>
<ul>
<li>for <strong>widely used scRNA-seq objects</strong>
(<code>SeuratObject</code> (<code>Seurat</code> package),
<code>seuratobject</code> (<code>Seurat v2</code> package),
<code>SingleCellExperiment</code> (<code>SingleCellExperiment</code>
package), <code>cell_data_set</code> (<code>Monocle v3</code> package),
<code>CellDataSet</code> (<code>Monocle</code> package)) and
<strong>bulk RNA-seq objects</strong> (<code>DESeqDataSet</code>
(<code>DESeq2</code> package), <code>DGEList</code> (<code>edgeR</code>
package)):</li>
</ul>
<table class="table">
<colgroup>
<col width="9%">
<col width="14%">
<col width="34%">
<col width="11%">
<col width="13%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th><strong>Object</strong></th>
<th><strong>raw count matrix</strong></th>
<th><strong>normalized count matrix</strong></th>
<th><strong>scaled count matrix</strong></th>
<th><strong>metadata</strong></th>
<th><strong>notes</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>SeuratObject</strong></td>
<td>slot <code>counts</code>
</td>
<td>slot <code>data</code>
</td>
<td>slot <code>scale.data</code>
</td>
<td><code>obj@meta.data</code></td>
<td>support multiple assays (e.g.: RNA, integrated)</td>
</tr>
<tr class="even">
<td><strong>seuratobject</strong></td>
<td><code>obj@raw.data</code></td>
<td><code>obj@data</code></td>
<td><code>obj@scale.data</code></td>
<td><code>obj@meta.data</code></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>SingleCellExperiment</strong></td>
<td>assay <code>counts</code>
</td>
<td>assay <code>logcounts</code>
</td>
<td>assay <code>scaledata</code>/<code>scale.data</code>
</td>
<td><code>SingleCellExperiment::colData(obj)</code></td>
<td>support multiple experiments</td>
</tr>
<tr class="even">
<td><strong>cell_data_set</strong></td>
<td>assay <code>counts</code>
</td>
<td>assay <code>logcounts</code>
</td>
<td>assay <code>scaledata</code>/<code>scale.data</code>
</td>
<td><code>SingleCellExperiment::colData(obj)</code></td>
<td>support multiple experiments</td>
</tr>
<tr class="odd">
<td><strong>CellDataSet</strong></td>
<td>empty or <code>Biobase::exprs(obj)</code>
</td>
<td>empty or <code>Biobase::exprs(obj)</code>
</td>
<td>empty or <code>Biobase::exprs(obj)</code>
</td>
<td><code>Biobase::pData(obj)</code></td>
<td></td>
</tr>
<tr class="even">
<td><strong>DESeqDataSet</strong></td>
<td><code>DESeq2::counts(obj, normalized = FALSE)</code></td>
<td>(<code>obj &lt;- DESeq2::estimateSizeFactors(obj)</code>)
<code>DESeq2::counts(obj, normalized = TRUE)</code>
</td>
<td>empty</td>
<td><code>obj@colData</code></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>DGEList</strong></td>
<td><code>obj$counts</code></td>
<td>
<code>obj &lt;- edgeR::calcNormFactors(obj, method = "TMM")</code>;
<code>count.mat.list[["data"]] &lt;- edgeR::cpm(obj)</code>
</td>
<td>empty</td>
<td><code>obj$samples</code></td>
<td></td>
</tr>
</tbody>
</table>
<ul>
<li>for <strong>non-standard objects (<code>matrix</code>,
<code>data.frame</code>, <code>dgCMatrix</code>, <code>dgRMatrix</code>,
<code>dgTMatrix</code>)</strong>:
<ul>
<li>if <code>matrix</code>/<code>data.frame</code>:
<ul>
<li>if the number of columns is greater than or equal to three:
<ul>
<li>if all columns values are in the same class and the class is
numeric/integer, treat this object as matrix</li>
<li>else if starting from the second column, all column values are
numeric/integer, treat this object as matrix</li>
<li>else if column names contain pattern
<code>"sample|name|cell|id|library|well|barcode|index|type|condition|treat|group"</code>,
treat this object as metadata/annotation</li>
<li>else display the structure of the object and load the object to
R.</li>
</ul>
</li>
<li>else if column names contain pattern
<code>"sample|name|cell|id|library|well|barcode|index|type|condition|treat|group"</code>,
treat this object as metadata/annotation</li>
<li>else display the structure of the object and load the object to
R.</li>
</ul>
</li>
<li>else if
<code>dgCMatrix</code>/<code>dgRMatrix</code>/<code>dgTMatrix</code>,
treat this object as matrix</li>
<li>else print the first six rows of the object.</li>
</ul>
</li>
</ul>
<div class="alert alert-success" role="alert">
<p>The first way has a higher priority; that is, if valid scRNA-seq/bulk
RNA-seq objects (containing comprehensive information) are detected,
skip the second way.</p>
</div>
<div class="section level3">
<h3 id="example-gse244572-seuratobject-with-multiple-assays">Example <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE244572" class="external-link">GSE244572</a>:
<code>SeuratObject</code> with multiple assays<a class="anchor" aria-label="anchor" href="#example-gse244572-seuratobject-with-multiple-assays"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the object</span></span>
<span><span class="fu"><a href="../reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE244572"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span>, <span class="st">"h5ad"</span>, <span class="st">"loom"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># process the object</span></span>
<span><span class="va">GSE244572.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadRData.html">LoadRData</a></span><span class="op">(</span></span>
<span>  rdata <span class="op">=</span> <span class="st">"GSE244572/GSE244572_RPE_CITESeq.RData"</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"seurat"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"cell_data_set"</span>, <span class="st">"CellDataSet"</span>, <span class="st">"DESeqDataSet"</span>, <span class="st">"DGEList"</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span>, return.obj <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # message:</span></span>
<span><span class="co"># The object classes stored in RData: Seurat.</span></span>
<span><span class="co">#      Class</span></span>
<span><span class="co"># obj Seurat</span></span>
<span><span class="co"># Detect 1 object(s) in given format(s): Seurat, seurat, SingleCellExperiment, cell_data_set, CellDataSet, DESeqDataSet, DGEList.</span></span>
<span><span class="co"># Load object: obj (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: obj (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 4.0.4, with assay(s): RNA, ADT, nADT, SCT, integrated, IADT.</span></span>
<span></span>
<span><span class="co"># returned value</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list objects in the global environment</span></span>
<span><span class="co"># [1] "GSE244572.list" "obj"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE244572.list</span><span class="op">)</span> <span class="co"># one valid object</span></span>
<span><span class="co"># [1] "obj"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE244572.list</span><span class="op">$</span><span class="va">obj</span><span class="op">$</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#                  orig.ident nCount_RNA nFeature_RNA sample percent.mt nCount_ADT nFeature_ADT donor time   rpe nCount_nADT nFeature_nADT nCount_SCT nFeature_SCT integrated_snn_res.0.5 seurat_clusters</span></span>
<span><span class="co"># CCTAGATTAAT_1 SeuratProject     487751         3922 318_2W   13.34062     387973          121   318   2W adult          22            14     214660         3175                      5               0</span></span>
<span><span class="co"># GTTGAGCTGCG_1 SeuratProject     124013         1477 318_2W   13.27925     312671          110   318   2W adult          21            14     215337         1475                      2               4</span></span>
<span><span class="co"># GTTGATCTTCT_1 SeuratProject      95909         1344 318_2W   15.60333      44974          100   318   2W adult          35            27     214940         1359                      0               2</span></span>
<span><span class="co"># TAACCTTGAGT_1 SeuratProject     328530         2361 319_2W   17.52930     279285          128   319   2W adult          23            18     215249         2349                      2               0</span></span>
<span><span class="co"># CCTCCGCCTGC_1 SeuratProject     702493         4766 319_2W   22.13887    1579539          143   319   2W adult          32            17     215407         4027                      1               6</span></span>
<span><span class="co"># CCATATACGAC_1 SeuratProject     497485         3694 319_2W   14.13791     873609          117   319   2W adult          23            12     214400         2913                      1               4</span></span>
<span><span class="co">#               integrated.weight wsnn_res.0.5 wsnn_res.0.2</span></span>
<span><span class="co"># CCTAGATTAAT_1         0.7954453            3            0</span></span>
<span><span class="co"># GTTGAGCTGCG_1         0.6762115            5            4</span></span>
<span><span class="co"># GTTGATCTTCT_1         0.9341439            0            2</span></span>
<span><span class="co"># TAACCTTGAGT_1         0.7959367            3            0</span></span>
<span><span class="co"># CCTCCGCCTGC_1         0.5048201           10            6</span></span>
<span><span class="co"># CCATATACGAC_1         0.7556721            5            4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE244572.list</span><span class="op">$</span><span class="va">obj</span><span class="op">$</span><span class="va">count.mat</span><span class="op">)</span> <span class="co"># list of six assays</span></span>
<span><span class="co"># [1] "RNA"        "ADT"        "nADT"       "SCT"        "integrated" "IADT"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE244572.list</span><span class="op">$</span><span class="va">obj</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span> <span class="co"># list of slot(s)</span></span>
<span><span class="co"># [1] "counts"</span></span>
<span><span class="va">GSE244572.list</span><span class="op">$</span><span class="va">obj</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># count matrix</span></span>
<span><span class="co"># 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#            CCTAGATTAAT_1 GTTGAGCTGCG_1 GTTGATCTTCT_1 TAACCTTGAGT_1 CCTCCGCCTGC_1</span></span>
<span><span class="co"># WASH7P                 .             .             .             .             .</span></span>
<span><span class="co"># CICP27                 .             .             .             .             .</span></span>
<span><span class="co"># AL627309.6             .             .             .             .             .</span></span>
<span><span class="co"># AL627309.5             .             .             .             .             .</span></span>
<span><span class="co"># FO538757.1             .             .             .             .             .</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="example-gse249307-multiple-seuratobject-objects">Example <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE249307" class="external-link">GSE249307</a>:
multiple <code>SeuratObject</code> objects<a class="anchor" aria-label="anchor" href="#example-gse249307-multiple-seuratobject-objects"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the object</span></span>
<span><span class="fu"><a href="../reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE249307"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span>, <span class="st">"h5ad"</span>, <span class="st">"loom"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># process the object</span></span>
<span><span class="va">GSE249307.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadRData.html">LoadRData</a></span><span class="op">(</span></span>
<span>  rdata <span class="op">=</span> <span class="st">"GSE249307/GSE249307_scRNA_seurat_data.RData"</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"seurat"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"cell_data_set"</span>, <span class="st">"CellDataSet"</span>, <span class="st">"DESeqDataSet"</span>, <span class="st">"DGEList"</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # message:</span></span>
<span><span class="co"># The object classes stored in RData: Seurat.</span></span>
<span><span class="co">#                          Class</span></span>
<span><span class="co"># processed_seurat_object Seurat</span></span>
<span><span class="co"># raw_seurat_object       Seurat</span></span>
<span><span class="co"># Detect 2 object(s) in given format(s): Seurat, seurat, SingleCellExperiment, cell_data_set, CellDataSet, DESeqDataSet, DGEList.</span></span>
<span><span class="co"># Load object: processed_seurat_object (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: processed_seurat_object (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 3.0.0.9000, with assay(s): RNA.</span></span>
<span><span class="co"># Load object: raw_seurat_object (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: raw_seurat_object (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 3.0.0.9000, with assay(s): RNA.</span></span>
<span></span>
<span><span class="co"># returned value</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list objects in the global environment</span></span>
<span><span class="co"># [1] "GSE249307.list"          "processed_seurat_object" "raw_seurat_object"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE249307.list</span><span class="op">)</span> <span class="co"># two valid objects</span></span>
<span><span class="co"># [1] "processed_seurat_object" "raw_seurat_object"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE249307.list</span><span class="op">$</span><span class="va">raw_seurat_object</span><span class="op">$</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#                               orig.ident nCount_RNA nFeature_RNA sample species umi_count_50dup rnd1_well rnd2_well rnd3_well valid library        cell_barcode cells_unique treatment1  treatment2 cytokine_timepoint</span></span>
<span><span class="co"># lib1_AGATCGCACATCAAGT_42 lib1_2_analysis     356571        13917 samp43    hg38        689652.3        42         7        24     1    lib1 AGATCGCACATCAAGT_42       unique  IL17+TNFa  Virus_6hpi               late</span></span>
<span><span class="co"># lib1_ACTATGCAATAGCGAC_22 lib1_2_analysis     198322        11831 samp23    hg38        383579.2        22        79        22     1    lib1 ACTATGCAATAGCGAC_22       unique    Vehicle  Virus_6hpi           baseline</span></span>
<span><span class="co"># lib1_CGACTGGAGACAGTGC_23 lib1_2_analysis     187640        11865 samp24    hg38        362918.9        23        92        40     1    lib1 CGACTGGAGACAGTGC_23       unique       IL13  Virus_6hpi              early</span></span>
<span><span class="co"># lib1_AACGCTTAGTACGCAA_24 lib1_2_analysis     177905        11530 samp25    hg38        344090.2        24        55        14     1    lib1 AACGCTTAGTACGCAA_24       unique  IL17+TNFa  Virus_6hpi              early</span></span>
<span><span class="co"># lib1_ACACGACCAGTCACTA_11 lib1_2_analysis     156939        10976 samp12    hg38        303539.4        11        26        73     1    lib1 ACACGACCAGTCACTA_11       unique    Vehicle Virus_72hpi           baseline</span></span>
<span><span class="co"># lib1_TCTTCACAATAGCGAC_42 lib1_2_analysis     142185        11049 samp43    hg38        275003.3        42        79        62     1    lib1 TCTTCACAATAGCGAC_42       unique  IL17+TNFa  Virus_6hpi               late</span></span>
<span><span class="co">#                          donor</span></span>
<span><span class="co"># lib1_AGATCGCACATCAAGT_42     C</span></span>
<span><span class="co"># lib1_ACTATGCAATAGCGAC_22     B</span></span>
<span><span class="co"># lib1_CGACTGGAGACAGTGC_23     B</span></span>
<span><span class="co"># lib1_AACGCTTAGTACGCAA_24     B</span></span>
<span><span class="co"># lib1_ACACGACCAGTCACTA_11     A</span></span>
<span><span class="co"># lib1_TCTTCACAATAGCGAC_42     C</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE249307.list</span><span class="op">$</span><span class="va">raw_seurat_object</span><span class="op">$</span><span class="va">count.mat</span><span class="op">)</span> <span class="co"># list of available assay</span></span>
<span><span class="co"># [1] "RNA"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE249307.list</span><span class="op">$</span><span class="va">raw_seurat_object</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span> <span class="co"># list of slot(s)</span></span>
<span><span class="co"># [1] "counts"</span></span>
<span><span class="va">GSE249307.list</span><span class="op">$</span><span class="va">raw_seurat_object</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># count matrix</span></span>
<span><span class="co"># 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#          lib1_AGATCGCACATCAAGT_42 lib1_ACTATGCAATAGCGAC_22 lib1_CGACTGGAGACAGTGC_23 lib1_AACGCTTAGTACGCAA_24 lib1_ACACGACCAGTCACTA_11</span></span>
<span><span class="co"># A1BG                            .                        .                        .                        .                        .</span></span>
<span><span class="co"># A1BG-AS1                        1                        .                        1                        .                        .</span></span>
<span><span class="co"># A1CF                            .                        .                        .                        .                        .</span></span>
<span><span class="co"># A2M                             .                        .                        .                        .                        .</span></span>
<span><span class="co"># A2ML1                           .                        .                        .                        .                        .</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="example-gse282783-seuratobject-and-non-standard-objects">Example <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE282783" class="external-link">GSE282783</a>:
<code>SeuratObject</code> and non-standard objects<a class="anchor" aria-label="anchor" href="#example-gse282783-seuratobject-and-non-standard-objects"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the object</span></span>
<span><span class="fu"><a href="../reference/ParseGEOProcessed.html">ParseGEOProcessed</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE282783"</span>, timeout <span class="op">=</span> <span class="fl">360000</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span>, <span class="st">"h5ad"</span>, <span class="st">"loom"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># process the object</span></span>
<span><span class="va">GSE282783.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadRData.html">LoadRData</a></span><span class="op">(</span></span>
<span>  rdata <span class="op">=</span> <span class="st">"GSE282783/GSE282783_E16_FT_E17_hGFAP-Cre_mek12dcko.Rdata"</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"seurat"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"cell_data_set"</span>, <span class="st">"CellDataSet"</span>, <span class="st">"DESeqDataSet"</span>, <span class="st">"DGEList"</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # message:</span></span>
<span><span class="co"># The object classes stored in RData: character, data.frame, IntegrationAnchorSet, Seurat, function.</span></span>
<span><span class="co">#                                Class</span></span>
<span><span class="co"># all.genes                  character</span></span>
<span><span class="co"># Cluster                   data.frame</span></span>
<span><span class="co"># g2m.genes                 data.frame</span></span>
<span><span class="co"># immune.anchors  IntegrationAnchorSet</span></span>
<span><span class="co"># immune.combined               Seurat</span></span>
<span><span class="co"># Mapkdcko_CTR                  Seurat</span></span>
<span><span class="co"># Mapkdcko_EXP                  Seurat</span></span>
<span><span class="co"># MarkerPlot                  function</span></span>
<span><span class="co"># s.genes                   data.frame</span></span>
<span><span class="co"># Detect 3 object(s) in given format(s): Seurat, seurat, SingleCellExperiment, cell_data_set, CellDataSet, DESeqDataSet, DGEList.</span></span>
<span><span class="co"># Load object: immune.combined (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: immune.combined (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 4.0.4, with assay(s): RNA, integrated.</span></span>
<span><span class="co"># Load object: Mapkdcko_CTR (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: Mapkdcko_CTR (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 4.0.4, with assay(s): RNA.</span></span>
<span><span class="co"># Load object: Mapkdcko_EXP (Seurat) to global environment!</span></span>
<span><span class="co"># Extract count matrix and metadata (if available) from: Mapkdcko_EXP (Seurat).</span></span>
<span><span class="co"># Detect Seurat version: 4.0.4, with assay(s): RNA.</span></span>
<span></span>
<span><span class="co"># returned value</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list objects in the global environment</span></span>
<span><span class="co"># [1] "GSE282783.list"  "immune.combined" "Mapkdcko_CTR"    "Mapkdcko_EXP"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE282783.list</span><span class="op">)</span> <span class="co"># three valid objects</span></span>
<span><span class="co"># [1] "immune.combined" "Mapkdcko_CTR"    "Mapkdcko_EXP"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE282783.list</span><span class="op">$</span><span class="va">immune.combined</span><span class="op">$</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#                        orig.ident nCount_RNA nFeature_RNA       Sample      S.Score   G2M.Score Phase    old.ident integrated_snn_res.0.6 seurat_clusters</span></span>
<span><span class="co"># AAACCCAAGCACGTCC-1_1 Mapkdcko_CTR       3584         1586 Mapkdcko_CTR -0.010667752 -0.10155295    G1 Mapkdcko_CTR                      0               0</span></span>
<span><span class="co"># AAACCCAAGTGCAGGT-1_1 Mapkdcko_CTR      11187         3232 Mapkdcko_CTR  0.544009405  0.92782765   G2M Mapkdcko_CTR                      9               9</span></span>
<span><span class="co"># AAACCCACAACGTAAA-1_1 Mapkdcko_CTR       6498         2217 Mapkdcko_CTR -0.138904079 -0.09983576    G1 Mapkdcko_CTR                      2               2</span></span>
<span><span class="co"># AAACCCACAACGTTAC-1_1 Mapkdcko_CTR       4531         1873 Mapkdcko_CTR -0.008647976 -0.16412585    G1 Mapkdcko_CTR                      0               0</span></span>
<span><span class="co"># AAACCCACACTGCGAC-1_1 Mapkdcko_CTR       6588         2568 Mapkdcko_CTR -0.175856885 -0.19708486    G1 Mapkdcko_CTR                      5               5</span></span>
<span><span class="co"># AAACCCAGTACGATCT-1_1 Mapkdcko_CTR       3790         1616 Mapkdcko_CTR -0.101067444 -0.06548076    G1 Mapkdcko_CTR                      0               0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE282783.list</span><span class="op">$</span><span class="va">immune.combined</span><span class="op">$</span><span class="va">count.mat</span><span class="op">)</span> <span class="co"># list of available assays</span></span>
<span><span class="co"># [1] "RNA"        "integrated"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GSE282783.list</span><span class="op">$</span><span class="va">immune.combined</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span> <span class="co"># list of slot(s)</span></span>
<span><span class="co"># [1] "counts"</span></span>
<span><span class="va">GSE282783.list</span><span class="op">$</span><span class="va">immune.combined</span><span class="op">$</span><span class="va">count.mat</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># count matrix</span></span>
<span><span class="co"># 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#         AAACCCAAGCACGTCC-1_1 AAACCCAAGTGCAGGT-1_1 AAACCCACAACGTAAA-1_1 AAACCCACAACGTTAC-1_1 AAACCCACACTGCGAC-1_1</span></span>
<span><span class="co"># Xkr4                       .                    .                    .                    .                    .</span></span>
<span><span class="co"># Gm1992                     .                    .                    .                    .                    .</span></span>
<span><span class="co"># Gm37381                    .                    .                    .                    .                    .</span></span>
<span><span class="co"># Rp1                        .                    .                    .                    .                    .</span></span>
<span><span class="co"># Rp1.1                      .                    .                    .                    .                    .</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="example-simulated-non-standard-objects">Example: simulated non-standard objects<a class="anchor" aria-label="anchor" href="#example-simulated-non-standard-objects"></a>
</h3>
<p>Generate <code>RData</code> file containing a mixture of non-standard
objects:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># dgCMatrix</span></span>
<span><span class="va">sparse.mat</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="fu">SeuratObject</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/SeuratObject/man/pbmc_small.html" class="external-link">pbmc_small</a></span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="co"># count matrix and metadata from GSE297431</span></span>
<span><span class="va">GSE297431.meta.supp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE297431"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.idx <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">GSE297431.cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE297431"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, supp.type <span class="op">=</span> <span class="st">"count"</span>, load2R <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co"># move rownames to dataframe (first column)</span></span>
<span><span class="va">GSE297431.cnt.row2col</span> <span class="op">&lt;-</span> <span class="va">GSE297431.cnt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># dataframe to matrix</span></span>
<span><span class="va">GSE297431.cnt.mat</span> <span class="op">&lt;-</span> <span class="va">GSE297431.cnt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># list (noise)</span></span>
<span><span class="va">cc.genes</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="va"><a href="https://satijalab.org/seurat/reference/cc.genes.html" class="external-link">cc.genes</a></span></span>
<span><span class="co"># dataframe (noise)</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span><span class="op">)</span></span>
<span><span class="co"># vector (noise)</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span><span class="co"># save</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">sparse.mat</span>, <span class="va">GSE297431.meta.supp</span>, <span class="va">GSE297431.cnt</span>, <span class="va">GSE297431.cnt.row2col</span>, <span class="va">GSE297431.cnt.mat</span>, <span class="va">cc.genes</span>, <span class="va">s.genes</span>, <span class="va">g2m.genes</span>,</span>
<span>  file <span class="op">=</span> <span class="st">"simulated_non_standard_objects.RData"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Dissect and extract the generated <code>RData</code> file:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># process the object</span></span>
<span><span class="va">non.standard.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadRData.html">LoadRData</a></span><span class="op">(</span></span>
<span>  rdata <span class="op">=</span> <span class="st">"simulated_non_standard_objects.RData"</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"seurat"</span>, <span class="st">"SingleCellExperiment"</span>, <span class="st">"cell_data_set"</span>, <span class="st">"CellDataSet"</span>, <span class="st">"DESeqDataSet"</span>, <span class="st">"DGEList"</span><span class="op">)</span>,</span>
<span>  return.obj <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># # message:</span></span>
<span><span class="co"># The object classes stored in RData: list, character, data.frame, matrix, array, dgCMatrix.</span></span>
<span><span class="co">#                               Class</span></span>
<span><span class="co"># cc.genes                       list</span></span>
<span><span class="co"># g2m.genes                 character</span></span>
<span><span class="co"># GSE297431.cnt            data.frame</span></span>
<span><span class="co"># GSE297431.cnt.mat     matrix, array</span></span>
<span><span class="co"># GSE297431.cnt.row2col    data.frame</span></span>
<span><span class="co"># GSE297431.meta.supp      data.frame</span></span>
<span><span class="co"># s.genes                  data.frame</span></span>
<span><span class="co"># sparse.mat                dgCMatrix</span></span>
<span><span class="co"># No valid object in given format(s): Seurat, seurat, SingleCellExperiment, cell_data_set, CellDataSet, DESeqDataSet, DGEList. Now we will guess the type!</span></span>
<span><span class="co"># The slot parameter does not work here!</span></span>
<span><span class="co"># cc.genes is list.</span></span>
<span><span class="co"># $s.genes</span></span>
<span><span class="co">#  [1] "MCM5"     "PCNA"     "TYMS"     "FEN1"     "MCM2"     "MCM4"     "RRM1"     "UNG"      "GINS2"    "MCM6"     "CDCA7"    "DTL"      "PRIM1"    "UHRF1"    "MLF1IP"   "HELLS"    "RFC2"     "RPA2"     "NASP"</span></span>
<span><span class="co"># [20] "RAD51AP1" "GMNN"     "WDR76"    "SLBP"     "CCNE2"    "UBR7"     "POLD3"    "MSH2"     "ATAD2"    "RAD51"    "RRM2"     "CDC45"    "CDC6"     "EXO1"     "TIPIN"    "DSCC1"    "BLM"      "CASP8AP2" "USP1"</span></span>
<span><span class="co"># [39] "CLSPN"    "POLA1"    "CHAF1B"   "BRIP1"    "E2F8"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># $g2m.genes</span></span>
<span><span class="co">#  [1] "HMGB2"   "CDK1"    "NUSAP1"  "UBE2C"   "BIRC5"   "TPX2"    "TOP2A"   "NDC80"   "CKS2"    "NUF2"    "CKS1B"   "MKI67"   "TMPO"    "CENPF"   "TACC3"   "FAM64A"  "SMC4"    "CCNB2"   "CKAP2L"  "CKAP2"   "AURKB"</span></span>
<span><span class="co"># [22] "BUB1"    "KIF11"   "ANP32E"  "TUBB4B"  "GTSE1"   "KIF20B"  "HJURP"   "CDCA3"   "HN1"     "CDC20"   "TTK"     "CDC25C"  "KIF2C"   "RANGAP1" "NCAPD2"  "DLGAP5"  "CDCA2"   "CDCA8"   "ECT2"    "KIF23"   "HMMR"</span></span>
<span><span class="co"># [43] "AURKA"   "PSRC1"   "ANLN"    "LBR"     "CKAP5"   "CENPE"   "CTCF"    "NEK2"    "G2E3"    "GAS2L3"  "CBX5"    "CENPA"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># g2m.genes is character.</span></span>
<span><span class="co"># [1] "HMGB2"  "CDK1"   "NUSAP1" "UBE2C"  "BIRC5"  "TPX2"</span></span>
<span><span class="co"># GSE297431.cnt has 107 columns and each column is numerical! Most likely a count matrix!</span></span>
<span><span class="co"># GSE297431.cnt.mat has 107 columns and each column is numerical! Most likely a count matrix!</span></span>
<span><span class="co"># GSE297431.cnt.row2col has 108 columns, all of which are numerical except for the first column! Maybe a count matrix!</span></span>
<span><span class="co"># Detect possible sample metadata keys: Sample_ID, Type in GSE297431.meta.supp. Maybe metadata/annotation!</span></span>
<span><span class="co"># Can not determine if s.genes is metadata/annotation. Load to the global environment, please manually check!</span></span>
<span><span class="co"># 'data.frame': 43 obs. of  1 variable:</span></span>
<span><span class="co">#  $ gene: chr  "MCM5" "PCNA" "TYMS" "FEN1" ...</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># sparse.mat is a sparse matrix. Most likely a count matrix!</span></span>
<span></span>
<span><span class="co"># returned value</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list objects in the global environment</span></span>
<span><span class="co"># [1] "non.standard.list" "s.genes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">non.standard.list</span><span class="op">)</span> <span class="co"># two elements: count matrix and metadata</span></span>
<span><span class="co"># [1] "count" "meta"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">non.standard.list</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span> <span class="co"># list of available metadata</span></span>
<span><span class="co"># [1] "GSE297431.meta.supp"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">non.standard.list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">GSE297431.meta.supp</span><span class="op">)</span></span>
<span><span class="co">#             Sample_ID  Batch  Plate   Type Growths Class</span></span>
<span><span class="co"># 1   Plate1_mut_A1_S70 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 2 Plate1_mut_A11_S135 batch2 Plate1 mutant       3    M3</span></span>
<span><span class="co"># 3 Plate1_mut_A12_S141 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 4   Plate1_mut_A2_S77 batch2 Plate1 mutant       1    M1</span></span>
<span><span class="co"># 5   Plate1_mut_A3_S84 batch2 Plate1 mutant       2    M2</span></span>
<span><span class="co"># 6   Plate1_mut_A4_S91 batch2 Plate1 mutant       2    M2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">non.standard.list</span><span class="op">$</span><span class="va">count</span><span class="op">)</span> <span class="co"># list of available count matrices</span></span>
<span><span class="co"># [1] "GSE297431.cnt"         "GSE297431.cnt.mat"     "GSE297431.cnt.row2col" "sparse.mat"</span></span>
<span><span class="va">non.standard.list</span><span class="op">$</span><span class="va">count</span><span class="op">$</span><span class="va">GSE297431.cnt</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># count matrix</span></span>
<span><span class="co">#       Plate1_mut_A1_S70 Plate1_mut_A11_S135 Plate1_mut_A12_S141 Plate1_mut_A2_S77 Plate1_mut_A3_S84</span></span>
<span><span class="co"># Gnai3               382                 201                 279               261                 8</span></span>
<span><span class="co"># Pbsn                  0                   0                   0                 0                 0</span></span>
<span><span class="co"># Cdc45               117                  86                  56               230                 7</span></span>
<span><span class="co"># Scml2               268                 116                 204               105                31</span></span>
<span><span class="co"># Apoh                  0                   0                   0                 0                 0</span></span></code></pre></div>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
