<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GEfetch2R">
<title>DownloadMatrices • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DownloadMatrices">
<meta property="og:description" content="GEfetch2R">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a>
    <a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a>
    <a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a>
    <a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>DownloadMatrices</h1>
            
            <h4 data-toc-skip class="date">2023-07-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/DownloadMatrices.Rmd" class="external-link"><code>vignettes/DownloadMatrices.Rmd</code></a></small>
      <div class="d-none name"><code>DownloadMatrices.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> provides functions for users to download
<strong>count matrices</strong> and <strong>annotations</strong>
(e.g. cell type annotation and composition) from GEO and some
single-cell databases (e.g. <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> and <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>).
<code>GEfetch2R</code> also supports loading the downloaded data to
<code>Seurat</code>.</p>
<p>Until now, the public resources supported and the returned
values:</p>
<table class="table">
<thead><tr>
<th>
Resources
</th>
<th>
URL
</th>
<th>
Download Type
</th>
<th>
Returned values
</th>
</tr></thead>
<tbody>
<tr>
<td>
GEO
</td>
<td>
<a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a>
</td>
<td>
count matrix
</td>
<td>
SeuratObject or count matrix for bulk RNA-seq
</td>
</tr>
<tr>
<td>
PanglaoDB
</td>
<td>
<a href="https://panglaodb.se/index.html" class="external-link uri">https://panglaodb.se/index.html</a>
</td>
<td>
count matrix and annotation
</td>
<td>
SeuratObject
</td>
</tr>
<tr>
<td>
UCSC Cell Browser
</td>
<td>
<a href="https://cells.ucsc.edu/" class="external-link uri">https://cells.ucsc.edu/</a>
</td>
<td>
count matrix and annotation
</td>
<td>
SeuratObject
</td>
</tr>
</tbody>
</table>
<hr>
</div>
<div class="section level2">
<h2 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is
an international public repository that archives and freely distributes
microarray, next-generation sequencing, and other forms of
high-throughput functional genomics data submitted by the research
community.</a> It provides a very convenient way for users to explore
and select interested scRNA-seq datasets.</p>
<pre><code><span><span class="co">## Warning: replacing previous import 'LoomExperiment::import' by</span></span>
<span><span class="co">## 'reticulate::import' when loading 'GEfetch2R'</span></span></code></pre>
<div class="section level3">
<h3 id="extract-metadata">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ExtractGEOMeta</code> to
extract sample metadata, including sample title, source name/tissue,
description, cell type, treatment, paper title, paper abstract,
organism, protocol, data processing methods, et al.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">GEfetch2R</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract metadata of specified platform</span></span>
<span><span class="va">GSE200257.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="st">"GPL24676"</span><span class="op">)</span></span>
<span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata of all platforms</span></span>
<span><span class="va">GSE94820.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>Show the metadata:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSE94820.meta</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="download-matrix-and-load-to-seurat">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat"></a>
</h3>
<p>After manually check the extracted metadata, users can
<strong>download count matrix</strong> and <strong>load the count
matrix</strong> to Seurat with <code>ParseGEO</code>.</p>
<p>For count matrix, <code>ParseGEO</code> supports downloading the
matrix from <strong>supplementary files</strong> and extracting from
<code>ExpressionSet</code>, users can control the source by specifying
<code>down.supp</code> or detecting automatically (<code>ParseGEO</code>
will extract the count matrix from <code>ExpressionSet</code> first, if
the count matrix is NULL or contains non-integer values,
<code>ParseGEO</code> will download supplementary files). While the
supplementary files have two main types: single count matrix file
containing all cells and CellRanger-style outputs (barcode, matrix,
feature/gene), users are required to choose the type of supplementary
files with <code>supp.type</code>.</p>
<p>With the count matrix, <code>ParseGEO</code> will load the matrix to
Seurat automatically. If multiple samples available, users can choose to
merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cellranger out</span></span>
<span><span class="va">GSE200257.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/Volumes/soyabean/GEfetch2R/dwonload_geo"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># count matrix</span></span>
<span><span class="va">GSE94820.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="st">"count"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the <code>GSE200257.seu</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for cellranger output</span></span>
<span><span class="va">GSE200257.seu</span></span></code></pre></div>
<p>Show the <code>GSE94820.seu</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for count matrix</span></span>
<span><span class="va">GSE94820.seu</span></span></code></pre></div>
<p>The structure of downloaded count matrix for 10x:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> /Volumes/soyabean/GEfetch2R/dwonload_geo</span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/Volumes/soyabean/GEfetch2R/dwonload_geo</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025652_1</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #00BB00; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025653_2</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #00BB00; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">GSM6025654_3</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   ├── <span style="color: #00BB00; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #00BB00; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## └── <span style="color: #0000BB; font-weight: bold;">GSM6025655_4</span></span></span>
<span><span class="co">##     ├── <span style="color: #00BB00; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">##     ├── <span style="color: #00BB00; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">##     └── <span style="color: #00BB00; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## 5 directories, 12 files</span></span></code></pre>
<p><strong>For bulk RNA-seq</strong>, set
<code>data.type = "bulk"</code> in <code>ParseGEO</code>, this will
return count matrix.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="panglaodb">PanglaoDB<a class="anchor" aria-label="anchor" href="#panglaodb"></a>
</h2>
<p><a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> is a database
which contains scRNA-seq datasets from mouse and human. Up to now, it
contains <strong>5,586,348 cells</strong> from <strong>1368 datasets
(1063 from Mus musculus and 305 from Homo sapiens)</strong>. It has well
organized metadata for every dataset, including tissue, protocol,
species, number of cells and cell type annotation (computationally
identified). Daniel Osorio has developed <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a> to access
<a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> data, the
functions of <code>GEfetch2R</code> here are based on <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a>.</p>
<p>Since <a href="https://panglaodb.se/about.html" class="external-link">PanglaoDB</a> is no
longer maintained, <code>GEfetch2R</code> has cached all metadata and
cell type composition and use these cached data by default to
accelerate, users can access the cached data with
<code>PanglaoDBMeta</code> (all metadata) and
<code>PanglaoDBComposition</code> (all cell type composition).</p>
<div class="section level3">
<h3 id="summary-attributes">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use cached metadata</span></span>
<span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $species</span></span>
<span><span class="co">##          Value  Num     Key</span></span>
<span><span class="co">## 1 Mus musculus 1063 species</span></span>
<span><span class="co">## 2 Homo sapiens  305 species</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $protocol</span></span>
<span><span class="co">##           Value  Num      Key</span></span>
<span><span class="co">## 1  10x chromium 1046 protocol</span></span>
<span><span class="co">## 2      drop-seq  204 protocol</span></span>
<span><span class="co">## 3 microwell-seq   74 protocol</span></span>
<span><span class="co">## 4    Smart-seq2   26 protocol</span></span>
<span><span class="co">## 5   C1 Fluidigm   16 protocol</span></span>
<span><span class="co">## 6       CEL-seq    1 protocol</span></span>
<span><span class="co">## 7       inDrops    1 protocol</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-metadata-1">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-1"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ExtractPanglaoDBMeta</code> to
select interested datasets with specified <strong>species</strong>,
<strong>protocol</strong>, <strong>tissue</strong> and <strong>cell
number</strong> (The available values of these attributes can be
obtained with <code>StatDBAttribute</code>). User can also choose to
whether to add cell type annotation to every dataset with
<code>show.cell.type</code>.</p>
<p><code>GEfetch2R</code> uses cached metadata and cell type composition
by default, users can change this by setting
<code>local.data = FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>,</span>
<span>  show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the metadata:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         SRA        SRS                             Tissue     Protocol</span></span>
<span><span class="co">## 1 SRA550660 SRS2089635 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 2 SRA550660 SRS2089636 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 3 SRA550660 SRS2089638 Peripheral blood mononuclear cells 10x chromium</span></span>
<span><span class="co">## 4 SRA605365 SRS2492922            Nasal airway epithelium 10x chromium</span></span>
<span><span class="co">## 5 SRA608611 SRS2517316                   Lung progenitors 10x chromium</span></span>
<span><span class="co">## 6 SRA608353 SRS2517519           Hepatocellular carcinoma 10x chromium</span></span>
<span><span class="co">##        Species Cells</span></span>
<span><span class="co">## 1 Homo sapiens  1860</span></span>
<span><span class="co">## 2 Homo sapiens  1580</span></span>
<span><span class="co">## 3 Homo sapiens  1818</span></span>
<span><span class="co">## 4 Homo sapiens  1932</span></span>
<span><span class="co">## 5 Homo sapiens  1077</span></span>
<span><span class="co">## 6 Homo sapiens  1230</span></span>
<span><span class="co">##                                                                      CellType</span></span>
<span><span class="co">## 1                                                           Unknown, NK cells</span></span>
<span><span class="co">## 2                              Unknown, T cells, Plasmacytoid dendritic cells</span></span>
<span><span class="co">## 3 Unknown, Gamma delta T cells, Dendritic cells, Plasmacytoid dendritic cells</span></span>
<span><span class="co">## 4       Luminal epithelial cells, Basal cells, Keratinocytes, Ependymal cells</span></span>
<span><span class="co">## 5                                           Unknown, Hepatocytes, Basal cells</span></span>
<span><span class="co">## 6                                        Unknown, Hepatocytes, Foveolar cells</span></span>
<span><span class="co">##   CellNum</span></span>
<span><span class="co">## 1    1860</span></span>
<span><span class="co">## 2    1580</span></span>
<span><span class="co">## 3    1818</span></span>
<span><span class="co">## 4    1932</span></span>
<span><span class="co">## 5    1077</span></span>
<span><span class="co">## 6    1230</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-cell-type-composition">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition"></a>
</h3>
<p><code>GEfetch2R</code> provides
<code>ExtractPanglaoDBComposition</code> to extract cell type annotation
and composition (use cached data by default to accelerate, users can
change this by setting <code>local.data = FALSE</code>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the extracted cell type annotation and composition:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa.composition</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           SRA        SRS                        Tissue     Protocol</span></span>
<span><span class="co">## 1.1 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.2 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.3 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.4 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.5 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">## 1.6 SRA553822 SRS2119548 Cultured embryonic stem cells 10x chromium</span></span>
<span><span class="co">##          Species Cluster Cells Cell Type</span></span>
<span><span class="co">## 1.1 Homo sapiens       0  1572   Unknown</span></span>
<span><span class="co">## 1.2 Homo sapiens       1   563   Unknown</span></span>
<span><span class="co">## 1.3 Homo sapiens       2   280   Unknown</span></span>
<span><span class="co">## 1.4 Homo sapiens       3   270   Unknown</span></span>
<span><span class="co">## 1.5 Homo sapiens       4   220   Unknown</span></span>
<span><span class="co">## 1.6 Homo sapiens       5   192   Unknown</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="download-matrix-and-load-to-seurat-1">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat-1"></a>
</h3>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParsePanglaoDB</code> to <strong>download count
matrix</strong> and <strong>load the count matrix</strong> to Seurat.
With available cell type annotation, uses can filter datasets without
specified cell type with <code>cell.type</code>. Users can also
include/exclude cells expressing specified genes with
<code>include.gene</code>/<code>exclude.gene</code>.</p>
<p>With the count matrix, <code>ParsePanglaoDB</code> will load the
matrix to Seurat automatically. If multiple datasets available, users
can choose to merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Show the returned <code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.seu</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 25917 features across 4996 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (25917 features, 0 variable features)</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="ucsc-cell-browser">UCSC Cell Browser<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser"></a>
</h2>
<p>The <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a> is a
web-based tool that allows scientists to interactively visualize
scRNA-seq datasets. It contains <strong>1267 single cell
datasets</strong> from <strong>25 different species</strong>. And, it is
<strong>organized with the hierarchical structure</strong>, which can
help users quickly locate the datasets they are interested in.</p>
<div class="section level3">
<h3 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ShowCBDatasets</code> to show
all available datasets. Due to the large number of datasets,
<code>ShowCBDatasets</code> enables users to perform <em>lazy load</em>
of dataset json files instead of downloading the json files online
(time-consuming!!!). This <em>lazy load</em> requires users to provide
<code>json.folder</code> to save json files and set
<code>lazy = TRUE</code> (for the first time of run,
<code>ShowCBDatasets</code> will download current json files to
<code>json.folder</code>, for next time of run, with
<code>lazy = TRUE</code>, <code>ShowCBDatasets</code> will load the
downloaded json files from <code>json.folder</code>.). And,
<code>ShowCBDatasets</code> supports updating the local datasets with
<code>update = TRUE</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first time run, the json files are stored under json.folder</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = "/Volumes/soyabean/GEfetch2R/cell_browser/json", update = TRUE)</span></span>
<span></span>
<span><span class="co"># second time run, load the downloaded json files</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="st">"/Volumes/soyabean/GEfetch2R/cell_browser/json"</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># always read online</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)</span></span></code></pre></div>
<p>Show the metadata:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                 name                          shortLabel</span></span>
<span><span class="co">## 1      ad-aging-brain/ad-aging-brain Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">## 2     ad-aging-brain/ad-atac/erosion Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">## 3          ad-aging-brain/ad-atac/hq Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">## 4 ad-aging-brain/ad-atac/integration Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">## 5      ad-aging-brain/ad-atac/rna-hq Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">## 6    ad-aging-brain/microglia-states Aging Brain and Alzheimer's Disease</span></span>
<span><span class="co">##                                                            subLabel tags</span></span>
<span><span class="co">## 1             AD and Aging Prefrontal Cortex Across 427 individuals  10x</span></span>
<span><span class="co">## 2                                  snATAC-seq of Epigenomic Erosion  10x</span></span>
<span><span class="co">## 3 snATAC-seq - AD and aging prefrontal cortex across 92 individuals  10x</span></span>
<span><span class="co">## 4                               Integrated snRNA-seq and snATAC-seq  10x</span></span>
<span><span class="co">## 5  snRNA-seq - AD and aging prefrontal cortex across 92 individuals  10x</span></span>
<span><span class="co">## 6                                Microglia States In The Aged Brain  10x</span></span>
<span><span class="co">##                         body_parts                   diseases</span></span>
<span><span class="co">## 1 brain, cortex, prefrontal cortex        Alzheimer’s disease</span></span>
<span><span class="co">## 2                            brain Alzheimer's disease|parent</span></span>
<span><span class="co">## 3                            brain Alzheimer's disease|parent</span></span>
<span><span class="co">## 4                            brain Alzheimer's disease|parent</span></span>
<span><span class="co">## 5                            brain Alzheimer's disease|parent</span></span>
<span><span class="co">## 6 brain, cortex, prefrontal cortex        Alzheimer’s disease</span></span>
<span><span class="co">##                   organisms      projects life_stages domains sources</span></span>
<span><span class="co">## 1        Human (H. sapiens)        ROSMAP               atlas      NA</span></span>
<span><span class="co">## 2 Human (H. sapiens)|parent ROSMAP|parent                          NA</span></span>
<span><span class="co">## 3 Human (H. sapiens)|parent ROSMAP|parent                          NA</span></span>
<span><span class="co">## 4 Human (H. sapiens)|parent ROSMAP|parent                          NA</span></span>
<span><span class="co">## 5 Human (H. sapiens)|parent ROSMAP|parent                          NA</span></span>
<span><span class="co">## 6        Human (H. sapiens)        ROSMAP               atlas      NA</span></span>
<span><span class="co">##   sampleCount assays              matrix               barcode</span></span>
<span><span class="co">## 1     2327742        ad430_matrix.mtx.gz ad430_barcodes.tsv.gz</span></span>
<span><span class="co">## 2      437098              matrix.mtx.gz       barcodes.tsv.gz</span></span>
<span><span class="co">## 3      171119              matrix.mtx.gz       barcodes.tsv.gz</span></span>
<span><span class="co">## 4      586083              matrix.mtx.gz       barcodes.tsv.gz</span></span>
<span><span class="co">## 5      414964              matrix.mtx.gz       barcodes.tsv.gz</span></span>
<span><span class="co">## 6      152459              matrix.mtx.gz       barcodes.tsv.gz</span></span>
<span><span class="co">##                 feature matrixType</span></span>
<span><span class="co">## 1 ad430_features.tsv.gz        10x</span></span>
<span><span class="co">## 2       features.tsv.gz        10x</span></span>
<span><span class="co">## 3       features.tsv.gz        10x</span></span>
<span><span class="co">## 4       features.tsv.gz        10x</span></span>
<span><span class="co">## 5       features.tsv.gz        10x</span></span>
<span><span class="co">## 6       features.tsv.gz        10x</span></span>
<span><span class="co">##                                                                                                                        title</span></span>
<span><span class="co">## 1 Single-cell atlas reveals correlates of high cognitive function, dementia, and resilience to Alzheimer's disease pathology</span></span>
<span><span class="co">## 2                       Epigenomic dissection of Alzheimer's disease pinpoints causal variants and reveals epigenome erosion</span></span>
<span><span class="co">## 3                       Epigenomic dissection of Alzheimer's disease pinpoints causal variants and reveals epigenome erosion</span></span>
<span><span class="co">## 4                       Epigenomic dissection of Alzheimer's disease pinpoints causal variants and reveals epigenome erosion</span></span>
<span><span class="co">## 5                       Epigenomic dissection of Alzheimer's disease pinpoints causal variants and reveals epigenome erosion</span></span>
<span><span class="co">## 6                                                         Human Microglial State Dynamics in Alzheimer's Disease Progression</span></span>
<span><span class="co">##   paper</span></span>
<span><span class="co">## 1      </span></span>
<span><span class="co">## 2      </span></span>
<span><span class="co">## 3      </span></span>
<span><span class="co">## 4      </span></span>
<span><span class="co">## 5      </span></span>
<span><span class="co">## 6      </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      abstract</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                           \nAlzheimer's disease (AD) is the most common cause of dementia worldwide, but\nthe molecular and cellular mechanisms underlying cognitive impairment remain\npoorly understood. To address this, we generated a single-cell transcriptomic\natlas of the aged human prefrontal cortex covering 2.3 million cells from\npost-mortem human brain samples of 427 individuals with varying degrees of AD\npathology and cognitive impairment. Our analyses identified AD\npathology-associated alterations shared between excitatory neuron subtypes,\nrevealed a coordinated increase of the cohesin complex and DNA damage response\nfactors in excitatory neurons and in oligodendrocytes, and uncovered genes and\npathways associated with high cognitive function, dementia, and resilience to\nAD pathology. Furthermore, we identified selectively vulnerable somatostatin\ninhibitory neuron subtypes depleted in AD, discovered two distinct groups of\ninhibitory neurons that were more abundant in individuals with preserved high\ncognitive function late in life, and uncovered a link between inhibitory\nneurons and resilience to AD pathology.\n</span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                \nRecent work has identified dozens of non-coding loci for Alzheimer's disease\n(AD) risk, but their mechanisms and AD transcriptional regulatory circuitry are\npoorly understood. Here, we profile epigenomic and transcriptomic landscapes of\n850k nuclei from prefrontal cortexes of 92 individuals with and without AD to\nbuild a map of the brain regulome, including epigenomic profiles,\ntranscriptional regulators, co-accessibility modules, and peak-to-gene links in\na cell-type-specific manner. We develop methods for multimodal integration and\ndetecting regulatory modules using peak-to-gene linking. We show AD risk loci\nare enriched in microglial enhancers and for specific TFs including SPI1, ELF2,\nand RUNX1. We detect 9,628 cell-type-specific ATAC-QTL loci, which we integrate\nalongside peak-to-gene links to prioritize AD variant regulatory circuits. We\nreport differential accessibility of regulatory modules in late-AD in glia and\nin early-AD in neurons. Strikingly, late-stage AD brains show global epigenome\ndysregulation indicative of epigenome erosion and cell identity loss.\n</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                \nRecent work has identified dozens of non-coding loci for Alzheimer's disease\n(AD) risk, but their mechanisms and AD transcriptional regulatory circuitry are\npoorly understood. Here, we profile epigenomic and transcriptomic landscapes of\n850k nuclei from prefrontal cortexes of 92 individuals with and without AD to\nbuild a map of the brain regulome, including epigenomic profiles,\ntranscriptional regulators, co-accessibility modules, and peak-to-gene links in\na cell-type-specific manner. We develop methods for multimodal integration and\ndetecting regulatory modules using peak-to-gene linking. We show AD risk loci\nare enriched in microglial enhancers and for specific TFs including SPI1, ELF2,\nand RUNX1. We detect 9,628 cell-type-specific ATAC-QTL loci, which we integrate\nalongside peak-to-gene links to prioritize AD variant regulatory circuits. We\nreport differential accessibility of regulatory modules in late-AD in glia and\nin early-AD in neurons. Strikingly, late-stage AD brains show global epigenome\ndysregulation indicative of epigenome erosion and cell identity loss.\n</span></span>
<span><span class="co">## 4                                                                                                                                                                                                                                                                \nRecent work has identified dozens of non-coding loci for Alzheimer's disease\n(AD) risk, but their mechanisms and AD transcriptional regulatory circuitry are\npoorly understood. Here, we profile epigenomic and transcriptomic landscapes of\n850k nuclei from prefrontal cortexes of 92 individuals with and without AD to\nbuild a map of the brain regulome, including epigenomic profiles,\ntranscriptional regulators, co-accessibility modules, and peak-to-gene links in\na cell-type-specific manner. We develop methods for multimodal integration and\ndetecting regulatory modules using peak-to-gene linking. We show AD risk loci\nare enriched in microglial enhancers and for specific TFs including SPI1, ELF2,\nand RUNX1. We detect 9,628 cell-type-specific ATAC-QTL loci, which we integrate\nalongside peak-to-gene links to prioritize AD variant regulatory circuits. We\nreport differential accessibility of regulatory modules in late-AD in glia and\nin early-AD in neurons. Strikingly, late-stage AD brains show global epigenome\ndysregulation indicative of epigenome erosion and cell identity loss.\n</span></span>
<span><span class="co">## 5                                                                                                                                                                                                                                                                \nRecent work has identified dozens of non-coding loci for Alzheimer's disease\n(AD) risk, but their mechanisms and AD transcriptional regulatory circuitry are\npoorly understood. Here, we profile epigenomic and transcriptomic landscapes of\n850k nuclei from prefrontal cortexes of 92 individuals with and without AD to\nbuild a map of the brain regulome, including epigenomic profiles,\ntranscriptional regulators, co-accessibility modules, and peak-to-gene links in\na cell-type-specific manner. We develop methods for multimodal integration and\ndetecting regulatory modules using peak-to-gene linking. We show AD risk loci\nare enriched in microglial enhancers and for specific TFs including SPI1, ELF2,\nand RUNX1. We detect 9,628 cell-type-specific ATAC-QTL loci, which we integrate\nalongside peak-to-gene links to prioritize AD variant regulatory circuits. We\nreport differential accessibility of regulatory modules in late-AD in glia and\nin early-AD in neurons. Strikingly, late-stage AD brains show global epigenome\ndysregulation indicative of epigenome erosion and cell identity loss.\n</span></span>
<span><span class="co">## 6 \nAltered microglial states affect neuroinflammation, neurodegeneration, and\ndisease, but remain poorly understood. Here, we report 194k single-nucleus\nmicroglial transcriptomes and epigenomes across 443 human subjects, and diverse\nAlzheimer's disease (AD) pathological phenotypes. We annotate 12 microglial\ntranscriptional states, including AD-dysregulated homeostatic, inflammatory,\nand lipid-processing states. We identify 1,542 AD-differentially-expressed\ngenes, including both microglia-state-specific and disease-stage-specific\nalterations. By integrating epigenomic, transcriptomic, and motif information,\nwe infer upstream regulators of microglial cell states, gene-regulatory\nnetworks, enhancer-gene links, and transcription factor-driven microglial state\ntransitions. We demonstrate that ectopic expression of our predicted\nhomeostatic-state activators induces homeostatic features in human iPSC-derived\nmicroglia-like cells, while inhibiting activators of inflammation can block\ninflammatory progression. Lastly, we pinpoint the expression of AD-risk genes\nin microglial states and differential expression of AD-risk genes and their\nregulators during AD progression. Overall, we provide insights underlying\nmicroglial states, including state-specific and AD-stage-specific microglial\nalterations at unprecedented resolution. \n</span></span>
<span><span class="co">##   unit                         coords</span></span>
<span><span class="co">## 1      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">## 2      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">## 3      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">## 4      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">## 5      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">## 6      UMAP_coordinates.coords.tsv.gz</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  methods</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n&lt;p&gt;\nWe selected 427 individuals from the Religious Orders Study and Rush Memory and\nAging Project (ROSMAP), both ongoing longitudinal clinical-pathologic cohort\nstudies of aging and dementia, in which all participants are brain donors. The\nstudies include clinical data collected annually, detailed post- mortem\npathological evaluations, and extensive genetic, epigenomic, transcriptomic,\nproteomic, and metabolomic bulk-tissue profiling (DA et al., 2018) .\nIndividuals were balanced between sexes (male:female ratio 212:215). Informed\nconsent was obtained from each subject, and the Religious Orders Study and Rush\nMemory and Aging Project were each approved by an Institutional Review Board\n(IRB) of Rush University Medical Center. Participants also signed an Anatomic\nGift Act, and a repository consent to allow their data to be repurposed.\n\n&lt;p&gt;\nFor droplet-based snRNA-seq, libraries were prepared using the Chromium Single\nCell 3′ Reagent Kits v3 according to the manufacturer's protocol (10x\nGenomics). The generated snRNA-seq libraries were sequenced using NextSeq\n500/550 High Output v2 kits (150 cycles) or NovaSeq 6000 S2 Reagent Kits. Gene\ncounts were obtained by aligning reads to the GRCh38 genome using Cell Ranger\nsoftware (v.3.0.2) (10x Genomics) (Zheng et al., 2017). To account for\nunspliced nuclear transcripts, reads mapping to pre-mRNA were counted. After\nquantification of pre-mRNA using the Cell Ranger count pipeline, the Cell\nRanger aggr pipeline was used to aggregate all libraries (without equalizing\nthe read depth between groups) to generate a gene-count matrix. The Cell Ranger\n3.0 default parameters were used to call cell barcodes. We used SCANPY (Wolf,\nAngerer and Theis, 2018) to process and cluster the expression profiles and\ninfer cell identities of major cell classes. To remove doublets and\npoor-quality cells, cells were excluded from subsequent analysis if they were\nextreme outliers (observations outside the range [Q1 – k(Q3 – Q1),Q3 +\nk(Q3-Q1)], with k=3 and Q1 and Q3 as the lower and upper quartiles) in terms of\nnumber of genes, number of unique molecular identifiers (UMIs), and percentage\nof mitochondrial genes. In addition, we called doublets using Scrublet (Wolock,\nLopez and Klein, 2019) and flagged and removed cells that were labeled as\ndoublets.\n</span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n&lt;p&gt;\nWe selected 427 individuals from ROSMAP (Mathys et al., co-submitted) and\nperformed snRNA-seq on the prefrontal cortex region of these samples. For 92 of\nthe samples, we carried out snATAC profiling as well. The 92 samples are with a\nvariety of AD-related pathological and cognitive measurements, and are grouped\ninto 48 controls (22 female; 26 male), 29 early-stage AD (15 female; 14 male)\nand 15 late-stage AD (9 female; 6 male) samples. The ages between AD groups are\nmatched, without significant difference).  \n\n&lt;p&gt;\nWe used "cellranger-atac mkfastq" (V1.1.0) to demultiplex BCL files into Fastq\nraw sequencing data, and then ran "cellranger-atac count" to map the reads onto\nhuman reference genome (GRCh38) and obtain the fragment file of each sample. We\nthen utilized ArchR (V1.0.1) to process the snATAC-seq data, using fragment\nfiles as input (Granja et al. 2021). We performed the first round of doublet\nremoval within each sample using the "filterDoublets" function.\n\n&lt;p&gt;\nFor the "high-quality cell" analysis, we selected the cells with TSS enrichment\n&gt; 6 and number of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=4, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We estimated gene expression score using ArchR, and then\nquantified the cell type signature of each cluster based on PsychENCODE marker\ngenes (PsychENCODE Consortium et al. 2015). We calculated the mean of gene\nscores of all the cell marker genes in each cell type for each cluster, which\nwere then used to assign cell types. We performed a second round of doublet\nremoval at cluster level by discarding the cell clusters that show ambiguous\ncell type marker signal and meanwhile locate between two clusters that show\nclear cell type signature.\n\n&lt;p&gt;\nFor the "erosion" analysis, we selected the cells with TSS enrichment &gt; 1 and\nnumber of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=3, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We performed a second round of doublet removal at cluster\nlevel using the same strategy as "TSS enrichment &gt; 6" analysis described above. \n</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n&lt;p&gt;\nWe selected 427 individuals from ROSMAP (Mathys et al., co-submitted) and\nperformed snRNA-seq on the prefrontal cortex region of these samples. For 92 of\nthe samples, we carried out snATAC profiling as well. The 92 samples are with a\nvariety of AD-related pathological and cognitive measurements, and are grouped\ninto 48 controls (22 female; 26 male), 29 early-stage AD (15 female; 14 male)\nand 15 late-stage AD (9 female; 6 male) samples. The ages between AD groups are\nmatched, without significant difference).  \n\n&lt;p&gt;\nWe used "cellranger-atac mkfastq" (V1.1.0) to demultiplex BCL files into Fastq\nraw sequencing data, and then ran "cellranger-atac count" to map the reads onto\nhuman reference genome (GRCh38) and obtain the fragment file of each sample. We\nthen utilized ArchR (V1.0.1) to process the snATAC-seq data, using fragment\nfiles as input (Granja et al. 2021). We performed the first round of doublet\nremoval within each sample using the "filterDoublets" function.\n\n&lt;p&gt;\nFor the "high-quality cell" analysis, we selected the cells with TSS enrichment\n&gt; 6 and number of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=4, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We estimated gene expression score using ArchR, and then\nquantified the cell type signature of each cluster based on PsychENCODE marker\ngenes (PsychENCODE Consortium et al. 2015). We calculated the mean of gene\nscores of all the cell marker genes in each cell type for each cluster, which\nwere then used to assign cell types. We performed a second round of doublet\nremoval at cluster level by discarding the cell clusters that show ambiguous\ncell type marker signal and meanwhile locate between two clusters that show\nclear cell type signature.\n\n&lt;p&gt;\nFor the "erosion" analysis, we selected the cells with TSS enrichment &gt; 1 and\nnumber of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=3, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We performed a second round of doublet removal at cluster\nlevel using the same strategy as "TSS enrichment &gt; 6" analysis described above. \n</span></span>
<span><span class="co">## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n&lt;p&gt;\nWe selected 427 individuals from ROSMAP (Mathys et al., co-submitted) and\nperformed snRNA-seq on the prefrontal cortex region of these samples. For 92 of\nthe samples, we carried out snATAC profiling as well. The 92 samples are with a\nvariety of AD-related pathological and cognitive measurements, and are grouped\ninto 48 controls (22 female; 26 male), 29 early-stage AD (15 female; 14 male)\nand 15 late-stage AD (9 female; 6 male) samples. The ages between AD groups are\nmatched, without significant difference).  \n\n&lt;p&gt;\nWe used "cellranger-atac mkfastq" (V1.1.0) to demultiplex BCL files into Fastq\nraw sequencing data, and then ran "cellranger-atac count" to map the reads onto\nhuman reference genome (GRCh38) and obtain the fragment file of each sample. We\nthen utilized ArchR (V1.0.1) to process the snATAC-seq data, using fragment\nfiles as input (Granja et al. 2021). We performed the first round of doublet\nremoval within each sample using the "filterDoublets" function.\n\n&lt;p&gt;\nFor the "high-quality cell" analysis, we selected the cells with TSS enrichment\n&gt; 6 and number of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=4, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We estimated gene expression score using ArchR, and then\nquantified the cell type signature of each cluster based on PsychENCODE marker\ngenes (PsychENCODE Consortium et al. 2015). We calculated the mean of gene\nscores of all the cell marker genes in each cell type for each cluster, which\nwere then used to assign cell types. We performed a second round of doublet\nremoval at cluster level by discarding the cell clusters that show ambiguous\ncell type marker signal and meanwhile locate between two clusters that show\nclear cell type signature.\n\n&lt;p&gt;\nFor the "erosion" analysis, we selected the cells with TSS enrichment &gt; 1 and\nnumber of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=3, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We performed a second round of doublet removal at cluster\nlevel using the same strategy as "TSS enrichment &gt; 6" analysis described above. \n</span></span>
<span><span class="co">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n&lt;p&gt;\nWe selected 427 individuals from ROSMAP (Mathys et al., co-submitted) and\nperformed snRNA-seq on the prefrontal cortex region of these samples. For 92 of\nthe samples, we carried out snATAC profiling as well. The 92 samples are with a\nvariety of AD-related pathological and cognitive measurements, and are grouped\ninto 48 controls (22 female; 26 male), 29 early-stage AD (15 female; 14 male)\nand 15 late-stage AD (9 female; 6 male) samples. The ages between AD groups are\nmatched, without significant difference).  \n\n&lt;p&gt;\nWe used "cellranger-atac mkfastq" (V1.1.0) to demultiplex BCL files into Fastq\nraw sequencing data, and then ran "cellranger-atac count" to map the reads onto\nhuman reference genome (GRCh38) and obtain the fragment file of each sample. We\nthen utilized ArchR (V1.0.1) to process the snATAC-seq data, using fragment\nfiles as input (Granja et al. 2021). We performed the first round of doublet\nremoval within each sample using the "filterDoublets" function.\n\n&lt;p&gt;\nFor the "high-quality cell" analysis, we selected the cells with TSS enrichment\n&gt; 6 and number of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=4, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We estimated gene expression score using ArchR, and then\nquantified the cell type signature of each cluster based on PsychENCODE marker\ngenes (PsychENCODE Consortium et al. 2015). We calculated the mean of gene\nscores of all the cell marker genes in each cell type for each cluster, which\nwere then used to assign cell types. We performed a second round of doublet\nremoval at cluster level by discarding the cell clusters that show ambiguous\ncell type marker signal and meanwhile locate between two clusters that show\nclear cell type signature.\n\n&lt;p&gt;\nFor the "erosion" analysis, we selected the cells with TSS enrichment &gt; 1 and\nnumber of fragments between 1000 to 100,000. We performed Iterative LSI\ndimension reduction and clustering using a 500 bp tile matrix, with parameters\n"iterations=3, resolution=0.2, varFeat=50000". Cell embedding visualization was\nperformed using UMAP. We performed a second round of doublet removal at cluster\nlevel using the same strategy as "TSS enrichment &gt; 6" analysis described above. \n</span></span>
<span><span class="co">## 6 \n&lt;p&gt;\nNuclei isolation from frozen postmortem brain tissue: We isolated nuclei from\nfrozen postmortem brain tissue as previously described (Mathys et al., Nature\n2019) with some modifications. Briefly, we homogenized the brain tissue in 700\nμL Homogenization buffer and filtered the homogenate through a 40 μm cell\nstrainer (Corning, NY), added 450 μL Working solution and loaded it as a 25%\nOptiPrep solution on top of a 30%/40% OptiPrep density gradient (750 μL 30%\nOptiPrep solution, 300 μL 40% OptiPrep solution). We separated the nuclei by\ncentrifugation using a fixed rotor, tabletop centrifuge (5 minutes, 10000g,\n4˚C). We collected the nuclei pellet at the 30%/40% interphase, transferred it\non a new tube, washed it twice with 1mL ice-cold PBS containing 0.04% BSA\n(centrifuged 3 minutes, 300g, 4˚C) and finally resuspended it in 100 μL PBS\ncontaining 0.04% BSA. After counting, we diluted the nuclei to a concentration\nof 1000 nuclei per μL. We used the isolated nuclei for the droplet-based 10x\nscRNA-seq assay, targeting 5000 nuclei per brain region and individual, and\nprepared libraries using the Chromium Single-Cell 3′ Reagent Kits v3 (10x\nGenomics, Pleasanton, CA) according to the manufacturer's protocol. We\nsequenced pooled libraries using the NovaSeq 6000 S2 sequencing kits (100\ncycles, Illumina).\n\n&lt;p&gt;\nsnRNA-seq data processing: We mapped the raw reads to human reference genome\nversion GRCh38 and quantified unique molecular identifiers (UMIs) counts for\neach gene in each cell using CellRanger software v3.0.1 (10x Genomics). We\npre-processed this count matrix (gene-cell) using the Seurat R package\nv.4.0.3). We kept the cells with more than 500 UMIs and less than 5%\nmitochondrial genes, and genes with expression at least in 50 cells for further\nanalysis. We normalized the counts by the total UMI counts for each cell,\nmultiplied by 10,000, and then log-transformed. We used the top 2,000 highly\nvariable genes for principal component analysis (PCA) and  the top 30 principal\ncomponents (PCs) as inputs to perform UMAP. We used Harmony for batch\ncorrection. We used the resolution as 0.5 to identify clusters. We used\nDoubletFinder to estimate the potential doublets formed by two or more cells\nbased on the by default parameters. The cells with high doublet scores (0.2 as\ncutoff) were removed for further analysis.After generating clusters, one\ncluster showing high expression of markers of two or more cell types was also\ntreated as doublets and removed for further analysis.\n\n&lt;p&gt;\nIn Silico Sorting to enrich immune cells and cell type annotation in snRNA-seq\ndata: For the full datasets with all cell types (2.8 million cells), we first\nannotated the cell type for each cluster based on three widely-used canonical\nmarkers of major cell types in the brain (including excitatory and inhibitory\nneurons, astrocytes, oligodendrocytes, OPCs, microglia and vascular cells) and\na list of markers for immune cells. We also tested the enrichment of a large\nset of markers in highly expressed genes for each cluster to confirm the\nannotation based on several marker genes. We next calculated the cell type\nscores (i.e. astrocyte, oligodendrocyte, microglia, etc) for each cell, which\nwere represented by the average expression of a group of markers for each cell\ntype. The cells were then selected as microglia/immune cells for further\nintegrative analysis if and only if (1) the clusters that the cells belong to\nwere annotated as microglia/immune cells; and (2) the cells had the highest\nscore for microglia/immune cells, and 3) the score for microglia/immune cells\nwas 2-fold higher than the second highest score. For the selected\nmicroglia/immune cells, we followed the same pipeline to perform dimensional\nreduction and clustering with the same parameters as full datasets. We used the\nWilcoxon rank-sum test in Seurat with customized parameters (min.pct = 0.25,\nlogfc.threshold = 0.25) to identify highly expressed genes for each cluster\ncompared to all cells from other clusters.\n</span></span>
<span><span class="co">##   geo</span></span>
<span><span class="co">## 1    </span></span>
<span><span class="co">## 2    </span></span>
<span><span class="co">## 3    </span></span>
<span><span class="co">## 4    </span></span>
<span><span class="co">## 5    </span></span>
<span><span class="co">## 6</span></span></code></pre>
<p>The number of datasets and all available species:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the number of datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1318</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># available species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\|parent"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">ucsc.cb.samples</span><span class="op">$</span><span class="va">organisms</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Human (H. sapiens)"                 "Mouse (M. musculus)"               </span></span>
<span><span class="co">##  [3] "Rhesus macaque (M. mulatta)"        "Chimp (P. troglodytes)"            </span></span>
<span><span class="co">##  [5] "Brine Shrimp (A. franciscana)"      "Canis lupus familiaris"            </span></span>
<span><span class="co">##  [7] "Dog (C. familiaris)"                "Human (H. Sapiens)"                </span></span>
<span><span class="co">##  [9] "C. intestinalis"                    "C. robusta"                        </span></span>
<span><span class="co">## [11] "Zebrafish (D. rerio)"               "Fruit fly (D. melanogaster)"       </span></span>
<span><span class="co">## [13] "Horse (E. caballus)"                "Hydra vulgaris"                    </span></span>
<span><span class="co">## [15] "Capitella teleta"                   "Spongilla lacustris"               </span></span>
<span><span class="co">## [17] "H. symbiolongicarpus"               "X. tropicalis"                     </span></span>
<span><span class="co">## [19] "Marmoset (C. jacchus)"              "P. leidyi"                         </span></span>
<span><span class="co">## [21] "Bonobo (P. paniscus)"               "Rat (R. norvegicus)"               </span></span>
<span><span class="co">## [23] "S. mansoni"                         "Starlet Sea Anemone (N. vectensis)"</span></span>
<span><span class="co">## [25] "Nematostella vectensis"             "Sea urchin (S. purpuratus)"        </span></span>
<span><span class="co">## [27] "Mouse lemur (T. microcebus)"        "Human-Mouse Xenograft"             </span></span>
<span><span class="co">## [29] "Xenopus laevis"</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="summary-attributes-1">Summary attributes<a class="anchor" aria-label="anchor" href="#summary-attributes-1"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell
Browser</a>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"UCSC"</span>, combine <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 301 × 3</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   organisms [29]</span></span></span>
<span><span class="co">##    organisms           body_parts        Num</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> human (h. sapiens)  brain             171</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> human (h. sapiens)  eye               111</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> human (h. sapiens)  retina            109</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> mouse (m. musculus) brain              96</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> human (h. sapiens)  lung               63</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> human (h. sapiens)  muscle             46</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> human (h. sapiens)  skeletal muscle    40</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> human (h. sapiens)  cortex             38</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> human (h. sapiens)  blood              34</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> human (h. sapiens)  heart              32</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 291 more rows</span></span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-metadata-2">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-2"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ExtractCBDatasets</code> to
filter metadata with <strong>collection</strong>,
<strong>sub-collection</strong>, <strong>organ</strong>, <strong>disease
status</strong>, <strong>organism</strong>, <strong>project</strong> and
<strong>cell number</strong> (The available values of these attributes
can be obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>). All attributes except cell number support fuzzy match
with <code>fuzzy.match</code>, this is useful when selecting
datasets.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBDatasets.html">ExtractCBDatasets</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, organ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"skeletal muscle"</span><span class="op">)</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Human (H. sapiens)"</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the metadata:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                 name      shortLabel</span></span>
<span><span class="co">## 1 skeletal-muscle/embryonic/embryonic-wk7-8-myogenic Skeletal Muscle</span></span>
<span><span class="co">## 2       skeletal-muscle/fetal/fetal-wk12-14-hindlimb Skeletal Muscle</span></span>
<span><span class="co">## 3           skeletal-muscle/in-vitro/hx-protocol-wk4 Skeletal Muscle</span></span>
<span><span class="co">## 4  skeletal-muscle/in-vitro/hx-protocol-wk6-myogenic Skeletal Muscle</span></span>
<span><span class="co">## 5  skeletal-muscle/in-vitro/hx-protocol-wk8-myogenic Skeletal Muscle</span></span>
<span><span class="co">## 6         skeletal-muscle/juvenile/juvenile-hindlimb Skeletal Muscle</span></span>
<span><span class="co">##                             subLabel tags                     body_parts</span></span>
<span><span class="co">## 1 Embryonic Week 7-8 Myogenic Subset      muscle, skeletal muscle|parent</span></span>
<span><span class="co">## 2   Fetal Week 12-14 Hindlimb Muscle      muscle, skeletal muscle|parent</span></span>
<span><span class="co">## 3         HX Protocol Week 4 Culture      muscle, skeletal muscle|parent</span></span>
<span><span class="co">## 4 HX Protocol Week 6 Myogenic Subset      muscle, skeletal muscle|parent</span></span>
<span><span class="co">## 5 HX Protocol Week 8 Myogenic Subset      muscle, skeletal muscle|parent</span></span>
<span><span class="co">## 6           Juvenile Hindlimb Muscle      muscle, skeletal muscle|parent</span></span>
<span><span class="co">##         diseases                 organisms projects life_stages domains sources</span></span>
<span><span class="co">## 1 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">## 2 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">## 3 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">## 4 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">## 5 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">## 6 Healthy|parent Human (H. sapiens)|parent                                   NA</span></span>
<span><span class="co">##   sampleCount assays            matrix barcode feature matrixType</span></span>
<span><span class="co">## 1        1448        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 2        1545        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 3        1562        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 4        1598        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 5        1350        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">## 6        1982        exprMatrix.tsv.gz                     matrix</span></span>
<span><span class="co">##                                title paper</span></span>
<span><span class="co">## 1 Embryonic Week 7-8 Myogenic Subset      </span></span>
<span><span class="co">## 2   Fetal Week 12-14 Hindlimb Muscle      </span></span>
<span><span class="co">## 3         HX Protocol Week 4 Culture      </span></span>
<span><span class="co">## 4 HX Protocol Week 6 Myogenic Subset      </span></span>
<span><span class="co">## 5 HX Protocol Week 8 Myogenic Subset      </span></span>
<span><span class="co">## 6           Juvenile Hindlimb Muscle      </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                           abstract</span></span>
<span><span class="co">## 1 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from hindlimbs\nof 7-8 week human embryos (Carnegie Stage: CS19-21). Cells are subjected to\nre-clustering to reveal potential heterogeneity within the myogenic subset.\nThree independent biological samples are included.\n</span></span>
<span><span class="co">## 2                                             \nSingle cell transcriptomes of the total skeletal muscle tissues of\nhindlimbs from human fetuses of 12-14 weeks of development. Hematopoietic and\nendothelial cells are pre-excluded by FACS. Two independent biological samples\nare included.\n</span></span>
<span><span class="co">## 3                           \nSingle cell transcriptomes of cell cultures differentiated from human\npluripotent stem cells (hPSCs) following the HX protocol (Xi et al., Cell Rep,\n2017) for 4 weeks. Cells are either non-enriched or enriched using an\nendogenous PAX7-driven GFP reporter.\n</span></span>
<span><span class="co">## 4 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from human\npluripotent stem cell (hPSCs) differentiated for 6 weeks using the HX protocol\n(Xi et al., Cell Rep, 2017). Cells are subjected to re-clustering to reveal\npotential heterogeneity within the myogenic subset.\n</span></span>
<span><span class="co">## 5 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from human\npluripotent stem cell (hPSCs) differentiated for 8 weeks using the HX protocol\n(Xi et al., Cell Rep, 2017). Cells are subjected to re-clustering to reveal\npotential heterogeneity within the myogenic subset.\n</span></span>
<span><span class="co">## 6                                                           \nSingle cell transcriptomes of the gastrocnemius or quadriceps muscles from\nhuman juveniles. Hematopoietic and endothelial cells are pre-excluded by FACS.\nTwo independent biological samples at 7 and 11 years old are included.\n</span></span>
<span><span class="co">##   unit                    coords</span></span>
<span><span class="co">## 1      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">## 2      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">## 3      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">## 4      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">## 5      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">## 6      Seurat_tsne.coords.tsv.gz</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                methods</span></span>
<span><span class="co">## 1 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 2 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 3 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 4 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 5 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 6 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">##   geo</span></span>
<span><span class="co">## 1    </span></span>
<span><span class="co">## 2    </span></span>
<span><span class="co">## 3    </span></span>
<span><span class="co">## 4    </span></span>
<span><span class="co">## 5    </span></span>
<span><span class="co">## 6</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="extract-cell-type-composition-1">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition-1"></a>
</h3>
<p><code>GEfetch2R</code> provides <code>ExtractCBComposition</code> to
extract cell type annotation and composition.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span></span>
<span>  json.folder <span class="op">=</span> <span class="st">"/Volumes/soyabean/GEfetch2R/cell_browser/json"</span>,</span>
<span>  sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the extracted cell type annotation and composition:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hbb.sample.ct</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        shortLabel                           subLabel  CellType Num tags</span></span>
<span><span class="co">## 1 Skeletal Muscle Embryonic Week 7-8 Myogenic Subset        MP 785     </span></span>
<span><span class="co">## 2 Skeletal Muscle Embryonic Week 7-8 Myogenic Subset        MB 303     </span></span>
<span><span class="co">## 3 Skeletal Muscle Embryonic Week 7-8 Myogenic Subset SkM.Mesen 264     </span></span>
<span><span class="co">## 4 Skeletal Muscle Embryonic Week 7-8 Myogenic Subset        MC  96     </span></span>
<span><span class="co">## 5 Skeletal Muscle   Fetal Week 12-14 Hindlimb Muscle       MSC 822     </span></span>
<span><span class="co">## 6 Skeletal Muscle   Fetal Week 12-14 Hindlimb Muscle       SkM 554     </span></span>
<span><span class="co">##                       body_parts       diseases                 organisms</span></span>
<span><span class="co">## 1 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">## 2 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">## 3 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">## 4 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">## 5 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">## 6 muscle, skeletal muscle|parent Healthy|parent Human (H. sapiens)|parent</span></span>
<span><span class="co">##   projects life_stages domains sources sampleCount assays</span></span>
<span><span class="co">## 1                                   NA        1448       </span></span>
<span><span class="co">## 2                                   NA        1448       </span></span>
<span><span class="co">## 3                                   NA        1448       </span></span>
<span><span class="co">## 4                                   NA        1448       </span></span>
<span><span class="co">## 5                                   NA        1545       </span></span>
<span><span class="co">## 6                                   NA        1545       </span></span>
<span><span class="co">##                                title paper</span></span>
<span><span class="co">## 1 Embryonic Week 7-8 Myogenic Subset      </span></span>
<span><span class="co">## 2 Embryonic Week 7-8 Myogenic Subset      </span></span>
<span><span class="co">## 3 Embryonic Week 7-8 Myogenic Subset      </span></span>
<span><span class="co">## 4 Embryonic Week 7-8 Myogenic Subset      </span></span>
<span><span class="co">## 5   Fetal Week 12-14 Hindlimb Muscle      </span></span>
<span><span class="co">## 6   Fetal Week 12-14 Hindlimb Muscle      </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                           abstract</span></span>
<span><span class="co">## 1 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from hindlimbs\nof 7-8 week human embryos (Carnegie Stage: CS19-21). Cells are subjected to\nre-clustering to reveal potential heterogeneity within the myogenic subset.\nThree independent biological samples are included.\n</span></span>
<span><span class="co">## 2 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from hindlimbs\nof 7-8 week human embryos (Carnegie Stage: CS19-21). Cells are subjected to\nre-clustering to reveal potential heterogeneity within the myogenic subset.\nThree independent biological samples are included.\n</span></span>
<span><span class="co">## 3 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from hindlimbs\nof 7-8 week human embryos (Carnegie Stage: CS19-21). Cells are subjected to\nre-clustering to reveal potential heterogeneity within the myogenic subset.\nThree independent biological samples are included.\n</span></span>
<span><span class="co">## 4 \nSingle cell transcriptomes of the skeletal muscle (SkM) lineage from hindlimbs\nof 7-8 week human embryos (Carnegie Stage: CS19-21). Cells are subjected to\nre-clustering to reveal potential heterogeneity within the myogenic subset.\nThree independent biological samples are included.\n</span></span>
<span><span class="co">## 5                                             \nSingle cell transcriptomes of the total skeletal muscle tissues of\nhindlimbs from human fetuses of 12-14 weeks of development. Hematopoietic and\nendothelial cells are pre-excluded by FACS. Two independent biological samples\nare included.\n</span></span>
<span><span class="co">## 6                                             \nSingle cell transcriptomes of the total skeletal muscle tissues of\nhindlimbs from human fetuses of 12-14 weeks of development. Hematopoietic and\nendothelial cells are pre-excluded by FACS. Two independent biological samples\nare included.\n</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                methods</span></span>
<span><span class="co">## 1 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 2 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 3 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 4 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 5 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">## 6 &lt;h3&gt;Single cell preparation, cell capture and library construction&lt;/h3&gt;\n&lt;p&gt;\nWhole hindlimbs of developmental week 5-9 human embryos and fetuses (feet\nexcluded for week 7.75-9), total hindlimb skeletal muscles of week 12-18 human\nfetuses, and gastrocnemius or quadriceps muscles from juvenile and adult human\nsubjects were digested into single cells. Dissociated cells were either sorted\n(fetal week 12 and above) to exclude the hematopoietic and endothelial lineages\nor directly used for downstream processing. For directed myogenic\ndifferentiation, human pluripotent stem cells (hPSCs) were differentiated using\nthree published protocols (HX protocol: Xi et al., Cell Rep, 2017; JC protocol:\nChal et al., Nat Biotechnol, 2015; MS protocol: Shelton et al., Stem Cell Rep,\n2014). Cell cultures were dissociated at indicated time points. Cells were\neither not enriched or sorted based on an endogenous PAX7-driven GFP reporter\nor cell surface combination of ERBB3+/NGFR+/HNK1- (Hicks, et al., Nat Cell\nBiol, 2018).\n&lt;/p&gt;\n\n&lt;p&gt;\nPrepared single cell solutions were subjected to single cell capture and\ndroplet formation following instructions in the online Drop-seq protocol v.3.1\n(&lt;a href="http://mccarrolllab.org/download/905/"\ntarget="_blank"&gt;http://mccarrolllab.org/download/905/&lt;/a&gt;) and those published\nin the original Drop-seq paper (Macosko et al., Cell, 2015).\n&lt;/p&gt;\n\n&lt;h3&gt;Raw read processing and computational analysis of single cell transcriptomes&lt;/h3&gt;\n&lt;p&gt;\nThe raw sequencing reads were processed using the Drop-seq_tools-1.13 pipeline\nfrom the McCaroll lab\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/releases/tag/v1.13&lt;/a&gt;),\nfollowing the general guidelines from the Drop-seq Alignment Cookbook v1.2\n(&lt;a href="https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf"\ntarget="_blank"&gt;https://github.com/broadinstitute/Drop-seq/files/2425535/Drop-seqAlignmentCookbookv1.2Jan2016.pdf&lt;/a&gt;)\n(Macosko et al., Cell, 2015). Indexed reads were aligned to the human reference\ngenome hg19 and digital gene expression matrices (DGEs) were generated by\ncounting the number of unique transcripts for each gene associated with each\ncell barcodes.\n&lt;/p&gt;\n\n&lt;p&gt;\nDownstream computational analysis of scRNA-seq data was mainly performed using\nthe R package Seurat v2.3.3 (&lt;a href="https://github.com/satijalab/seurat/releases/tag/v2.3.3"\ntarget="_blank"&gt;https://github.com/satijalab/seurat/releases/tag/v2.3.3&lt;/a&gt;)\n(Butler et al., Nat Biotechnol, 2018) by largely following the standard guidelines from the Satija\nlab (&lt;a href="https://satijalab.org/seurat/" target="_blank"&gt;https://satijalab.org/seurat/&lt;/a&gt;).\nSeurat objects were generated with DGEs\nconstructed as described above. Violin plots of number of expressed genes and\nunique transcripts (nGene and nUMI, respectively) of each cell were generated\nand outliers with too high or too low nGene/nUMI were removed to exclude\npotential cell doublets/aggregates or low quality cells/cell debris,\nrespectively, based on each object/sample. Raw count data were normalized and\nscaled with regression on effects from cell cycle (Tirosh et al., Science,\n2016) and dissociation-associated stress response (van den Brink et al., Nat\nMethods, 2017). Highly variable genes from each dataset were obtained to\ncalculate the top variable PCs, which were further used in tSNE dimensional\nreduction and cell clustering. The skeletal muscle cell cluster from individual\ndataset was further isolated in silico and subjected to re-clustering to reveal\npotential myogenic subpopulations. Skeletal muscle stem and progenitor cells\nfrom different stage in vivo human samples were computationally purified and\nassembled to construct a trajectory of human myogenesis using diffusion map in\nSeurat. Myogenic progenitors derived from hPSCs in vitro were aligned to the\ntrajectory to determine their developmental identities.\n&lt;/p&gt;\n</span></span>
<span><span class="co">##   geo</span></span>
<span><span class="co">## 1    </span></span>
<span><span class="co">## 2    </span></span>
<span><span class="co">## 3    </span></span>
<span><span class="co">## 4    </span></span>
<span><span class="co">## 5    </span></span>
<span><span class="co">## 6</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="load-the-online-datasets-to-seurat">Load the online datasets to Seurat<a class="anchor" aria-label="anchor" href="#load-the-online-datasets-to-seurat"></a>
</h3>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParseCBDatasets</code> to <strong>load the online count
matrix</strong> to Seurat. All the attributes available in
<code>ExtractCBDatasets</code> are also same here. Please note that the
loading process provided by <code>ParseCBDatasets</code> will load the
online count matrix instead of downloading it to local. If multiple
datasets available, users can choose to merge the SeuratObject with
<code>merge</code>.</p>
<p><code>ParseCBDatasets</code> supports <strong>extracting subset with
metadata and gene</strong>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parse the whole datasets</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span>
<span><span class="co"># subset metadata and gene</span></span>
<span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span></span>
<span>  sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span>, obs.value.filter <span class="op">=</span> <span class="st">"Cell.Type == 'MP' &amp; Phase == 'G2M'"</span>,</span>
<span>  include.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"PAX7"</span>, <span class="st">"MYF5"</span>, <span class="st">"C1QTNF3"</span>, <span class="st">"MYOD1"</span>, <span class="st">"MYOG"</span>, <span class="st">"RASSF4"</span>, <span class="st">"MYH3"</span>, <span class="st">"MYL4"</span>,</span>
<span>    <span class="st">"TNNT3"</span>, <span class="st">"PDGFRA"</span>, <span class="st">"OGN"</span>, <span class="st">"COL3A1"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Show the returned <code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.seu</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 14 features across 5684 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (14 features, 0 variable features)</span></span></code></pre>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
