<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GEfetch2R">
<title>GEfetch2R - Access and Format Single-cell RNA-seq Datasets from Public Resources
 • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GEfetch2R - Access and Format Single-cell RNA-seq Datasets from Public Resources
">
<meta property="og:description" content="GEfetch2R">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a>
    <a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a>
    <a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a>
    <a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="GEfetch2R_files/font-awesome-6.4.0/css/all.min.css" rel="stylesheet">
<link href="GEfetch2R_files/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>GEfetch2R - Access and Format Single-cell RNA-seq Datasets from Public Resources
</h1>
                        <h4 data-toc-skip class="author">Yabing
Song</h4>
                        
            <h4 data-toc-skip class="date">3 March 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/GEfetch2R.Rmd" class="external-link"><code>vignettes/GEfetch2R.Rmd</code></a></small>
      <div class="d-none name"><code>GEfetch2R.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> is designed to accelerate users download and
prepare single-cell datasets from public resources. It can be used
to:</p>
<ul>
<li>
<strong>Download fastq files</strong> from <code>GEO/SRA</code>,
<strong>foramt fastq files</strong> to standard style that can be
identified by 10x softwares (e.g. CellRanger).</li>
<li>
<strong>Download bam files</strong> from <code>GEO/SRA</code>,
support <strong>downloading original 10x generated bam files (with
custom tags) and normal bam files</strong>, and <strong>convert bam
files to fastq files</strong>.</li>
<li>Download scRNA-seq <strong>matrix</strong> and <strong>annotation
(e.g. cell type)</strong> information from <code>GEO</code>,
<code>PanglanDB</code> and <code>UCSC Cell Browser</code>, <strong>load
the downnloaded matrix to <code>Seurat</code></strong>.</li>
<li>Download processed objects from <code>Zeenodo</code> and
<code>CELLxGENE</code>.</li>
<li>
<strong>Formats conversion between widely used single cell
objects</strong> (<code>SeuratObject</code>, <code>AnnData</code>,
<code>SingleCellExperiment</code>,
<code>CellDataSet/cell_data_set</code> and <code>loom</code>).</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>GEfetch2R</code> is an R package distributed as part of the <a href="https://cran.r-project.org/" class="external-link">CRAN</a>. To install the package,
start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install via CRAN (v0.5.0) # old version, it's better to install via Github</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"GEfetch2R"</span><span class="op">)</span></span>
<span><span class="co"># if you install from CRAN, you should install the following packages</span></span>
<span><span class="co"># install.packages("devtools") #In case you have not installed it.</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"alexvpickering/GEOfastq"</span><span class="op">)</span> <span class="co"># download fastq</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cellgeni/sceasy"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mojaveazure/seurat-disk"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/seurat-wrappers"</span><span class="op">)</span> <span class="co"># format conversion</span></span>
<span></span>
<span><span class="co"># install via Github (v0.5.0)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"showteeth/GEfetch2R"</span><span class="op">)</span></span></code></pre></div>
<p>For <strong>data structures conversion</strong>,
<code>GEfetch2R</code> requires several python pcakages, you can install
with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install python packages</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda loompy anndata</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install anndata loompy</span></code></pre></div>
<p>In general, it is <strong>recommended</strong> to install from <a href="https://github.com/showteeth/GEfetch2R" class="external-link">Github repository</a>
(update more timely).</p>
<p>Once <code>GEfetch2R</code> is installed, it can be loaded by the
following command.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">"GEfetch2R"</a></span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="downloas-fastq-and-bam">Downloas fastq and bam<a class="anchor" aria-label="anchor" href="#downloas-fastq-and-bam"></a>
</h2>
<p>Since the downloading process is time-consuming, we provide the
commands used to illustrate the usage.</p>
<div class="section level3">
<h3 id="downloas-fastq">Downloas fastq<a class="anchor" aria-label="anchor" href="#downloas-fastq"></a>
</h3>
<div class="section level4">
<h4 id="prepare-run-number">Prepare run number<a class="anchor" aria-label="anchor" href="#prepare-run-number"></a>
</h4>
<p>For fastq files stored in SRA, <code>GEfetch2R</code> can extract
sample information and run number with GEO accession number or users can
also provide a dataframe contains the run number of interested
samples.</p>
<p>Extract all samples under <code>GSE130636</code> and the platform is
<code>GPL20301</code> (use <code>platform = NULL</code> for all
platforms):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE130636"</span>, platform <span class="op">=</span> <span class="st">"GPL20301"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-sra">Download sra<a class="anchor" aria-label="anchor" href="#download-sra"></a>
</h4>
<p>With the dataframe contains gsm and run number,
<code>GEfetch2R</code> will download sra files using
<code>prefetch</code>. The returned result is a dataframe contains
failed runs. If not <code>NULL</code>, users can re-run
<code>DownloadSRA</code> by setting <code>gsm.df</code> to the returned
result.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE130636.runs</span> <span class="op">&lt;-</span> <span class="va">GSE130636.runs</span><span class="op">[</span><span class="va">GSE130636.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR9004346"</span>, <span class="st">"SRR9004351"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download, you may need to set prefetch.path</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE130636.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadSRA.html">DownloadSRA</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE130636.runs</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE130636.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be:
<code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level4">
<h4 id="split-fastq">Split fastq<a class="anchor" aria-label="anchor" href="#split-fastq"></a>
</h4>
<p>After obtaining the sra files, <code>GEfetch2R</code> provides
function <code>SplitSRA</code> to split sra files to fastq files using
<code>parallel-fastq-dump</code> (parallel, fastest and gzip output),
<code>fasterq-dump</code> (parallel, fast but unzipped output) and
<code>fastq-dump</code> (slowest and gzip output).</p>
<p>For fastqs generated with 10x Genomics, <code>SplitSRA</code> can
identify read1, read2 and index files and format the read1 and read2 to
10x required format (<code>sample1_S1_L001_R1_001.fastq.gz</code> and
<code>sample1_S1_L001_R2_001.fastq.gz</code>). In detail, the file with
read length 26 or 28 is considered as read1, the files with read length
8 or 10 are considered as index files and the remain file is considered
as read2. The read length rules is from <a href="https://www.10xgenomics.com/cn/support/single-cell-gene-expression/documentation/steps/sequencing/sequencing-requirements-for-single-cell-3" class="external-link">Sequencing
Requirements for Single Cell 3’</a> and <a href="https://www.10xgenomics.com/cn/support/single-cell-immune-profiling/documentation/steps/sequencing/sequencing-requirements-for-single-cell-v-d-j" class="external-link">Sequencing
Requirements for Single Cell V(D)J</a>.</p>
<p>The returned result is a vector of failed sra files. If not
<code>NULL</code>, users can re-run <code>SplitSRA</code> by setting
<code>sra.path</code> to the returned result.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># parallel-fastq-dump requires sratools.path</span></span>
<span><span class="co"># you may need to set split.cmd.path and sratools.path</span></span>
<span><span class="va">sra.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE130636.split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SplitSRA.html">SplitSRA</a></span><span class="op">(</span></span>
<span>  sra.folder <span class="op">=</span> <span class="va">sra.folder</span>,</span>
<span>  fastq.type <span class="op">=</span> <span class="st">"10x"</span>, split.cmd.threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="download-bam">Download bam<a class="anchor" aria-label="anchor" href="#download-bam"></a>
</h3>
<div class="section level4">
<h4 id="prepare-run-number-1">Prepare run number<a class="anchor" aria-label="anchor" href="#prepare-run-number-1"></a>
</h4>
<p><code>GEfetch2R</code> can extract sample information and run number
with GEO accession number or users can also provide a dataframe contains
the run number of interested samples.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractRun.html">ExtractRun</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE138266"</span>, platform <span class="op">=</span> <span class="st">"GPL18573"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-bam-1">Download bam<a class="anchor" aria-label="anchor" href="#download-bam-1"></a>
</h4>
<p>With the dataframe contains gsm and run number,
<code>GEfetch2R</code> provides <code>DownloadBam</code> to download bam
files using <code>prefetch</code>. It suooorts 10x generated bam files
and normal bam files.</p>
<ul>
<li>10x generated bam: While bam files generated from 10x softwares
(e.g. CellRanger) contain custom tags which are not kept when using
default parameters of <code>prefetch</code>, <code>GEfetch2R</code> adds
<code>--type TenX</code> to make sure the downloaded bam files contain
these tags.</li>
<li>normal bam: For normal bam files, <code>DownloadBam</code> will
download sra files first and then convert sra files to bam files with
<code>sam-dump</code>. After testing the efficiency of
<code>prefetch</code> + <code>sam-dump</code> and <code>sam-dump</code>,
the former is much faster than the latter (52G sra and 72G bam
files):</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # use prefetch to download sra file</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># prefetch -X 60G SRR1976036</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    117m26.334s</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    16m42.062s</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 3m28.295s</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # use sam-dump to convert sra to bam</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036.sra | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># # real    536m2.721s</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># # user    749m41.421s</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># # sys 20m49.069s</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># use sam-dump to download bam directly</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co"># time (sam-dump SRR1976036 | samtools view -bS - -o SRR1976036.bam)</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># # more than 36hrs only get ~3G bam files, too slow</span></span></code></pre></div>
<p>The returned result is a dataframe containing failed runs (either
failed to download sra files or failed to convert to bam files for
normal bam; failed to download bam files for 10x generated bam). If not
<code>NULL</code>, users can re-run <code>DownloadBam</code> by setting
<code>gsm.df</code> to the returned result. The following is an example
to download 10x generated bam file:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a small test</span></span>
<span><span class="va">GSE138266.runs</span> <span class="op">&lt;-</span> <span class="va">GSE138266.runs</span><span class="op">[</span><span class="va">GSE138266.runs</span><span class="op">$</span><span class="va">run</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR10211566"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># download, you may need to set prefetch.path</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE138266.down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DownloadBam.html">DownloadBam</a></span><span class="op">(</span></span>
<span>  gsm.df <span class="op">=</span> <span class="va">GSE138266.runs</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># GSE138266.down is null or dataframe contains failed runs</span></span></code></pre></div>
<p>The <code>out.folder</code> structure will be:
<code>gsm_number/run_number</code>.</p>
<hr>
</div>
<div class="section level4">
<h4 id="convert-bam-to-fastq">Convert bam to fastq<a class="anchor" aria-label="anchor" href="#convert-bam-to-fastq"></a>
</h4>
<p>With downloaded bam files, <code>GEfetch2R</code> provides function
<code>Bam2Fastq</code> to convert bam files to fastq files. For bam
files generated from 10x softwares, <code>Bam2Fastq</code> utilizes
<code>bamtofastq</code> tool developed by 10x Genomics.</p>
<p>The returned result is a vector of bam files failed to convert to
fastq files. If not <code>NULL</code>, users can re-run
<code>Bam2Fastq</code> by setting <code>bam.path</code> to the returned
result.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># you may need to set bamtofastq.path and bamtofastq.paras</span></span>
<span><span class="va">GSE138266.convert</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Bam2Fastq.html">Bam2Fastq</a></span><span class="op">(</span></span>
<span>  bam.folder <span class="op">=</span> <span class="va">bam.folder</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="download-count-matrix">Download count matrix<a class="anchor" aria-label="anchor" href="#download-count-matrix"></a>
</h2>
<p><code>GEfetch2R</code> provides functions for users to download
<strong>count matrices</strong> and <strong>annotations</strong>
(e.g. cell type annotation and composition) from GEO and some
single-cell databases (e.g. <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> and <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a>).
<code>GEfetch2R</code> also supports loading the downloaded data to
<code>Seurat</code>.</p>
<p>Until now, the public resources supported and the returned
results:</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="30%">
<col width="12%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th>Resources</th>
<th>URL</th>
<th>Download Type</th>
<th>Returned results</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>GEO</td>
<td><a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/</a></td>
<td>count matrix</td>
<td>SeuratObject or count matrix for bulk RNA-seq</td>
</tr>
<tr class="even">
<td>PanglaoDB</td>
<td><a href="https://panglaodb.se/index.html" class="external-link uri">https://panglaodb.se/index.html</a></td>
<td>count matrix</td>
<td>SeuratObject</td>
</tr>
<tr class="odd">
<td>UCSC Cell Browser</td>
<td><a href="https://cells.ucsc.edu/" class="external-link uri">https://cells.ucsc.edu/</a></td>
<td>count matrix</td>
<td>SeuratObject</td>
</tr>
</tbody>
</table>
<hr>
<div class="section level3">
<h3 id="geo">GEO<a class="anchor" aria-label="anchor" href="#geo"></a>
</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html" class="external-link">GEO is
an international public repository that archives and freely distributes
microarray, next-generation sequencing, and other forms of
high-throughput functional genomics data submitted by the research
community.</a> It provides a very convenient way for users to explore
and select interested scRNA-seq datasets.</p>
<div class="section level4">
<h4 id="extract-metadata">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractGEOMeta</code> to
extract sample metadata, including sample title, source name/tissue,
description, cell type, treatment, paper title, paper abstract,
organism, protocol, data processing methods, et al.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract metadata of specified platform</span></span>
<span><span class="va">GSE200257.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="st">"GPL24676"</span><span class="op">)</span></span>
<span><span class="co"># set VROOM_CONNECTION_SIZE to avoid error: Error: The size of the connection buffer (786432) was not large enough</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">"VROOM_CONNECTION_SIZE"</span> <span class="op">=</span> <span class="fl">131072</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co"># extract metadata of all platforms</span></span>
<span><span class="va">GSE94820.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGEOMeta.html">ExtractGEOMeta</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-matrix-and-load-to-seurat">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat"></a>
</h4>
<p>After manually check the extracted metadata, users can
<strong>download count matrix</strong> and <strong>load the count
matrix</strong> to Seurat with <code>ParseGEO</code>.</p>
<p>For count matrix, <code>ParseGEO</code> supports downloading the
matrix from <strong>supplementary files</strong> and extracting from
<code>ExpressionSet</code>, users can control the source by specifying
<code>down.supp</code> or detecting automatically (<code>ParseGEO</code>
will extract the count matrix from <code>ExpressionSet</code> first, if
the count matrix is NULL or contains non-integer values,
<code>ParseGEO</code> will download supplementary files). While the
supplementary files have two main types: single count matrix file
containing all cells and CellRanger-style outputs (barcode, matrix,
feature/gene), users are required to choose the type of supplementary
files with <code>supp.type</code>.</p>
<p>With the count matrix, <code>ParseGEO</code> will load the matrix to
Seurat automatically. If multiple samples available, users can choose to
merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for cellranger output</span></span>
<span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GSE200257.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE200257"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># for count matrix, no need to specify out.folder, download count matrix to tmp folder</span></span>
<span><span class="va">GSE94820.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseGEO.html">ParseGEO</a></span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, platform <span class="op">=</span> <span class="cn">NULL</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span></code></pre></div>
<p><strong>For bulk RNA-seq</strong>, set
<code>data.type = "bulk"</code> in <code>ParseGEO</code>, this will
return count matrix.</p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="panglaodb">PanglaoDB<a class="anchor" aria-label="anchor" href="#panglaodb"></a>
</h3>
<p><a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> is a database
which contains scRNA-seq datasets from mouse and human. Up to now, it
contains <strong>5,586,348 cells</strong> from <strong>1368 datasets
(1063 from Mus musculus and 305 from Homo sapiens)</strong>. It has well
organized metadata for every dataset, including tissue, protocol,
species, number of cells and cell type annotation (computationally
identified). Daniel Osorio has developed <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a> to access
<a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a> data, the
functions of <code>GEfetch2R</code> here are based on <a href="https://github.com/dosorio/rPanglaoDB/" class="external-link">rPanglaoDB</a>.</p>
<p>Since <a href="https://panglaodb.se/about.html" class="external-link">PanglaoDB</a> is no
longer maintained, <code>GEfetch2R</code> has cached all metadata and
cell type composition and use these cached data by default to
accelerate, users can access the cached data with
<code>PanglaoDBMeta</code> (all metadata) and
<code>PanglaoDBComposition</code> (all cell type composition).</p>
<div class="section level4">
<h4 id="summarise-attributes">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://panglaodb.se/index.html" class="external-link">PanglaoDB</a>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use cached metadata</span></span>
<span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PanglaoDBMeta</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"protocol"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"PanglaoDB"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-metadata-1">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractPanglaoDBMeta</code> to
select interested datasets with specified <strong>species</strong>,
<strong>protocol</strong>, <strong>tissue</strong> and <strong>cell
number</strong> (The available values of these attributes can be
obtained with <code>StatDBAttribute</code>). User can also choose to
whether to add cell type annotation to every dataset with
<code>show.cell.type</code>.</p>
<p><code>GEfetch2R</code> uses cached metadata and cell type composition
by default, users can change this by setting
<code>local.data = FALSE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span>,</span>
<span>  show.cell.type <span class="op">=</span> <span class="cn">TRUE</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-cell-type-composition">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition"></a>
</h4>
<p><code>GEfetch2R</code> provides
<code>ExtractPanglaoDBComposition</code> to extract cell type annotation
and composition (use cached data by default to accelerate, users can
change this by setting <code>local.data = FALSE</code>).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hsa.composition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, protocol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smart-seq2"</span>, <span class="st">"10x chromium"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-matrix-and-load-to-seurat-1">Download matrix and load to Seurat<a class="anchor" aria-label="anchor" href="#download-matrix-and-load-to-seurat-1"></a>
</h4>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParsePanglaoDB</code> to <strong>download count
matrix</strong> and <strong>load the count matrix</strong> to Seurat.
With available cell type annotation, uses can filter datasets without
specified cell type with <code>cell.type</code>. Users can also
include/exclude cells expressing specified genes with
<code>include.gene</code>/<code>exclude.gene</code>.</p>
<p>With the count matrix, <code>ParsePanglaoDB</code> will load the
matrix to Seurat automatically. If multiple datasets available, users
can choose to merge the SeuratObject with <code>merge</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># small test</span></span>
<span><span class="va">hsa.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParsePanglaoDB.html">ParsePanglaoDB</a></span><span class="op">(</span><span class="va">hsa.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>, merge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="ucsc-cell-browser">UCSC Cell Browser<a class="anchor" aria-label="anchor" href="#ucsc-cell-browser"></a>
</h3>
<p>The <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell Browser</a> is a
web-based tool that allows scientists to interactively visualize
scRNA-seq datasets. It contains <strong>1040 single cell
datasets</strong> from <strong>17 different species</strong>. And, it is
<strong>organized with the hierarchical structure</strong>, which can
help users quickly locate the datasets they are interested in.</p>
<div class="section level4">
<h4 id="show-available-datasets">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ShowCBDatasets</code> to show
all available datasets. Due to the large number of datasets,
<code>ShowCBDatasets</code> enables users to perform <em>lazy load</em>
of dataset json files instead of downloading the json files online
(time-consuming!!!). This <em>lazy load</em> requires users to provide
<code>json.folder</code> to save json files and set
<code>lazy = TRUE</code> (for the first time of run,
<code>ShowCBDatasets</code> will download current json files to
<code>json.folder</code>, for next time of run, with
<code>lazy = TRUE</code>, <code>ShowCBDatasets</code> will load the
downloaded json files from <code>json.folder</code>.). And,
<code>ShowCBDatasets</code> supports updating the local datasets with
<code>update = TRUE</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">json.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># first time run, the json files are stored under json.folder</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = TRUE, json.folder = json.folder, update = TRUE)</span></span>
<span></span>
<span><span class="co"># second time run, load the downloaded json files</span></span>
<span><span class="va">ucsc.cb.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCBDatasets.html">ShowCBDatasets</a></span><span class="op">(</span>lazy <span class="op">=</span> <span class="cn">TRUE</span>, json.folder <span class="op">=</span> <span class="va">json.folder</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># always read online</span></span>
<span><span class="co"># ucsc.cb.samples = ShowCBDatasets(lazy = FALSE)</span></span></code></pre></div>
<p>The number of datasets and all available species:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the number of datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ucsc.cb.samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># available species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\|parent"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">ucsc.cb.samples</span><span class="op">$</span><span class="va">organisms</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="summarise-attributes-1">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://cells.ucsc.edu/?#" class="external-link">UCSC Cell
Browser</a>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"organ"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"UCSC"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-metadata-2">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-2"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCBDatasets</code> to
filter metadata with <strong>collection</strong>,
<strong>sub-collection</strong>, <strong>organ</strong>, <strong>disease
status</strong>, <strong>organism</strong>, <strong>project</strong> and
<strong>cell number</strong> (The available values of these attributes
can be obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>). All attributes except cell number support fuzzy match
with <code>fuzzy.match</code>, this is useful when selecting
datasets.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBDatasets.html">ExtractCBDatasets</a></span><span class="op">(</span>all.samples.df <span class="op">=</span> <span class="va">ucsc.cb.samples</span>, organ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"blood"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Human (H. sapiens)"</span>, cell.num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-cell-type-composition-1">Extract cell type composition<a class="anchor" aria-label="anchor" href="#extract-cell-type-composition-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCBComposition</code> to
extract cell type annotation and composition.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCBComposition.html">ExtractCBComposition</a></span><span class="op">(</span>json.folder <span class="op">=</span> <span class="va">json.folder</span>, sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="load-the-online-datasets-to-seurat">Load the online datasets to Seurat<a class="anchor" aria-label="anchor" href="#load-the-online-datasets-to-seurat"></a>
</h4>
<p>After manually check the extracted metadata, <code>GEfetch2R</code>
provides <code>ParseCBDatasets</code> to <strong>load the online count
matrix</strong> to Seurat. All the attributes available in
<code>ExtractCBDatasets</code> are also same here. Please note that the
loading process provided by <code>ParseCBDatasets</code> will load the
online count matrix instead of downloading it to local. If multiple
datasets available, users can choose to merge the SeuratObject with
<code>merge</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbb.sample.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseCBDatasets.html">ParseCBDatasets</a></span><span class="op">(</span>sample.df <span class="op">=</span> <span class="va">hbb.sample.df</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="download-object">Download object<a class="anchor" aria-label="anchor" href="#download-object"></a>
</h2>
<p><code>GEfetch2R</code> provides functions for users to download
processed single-cell RNA-seq data from <a href="https://zenodo.org/" class="external-link">Zenodo</a>, <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> and <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>, including
<code>RDS</code>, <code>RData</code>, <code>h5ad</code>,
<code>h5</code>, <code>loom</code> objects.</p>
<p>Until now, the public resources supported and the returned
results:</p>
<table class="table">
<colgroup>
<col width="15%">
<col width="29%">
<col width="33%">
<col width="21%">
</colgroup>
<thead><tr class="header">
<th>Resources</th>
<th>URL</th>
<th>Download Type</th>
<th>Returned results</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Zenodo</td>
<td><a href="https://zenodo.org/" class="external-link uri">https://zenodo.org/</a></td>
<td>count matrix, rds, rdata, h5ad, et al.</td>
<td>NULL or failed datasets</td>
</tr>
<tr class="even">
<td>CELLxGENE</td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link uri">https://cellxgene.cziscience.com/</a></td>
<td>rds, h5ad</td>
<td>NULL or failed datasets</td>
</tr>
<tr class="odd">
<td>Human Cell Atlas</td>
<td><a href="https://www.humancellatlas.org/" class="external-link uri">https://www.humancellatlas.org/</a></td>
<td>rds, rdata, h5, h5ad, loom</td>
<td>NULL or failed projects</td>
</tr>
</tbody>
</table>
<hr>
<div class="section level3">
<h3 id="zenodo">Zenodo<a class="anchor" aria-label="anchor" href="#zenodo"></a>
</h3>
<p><a href="https://zenodo.org/" class="external-link">Zenodo</a> contains various types of
processed objects, such as <code>SeuratObject</code> which has been
clustered and annotated, <code>AnnData</code> which contains processed
results generated by <code>scanpy</code>.</p>
<div class="section level4">
<h4 id="extract-metadata-3">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-3"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractZenodoMeta</code> to
extract dataset metadata, including dataset title, description,
available files and corresponding md5. Please note that when the dataset
is restricted access, the returned dataframe will be empty.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single doi</span></span>
<span><span class="va">zebrafish.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="st">"10.5281/zenodo.7243603"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># vector dois</span></span>
<span><span class="va">multi.dois</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractZenodoMeta.html">ExtractZenodoMeta</a></span><span class="op">(</span>doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-object-1">Download object<a class="anchor" aria-label="anchor" href="#download-object-1"></a>
</h4>
<p>After manually check the extracted metadata, users can
<strong>download the specified objects</strong> with
<code>ParseZenodo</code>. The downloaded objects are controlled by
<code>file.ext</code> and <strong>the provided object formats should be
in lower case (e.g. rds/rdata/h5ad).</strong></p>
<p>The returned result is a dataframe containing failed objects. If not
<code>NULL</code>, users can re-run <code>ParseZenodo</code> by setting
<code>doi.df</code> to the returned result.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">multi.dois.parse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ParseZenodo.html">ParseZenodo</a></span><span class="op">(</span></span>
<span>  doi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1111"</span>, <span class="st">"10.5281/zenodo.7243603"</span>, <span class="st">"10.5281/zenodo.7244441"</span><span class="op">)</span>,</span>
<span>  file.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rdata"</span>, <span class="st">"rds"</span><span class="op">)</span>, out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h3>
<p>The <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a> is a
web server contains <strong>910</strong> single-cell datasets, users can
explore, download and upload own datasets. The downloaded datasets
provided by <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a>
have two formats: <code>h5ad (AnnData v0.8)</code> and
<code>rds (Seurat v4)</code>.</p>
<div class="section level4">
<h4 id="show-available-datasets-1">Show available datasets<a class="anchor" aria-label="anchor" href="#show-available-datasets-1"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ShowCELLxGENEDatasets</code> to
extract dataset metadata, including dataset title, description, contact,
organism, ethnicity, sex, tissue, disease, assay, suspension type, cell
type, et al.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all available datasets</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="summarise-attributes-2">Summarise attributes<a class="anchor" aria-label="anchor" href="#summarise-attributes-2"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>StatDBAttribute</code> to
summary attributes of <a href="https://cellxgene.cziscience.com/" class="external-link">CELLxGENE</a>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StatDBAttribute.html">StatDBAttribute</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"organism"</span>, <span class="st">"sex"</span><span class="op">)</span>, database <span class="op">=</span> <span class="st">"CELLxGENE"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="extract-metadata-4">Extract metadata<a class="anchor" aria-label="anchor" href="#extract-metadata-4"></a>
</h4>
<p><code>GEfetch2R</code> provides <code>ExtractCELLxGENEMeta</code> to
filter dataset metadata, the available values of attributes can be
obtained with <code>StatDBAttribute</code> except <strong>cell
number</strong>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">human.10x.cellxgene.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCELLxGENEMeta.html">ExtractCELLxGENEMeta</a></span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="download-object-2">Download object<a class="anchor" aria-label="anchor" href="#download-object-2"></a>
</h4>
<p>After manually check the extracted metadata, users can
<strong>download the specified objects</strong> with
<code>ParseCELLxGENE</code>. The downloaded objects are controlled by
<code>file.ext</code> (choose from <code>"rds"</code> and
<code>"h5ad"</code>).</p>
<p>The returned result is a dataframe containing failed datasets. If not
<code>NULL</code>, users can re-run <code>ParseCELLxGENE</code> by
setting <code>meta</code> to the returned result.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ParseCELLxGENE.html">ParseCELLxGENE</a></span><span class="op">(</span></span>
<span>  meta <span class="op">=</span> <span class="va">human.10x.cellxgene.meta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>, file.ext <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="va">out.folder</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="format-conversion">Format conversion<a class="anchor" aria-label="anchor" href="#format-conversion"></a>
</h2>
<p>There are many tools have been developed to process scRNA-seq data,
such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">Scanpy</a>,
<a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
and <a href="http://cole-trapnell-lab.github.io/monocle-release/" class="external-link">Monocle</a>.
These tools have their own objects, such as <code>Anndata</code> of
<code>Scanpy</code>, <code>SeuratObject</code> of <code>Seurat</code>,
<code>SingleCellExperiment</code> of <code>scran</code> and
<code>CellDataSet</code>/<code>cell_data_set</code> of
<code>Monocle2</code>/<code>Monocle3</code>. There are also some file
format designed for large omics datasets, such as <a href="http://loompy.org/" class="external-link">loom</a>. To perform a comprehensive scRNA-seq
data analysis, we usually need to combine multiple tools, which means we
need to perform object conversion frequently. To facilitate user
analysis of scRNA-seq data, <code>GEfetch2R</code> provides multiple
functions to perform object conversion between widely used tools and
formats. The object conversion implemented in <code>GEfetch2R</code> has
two main advantages:</p>
<ul>
<li>
<strong>one-step conversion between different objects</strong>.
There will be no conversion to intermediate objects, thus preventing
unnecessary information loss.</li>
<li>
<strong>tools used for object conversion are developed by the team
of the source/destination object as far as possible</strong>. For
example, we use <code>SeuratDisk</code> to convert SeuratObject to loom,
use <code>zellkonverter</code> to perform conversion between
<code>SingleCellExperiment</code> and <code>Anndata</code>. When there
is no such tools, we use <code>sceasy</code> to perform conversion.</li>
</ul>
<hr>
<div class="section level3">
<h3 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h3>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># pbmc_small</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span> <span class="co"># seger</span></span></code></pre></div>
<p><code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># object</span></span>
<span><span class="va">pbmc_small</span></span></code></pre></div>
<p><code>SingleCellExperiment</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html" class="external-link">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="convert-seuratobject-to-other-objects">Convert SeuratObject to other objects<a class="anchor" aria-label="anchor" href="#convert-seuratobject-to-other-objects"></a>
</h3>
<p>Here, we will convert SeuratObject to
<code>SingleCellExperiment</code>,
<code>CellDataSet</code>/<code>cell_data_set</code>,
<code>Anndata</code>, <code>loom</code>.</p>
<div class="section level4">
<h4 id="seuratobject-to-singlecellexperiment">SeuratObject to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#seuratobject-to-singlecellexperiment"></a>
</h4>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"SCE"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="seuratobject-to-celldatasetcell_data_set">SeuratObject to CellDataSet/cell_data_set<a class="anchor" aria-label="anchor" href="#seuratobject-to-celldatasetcell_data_set"></a>
</h4>
<p>To <code>CellDataSet</code> (The conversion is performed with
functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("monocle") # reuqire monocle</span></span>
<span><span class="va">cds.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, to <span class="op">=</span> <span class="st">"CellDataSet"</span><span class="op">)</span></span></code></pre></div>
<p>To <code>cell_data_set</code> (The conversion is performed with
functions implemented in <code>SeuratWrappers</code>):</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github('cole-trapnell-lab/monocle3') # reuqire monocle3</span></span>
<span><span class="va">cds3.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"cell_data_set"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="seuratobject-to-anndata">SeuratObject to AnnData<a class="anchor" aria-label="anchor" href="#seuratobject-to-anndata"></a>
</h4>
<p><code>AnnData</code> is a Python object, <code>reticulate</code> is
used to communicate between Python and R. User should create a Python
environment which contains <code>anndata</code> package and specify the
environment path with <code>conda.path</code> to ensure the exact usage
of this environment.</p>
<p>The conversion is performed with functions implemented in
<code>sceasy</code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove pbmc_small.h5ad first</span></span>
<span><span class="va">anndata.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"pbmc_small_"</span>, fileext <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span></span>
<span><span class="co"># you may need to set conda.path</span></span>
<span><span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"AnnData"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="va">anndata.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="seuratobject-to-loom">SeuratObject to loom<a class="anchor" aria-label="anchor" href="#seuratobject-to-loom"></a>
</h4>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code>:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"pbmc_small_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"loom"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="convert-other-objects-to-seuratobject">Convert other objects to SeuratObject<a class="anchor" aria-label="anchor" href="#convert-other-objects-to-seuratobject"></a>
</h3>
<div class="section level4">
<h4 id="singlecellexperiment-to-seuratobject">SingleCellExperiment to SeuratObject<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-seuratobject"></a>
</h4>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sce.obj</span>, from <span class="op">=</span> <span class="st">"SCE"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="celldatasetcell_data_set-to-seuratobject">CellDataSet/cell_data_set to SeuratObject<a class="anchor" aria-label="anchor" href="#celldatasetcell_data_set-to-seuratobject"></a>
</h4>
<p><code>CellDataSet</code> to <code>SeuratObject</code> (The conversion
is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds.obj</span>, from <span class="op">=</span> <span class="st">"CellDataSet"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<p><code>cell_data_set</code> to <code>SeuratObject</code> (The
conversion is performed with functions implemented in
<code>Seurat</code>):</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds3.obj</span>, from <span class="op">=</span> <span class="st">"cell_data_set"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="anndata-to-seuratobject">AnnData to SeuratObject<a class="anchor" aria-label="anchor" href="#anndata-to-seuratobject"></a>
</h4>
<p><code>AnnData</code> is a Python object, <code>reticulate</code> is
used to communicate between Python and R. User should create a Python
environment which contains <code>anndata</code> package and specify the
environment path with <code>conda.path</code> to ensure the exact usage
of this environment.</p>
<p>The conversion is performed with functions implemented in
<code>sceasy</code>:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># you may need to set conda.path</span></span>
<span><span class="va">seu.obj.h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="va">anndata.file</span>, from <span class="op">=</span> <span class="st">"AnnData"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="loom-to-seuratobject">loom to SeuratObject<a class="anchor" aria-label="anchor" href="#loom-to-seuratobject"></a>
</h4>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code> and <code>Seurat</code>:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loom will lose reduction</span></span>
<span><span class="va">seu.obj.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>loom.file <span class="op">=</span> <span class="va">loom.file</span>, from <span class="op">=</span> <span class="st">"loom"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="conversion-between-singlecellexperiment-and-anndata">Conversion between SingleCellExperiment and AnnData<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-anndata"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>zellkonverter</code>.</p>
<div class="section level4">
<h4 id="singlecellexperiment-to-anndata">SingleCellExperiment to AnnData<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-anndata"></a>
</h4>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.h5ad first</span></span>
<span><span class="va">seger.anndata.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"seger_"</span>, fileext <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/SCEAnnData.html">SCEAnnData</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"AnnData"</span>, sce <span class="op">=</span> <span class="va">seger</span>, X_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="va">seger.anndata.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="anndata-to-singlecellexperiment">AnnData to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#anndata-to-singlecellexperiment"></a>
</h4>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.anndata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SCEAnnData.html">SCEAnnData</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"AnnData"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="va">seger.anndata.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="conversion-between-singlecellexperiment-and-loom">Conversion between SingleCellExperiment and loom<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-loom"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>LoomExperiment</code>.</p>
<div class="section level4">
<h4 id="singlecellexperiment-to-loom">SingleCellExperiment to loom<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-loom"></a>
</h4>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.loom first</span></span>
<span><span class="va">seger.loom.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"seger_"</span>, fileext <span class="op">=</span> <span class="st">".loom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"loom"</span>, sce <span class="op">=</span> <span class="va">seger</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="loom-to-singlecellexperiment">loom to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#loom-to-singlecellexperiment"></a>
</h4>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"loom"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="va">seger.loom.file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
