<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GEfetch2R - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools
 • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GEfetch2R - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="GEfetch2R_files/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="GEfetch2R_files/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>GEfetch2R - Access Single-cell/Bulk RNA-seq Data from Public Repositories and Benchmark the Subsequent Format Conversion Tools
</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/GEfetch2R.Rmd" class="external-link"><code>vignettes/GEfetch2R.Rmd</code></a></small>
      <div class="d-none name"><code>GEfetch2R.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start">
<a href="https://showteeth.github.io/GEfetch2R/articles/QuickStart.html"><mark>Quick
start</mark></a><a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GEfetch2R</code> is designed to accelerate users’ downloading
and preparation of single-cell/bulk RNA-seq datasets from public
resources. It can be used to:</p>
<p>For <strong>bulk RNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from
<code>SRA/ENA</code> with <strong>GEO accession</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files
from <code>SRA/ENA</code>
</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to
<code>fastq</code> files</li>
<li><strong>Convert <code>bam</code> to <code>fastq</code>
files</strong></li>
<li><strong>Mapping with <code>STAR</code> and load the output to
<code>DESeq2</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with
<strong>GEO accession</strong>:
<ul>
<li><strong>Generate count matrix from supplementary files</strong></li>
<li><strong>Load the count matrix to <code>DESeq2</code></strong></li>
</ul>
</li>
</ul>
<p>For <strong>scRNA-seq</strong>:</p>
<ul>
<li>Download <strong>raw data (<code>sra/fastq/bam</code>)</strong> from
<code>SRA/ENA</code> with <strong>GEO accession</strong>:
<ul>
<li>
<strong>Parallel</strong> download <code>sra/fastq/bam</code> files
from <code>SRA/ENA</code>
</li>
<li>
<strong>Foramt downloaded <code>fastq</code> files</strong> to
standard style that can be identified by 10x softwares
(e.g. <code>CellRanger</code>).</li>
<li>
<strong>Support Aspera</strong> (<code>ENA</code>)</li>
<li>
<strong>Parallel</strong> split <code>sra</code> to
<code>fastq</code> files</li>
<li><strong>Download original 10x generated <code>bam</code> files (with
custom tags)</strong></li>
<li><strong>Convert <code>bam</code> to <code>fastq</code> files
(<code>samtools/bamtofastq_linux</code>)</strong></li>
<li><strong>Mapping with <code>STAR/CellRanger</code> and load the
output to <code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>count matrix</strong>:
<ul>
<li>Download <strong>count matrix</strong> from <code>GEO</code> with
<strong>GEO accession</strong>
</li>
<li>Download <strong>count matrix</strong> and <strong>annotation
(e.g. cell type)</strong> information from <code>PanglaoDB</code> and
<code>UCSC Cell Browser</code> with <strong>key words (filter
criteria)</strong>
</li>
<li>
<strong>Extract subset with annotation and gene
(<code>PanglaoDB</code> and
<code>UCSC Cell Browser</code>)</strong><br>
</li>
<li><strong>Load the count matrix and annotation to
<code>Seurat</code></strong></li>
</ul>
</li>
<li>Download <strong>processed object</strong>:
<ul>
<li>Download <strong>processed object</strong> from <code>GEO</code>
with <strong>GEO accession</strong>
</li>
<li>Download <strong>processed object</strong> from <code>Zenodo</code>
with <strong>DOI</strong>
</li>
<li>Download <strong>processed object</strong> from
<code>CELLxGENE</code> and <code>Human Cell Atlas</code> with
<strong>key words (filter criteria)</strong>
</li>
<li>
<strong>Parallel</strong> download</li>
<li><strong>Extract subset with annotation and gene
(<code>CELLxGENE</code>)</strong></li>
<li><strong>Load the processed object
(<code>rds</code>/<code>rdata</code>) to
<code>Seurat</code></strong></li>
<li>
<strong>Extract count matrix and metadata from processed object
(<code>rds</code>/<code>rdata</code>)</strong>, including
<code>Seurat</code>, <code>seurat</code>,
<code>SingleCellExperiment</code>, <code>cell_data_set</code>,
<code>CellDataSet</code>, <code>DESeqDataSet</code>,
<code>DGEList</code>.</li>
</ul>
</li>
<li>
<strong>Format conversion and benchmark related tools</strong>:
<ul>
<li>
<strong>Format conversion between widely used single cell
objects</strong> (<code>SeuratObject</code>, <code>AnnData</code>,
<code>SingleCellExperiment</code>,
<code>CellDataSet/cell_data_set</code> and <code>loom</code>)</li>
<li>
<strong>Benchmark the tools (more than two tools available for the
conversion)</strong>:
<ul>
<li>
<code>SeuratObject</code> to <code>AnnData</code>:
<code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>
</li>
<li>
<code>SingleCellExperiment</code> to <code>AnnData</code>:
<code>sceasy</code>, <code>scDIOR</code>,
<code>zellkonverter</code>
</li>
<li>
<code>AnnData</code> to <code>SeuratObject</code>:
<code>SeuratDisk</code>, <code>sceasy</code>, <code>scDIOR</code>,
<code>schard</code>
</li>
<li>
<code>AnnData</code> to <code>SingleCellExperiment</code>:
<code>schard</code>, <code>scDIOR</code>,
<code>zellkonverter</code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="framework">Framework<a class="anchor" aria-label="anchor" href="#framework"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_overview.png" title="GEfetch2R_framework" alt="GEfetch2R_framework"></p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="vignette">Vignette<a class="anchor" aria-label="anchor" href="#vignette"></a>
</h2>
<p>Detailed usage is available in <a href="https://showteeth.github.io/GEfetch2R/">website</a>:</p>
<ul>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadRaw.html" class="external-link">Download
raw data (SRA/ENA)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadMatrices.html" class="external-link">Download
count matrix and annotaion (GEO, PanglaoDB, UCSC Cell Browser)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/DownloadObjects.html" class="external-link">Download
processed data (GEO, Zenodo, CELLxGENE, Human Cell Atlas)</a></li>
<li><a href="https://showteeth.github.io/scfetch/articles/ObjectConversion.html" class="external-link">Format
conversion and benchmark (AnnData, SeuratObject, SingleCellExperiment,
CellDataSet/cell_data_set, loom)</a></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="comparison-with-similar-tools">Comparison with similar tools<a class="anchor" aria-label="anchor" href="#comparison-with-similar-tools"></a>
</h2>
<div align="center">
<p><img src="reference/figures/GEfetch2R_compare.png" title="GEfetch2R_comparison" alt="GEfetch2R_comparison"></p>
</div>
<hr>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
