<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scfetch">
<title>ObjectConversion â€¢ scfetch</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ObjectConversion">
<meta property="og:description" content="scfetch">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scfetch</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/DownloadRaw.html">Download fastq/bam files</a>
    <a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrices</a>
    <a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed objects</a>
    <a class="dropdown-item" href="../articles/ObjectConversion.html">Convert formats between objects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/showteeth/scfetch">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>ObjectConversion</h1>
            
            <h4 data-toc-skip class="date">2023-07-30</h4>
      
      
      <div class="d-none name"><code>ObjectConversion.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>There are many tools have been developed to process scRNA-seq data,
such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">Scanpy</a>,
<a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
and <a href="http://cole-trapnell-lab.github.io/monocle-release/" class="external-link">Monocle</a>.
These tools have their own objects, such as <code>Anndata</code> of
<code>Scanpy</code>, <code>SeuratObject</code> of <code>Seurat</code>,
<code>SingleCellExperiment</code> of <code>scran</code> and
<code>CellDataSet</code>/<code>cell_data_set</code> of
<code>Monocle2</code>/<code>Monocle3</code>. There are also some file
format designed for large omics datasets, such as <a href="http://loompy.org/" class="external-link">loom</a>. To perform a comprehensive scRNA-seq
data analysis, we usually need to combine multiple tools, which means we
need to perform object conversion frequently. To facilitate user
analysis of scRNA-seq data, <code>scfetch</code> provides multiple
functions to perform object conversion between widely used tools and
formats. The object conversion implemented in <code>scfetch</code> has
two main advantages:</p>
<ul>
<li>
<strong>one-step conversion between different objects</strong>.
There will be no conversion to intermediate objects, thus preventing
unnecessary information loss.</li>
<li>
<strong>tools used for object conversion are developed by the team
of the source/destination object as far as possible</strong>. For
example, we use <code>SeuratDisk</code> to convert SeuratObject to loom,
use <code>zellkonverter</code> to perform conversion between
<code>SingleCellExperiment</code> and <code>Anndata</code>. When there
is no such tools, we use <code>sceasy</code> to perform conversion.</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scfetch</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Setting options('download.file.method.GEOquery'='auto')</span></span></code></pre>
<pre><code><span><span class="co">## Setting options('GEOquery.inmemory.gpl'=FALSE)</span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'SeuratDisk':</span></span>
<span><span class="co">##   method            from  </span></span>
<span><span class="co">##   as.sparse.H5Group Seurat</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># pbmc_small</span></span></code></pre></div>
<pre><code><span><span class="co">## Attaching SeuratObject</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span> <span class="co"># seger</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SingleCellExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Warning: package 'BiocGenerics' was built under R version 4.0.5</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:parallel':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></span>
<span><span class="co">##     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></span>
<span><span class="co">##     parLapplyLB, parRapply, parSapply, parSapplyLB</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></span>
<span><span class="co">##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></span>
<span><span class="co">##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></span>
<span><span class="co">##     union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Warning: package 'GenomeInfoDb' was built under R version 4.0.5</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SummarizedExperiment'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:SeuratObject':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Assays</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Seurat':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Assays</span></span></code></pre>
<pre><code><span><span class="co">## Possible Ensembl SSL connectivity problems detected.</span></span>
<span><span class="co">## Please see the 'Connection Troubleshooting' section of the biomaRt vignette</span></span>
<span><span class="co">## vignette('accessing_ensembl', package = 'biomaRt')Error in curl::curl_fetch_memory(url, handle = handle) : </span></span>
<span><span class="co">##   SSL peer certificate or SSH remote key was not OK: [uswest.ensembl.org] SSL certificate problem: certificate has expired</span></span></code></pre>
<p><code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># object</span></span>
<span><span class="va">pbmc_small</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 20 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, tsne</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span><span class="co">## ATGCCAGAACGACT SeuratProject         70           47               0</span></span>
<span><span class="co">## CATGGCCTGTGCAT SeuratProject         85           52               0</span></span>
<span><span class="co">## GAACCTGATGAACC SeuratProject         87           50               1</span></span>
<span><span class="co">## TGACTGGATTCTCA SeuratProject        127           56               0</span></span>
<span><span class="co">## AGTCAGACTGCACA SeuratProject        173           53               0</span></span>
<span><span class="co">## TCTGATACACGTGT SeuratProject         70           48               0</span></span>
<span><span class="co">##                letter.idents groups RNA_snn_res.1</span></span>
<span><span class="co">## ATGCCAGAACGACT             A     g2             0</span></span>
<span><span class="co">## CATGGCCTGTGCAT             A     g1             0</span></span>
<span><span class="co">## GAACCTGATGAACC             B     g2             0</span></span>
<span><span class="co">## TGACTGGATTCTCA             A     g2             0</span></span>
<span><span class="co">## AGTCAGACTGCACA             A     g2             0</span></span>
<span><span class="co">## TCTGATACACGTGT             A     g1             0</span></span></code></pre>
<p><code>SingleCellExperiment</code>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html" class="external-link">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## snapshotDate(): 2020-10-27</span></span></code></pre>
<pre><code><span><span class="co">## see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span></code></pre>
<pre><code><span><span class="co">## loading from cache</span></span></code></pre>
<pre><code><span><span class="co">## see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span></code></pre>
<pre><code><span><span class="co">## loading from cache</span></span></code></pre>
<pre><code><span><span class="co">## see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span></code></pre>
<pre><code><span><span class="co">## loading from cache</span></span></code></pre>
<pre><code><span><span class="co">## snapshotDate(): 2020-10-27</span></span></code></pre>
<pre><code><span><span class="co">## see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span></code></pre>
<pre><code><span><span class="co">## loading from cache</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 26179 3514 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(26179): SGIP1 AZIN2 ... BIVM-ERCC5 eGFP</span></span>
<span><span class="co">## rowData names(2): symbol refseq</span></span>
<span><span class="co">## colnames(3514): HP1502401_N13 HP1502401_D14 ... HP1526901T2D_O11</span></span>
<span><span class="co">##   HP1526901T2D_A8</span></span>
<span><span class="co">## colData names(8): Source Name individual ... age body mass index</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## altExpNames(1): ERCC</span></span></code></pre>
<hr>
</div>
<div class="section level2">
<h2 id="convert-seuratobject-to-other-objects">Convert SeuratObject to other objects<a class="anchor" aria-label="anchor" href="#convert-seuratobject-to-other-objects"></a>
</h2>
<p>Here, we will convert SeuratObject to
<code>SingleCellExperiment</code>,
<code>CellDataSet</code>/<code>cell_data_set</code>,
<code>Anndata</code>, <code>loom</code>.</p>
<div class="section level3">
<h3 id="seuratobject-to-singlecellexperiment">SeuratObject to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#seuratobject-to-singlecellexperiment"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"SCE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert SeuratObject to SingleCellExperiment (suitable for scater)!</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 230 80 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(230): MS4A1 CD79B ... SPON2 S100B</span></span>
<span><span class="co">## rowData names(5): vst.mean vst.variance vst.variance.expected</span></span>
<span><span class="co">##   vst.variance.standardized vst.variable</span></span>
<span><span class="co">## colnames(80): ATGCCAGAACGACT CATGGCCTGTGCAT ... GGAACACTTCAGAC</span></span>
<span><span class="co">##   CTTGATTGATCTTC</span></span>
<span><span class="co">## colData names(8): orig.ident nCount_RNA ... RNA_snn_res.1 ident</span></span>
<span><span class="co">## reducedDimNames(2): PCA TSNE</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-celldatasetcell_data_set">SeuratObject to CellDataSet/cell_data_set<a class="anchor" aria-label="anchor" href="#seuratobject-to-celldatasetcell_data_set"></a>
</h3>
<p>To <code>CellDataSet</code> (The conversion is performed with
functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("monocle") # reuqire monocle</span></span>
<span><span class="va">cds.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, to <span class="op">=</span> <span class="st">"CellDataSet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert SeuratObject to CellDataSet (suitable for Monocle)!</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cds.obj</span></span></code></pre></div>
<pre><code><span><span class="co">## CellDataSet (storageMode: environment)</span></span>
<span><span class="co">## assayData: 230 features, 80 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: ATGCCAGAACGACT CATGGCCTGTGCAT ... CTTGATTGATCTTC (80</span></span>
<span><span class="co">##     total)</span></span>
<span><span class="co">##   varLabels: orig.ident nCount_RNA ... Size_Factor (8 total)</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: MS4A1 CD79B ... S100B (230 total)</span></span>
<span><span class="co">##   fvarLabels: vst.mean vst.variance ... gene_short_name (6 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:</span></span></code></pre>
<p>To <code>cell_data_set</code> (The conversion is performed with
functions implemented in <code>SeuratWrappers</code>):</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github('cole-trapnell-lab/monocle3') # reuqire monocle3</span></span>
<span><span class="va">cds3.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"cell_data_set"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert SeuratObject to cell_data_set (suitable for Monocle3)!</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Monocle 3 trajectories require cluster partitions, which Seurat does</span></span>
<span><span class="co">## not calculate. Please run 'cluster_cells' on your cell_data_set object</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cds3.obj</span></span></code></pre></div>
<pre><code><span><span class="co">## class: cell_data_set </span></span>
<span><span class="co">## dim: 230 80 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(230): MS4A1 CD79B ... SPON2 S100B</span></span>
<span><span class="co">## rowData names(5): vst.mean vst.variance vst.variance.expected</span></span>
<span><span class="co">##   vst.variance.standardized vst.variable</span></span>
<span><span class="co">## colnames(80): ATGCCAGAACGACT CATGGCCTGTGCAT ... GGAACACTTCAGAC</span></span>
<span><span class="co">##   CTTGATTGATCTTC</span></span>
<span><span class="co">## colData names(9): orig.ident nCount_RNA ... ident Size_Factor</span></span>
<span><span class="co">## reducedDimNames(2): PCA TSNE</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-anndata">SeuratObject to AnnData<a class="anchor" aria-label="anchor" href="#seuratobject-to-anndata"></a>
</h3>
<p><code>AnnData</code> is a Python object, <code>reticulate</code> is
used to communicate between Python and R. User should create a Python
environment which contains <code>anndata</code> package and specify the
environment path with <code>conda.path</code> to ensure the exact usage
of this environment.</p>
<p>The conversion is performed with functions implemented in
<code>sceasy</code>:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove pbmc_small.h5ad first</span></span>
<span><span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"AnnData"</span>, conda.path <span class="op">=</span> <span class="st">"/Applications/anaconda3"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/pbmc_small.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-loom">SeuratObject to loom<a class="anchor" aria-label="anchor" href="#seuratobject-to-loom"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code>:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"loom"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/pbmc_small.loom"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="convert-other-objects-to-seuratobject">Convert other objects to SeuratObject<a class="anchor" aria-label="anchor" href="#convert-other-objects-to-seuratobject"></a>
</h2>
<div class="section level3">
<h3 id="singlecellexperiment-to-seuratobject">SingleCellExperiment to SeuratObject<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-seuratobject"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sce.obj</span>, from <span class="op">=</span> <span class="st">"SCE"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert SingleCellExperiment to SeuratObject!</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">## underscore, setting key from PC__ to PC_</span></span></code></pre>
<pre><code><span><span class="co">## Warning: All keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">## underscore '_', setting key to PC_</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">## underscore, setting key from tSNE__ to tSNE_</span></span></code></pre>
<pre><code><span><span class="co">## Warning: All keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">## underscore '_', setting key to tSNE_</span></span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, tsne</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="celldatasetcell_data_set-to-seuratobject">CellDataSet/cell_data_set to SeuratObject<a class="anchor" aria-label="anchor" href="#celldatasetcell_data_set-to-seuratobject"></a>
</h3>
<p><code>CellDataSet</code> to <code>SeuratObject</code> (The conversion
is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds.obj</span>, from <span class="op">=</span> <span class="st">"CellDataSet"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert CellDataSet (Monocle) to SeuratObject!</span></span></code></pre>
<pre><code><span><span class="co">## Pulling expression data</span></span></code></pre>
<pre><code><span><span class="co">## Building Seurat object</span></span></code></pre>
<pre><code><span><span class="co">## Adding feature-level metadata</span></span></code></pre>
<pre><code><span><span class="co">## No dispersion information in CellDataSet object</span></span></code></pre>
<pre><code><span><span class="co">## No variable features present</span></span></code></pre>
<pre><code><span><span class="co">## Adding tSNE dimensional reduction</span></span></code></pre>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">##  1 dimensional reduction calculated: tsne</span></span></code></pre>
<p><code>cell_data_set</code> to <code>SeuratObject</code> (The
conversion is performed with functions implemented in
<code>Seurat</code>):</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cds3.obj</span>, from <span class="op">=</span> <span class="st">"cell_data_set"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert cell_data_set (Monocle3) to SeuratObject!</span></span></code></pre>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds3</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, tsne</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="anndata-to-seuratobject">AnnData to SeuratObject<a class="anchor" aria-label="anchor" href="#anndata-to-seuratobject"></a>
</h3>
<p><code>AnnData</code> is a Python object, <code>reticulate</code> is
used to communicate between Python and R. User should create a Python
environment which contains <code>anndata</code> package and specify the
environment path with <code>conda.path</code> to ensure the exact usage
of this environment.</p>
<p>The conversion is performed with functions implemented in
<code>sceasy</code>:</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/pbmc_small.h5ad"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"AnnData"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, conda.path <span class="op">=</span> <span class="st">"/Applications/anaconda3"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert AnnData to SeuratObject!</span></span></code></pre>
<pre><code><span><span class="co">## X -&gt; counts</span></span></code></pre>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.h5ad</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, tsne</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="loom-to-seuratobject">loom to SeuratObject<a class="anchor" aria-label="anchor" href="#loom-to-seuratobject"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code> and <code>Seurat</code>:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loom will lose reduction</span></span>
<span><span class="va">seu.obj.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>loom.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/pbmc_small.loom"</span>, from <span class="op">=</span> <span class="st">"loom"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert loom to SeuratObject!</span></span></code></pre>
<pre><code><span><span class="co">## Reading in /matrix</span></span></code></pre>
<pre><code><span><span class="co">## Storing /matrix as counts</span></span></code></pre>
<pre><code><span><span class="co">## Saving /matrix to assay 'RNA'</span></span></code></pre>
<pre><code><span><span class="co">## Loading graph RNA_snn</span></span></code></pre>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.loom</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (230 features, 0 variable features)</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="conversion-between-singlecellexperiment-and-anndata">Conversion between SingleCellExperiment and AnnData<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-anndata"></a>
</h2>
<p>The conversion is performed with functions implemented in
<code>zellkonverter</code>.</p>
<div class="section level3">
<h3 id="singlecellexperiment-to-anndata">SingleCellExperiment to AnnData<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-anndata"></a>
</h3>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.h5ad first</span></span>
<span><span class="fu"><a href="../reference/SCEAnnData.html">SCEAnnData</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"AnnData"</span>, sce <span class="op">=</span> <span class="va">seger</span>, X_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/seger.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="anndata-to-singlecellexperiment">AnnData to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#anndata-to-singlecellexperiment"></a>
</h3>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.anndata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SCEAnnData.html">SCEAnnData</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"AnnData"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  anndata.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/seger.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert AnnData to SingleCellExperiment.</span></span></code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.anndata</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 26179 3514 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(26179): SGIP1 AZIN2 ... BIVM-ERCC5 eGFP</span></span>
<span><span class="co">## rowData names(2): symbol refseq</span></span>
<span><span class="co">## colnames(3514): HP1502401_N13 HP1502401_D14 ... HP1526901T2D_O11</span></span>
<span><span class="co">##   HP1526901T2D_A8</span></span>
<span><span class="co">## colData names(8): Source Name individual ... age body mass index</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="conversion-between-singlecellexperiment-and-loom">Conversion between SingleCellExperiment and loom<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-loom"></a>
</h2>
<p>The conversion is performed with functions implemented in
<code>LoomExperiment</code>.</p>
<div class="section level3">
<h3 id="singlecellexperiment-to-loom">SingleCellExperiment to loom<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-loom"></a>
</h3>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.loom first</span></span>
<span><span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"loom"</span>, sce <span class="op">=</span> <span class="va">seger</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/seger.loom"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="loom-to-singlecellexperiment">loom to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#loom-to-singlecellexperiment"></a>
</h3>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"loom"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/seger.loom"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Convert loom to SingleCellExperiment.</span></span></code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 26179 3514 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(26179): SGIP1 AZIN2 ... BIVM-ERCC5.1 eGFP</span></span>
<span><span class="co">## rowData names(2): refseq symbol</span></span>
<span><span class="co">## colnames(3514): HP1502401_N13 HP1502401_D14 ... HP1526901T2D_O11</span></span>
<span><span class="co">##   HP1526901T2D_A8</span></span>
<span><span class="co">## colData names(8): Source.Name age ... sex single.cell.well.quality</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
