<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ObjectConversion • GEfetch2R</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ObjectConversion">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ObjectConversion</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/vignettes/ObjectConversion.Rmd" class="external-link"><code>vignettes/ObjectConversion.Rmd</code></a></small>
      <div class="d-none name"><code>ObjectConversion.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>There are many tools have been developed to process scRNA-seq data,
such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">Scanpy</a>,
<a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
and <a href="http://cole-trapnell-lab.github.io/monocle-release/" class="external-link">Monocle</a>.
These tools have their own objects, such as <code>AnnData</code> of
<code>Scanpy</code>, <code>SeuratObject</code> of <code>Seurat</code>,
<code>SingleCellExperiment</code> of <code>scran</code> and
<code>CellDataSet</code>/<code>cell_data_set</code> of
<code>Monocle2</code>/<code>Monocle3</code>. There are also some file
format designed for large omics datasets, such as <a href="http://loompy.org/" class="external-link">loom</a>. To perform a comprehensive scRNA-seq
data analysis, we usually need to combine multiple tools, which means we
need to perform object conversion frequently.</p>
<p>To facilitate user analysis of scRNA-seq data,
<code>GEfetch2R</code>:</p>
<ul>
<li>
<strong>benchmarked the format conversion tools</strong> and
provides some guides for tool selection under different
scenarios(<code>AnnData</code> -&gt; <code>SeuratObject</code>,
<code>SeuratObject</code> to <code>AnnData</code>, <code>AnnData</code>
to <code>SingleCellExperiment</code>, <code>SingleCellExperiment</code>
to <code>AnnData</code>)</li>
<li>provides multiple functions to <strong>perform format conversion
between widely used scRNA-seq objects</strong>
(<code>SeuratObject</code>, <code>AnnData</code>,
<code>SingleCellExperiment</code>,
<code>CellDataSet/cell_data_set</code> and <code>loom</code>)</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/showteeth/GEfetch2R" class="external-link">GEfetch2R</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># pbmc_small</span></span>
<span><span class="co"># library(scRNAseq) # seger</span></span></code></pre></div>
<p><code>SeuratObject</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># object</span></span>
<span><span class="va">pbmc_small</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 230 features across 80 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (230 features, 20 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, tsne</span></span>
<span></span>
<span><span class="co"># metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#                   orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span><span class="co"># ATGCCAGAACGACT SeuratProject         70           47               0             A     g2</span></span>
<span><span class="co"># CATGGCCTGTGCAT SeuratProject         85           52               0             A     g1</span></span>
<span><span class="co"># GAACCTGATGAACC SeuratProject         87           50               1             B     g2</span></span>
<span><span class="co"># TGACTGGATTCTCA SeuratProject        127           56               0             A     g2</span></span>
<span><span class="co"># AGTCAGACTGCACA SeuratProject        173           53               0             A     g2</span></span>
<span><span class="co"># TCTGATACACGTGT SeuratProject         70           48               0             A     g1</span></span>
<span><span class="co">#                RNA_snn_res.1</span></span>
<span><span class="co"># ATGCCAGAACGACT             0</span></span>
<span><span class="co"># CATGGCCTGTGCAT             0</span></span>
<span><span class="co"># GAACCTGATGAACC             0</span></span>
<span><span class="co"># TGACTGGATTCTCA             0</span></span>
<span><span class="co"># AGTCAGACTGCACA             0</span></span>
<span><span class="co"># TCTGATACACGTGT             0</span></span></code></pre></div>
<p><code>SingleCellExperiment</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># seger &lt;- scRNAseq::SegerstolpePancreasData()</span></span>
<span><span class="co"># load from local</span></span>
<span><span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/gefetch2r/doc/conversion/seger.rds"</span><span class="op">)</span></span>
<span><span class="va">seger</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 26179 3514</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(26179): SGIP1 AZIN2 ... BIVM-ERCC5 eGFP</span></span>
<span><span class="co"># rowData names(2): refseq symbol</span></span>
<span><span class="co"># colnames(3514): HP1502401_N13 HP1502401_D14 ... HP1526901T2D_O11 HP1526901T2D_A8</span></span>
<span><span class="co"># colData names(9): individual sex ... submitted single cell quality cell type</span></span>
<span><span class="co"># reducedDimNames(0):</span></span>
<span><span class="co"># altExpNames(1): ERCC</span></span></code></pre></div>
<p><code>AnnData</code> (<a href="https://github.com/showteeth/GEfetch2R/blob/main/man/benchmark/generate_pbmc3k_anndata.ipynb" class="external-link">generate_pbmc3k_anndata.ipynb</a>):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>pbmc3k <span class="op">=</span> sc.read(<span class="st">'~/gefetch2r/doc/conversion/pbmc3k.h5ad'</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>pbmc3k</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># AnnData object with n_obs × n_vars = 2638 × 1838</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#     obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'leiden'</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#     var: 'gene_ids', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#     uns: 'hvg', 'leiden', 'log1p', 'neighbors', 'pca', 'rank_genes_groups', 'umap'</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#     obsm: 'X_pca', 'X_umap'</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#     varm: 'PCs'</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#     layers: 'logcounts', 'rawcounts'</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#     obsp: 'connectivities', 'distances'</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="convert-seuratobject-to-other-objects">Convert SeuratObject to other objects<a class="anchor" aria-label="anchor" href="#convert-seuratobject-to-other-objects"></a>
</h2>
<p>Here, we will convert SeuratObject to
<code>SingleCellExperiment</code>,
<code>CellDataSet</code>/<code>cell_data_set</code>,
<code>AnnData</code>, <code>loom</code>.</p>
<div class="section level3">
<h3 id="seuratobject-to-singlecellexperiment">SeuratObject to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#seuratobject-to-singlecellexperiment"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"SCE"</span><span class="op">)</span></span>
<span><span class="va">sce.obj</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 230 80</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(2): counts logcounts</span></span>
<span><span class="co"># rownames(230): MS4A1 CD79B ... SPON2 S100B</span></span>
<span><span class="co"># rowData names(5): vst.mean vst.variance vst.variance.expected vst.variance.standardized</span></span>
<span><span class="co">#   vst.variable</span></span>
<span><span class="co"># colnames(80): ATGCCAGAACGACT CATGGCCTGTGCAT ... GGAACACTTCAGAC CTTGATTGATCTTC</span></span>
<span><span class="co"># colData names(8): orig.ident nCount_RNA ... RNA_snn_res.1 ident</span></span>
<span><span class="co"># reducedDimNames(2): PCA TSNE</span></span>
<span><span class="co"># altExpNames(0):</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-celldatasetcell_data_set">SeuratObject to CellDataSet/cell_data_set<a class="anchor" aria-label="anchor" href="#seuratobject-to-celldatasetcell_data_set"></a>
</h3>
<p>To <code>CellDataSet</code> (The conversion is performed with
functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BiocManager::install("monocle") # reuqire monocle</span></span>
<span><span class="va">cds.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, to <span class="op">=</span> <span class="st">"CellDataSet"</span><span class="op">)</span></span>
<span><span class="va">cds.obj</span></span>
<span><span class="co"># CellDataSet (storageMode: environment)</span></span>
<span><span class="co"># assayData: 230 features, 80 samples</span></span>
<span><span class="co">#   element names: exprs</span></span>
<span><span class="co"># protocolData: none</span></span>
<span><span class="co"># phenoData</span></span>
<span><span class="co">#   sampleNames: ATGCCAGAACGACT CATGGCCTGTGCAT ... CTTGATTGATCTTC (80 total)</span></span>
<span><span class="co">#   varLabels: orig.ident nCount_RNA ... Size_Factor (8 total)</span></span>
<span><span class="co">#   varMetadata: labelDescription</span></span>
<span><span class="co"># featureData</span></span>
<span><span class="co">#   featureNames: MS4A1 CD79B ... S100B (230 total)</span></span>
<span><span class="co">#   fvarLabels: vst.mean vst.variance ... gene_short_name (6 total)</span></span>
<span><span class="co">#   fvarMetadata: labelDescription</span></span>
<span><span class="co"># experimentData: use 'experimentData(object)'</span></span>
<span><span class="co"># Annotation:</span></span></code></pre></div>
<p>To <code>cell_data_set</code> (The conversion is performed with
functions implemented in <code>SeuratWrappers</code>):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github('cole-trapnell-lab/monocle3') # reuqire monocle3</span></span>
<span><span class="va">cds3.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span>seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"cell_data_set"</span><span class="op">)</span></span>
<span><span class="va">cds3.obj</span></span>
<span><span class="co"># class: cell_data_set</span></span>
<span><span class="co"># dim: 230 80</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(2): counts logcounts</span></span>
<span><span class="co"># rownames(230): MS4A1 CD79B ... SPON2 S100B</span></span>
<span><span class="co"># rowData names(0):</span></span>
<span><span class="co"># colnames(80): ATGCCAGAACGACT CATGGCCTGTGCAT ... GGAACACTTCAGAC CTTGATTGATCTTC</span></span>
<span><span class="co"># colData names(9): orig.ident nCount_RNA ... ident Size_Factor</span></span>
<span><span class="co"># reducedDimNames(2): PCA TSNE</span></span>
<span><span class="co"># mainExpName: RNA</span></span>
<span><span class="co"># altExpNames(0):</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-anndata">SeuratObject to AnnData<a class="anchor" aria-label="anchor" href="#seuratobject-to-anndata"></a>
</h3>
<p>There are multiple tools available for format conversion from
<code>SeuratObject</code> to <code>AnnData</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept
and usability</li>
<li>
<code>sceasy</code> has best performance in running time and disk
usage.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sceasy, generate pbmc_small_sceasy.h5ad</span></span>
<span><span class="fu"><a href="../reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, conda.path <span class="op">=</span> <span class="st">"~/anaconda3"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># AnnData object with n_obs × n_vars = 80 × 230</span></span>
<span><span class="co">#     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'letter.idents', 'groups', 'RNA_snn_res.1'</span></span>
<span><span class="co">#     var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'vst.variable'</span></span>
<span><span class="co">#     obsm: 'X_pca', 'X_tsne'</span></span>
<span></span>
<span><span class="co"># SeuratDisk, generate pbmc_small_SeuratDisk.h5Seurat, pbmc_small_SeuratDisk.h5ad</span></span>
<span><span class="fu"><a href="../reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Creating h5Seurat file for version 3.1.5.9900</span></span>
<span><span class="co"># Adding counts for RNA</span></span>
<span><span class="co"># Adding data for RNA</span></span>
<span><span class="co"># Adding scale.data for RNA</span></span>
<span><span class="co"># Adding variable features for RNA</span></span>
<span><span class="co"># Adding feature-level metadata for RNA</span></span>
<span><span class="co"># Adding cell embeddings for pca</span></span>
<span><span class="co"># Adding loadings for pca</span></span>
<span><span class="co"># Adding projected loadings for pca</span></span>
<span><span class="co"># Adding standard deviations for pca</span></span>
<span><span class="co"># Adding JackStraw information for pca</span></span>
<span><span class="co"># Adding cell embeddings for tsne</span></span>
<span><span class="co"># No loadings for tsne</span></span>
<span><span class="co"># No projected loadings for tsne</span></span>
<span><span class="co"># No standard deviations for tsne</span></span>
<span><span class="co"># No JackStraw data for tsne</span></span>
<span><span class="co"># Validating h5Seurat file</span></span>
<span><span class="co"># Adding scale.data from RNA as X</span></span>
<span><span class="co"># Transfering meta.features to var</span></span>
<span><span class="co"># Adding data from RNA as raw</span></span>
<span><span class="co"># Transfering meta.features to raw/var</span></span>
<span><span class="co"># Transfering meta.data to obs</span></span>
<span><span class="co"># Adding dimensional reduction information for pca</span></span>
<span><span class="co"># Adding feature loadings for pca</span></span>
<span><span class="co"># Adding dimensional reduction information for tsne</span></span>
<span><span class="co"># Adding RNA_snn as neighbors</span></span>
<span><span class="co"># [1] "~/gefetch2r/doc/conversion/pbmc_small_SeuratDisk.h5ad"</span></span>
<span></span>
<span><span class="co"># # scDIOR, generate pbmc_small_scDIOR.h5</span></span>
<span><span class="fu"><a href="../reference/Seu2AD.html">Seu2AD</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, save.scale <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># Warning message:</span></span>
<span><span class="co"># In matrix_to_h5(mat = slot(object = slot_assay, name = "scale.data"),  :</span></span>
<span><span class="co">#   sdata is dense matrix</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="seuratobject-to-loom">SeuratObject to loom<a class="anchor" aria-label="anchor" href="#seuratobject-to-loom"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ExportSeurat.html">ExportSeurat</a></span><span class="op">(</span></span>
<span>  seu.obj <span class="op">=</span> <span class="va">pbmc_small</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, to <span class="op">=</span> <span class="st">"loom"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc_small.loom"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert SeuratObject to loom!</span></span>
<span><span class="co"># Saving data from RNA as /matrix</span></span>
<span><span class="co">#   |========================================================================================| 100%</span></span>
<span><span class="co"># Adding slot counts for assay RNA</span></span>
<span><span class="co"># Adding layer counts</span></span>
<span><span class="co">#   |========================================================================================| 100%</span></span>
<span><span class="co"># Adding col attribute CellID</span></span>
<span><span class="co"># Adding col attribute orig.ident</span></span>
<span><span class="co"># Adding col attribute nCount_RNA</span></span>
<span><span class="co"># Adding col attribute nFeature_RNA</span></span>
<span><span class="co"># Adding col attribute RNA_snn_res.0.8</span></span>
<span><span class="co"># Adding col attribute letter.idents</span></span>
<span><span class="co"># Adding col attribute groups</span></span>
<span><span class="co"># Adding col attribute RNA_snn_res.1</span></span>
<span><span class="co"># Adding row attribute Gene</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="convert-other-objects-to-seuratobject">Convert other objects to SeuratObject<a class="anchor" aria-label="anchor" href="#convert-other-objects-to-seuratobject"></a>
</h2>
<div class="section level3">
<h3 id="singlecellexperiment-to-seuratobject">SingleCellExperiment to SeuratObject<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-seuratobject"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>Seurat</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span></span>
<span>  obj <span class="op">=</span> <span class="va">sce.obj</span>, from <span class="op">=</span> <span class="st">"SCE"</span>, count.assay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert SingleCellExperiment to SeuratObject!</span></span>
<span><span class="co"># The assay you provided: RNA is not in the cell_data_set object: . Change assay to NULL # (Seurat 4.4.0)</span></span>
<span><span class="co"># Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from PC__ to PC_</span></span>
<span><span class="co"># Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to PC_</span></span>
<span><span class="co"># Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from tSNE__ to tSNE_</span></span>
<span><span class="co"># Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to tSNE_</span></span>
<span><span class="va">seu.obj.sce</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 230 features across 80 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, tsne</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="celldatasetcell_data_set-to-seuratobject">CellDataSet/cell_data_set to SeuratObject<a class="anchor" aria-label="anchor" href="#celldatasetcell_data_set-to-seuratobject"></a>
</h3>
<p><code>CellDataSet</code> to <code>SeuratObject</code> (The conversion
is performed with functions implemented in <code>Seurat</code>):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span></span>
<span>  obj <span class="op">=</span> <span class="va">cds.obj</span>, from <span class="op">=</span> <span class="st">"CellDataSet"</span>,</span>
<span>  count.assay <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert CellDataSet (Monocle) to SeuratObject!</span></span>
<span><span class="co"># Pulling expression data</span></span>
<span><span class="co"># Building Seurat object</span></span>
<span><span class="co"># Adding feature-level metadata</span></span>
<span><span class="co"># No dispersion information in CellDataSet object</span></span>
<span><span class="co"># No variable features present</span></span>
<span><span class="co"># Adding tSNE dimensional reduction</span></span>
<span><span class="va">seu.obj.cds</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 230 features across 80 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">#  1 dimensional reduction calculated: tsne</span></span></code></pre></div>
<p><code>cell_data_set</code> to <code>SeuratObject</code> (The
conversion is performed with functions implemented in
<code>Seurat</code>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.obj.cds3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span></span>
<span>  obj <span class="op">=</span> <span class="va">cds3.obj</span>, from <span class="op">=</span> <span class="st">"cell_data_set"</span>,</span>
<span>  count.assay <span class="op">=</span> <span class="st">"counts"</span>, data.assay <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert cell_data_set (Monocle3) to SeuratObject!</span></span>
<span><span class="co"># The assay you provided: RNA is not in the cell_data_set object: . Change assay to NULL # (Seurat 4.4.0)</span></span>
<span><span class="va">seu.obj.cds3</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 230 features across 80 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (230 features, 0 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, tsne</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="anndata-to-seuratobject">AnnData to SeuratObject<a class="anchor" aria-label="anchor" href="#anndata-to-seuratobject"></a>
</h3>
<p>There are multiple tools available for format conversion from
<code>AnnData</code> to <code>SeuratObject</code>:</p>
<ul>
<li>
<code>scDIOR</code> is the best method in terms of information kept
(<strong><code>GEfetch2R</code> integrates <code>scDIOR</code> and
<code>SeuratDisk</code> to achieve the best performance in information
kept</strong>)</li>
<li>
<code>schard</code> is the best method in terms of usability</li>
<li>
<code>schard</code> and <code>sceasy</code> have comparable
performance when cell number below 200k, but <code>sceasy</code> has
better performance in scalability</li>
<li>
<code>sceasy</code> has better performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sceasy</span></span>
<span><span class="va">ann.sceasy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co"># X -&gt; scale.data; raw.X -&gt; data</span></span>
<span><span class="va">ann.sceasy</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 13714 features across 2638 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, umap</span></span>
<span></span>
<span><span class="co"># SeuratDisk</span></span>
<span><span class="va">ann.seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"SeuratDisk"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Warning: Unknown file type: h5ad</span></span>
<span><span class="co"># Creating h5Seurat file for version 3.1.5.9900</span></span>
<span><span class="co"># Adding X as scale.data</span></span>
<span><span class="co"># Adding raw/X as data</span></span>
<span><span class="co"># Adding raw/X as counts</span></span>
<span><span class="co"># Adding meta.features from raw/var</span></span>
<span><span class="co"># Adding dispersions from scaled feature-level metadata</span></span>
<span><span class="co"># Adding dispersions_norm from scaled feature-level metadata</span></span>
<span><span class="co"># Merging gene_ids from scaled feature-level metadata</span></span>
<span><span class="co"># Adding highly_variable from scaled feature-level metadata</span></span>
<span><span class="co"># Adding mean from scaled feature-level metadata</span></span>
<span><span class="co"># Merging mean_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding means from scaled feature-level metadata</span></span>
<span><span class="co"># Merging mt from scaled feature-level metadata</span></span>
<span><span class="co"># Merging n_cells from scaled feature-level metadata</span></span>
<span><span class="co"># Merging n_cells_by_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Merging pct_dropout_by_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding std from scaled feature-level metadata</span></span>
<span><span class="co"># Merging total_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding X_pca as cell embeddings for pca</span></span>
<span><span class="co"># Adding X_umap as cell embeddings for umap</span></span>
<span><span class="co"># Adding PCs as feature loadings fpr pca</span></span>
<span><span class="co"># Adding miscellaneous information for pca</span></span>
<span><span class="co"># Adding standard deviations for pca</span></span>
<span><span class="co"># Adding miscellaneous information for umap</span></span>
<span><span class="co"># Adding hvg to miscellaneous data</span></span>
<span><span class="co"># Adding leiden to miscellaneous data</span></span>
<span><span class="co"># Adding log1p to miscellaneous data</span></span>
<span><span class="co"># Adding rank_genes_groups to miscellaneous data</span></span>
<span><span class="co"># Adding layer logcounts as data in assay logcounts</span></span>
<span><span class="co"># Adding layer rawcounts as data in assay rawcounts</span></span>
<span><span class="co"># Validating h5Seurat file</span></span>
<span><span class="co"># Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co"># Initializing RNA with data</span></span>
<span><span class="co"># Adding counts for RNA</span></span>
<span><span class="co"># Adding scale.data for RNA</span></span>
<span><span class="co"># Adding feature-level metadata for RNA</span></span>
<span><span class="co"># Adding reduction pca</span></span>
<span><span class="co"># Adding cell embeddings for pca</span></span>
<span><span class="co"># Adding feature loadings for pca</span></span>
<span><span class="co"># Adding miscellaneous information for pca</span></span>
<span><span class="co"># Adding reduction umap</span></span>
<span><span class="co"># Adding cell embeddings for umap</span></span>
<span><span class="co"># Adding miscellaneous information for umap</span></span>
<span><span class="co"># Adding command information</span></span>
<span><span class="co"># Adding cell-level metadata</span></span>
<span><span class="va">ann.seu</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 13714 features across 2638 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, umap</span></span>
<span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">ann.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"scDIOR"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Warning: No columnames present in cell embeddings, setting to 'PCA_1:50'</span></span>
<span><span class="co"># Warning: No columnames present in cell embeddings, setting to 'UMAP_1:2'</span></span>
<span><span class="co"># Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="va">ann.scdior</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 17390 features across 2638 samples within 3 assays</span></span>
<span><span class="co"># Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#  2 other assays present: logcounts, rawcounts</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, umap</span></span>
<span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">ann.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"schard"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, use.raw <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna to rna_</span></span>
<span><span class="co"># Warning: Invalid name supplied, making object name syntactically valid. New object name is X_indexn_genesn_genes_by_countstotal_countstotal_counts_mtpct_counts_mtleiden; see ?make.names for more details on syntax validity</span></span>
<span><span class="va">ann.schard</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 13714 features across 2638 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#  2 dimensional reductions calculated: Xpca_, Xumap_</span></span>
<span></span>
<span><span class="co"># SeuratDisk+scDIOR</span></span>
<span><span class="va">ann.seuscdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2Seu.html">AD2Seu</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"SeuratDisk+scDIOR"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, load.assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Warning: Unknown file type: h5ad</span></span>
<span><span class="co"># Creating h5Seurat file for version 3.1.5.9900</span></span>
<span><span class="co"># Adding X as scale.data</span></span>
<span><span class="co"># Adding raw/X as data</span></span>
<span><span class="co"># Adding raw/X as counts</span></span>
<span><span class="co"># Adding meta.features from raw/var</span></span>
<span><span class="co"># Adding dispersions from scaled feature-level metadata</span></span>
<span><span class="co"># Adding dispersions_norm from scaled feature-level metadata</span></span>
<span><span class="co"># Merging gene_ids from scaled feature-level metadata</span></span>
<span><span class="co"># Adding highly_variable from scaled feature-level metadata</span></span>
<span><span class="co"># Adding mean from scaled feature-level metadata</span></span>
<span><span class="co"># Merging mean_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding means from scaled feature-level metadata</span></span>
<span><span class="co"># Merging mt from scaled feature-level metadata</span></span>
<span><span class="co"># Merging n_cells from scaled feature-level metadata</span></span>
<span><span class="co"># Merging n_cells_by_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Merging pct_dropout_by_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding std from scaled feature-level metadata</span></span>
<span><span class="co"># Merging total_counts from scaled feature-level metadata</span></span>
<span><span class="co"># Adding X_pca as cell embeddings for pca</span></span>
<span><span class="co"># Adding X_umap as cell embeddings for umap</span></span>
<span><span class="co"># Adding PCs as feature loadings fpr pca</span></span>
<span><span class="co"># Adding miscellaneous information for pca</span></span>
<span><span class="co"># Adding standard deviations for pca</span></span>
<span><span class="co"># Adding miscellaneous information for umap</span></span>
<span><span class="co"># Adding hvg to miscellaneous data</span></span>
<span><span class="co"># Adding leiden to miscellaneous data</span></span>
<span><span class="co"># Adding log1p to miscellaneous data</span></span>
<span><span class="co"># Adding rank_genes_groups to miscellaneous data</span></span>
<span><span class="co"># Adding layer logcounts as data in assay logcounts</span></span>
<span><span class="co"># Adding layer rawcounts as data in assay rawcounts</span></span>
<span><span class="co"># Validating h5Seurat file</span></span>
<span><span class="co"># Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co"># Initializing RNA with data</span></span>
<span><span class="co"># Adding counts for RNA</span></span>
<span><span class="co"># Adding scale.data for RNA</span></span>
<span><span class="co"># Adding feature-level metadata for RNA</span></span>
<span><span class="co"># Adding reduction pca</span></span>
<span><span class="co"># Adding cell embeddings for pca</span></span>
<span><span class="co"># Adding feature loadings for pca</span></span>
<span><span class="co"># Adding miscellaneous information for pca</span></span>
<span><span class="co"># Adding reduction umap</span></span>
<span><span class="co"># Adding cell embeddings for umap</span></span>
<span><span class="co"># Adding miscellaneous information for umap</span></span>
<span><span class="co"># Adding command information</span></span>
<span><span class="co"># Adding cell-level metadata</span></span>
<span><span class="co"># Warning: No columnames present in cell embeddings, setting to 'PCA_1:50'</span></span>
<span><span class="co"># Warning: No columnames present in cell embeddings, setting to 'UMAP_1:2'</span></span>
<span><span class="co"># Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="va">ann.seuscdior</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 17390 features across 2638 samples within 3 assays</span></span>
<span><span class="co"># Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#  2 other assays present: logcounts, rawcounts</span></span>
<span><span class="co">#  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="loom-to-seuratobject">loom to SeuratObject<a class="anchor" aria-label="anchor" href="#loom-to-seuratobject"></a>
</h3>
<p>The conversion is performed with functions implemented in
<code>SeuratDisk</code> and <code>Seurat</code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loom will lose reduction</span></span>
<span><span class="va">seu.obj.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportSeurat.html">ImportSeurat</a></span><span class="op">(</span>loom.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc_small.loom"</span>, from <span class="op">=</span> <span class="st">"loom"</span><span class="op">)</span></span>
<span><span class="co"># Convert loom to SeuratObject!</span></span>
<span><span class="co"># Reading in /matrix</span></span>
<span><span class="co"># Storing /matrix as counts</span></span>
<span><span class="co"># Saving /matrix to assay 'RNA'</span></span>
<span><span class="co"># Loading graph RNA_snn</span></span>
<span><span class="va">seu.obj.loom</span></span>
<span><span class="co"># An object of class Seurat</span></span>
<span><span class="co"># 230 features across 80 samples within 1 assay</span></span>
<span><span class="co"># Active assay: RNA (230 features, 0 variable features)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="conversion-between-singlecellexperiment-and-anndata">Conversion between SingleCellExperiment and AnnData<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-anndata"></a>
</h2>
<div class="section level3">
<h3 id="singlecellexperiment-to-anndata">SingleCellExperiment to AnnData<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-anndata"></a>
</h3>
<p>There are multiple tools available for format conversion from
<code>SingleCellExperiment</code> to <code>AnnData</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of
information kept and running time</li>
<li>
<code>scDIOR</code> is the best method in terms of usability and
disk usage (<strong><code>scDIOR</code> require <code>diopy</code> to
read <code>h5</code> file</strong>)</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># zellkonverter, output seger_zellkonverter.h5ad</span></span>
<span><span class="fu"><a href="../reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span></span>
<span>  sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"zellkonverter"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  conda.path <span class="op">=</span> <span class="st">"~/anaconda3"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># # read seger_zellkonverter.h5ad in python</span></span>
<span><span class="co"># import scanpy as sc</span></span>
<span><span class="co"># seger_zellkonverter = sc.read('~/gefetch2r/doc/conversion/seger_zellkonverter.h5ad')</span></span>
<span><span class="co"># seger_zellkonverter</span></span>
<span><span class="co"># AnnData object with n_obs × n_vars = 3514 × 26179</span></span>
<span><span class="co">#     obs: 'individual', 'sex', 'age', 'body mass index', 'clinical information', 'disease', 'single cell well quality', 'submitted single cell quality', 'cell type'</span></span>
<span><span class="co">#     var: 'refseq', 'symbol'</span></span>
<span><span class="co">#     uns: 'X_name'</span></span>
<span></span>
<span><span class="co"># sceasy, output seger_sceasy.h5ad</span></span>
<span><span class="fu"><a href="../reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span></span>
<span>  sce.obj <span class="op">=</span> <span class="va">seger</span>, method <span class="op">=</span> <span class="st">"sceasy"</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span>, conda.path <span class="op">=</span> <span class="st">"~/anaconda3"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># AnnData object with n_obs × n_vars = 3514 × 26179</span></span>
<span><span class="co">#     obs: 'individual', 'sex', 'age', 'body.mass.index', 'clinical.information', 'disease', 'single.cell.well.quality', 'submitted.single.cell.quality', 'cell.type'</span></span>
<span><span class="co">#     var: 'refseq', 'symbol'</span></span>
<span></span>
<span><span class="co"># scDIOR, output seger.scdior_scDIOR.h5</span></span>
<span><span class="va">seger.scdior</span> <span class="op">&lt;-</span> <span class="va">seger</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co"># scDIOR does not support varm in rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">seger.scdior</span><span class="op">)</span><span class="op">$</span><span class="va">varm</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="../reference/SCE2AD.html">SCE2AD</a></span><span class="op">(</span>sce.obj <span class="op">=</span> <span class="va">seger.scdior</span>, method <span class="op">=</span> <span class="st">"scDIOR"</span>, out.folder <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion"</span><span class="op">)</span></span>
<span><span class="co"># [1] "The first 'assayNames' defaults to 'X'"</span></span>
<span><span class="co"># [1] "RNA"</span></span>
<span><span class="co"># # read seger.scdior_scDIOR.h5 in python</span></span>
<span><span class="co"># import diopy</span></span>
<span><span class="co"># seger_scdior = diopy.input.read_h5(file = "~/gefetch2r/doc/conversion/seger.scdior_scDIOR.h5")</span></span>
<span><span class="co"># seger_scdior</span></span>
<span><span class="co"># AnnData object with n_obs × n_vars = 3514 × 26179</span></span>
<span><span class="co">#     obs: 'individual', 'sex', 'age', 'body mass index', 'clinical information', 'disease', 'single cell well quality', 'submitted single cell quality', 'cell type'</span></span>
<span><span class="co">#     var: 'refseq', 'symbol'</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="anndata-to-singlecellexperiment">AnnData to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#anndata-to-singlecellexperiment"></a>
</h3>
<p>There are multiple tools available for format conversion from
<code>AnnData</code> to <code>SingleCellExperiment</code>:</p>
<ul>
<li>
<code>zellkonverter</code> is the best method in terms of
information kept</li>
<li>
<code>schard</code> is the best method in terms of usability and
running time</li>
<li>
<code>schard</code> and <code>scDIOR</code> have comparable
performance in disk usage</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># zellkonverter</span></span>
<span><span class="va">sce.zell</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"zellkonverter"</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span>,</span>
<span>  use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"~/anaconda3"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce.zell</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 1838 2638</span></span>
<span><span class="co"># metadata(7): hvg leiden ... rank_genes_groups umap</span></span>
<span><span class="co"># assays(3): scale.data logcounts rawcounts</span></span>
<span><span class="co"># rownames(1838): TNFRSF4 CPSF3L ... S100B PRMT2</span></span>
<span><span class="co"># rowData names(14): gene_ids n_cells ... std varm</span></span>
<span><span class="co"># colnames(2638): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... TTTGCATGAGAGGC-1</span></span>
<span><span class="co">#   TTTGCATGCCTCAC-1</span></span>
<span><span class="co"># colData names(6): n_genes n_genes_by_counts ... pct_counts_mt leiden</span></span>
<span><span class="co"># reducedDimNames(2): X_pca X_umap</span></span>
<span><span class="co"># altExpNames(1): raw</span></span>
<span></span>
<span><span class="co"># scDIOR</span></span>
<span><span class="va">sce.scdior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"scDIOR"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  use.raw <span class="op">=</span> <span class="cn">TRUE</span>, conda.path <span class="op">=</span> <span class="st">"~/anaconda3"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce.scdior</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 13714 2638</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(1): X</span></span>
<span><span class="co"># rownames(13714): AL627309.1 AP006222.2 ... PNRC2-1 SRSF10-1</span></span>
<span><span class="co"># rowData names(7): gene_ids n_cells ... pct_dropout_by_counts</span></span>
<span><span class="co">#   total_counts</span></span>
<span><span class="co"># colnames(2638): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... TTTGCATGAGAGGC-1</span></span>
<span><span class="co">#   TTTGCATGCCTCAC-1</span></span>
<span><span class="co"># colData names(6): n_genes n_genes_by_counts ... pct_counts_mt leiden</span></span>
<span><span class="co"># reducedDimNames(2): pca umap</span></span>
<span><span class="co"># altExpNames(0):</span></span>
<span></span>
<span><span class="co"># schard</span></span>
<span><span class="va">sce.schard</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AD2SCE.html">AD2SCE</a></span><span class="op">(</span></span>
<span>  anndata.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/pbmc3k.h5ad"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"schard"</span>, use.raw <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce.schard</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 13714 2638</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(1): X</span></span>
<span><span class="co"># rownames(13714): AL627309.1 AP006222.2 ... PNRC2-1 SRSF10-1</span></span>
<span><span class="co"># rowData names(8): _index gene_ids ... pct_dropout_by_counts</span></span>
<span><span class="co">#   total_counts</span></span>
<span><span class="co"># colnames(2638): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... TTTGCATGAGAGGC-1</span></span>
<span><span class="co">#   TTTGCATGCCTCAC-1</span></span>
<span><span class="co"># colData names(7): _index n_genes ... pct_counts_mt leiden</span></span>
<span><span class="co"># reducedDimNames(2): X_pca X_umap</span></span>
<span><span class="co"># altExpNames(0):</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="conversion-between-singlecellexperiment-and-loom">Conversion between SingleCellExperiment and loom<a class="anchor" aria-label="anchor" href="#conversion-between-singlecellexperiment-and-loom"></a>
</h2>
<p>The conversion is performed with functions implemented in
<code>LoomExperiment</code>.</p>
<div class="section level3">
<h3 id="singlecellexperiment-to-loom">SingleCellExperiment to loom<a class="anchor" aria-label="anchor" href="#singlecellexperiment-to-loom"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove seger.loom first</span></span>
<span><span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>, to <span class="op">=</span> <span class="st">"loom"</span>, sce <span class="op">=</span> <span class="va">seger</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"~/gefetch2r/doc/conversion/seger.loom"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert SingleCellExperiment to loom.</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="loom-to-singlecellexperiment">loom to SingleCellExperiment<a class="anchor" aria-label="anchor" href="#loom-to-singlecellexperiment"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seger.loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SCELoom.html">SCELoom</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="st">"loom"</span>, to <span class="op">=</span> <span class="st">"SingleCellExperiment"</span>,</span>
<span>  loom.file <span class="op">=</span> <span class="st">"/Users/soyabean/Desktop/tmp/scdown/conversion/seger.loom"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert loom to SingleCellExperiment.</span></span>
<span><span class="va">seger.loom</span></span>
<span><span class="co"># class: SingleCellExperiment</span></span>
<span><span class="co"># dim: 26179 3514</span></span>
<span><span class="co"># metadata(0):</span></span>
<span><span class="co"># assays(1): counts</span></span>
<span><span class="co"># rownames(26179): SGIP1 AZIN2 ... BIVM-ERCC5.1 eGFP</span></span>
<span><span class="co"># rowData names(2): refseq symbol</span></span>
<span><span class="co"># colnames(3514): HP1502401_N13 HP1502401_D14 ... HP1526901T2D_O11 HP1526901T2D_A8</span></span>
<span><span class="co"># colData names(9): age body.mass.index ... single.cell.well.quality</span></span>
<span><span class="co">#   submitted.single.cell.quality</span></span>
<span><span class="co"># reducedDimNames(0):</span></span>
<span><span class="co"># altExpNames(0):</span></span></code></pre></div>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
