<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Extract Metadata of CELLxGENE Datasets with Attributes. — ExtractCELLxGENEMeta • GEfetch2R</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extract Metadata of CELLxGENE Datasets with Attributes. — ExtractCELLxGENEMeta"><meta name="description" content="Extract Metadata of CELLxGENE Datasets with Attributes."><meta property="og:description" content="Extract Metadata of CELLxGENE Datasets with Attributes."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Extract Metadata of CELLxGENE Datasets with Attributes.</h1>
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/R/cellxgene.R" class="external-link"><code>R/cellxgene.R</code></a></small>
      <div class="d-none name"><code>ExtractCELLxGENEMeta.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Extract Metadata of CELLxGENE Datasets with Attributes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ExtractCELLxGENEMeta</span><span class="op">(</span></span>
<span>  <span class="va">all.samples.df</span>,</span>
<span>  organism <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ethnicity <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sex <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tissue <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  disease <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  suspension.type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell.type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell.num <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-all-samples-df">all.samples.df<a class="anchor" aria-label="anchor" href="#arg-all-samples-df"></a></dt>
<dd><p>All detail information of CELLxGENE datasets, obtained with <code>ShowCELLxGENEDatasets</code>.</p></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>The organism of the datasets, choose from "Homo sapiens", "Mus musculus", "Callithrix jacchus",
"Macaca mulatta", "Sus scrofa domesticus", one or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-ethnicity">ethnicity<a class="anchor" aria-label="anchor" href="#arg-ethnicity"></a></dt>
<dd><p>The ethnicity of the datasets, choose from "Asian", "European", "unknown", "na", "African", "Bangladeshi",
"British", "Irish", "East Asian", "African American", "Hispanic or Latin American", "African American or Afro-Caribbean",
"European American", "Finnish", "Indian", "Japanese", "Korean", "Malaysian", "Singaporean Chinese", "American", "Pacific Islander",
"admixed ancestry", "Eskimo", "Han Chinese", "Greater Middle Eastern  (Middle Eastern, North African or Persian)", "multiethnic",
"Jewish Israeli", "South Asian", "Oceanian", "Chinese", one or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-sex">sex<a class="anchor" aria-label="anchor" href="#arg-sex"></a></dt>
<dd><p>The sex of the datasets, choose from "female", "male", "unknown", one or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-tissue">tissue<a class="anchor" aria-label="anchor" href="#arg-tissue"></a></dt>
<dd><p>The tissue of the datasets, obtain available values with <code>StatDBAttribute</code>. One or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-disease">disease<a class="anchor" aria-label="anchor" href="#arg-disease"></a></dt>
<dd><p>The disease of the datasets, obtain available values with <code>StatDBAttribute</code>. One or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>The assay of the datasets, choose from "10x 3' v1", "10x 3' v2", "10x 3' v3", "10x 3' transcription profiling",
"10x 5' v1", "10x 5' v2", "10x 5' transcription profiling", "10x multiome", "10x scATAC-seq", "sci-RNA-seq", "Drop-seq",
"Smart-seq", "Smart-seq2", "Smart-seq v4", "snmC-Seq2", "Visium Spatial Gene Expression", "Seq-Well", "Seq-Well S3", "Patch-seq",
"sci-Plex", "BD Rhapsody Targeted mRNA", "BD Rhapsody Whole Transcriptome Analysis", "Slide-seqV2", "GEXSCOPE technology", "inDrop",
"microwell-seq", "CEL-seq2", "STRT-seq", "DroNc-seq", "MERFISH", "scATAC-seq", "MARS-seq", "TruDrop", one or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-suspension-type">suspension.type<a class="anchor" aria-label="anchor" href="#arg-suspension-type"></a></dt>
<dd><p>The suspension type of the datasets, choose from "nucleus", "cell", "na", one or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-cell-type">cell.type<a class="anchor" aria-label="anchor" href="#arg-cell-type"></a></dt>
<dd><p>The cell type of the datasets, obtain available values with <code>StatDBAttribute</code>. One or multiple values. Default: NULL (All).</p></dd>


<dt id="arg-cell-num">cell.num<a class="anchor" aria-label="anchor" href="#arg-cell-num"></a></dt>
<dd><p>Cell number filter. If NULL, no filter; if one value, lower filter; if two values, low and high filter.
Deault: NULL(without filtering).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Dataframe contains filtered datasets.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>https://gist.github.com/ivirshup/f1a1603db69de3888eacb4bdb6a9317a</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># \donttest{</span></span>
<span><span class="co"># all available datasets</span></span>
<span><span class="va">all.cellxgene.datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># human 10x v2 and v3 datasets</span></span>
<span><span class="va">human.10x.cellxgene.meta</span> <span class="op">&lt;-</span> <span class="fu">ExtractCELLxGENEMeta</span><span class="op">(</span></span>
<span>  all.samples.df <span class="op">=</span> <span class="va">all.cellxgene.datasets</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x 3' v2"</span>, <span class="st">"10x 3' v3"</span><span class="op">)</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># }</span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

