<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ GEfetch2R</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AD2SCE.html">AD2SCE()</a></code>

        </dt>
        <dd>Convert AnnData to SingleCellExperiemnt using scDIOR/zellkonverter/schard.</dd>
      </dl><dl><dt>

          <code><a href="AD2Seu.html">AD2Seu()</a></code>

        </dt>
        <dd>Convert AnnData to SeuratObject using SeuratDisk/sceasy/scDIOR/schard/SeuratDisk+scDIOR.</dd>
      </dl><dl><dt>

          <code><a href="Bam2Fastq.html">Bam2Fastq()</a></code>

        </dt>
        <dd>Convert bam files to fastq files.</dd>
      </dl><dl><dt>

          <code><a href="CheckAPI.html">CheckAPI()</a></code>

        </dt>
        <dd>Check the Availability of APIs.</dd>
      </dl><dl><dt>

          <code><a href="DownloadBam.html">DownloadBam()</a></code>

        </dt>
        <dd>Download bam.</dd>
      </dl><dl><dt>

          <code><a href="DownloadFastq.html">DownloadFastq()</a></code>

        </dt>
        <dd>Download FASTQ Files.</dd>
      </dl><dl><dt>

          <code><a href="DownloadFastq2R.html">DownloadFastq2R()</a></code>

        </dt>
        <dd>Extract Run, Distinguish RNA-seq Type, Download Fastq, Perform Read Mapping, Load Output to R.</dd>
      </dl><dl><dt>

          <code><a href="DownloadSRA.html">DownloadSRA()</a></code>

        </dt>
        <dd>Download SRA.</dd>
      </dl><dl><dt>

          <code><a href="ExportSeurat.html">ExportSeurat()</a></code>

        </dt>
        <dd>Export SeuratObject to Other Formats.</dd>
      </dl><dl><dt>

          <code><a href="ExtractCBComposition.html">ExtractCBComposition()</a></code>

        </dt>
        <dd>Extract Cell Type Composition of UCSC Cell Browser Datasets.</dd>
      </dl><dl><dt>

          <code><a href="ExtractCBDatasets.html">ExtractCBDatasets()</a></code>

        </dt>
        <dd>Extract UCSC Cell Browser Datasets with Attributes.</dd>
      </dl><dl><dt>

          <code><a href="ExtractCELLxGENEMeta.html">ExtractCELLxGENEMeta()</a></code>

        </dt>
        <dd>Extract Metadata of CELLxGENE Datasets with Attributes.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOExp.html">ExtractGEOExp()</a></code>

        </dt>
        <dd>Extract Raw Count Matrix or Fortmat Supplementary Files to 10x.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOExpSupp.html">ExtractGEOExpSupp()</a></code>

        </dt>
        <dd>Extract Raw Count Matrix from Supplementary Files.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOExpSupp10x.html">ExtractGEOExpSupp10x()</a></code>

        </dt>
        <dd>Fortmat Supplementary Files to 10x.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOExpSupp10xSingle.html">ExtractGEOExpSupp10xSingle()</a></code>

        </dt>
        <dd>Fortmat Supplementary Files to 10x (separate files).</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOExpSuppAll.html">ExtractGEOExpSuppAll()</a></code>

        </dt>
        <dd>Extract Raw Count Matrix from Supplementary Files or Fortmat Supplementary Files to 10x.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOInfo.html">ExtractGEOInfo()</a></code>

        </dt>
        <dd>Extract GEO Study Information.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOMeta.html">ExtractGEOMeta()</a></code>

        </dt>
        <dd>Extract Sample Metadata from GEO.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOMetaSupp.html">ExtractGEOMetaSupp()</a></code>

        </dt>
        <dd>Extract Sample Metadata from Supplementary Files.</dd>
      </dl><dl><dt>

          <code><a href="ExtractGEOSubMeta.html">ExtractGEOSubMeta()</a></code>

        </dt>
        <dd>Extract Sample Metadata.</dd>
      </dl><dl><dt>

          <code><a href="ExtractHCACatalogs.html">ExtractHCACatalogs()</a></code>

        </dt>
        <dd>Show All Available Catalogs in Human Cell Atlas.</dd>
      </dl><dl><dt>

          <code><a href="ExtractHCAMeta.html">ExtractHCAMeta()</a></code>

        </dt>
        <dd>Extract Metadata of Human Cell Atlas Projects with Attributes.</dd>
      </dl><dl><dt>

          <code><a href="ExtractPanglaoDBComposition.html">ExtractPanglaoDBComposition()</a></code>

        </dt>
        <dd>Extract Cell Type Composition of PanglaoDB Datasets.</dd>
      </dl><dl><dt>

          <code><a href="ExtractPanglaoDBMeta.html">ExtractPanglaoDBMeta()</a></code>

        </dt>
        <dd>Extract Metadata of scRNA-seq Datasets in PanglaoDB.</dd>
      </dl><dl><dt>

          <code><a href="ExtractRun.html">ExtractRun()</a></code>

        </dt>
        <dd>Extract Runs with GEO Accession Number or GSM Number.</dd>
      </dl><dl><dt>

          <code><a href="ExtractZenodoMeta.html">ExtractZenodoMeta()</a></code>

        </dt>
        <dd>Prepare Dataframe with Zenodo DOIs.</dd>
      </dl><dl><dt>

          <code><a href="Fastq2R.html">Fastq2R()</a></code>

        </dt>
        <dd>Pipe FASTQ files to SeuratObject and DESeqDataSet.</dd>
      </dl><dl><dt>

          <code><a href="ImportSeurat.html">ImportSeurat()</a></code>

        </dt>
        <dd>Convert Other Formats to SeuratObject.</dd>
      </dl><dl><dt>

          <code><a href="LoadRData.html">LoadRData()</a></code>

        </dt>
        <dd>Load Objects, Count Matrix, and Metadata from RData.</dd>
      </dl><dl><dt>

          <code><a href="PanglaoDBComposition.html">PanglaoDBComposition</a></code>

        </dt>
        <dd>All Sample Composition of PanglaoDB Datasets</dd>
      </dl><dl><dt>

          <code><a href="PanglaoDBMeta.html">PanglaoDBMeta</a></code>

        </dt>
        <dd>All Sample Metadata of PanglaoDB Datasets</dd>
      </dl><dl><dt>

          <code><a href="ParseCBDatasets.html">ParseCBDatasets()</a></code>

        </dt>
        <dd>Download UCSC Cell Browser Datasets.</dd>
      </dl><dl><dt>

          <code><a href="ParseCELLxGENE.html">ParseCELLxGENE()</a></code>

        </dt>
        <dd>Download CELLxGENE Datasets and Return SeuratObject.</dd>
      </dl><dl><dt>

          <code><a href="ParseGEO.html">ParseGEO()</a></code>

        </dt>
        <dd>Download Matrix from GEO and Load to Seurat/DESeq2.</dd>
      </dl><dl><dt>

          <code><a href="ParseGEOProcessed.html">ParseGEOProcessed()</a></code>

        </dt>
        <dd>Download Processed Objects from GEO.</dd>
      </dl><dl><dt>

          <code><a href="ParseHCA.html">ParseHCA()</a></code>

        </dt>
        <dd>Download Human Cell Atlas Datasets.</dd>
      </dl><dl><dt>

          <code><a href="ParsePanglaoDB.html">ParsePanglaoDB()</a></code>

        </dt>
        <dd>Parse PanglaoDB Data.</dd>
      </dl><dl><dt>

          <code><a href="ParseZenodo.html">ParseZenodo()</a></code>

        </dt>
        <dd>Download Data with Zenodo DOI and Return SeuratObject.</dd>
      </dl><dl><dt>

          <code><a href="RunCellRanger.html">RunCellRanger()</a></code>

        </dt>
        <dd>Run CellRanger on Downloaded FASTQ Files.</dd>
      </dl><dl><dt>

          <code><a href="RunSTAR.html">RunSTAR()</a></code>

        </dt>
        <dd>Run STAR on Downloaded FASTQ Files.</dd>
      </dl><dl><dt>

          <code><a href="SCE2AD.html">SCE2AD()</a></code>

        </dt>
        <dd>Convert SingleCellExperiemnt to AnnData using sceasy/scDIOR/zellkonverter.</dd>
      </dl><dl><dt>

          <code><a href="SCEAnnData.html">SCEAnnData()</a></code>

        </dt>
        <dd>Data Format Conversion between SingleCellExperiment and AnnData.</dd>
      </dl><dl><dt>

          <code><a href="SCELoom.html">SCELoom()</a></code>

        </dt>
        <dd>Data Format Conversion between SingleCellExperiment and loom.</dd>
      </dl><dl><dt>

          <code><a href="Seu2AD.html">Seu2AD()</a></code>

        </dt>
        <dd>Convert SeuratObject to AnnData using SeuratDisk/sceasy/scDIOR.</dd>
      </dl><dl><dt>

          <code><a href="ShowCBDatasets.html">ShowCBDatasets()</a></code>

        </dt>
        <dd>Show All Available Datasets in UCSC Cell Browser.</dd>
      </dl><dl><dt>

          <code><a href="ShowCELLxGENEDatasets.html">ShowCELLxGENEDatasets()</a></code>

        </dt>
        <dd>Show All Available Datasets in CELLxGENE.</dd>
      </dl><dl><dt>

          <code><a href="ShowHCAProjects.html">ShowHCAProjects()</a></code>

        </dt>
        <dd>Show All Available Projects in Human Cell Atlas.</dd>
      </dl><dl><dt>

          <code><a href="SplitSRA.html">SplitSRA()</a></code>

        </dt>
        <dd>Split SRA to fastq Files and Format to 10x Standard Style.</dd>
      </dl><dl><dt>

          <code><a href="StatDBAttribute.html">StatDBAttribute()</a></code>

        </dt>
        <dd>Stat Database Attributes.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

