<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Download Matrix from GEO and Load to Seurat/DESeq2. — ParseGEO • GEfetch2R</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Download Matrix from GEO and Load to Seurat/DESeq2. — ParseGEO"><meta name="description" content="Download Matrix from GEO and Load to Seurat/DESeq2."><meta property="og:description" content="Download Matrix from GEO and Load to Seurat/DESeq2."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEfetch2R</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/QuickStart.html">Quick start</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadRaw.html">Download raw data (sra/fastq/bam)</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadMatrices.html">Download count matrix and annotation</a></li>
    <li><a class="dropdown-item" href="../articles/DownloadObjects.html">Download processed object</a></li>
    <li><a class="dropdown-item" href="../articles/ObjectConversion.html">Format conversion and benchmark</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/showteeth/GEfetch2R"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Download Matrix from GEO and Load to Seurat/DESeq2.</h1>
      <small class="dont-index">Source: <a href="https://github.com/showteeth/GEfetch2R/blob/HEAD/R/GEO.R" class="external-link"><code>R/GEO.R</code></a></small>
      <div class="d-none name"><code>ParseGEO.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Download Matrix from GEO and Load to Seurat/DESeq2.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ParseGEO</span><span class="op">(</span></span>
<span>  <span class="va">acce</span>,</span>
<span>  platform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  down.supp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  supp.idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  timeout <span class="op">=</span> <span class="fl">3600</span>,</span>
<span>  data.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sc"</span>, <span class="st">"bulk"</span><span class="op">)</span>,</span>
<span>  supp.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"count"</span>, <span class="st">"10x"</span>, <span class="st">"10xSingle"</span><span class="op">)</span>,</span>
<span>  file.regex <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extra.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"strand"</span>, <span class="st">"length"</span>, <span class="st">"width"</span>, <span class="st">"chromosome"</span>,</span>
<span>    <span class="st">"seqnames"</span>, <span class="st">"seqname"</span>, <span class="st">"chrom"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"seqid"</span>, <span class="st">"stop"</span><span class="op">)</span>,</span>
<span>  transpose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  accept.fmt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MEX"</span>, <span class="st">"h5"</span><span class="op">)</span>,</span>
<span>  gene2feature <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  load2R <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  merge <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  meta.data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fmu <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-acce">acce<a class="anchor" aria-label="anchor" href="#arg-acce"></a></dt>
<dd><p>GEO accession number.</p></dd>


<dt id="arg-platform">platform<a class="anchor" aria-label="anchor" href="#arg-platform"></a></dt>
<dd><p>Platform information/field. Disable when <code>down.supp</code> is TRUE. Default: NULL (disable).</p></dd>


<dt id="arg-down-supp">down.supp<a class="anchor" aria-label="anchor" href="#arg-down-supp"></a></dt>
<dd><p>Logical value, whether to download supplementary files to create count matrix. If TRUE, always
download supplementary files. If FALSE, use <code>ExpressionSet</code> (If contains non-integer or empty,
download supplementary files automatically). Default: FALSE.</p></dd>


<dt id="arg-supp-idx">supp.idx<a class="anchor" aria-label="anchor" href="#arg-supp-idx"></a></dt>
<dd><p>The index of supplementary files to download. This should be consistent with <code>platform</code>. Used when <code>supp.type</code> is 10x/count. Default: 1.</p></dd>


<dt id="arg-timeout">timeout<a class="anchor" aria-label="anchor" href="#arg-timeout"></a></dt>
<dd><p>Timeout for <code><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></code>. Default: 3600.</p></dd>


<dt id="arg-data-type">data.type<a class="anchor" aria-label="anchor" href="#arg-data-type"></a></dt>
<dd><p>The data type of the dataset, choose from "sc" (single-cell) and "bulk" (bulk). Default: "sc".</p></dd>


<dt id="arg-supp-type">supp.type<a class="anchor" aria-label="anchor" href="#arg-supp-type"></a></dt>
<dd><p>The type of downloaded supplementary files, choose from count (count matrix file or single count matrix file),
10x (cellranger output files in tar/gz supplementary files, contains barcodes, genes/features and matrix, e.g. GSE200257)
and 10xSingle (cellranger output files in supplementary files directly, e.g. GSE236082). Default: count.</p></dd>


<dt id="arg-file-regex">file.regex<a class="anchor" aria-label="anchor" href="#arg-file-regex"></a></dt>
<dd><p>The regex to extract correct count matrix files. Used when <code>supp.type</code> is count.
Default: NULL (when multiple file extensions are available in the downloaded tar file, use the first file extension).</p></dd>


<dt id="arg-extra-cols">extra.cols<a class="anchor" aria-label="anchor" href="#arg-extra-cols"></a></dt>
<dd><p>Extra columns to remove, e.g., "Chr", "Start", "End", "Strand", "Length" (featureCounts). Used when <code>supp.type</code> is count.
Default: "chr", "start", "end", "strand", "length", "width", "chromosome", "seqnames", "seqname", "chrom", "chromosome_name", "seqid", "stop".</p></dd>


<dt id="arg-transpose">transpose<a class="anchor" aria-label="anchor" href="#arg-transpose"></a></dt>
<dd><p>Logical value, whether to transpose the matrix. Used when the number of rows is less than the number of columns. Used when <code>supp.type</code> is count. Default: TRUE.</p></dd>


<dt id="arg-out-folder">out.folder<a class="anchor" aria-label="anchor" href="#arg-out-folder"></a></dt>
<dd><p>Output folder to save 10x files. Used when <code>supp.type</code> is 10x/10xSingle. Default: NULL (current working directory).</p></dd>


<dt id="arg-accept-fmt">accept.fmt<a class="anchor" aria-label="anchor" href="#arg-accept-fmt"></a></dt>
<dd><p>Vector of accepted 10x output format, MEX (barcode/feature/gene/matrix), h5 (h5/h5.gz). Used when <code>supp.type</code> is 10x/10xSingle. Default: c("MEX", "h5").</p></dd>


<dt id="arg-gene-feature">gene2feature<a class="anchor" aria-label="anchor" href="#arg-gene-feature"></a></dt>
<dd><p>Logical value, whether to rename <code>genes.tsv.gz</code> to <code>features.tsv.gz</code>. Used when <code>supp.type</code> is 10x/10xSingle. Default: TRUE.</p></dd>


<dt id="arg-load-r">load2R<a class="anchor" aria-label="anchor" href="#arg-load-r"></a></dt>
<dd><p>Logical value, whether to load the count matrix to R. Default: TRUE.</p></dd>


<dt id="arg-merge">merge<a class="anchor" aria-label="anchor" href="#arg-merge"></a></dt>
<dd><p>Logical value, whether to merge Seurat list when there are multiple 10x files (<code>supp.type</code> is 10x).
Used when <code>supp.type</code> is 10x/10xSingle. Default: FALSE.</p></dd>


<dt id="arg-meta-data">meta.data<a class="anchor" aria-label="anchor" href="#arg-meta-data"></a></dt>
<dd><p>Dataframe contains sample information for DESeqDataSet. Used when <code>data.type</code> is bulk. Default: NULL.</p></dd>


<dt id="arg-fmu">fmu<a class="anchor" aria-label="anchor" href="#arg-fmu"></a></dt>
<dd><p>Column of <code>meta.data</code> contains group information. Used when <code>data.type</code> is bulk. Default: NULL.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Parameters for <code>getGEO</code>. Used when <code>down.supp</code> is FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If <code>load2R</code> is FALSE, return count matrix. If <code>data.type</code> is "sc", return SeuratObject (if <code>merge</code> is TRUE), SeuratObject list (if <code>merge</code> is FALSE), NULL (no SeuratObject detected).
If <code>data.type</code> is "bulk", return DESeqDataSet.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span>
<span><span class="co"># the supp files are count matrix</span></span>
<span><span class="va">GSE94820.seu</span> <span class="op">&lt;-</span> <span class="fu">ParseGEO</span><span class="op">(</span>acce <span class="op">=</span> <span class="st">"GSE94820"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, supp.type <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span>
<span><span class="co"># the supp files are cellranger output files: barcodes, genes/features and matrix</span></span>
<span><span class="co"># need users to provide the output folder</span></span>
<span><span class="va">GSE200257.seu</span> <span class="op">&lt;-</span> <span class="fu">ParseGEO</span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE200257"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.idx <span class="op">=</span> <span class="fl">1</span>, supp.type <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/output/folder"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># need users to provide the output folder</span></span>
<span><span class="va">GSE236082.seu</span> <span class="op">&lt;-</span> <span class="fu">ParseGEO</span><span class="op">(</span></span>
<span>  acce <span class="op">=</span> <span class="st">"GSE236082"</span>, down.supp <span class="op">=</span> <span class="cn">TRUE</span>, supp.type <span class="op">=</span> <span class="st">"10xSingle"</span>,</span>
<span>  out.folder <span class="op">=</span> <span class="st">"/path/to/output/folder"</span></span>
<span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="co"># }</span></span></code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yabing Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

